#!/usr/bin/env python2

#  Copyright 2012 Unknown <diogo@arch>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#
#  Author: Diogo N. Silva
#  Version: 0.1
#  Last update: 11/02/14


################################################################################

"""
This module contains the text content for the informative popups that can be
accessed by clicking info (?) buttons throughout the app. The text content
will be stored in a single dictionary variable, whose format should be the
following:
   ..: key: string id of the informative content
   ..: value: list containing the actual informative content
      ..: 1st element: string with the title
      ..: 2nd element: string with text body

Both list elements will be kivy Label objects that support markdown syntax
"""

informative_storage = {
    "process_dataset": ["[b]Data set selection - Help[/b]", "The '[b]"
        "Data set[/b]' option offers a convenient way to perform "
        "operations on specific sets of taxa and/or files. User specified "
        "groups can be selected by toggling the taxon or file buttons in 'Menu"
        " > Open/View data' and selecting the 'Active taxa/files' option or "
        "by creating taxa groups in 'Menu > Taxa Groups' and selecting the "
        "name of the corresponding group."],
    "zorro": ["[b]ZORRO weights - Help[/b]", "(Concatenation only)\n\nThis "
        "option allows the user to"
        " concatenate alignment weight files that are generated by the "
        "ZORRO software [u][ref=http://journals.plos.org/plosone/article?id=10"
        ".1371/journal.pone.0030288](Wu M et al. 2012, PLoS ONE 7(1): "
        "e30288)[/ref][/u].\n\nIn order to maintain correspondence between the "
        "alignment and weight files, the"
        " weight file names should begin with the name of the corresponding "
        "alignment file and end with a suffix that must be provided. For"
        " example, if the files:\n\n -alignment_a.fas\n -alignment_b.fas\n"
        "\nare to be concatenated, then the corresponding zorro weight files "
        "should be:\n\n -alignment_a_zorro.txt\n -alignment_b_zorro.txt\n\n"
        "where '_zorro' is the suffix."],
    "collapse": ["[b]Collapse - Help", "The [b]Collapse[/b] option will merge"
        " all equal sequences into single haplotypes, so that the final "
        "alignment  contains only unique sequences. As a consequence, the taxa"
        " names will be replaced by haplotype names, which can be partially "
        "determined by the  user using the [b]Haplotype prefix[/b] option, and "
        "the correspondence between haplotype names and taxa names with the "
        "same sequence will be written into an *.haplotypes file."],
    "filter": ["[b]Gap/Missing Filter - Help", "The [b]Gap/Missing data[/b] "
        "option will perform two tasks:\n\n - Replacement of the gap symbols at"
        " the beginning and end of the alignments created by the "
        "alignment software;\n\n - Removal of alignment columns containing a "
        "proportion of gaps and/or missing data higher than the thresholds "
        "specified using the [b]Gap/Missing data filter[/b] option.\n\nThe "
        "[b]gap threshold[/b] refers only to gap characters found in the middle"
        " of the alignment. The [b]missing data threshold[/b] refers to the"
        " combination of gap and missing data characters found in any column of"
        "the alignment.\n\nIf only the first task is desired, both filter "
        "thresholds should be set to '100'."],
    "taxa_filter": ["Taxa Filter - Help", "The [b]Taxa Filter[/b] option allows"
        " alignments to be filtered depending on whether they contain or"
        " exclude a given set of taxa. When choosing the filter mode "
        " '[b]Contain[/b]', only those alignments containing at least the"
        " specified taxa in the taxa group will be saved. When choosing the "
        "'[b]Exclude[/b]' filter mode, alignments that contain [i]all[/i] the "
        "taxa specified in the taxa group will [i]not[/i] be saved."],
    "codon_filter": ["Codon Filter - Help", "The [b]Codon filter[/b] option"
        " is only available for nucleotide sequences, where it allows the "
        "filtering of codon positions in the alignments. By default, all "
        "positions are set to be saved but any combination can be specified to "
        "be filtered out."],
    "variation_filter": ["Variation Filter - Help", "The [b]variation filter["
        "/b] option allows alignments to be filtered according their amount "
        "or type of sequence variation."],
    "consensus": ["Alignment consensus - Help", "The alignment consensus "
        "option will merge all sequences in an alignment into a single "
        "consensus sequence. Variation in alignment columns can be either "
        "converted into the corresponding IUPAC symbols, soft masked or "
        "removed from the final consensus"],
    "consensus_mode": ["Consensus variation handling - Help", "The consensus "
        "mode option determines how variation in an alignment will be handled "
        "to create a consensus sequence. The options are:\n\n[b]IUPAC:[/b] "
        "Converts variable sites using the IUPAC ambiguity code.\n\n[b]Soft"
        " mask:[/b] Replaces variable sites with missing data symbol.\n\n[b]"
        "Remove:[/b] Removes variable sites.\n\n[b]First sequence:[/b] Uses"
        " the first sequence in each alignment"],
    "info_sequence_size": ["Distribution of sequence size - Help",
        "Plots the distribution of the average sequence size"
        " either for each species or for the entire active"
        " data set."],
    "info_nucleotide_proportion": ["Nucleotide or residue proportion - Help",
        "Plots the proportion of nucleotides (for DNA data) or residues (for"
        " amino acid data) either for each species of for the entire active "
        "data set."]
}

orthology_storage = {
    "Taxa distribution": "Distribution of the number of taxa across clusters",
    "Taxa coverage": "Shows available and missing ortholog data per taxon",
    "Gene copy distribution": "Distribution of gene copy numbers across "
        "clusters",
    "Taxa gene copies": "Cumulative number of multiple gene copies per taxon"
}