

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>trifusion.process.sequence &mdash; TriFusion 0.5.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="TriFusion 0.5.5 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> TriFusion
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../trifusion.html">TriFusion modules and API reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">TriFusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>trifusion.process.sequence</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for trifusion.process.sequence</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">The `sequence` module of TriFusion contains the main classes handing alignment</span>
<span class="sd">sequence data. These are :class:`.Alignment` and</span>
<span class="sd">`AlignmentList`. Here follows a</span>
<span class="sd">brief explanation of how these classes work and how to deal with the</span>
<span class="sd">sqlite database.</span>

<span class="sd">`Alignment` class</span>
<span class="sd">-----------------</span>

<span class="sd">The `Alignment` class is the main interface with single alignment files. It</span>
<span class="sd">can be viewed as the building block of an `AlignmentList` object, which</span>
<span class="sd">can have one or more `Alignment` objects. It contains all methods and</span>
<span class="sd">attributes that pertain to a given alignment and are used to retrieve</span>
<span class="sd">information</span>
<span class="sd">or modify it. **However, it is NOT meant to be used independently**, but rather</span>
<span class="sd">within the context of an `AlignmentList` object (See `AlignmentList class`_).</span>
<span class="sd">The data from each alignment is stored in a single sqlite database during</span>
<span class="sd">the execution of TriFusion or the TriSeq/TriStats CLI programs. The connection</span>
<span class="sd">to this database is automatically handled by `AlignmentList`</span>
<span class="sd">for all `Alignment` objects included in it. In this way, we can use the</span>
<span class="sd">`AlignmentList` class to handle the setup of the sqlite3 database, and</span>
<span class="sd">focus on single alignment data handling in general in this class.</span>

<span class="sd">The main types of methods defined in this class are:</span>

<span class="sd">Parsers</span>
<span class="sd">~~~~~~~</span>

<span class="sd">Parsing methods are defined with `_read_&lt;format&gt;` (e.g. `_read_fasta`) and</span>
<span class="sd">they are always called</span>
<span class="sd">from the `read_alignment` method. When an `Alignment` object is instantiated</span>
<span class="sd">with a path to an alignment file, it automatically detects the format of</span>
<span class="sd">the alignment and keeps that information on the `input_format` attribute.</span>
<span class="sd">The `read_alignment` method then calls the parsing method corresponding to</span>
<span class="sd">that format. That information is stored in a dictionary::</span>

<span class="sd">    parsing_methods = {</span>
<span class="sd">        &quot;phylip&quot;: self._read_phylip,</span>
<span class="sd">        &quot;fasta&quot;: self._read_fasta,</span>
<span class="sd">        &quot;loci&quot;: self._read_loci,</span>
<span class="sd">        &quot;nexus&quot;: self._read_nexus,</span>
<span class="sd">        &quot;stockholm&quot;: self._read_stockholm</span>
<span class="sd">    }</span>

<span class="sd">    # Call the appropriate method</span>
<span class="sd">    parsing_methods[self.input_format]()</span>

<span class="sd">Each format has its own parsing method (which can be modified directly). To</span>
<span class="sd">add a new format, it is necessary to add it to the automatic format</span>
<span class="sd">recognition in `trifusion.process.base.Base`. Then, create the new parsing</span>
<span class="sd">method, using the same `_read_&lt;format&gt;` notation and add it to the</span>
<span class="sd">`parsing_methods` dictionary in `read_alignment`.</span>

<span class="sd">Data fetching</span>
<span class="sd">~~~~~~~~~~~~~</span>

<span class="sd">To facilitate fetching alignment data from the database, several generators</span>
<span class="sd">and data retrieval methods are defined. These should always use the</span>
<span class="sd">`setup_intable` decorator and defined with the `table_suffix`, `table_name`</span>
<span class="sd">and `table` arguments. For instance, the `iter_sequences` method is a</span>
<span class="sd">generator that allows the iteration over the sequences in the alignment</span>
<span class="sd">and is defined as::</span>

<span class="sd">    @setup_intable</span>
<span class="sd">    def iter_sequences(self, table_suffix=&quot;&quot;, table_name=None, table=None):</span>

<span class="sd">When calling these methods, only the `table_suffix` and `table_name` have</span>
<span class="sd">to be provided. In fact, the value provided to the `table` argument</span>
<span class="sd">at calling time is ignored. The decorator will check the values of both</span>
<span class="sd">`table_suffix` and `table_name` and evaluate the database table that will</span>
<span class="sd">be used. This final table name will then be provided as</span>
<span class="sd">the `table` argument value within the decorator. In this way, these methods</span>
<span class="sd">can be called like::</span>

<span class="sd">    for seq in self.iter_sequences(table_suffix=&quot;_collapse&quot;):</span>
<span class="sd">        # Do stuff</span>

<span class="sd">In this case, the `setup_intable` decorator will append the `table_suffix`</span>
<span class="sd">to the name of the original alignment table. If `Alignment.table_name=&quot;main&quot;`,</span>
<span class="sd">then the final table name in this case will be &quot;main_collapse&quot;.</span>

<span class="sd">Alternatively, we can use `table_name`::</span>

<span class="sd">    for seq in self.iter_sequences(table_name=&quot;main_2&quot;):</span>
<span class="sd">        # Do stuff</span>

<span class="sd">In this case, the final table name will be &quot;main_2&quot;.</span>

<span class="sd">If the final table name does not exist, the method falls back to the original</span>
<span class="sd">table name defined by the `table_name` attribute.</span>

<span class="sd">Alignment modifiers</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Methods that perform modifications to the alignment are also defined here.</span>
<span class="sd">For example, the `collapse` method transforms the original alignment into a</span>
<span class="sd">new one that contains only unique sequences. An important factor to</span>
<span class="sd">take into account with alignment modifying methods, is that it may be</span>
<span class="sd">important to preserve the original alignment data for future operations.</span>
<span class="sd">In TriFusion, the original alignment must be available at all times since</span>
<span class="sd">users may perform any number of process executions in a single session.</span>
<span class="sd">Therefore, all methods that can potentially modify the original alignment</span>
<span class="sd">need to be decorated with the `setup_database` function, and must be</span>
<span class="sd">defined with at least the `table_in` and `table_out` arguments. The decorator</span>
<span class="sd">and these arguments will work together to determine the database&#39;s table</span>
<span class="sd">from where the data will be fetched, and to where the modified alignment</span>
<span class="sd">will be written. For</span>
<span class="sd">instance, the `collapse` method is defined as::</span>

<span class="sd">    @setup_database</span>
<span class="sd">    def collapse(..., table_in=None, table_out=&quot;collapsed&quot;,</span>
<span class="sd">                 use_main_table=False):</span>

<span class="sd">If we want to perform a collapse of the original alignment and store</span>
<span class="sd">the modified alignment in a new table, we could call `collapse` like::</span>

<span class="sd">    collapse(table_out=&quot;new_table&quot;)</span>

<span class="sd">The `setup_database` decorator interprets `table_in=None` as an instruction</span>
<span class="sd">to use the table</span>
<span class="sd">with the original alignment, and stores the modified alignment in</span>
<span class="sd">`&quot;new_table&quot;`.</span>

<span class="sd">However, we may want to perform a collapse operation after a previous</span>
<span class="sd">modification from other method. In that case, we can specify a</span>
<span class="sd">`table_in`::</span>

<span class="sd">    collapse(table_in=&quot;other_table&quot;, table_out=&quot;collapse&quot;)</span>

<span class="sd">One issue with this approach is that we do not know *a priori* which</span>
<span class="sd">operations will be requested by the user nor the order. If one execution</span>
<span class="sd">performs, say, a `consensus` and a `collapse`, the new table should be created</span>
<span class="sd">in `consensus` and then used as input in `collapse`. However, if only</span>
<span class="sd">`collapse`</span>
<span class="sd">is called, then the new table should only be created there. To solve this</span>
<span class="sd">issue, the `setup_database` decorator is smart about its arguments.</span>
<span class="sd">We can create a sequence of operations with the same `table_in` and `table_out`</span>
<span class="sd">arguments::</span>

<span class="sd">    new_table = &quot;master_table&quot;</span>
<span class="sd">    if &quot;consensus&quot; in operations:</span>
<span class="sd">        consensus(table_in=new_table, table_out=new_table)</span>
<span class="sd">    if &quot;collapse&quot; in operations:</span>
<span class="sd">        collapse(table_in=new_table, table_out=new_table)</span>

<span class="sd">In this simple pipeline, the user may perform either a `consensus`, a</span>
<span class="sd">`collapse`, or both. When the first method is called, the</span>
<span class="sd">`setup_database` decorator will check if the table</span>
<span class="sd">provided in `table_in` exists. In the first called method it will not exist,</span>
<span class="sd">so instead of returning an error, it falls back to the original alignment</span>
<span class="sd">table and then writes the modified alignment to &quot;master_table&quot;. In the second</span>
<span class="sd">method, `table_in` already exists, so it fetches alignment data from the</span>
<span class="sd">&quot;master_table&quot;. This will work whether these methods are called individually</span>
<span class="sd">or in combination.</span>

<span class="sd">When there is no need to keep the original alignment data (in single</span>
<span class="sd">execution of TriSeq, for instance), the special `use_main_table` argument</span>
<span class="sd">can be provided to tell the method to use the original table as the input and</span>
<span class="sd">output table. If this argument is True, it supersedes</span>
<span class="sd">any information provided by `table_in` or `table_out`::</span>

<span class="sd">    collapse(use_main_table=True)</span>

<span class="sd">Writers</span>
<span class="sd">~~~~~~~</span>

<span class="sd">Like parsers, writer methods are defined with `_write_&lt;format&gt;` and are</span>
<span class="sd">always called from the `write_to_file` method. When the `write_to_file`</span>
<span class="sd">method is called, a list with the requested output formats is also provided.</span>
<span class="sd">Then, for each format specified in the argument, the corresponding writer</span>
<span class="sd">method is called. That method is responsible for fetching the data from</span>
<span class="sd">the database and write it to an output file in the appropriate format.</span>
<span class="sd">The map between the formats and the methods is stored in a dictionary::</span>

<span class="sd">    write_methods = {</span>
<span class="sd">        &quot;fasta&quot;: self._write_fasta,</span>
<span class="sd">        &quot;phylip&quot;: self._write_phylip,</span>
<span class="sd">        &quot;nexus&quot;: self._write_nexus,</span>
<span class="sd">        &quot;stockholm&quot;: self._write_stockholm,</span>
<span class="sd">        &quot;gphocs&quot;: self._write_gphocs,</span>
<span class="sd">        &quot;ima2&quot;: self._write_ima2,</span>
<span class="sd">        &quot;mcmctree&quot;: self._write_mcmctree</span>
<span class="sd">    }</span>

<span class="sd">    # Call apropriate method for each output format</span>
<span class="sd">    for fmt in output_format:</span>
<span class="sd">        write_methods[fmt](output_file, **kwargs)</span>

<span class="sd">The `output_file` and a `kwargs` dictionary are provided as arguments</span>
<span class="sd">to each of these methods. The `kwargs` dictionary contains all keyword</span>
<span class="sd">arguments used when calling `write_to_file` and each writer method</span>
<span class="sd">fetches the ones relevant to the format. For instance, in the beggining</span>
<span class="sd">of the `_write_fasta` method, the relevant keyword arguments are retrieved::</span>

<span class="sd">    # Get relevant keyword arguments</span>
<span class="sd">    ld_hat = kwargs.get(&quot;ld_hat&quot;, False)</span>
<span class="sd">    interleave = kwargs.get(&quot;interleave&quot;, False)</span>
<span class="sd">    table_suffix = kwargs.get(&quot;table_suffix&quot;, None)</span>
<span class="sd">    table_name = kwargs.get(&quot;table_name&quot;, None)</span>
<span class="sd">    ns_pipe = kwargs.get(&quot;ns_pipe&quot;, None)</span>
<span class="sd">    pbar = kwargs.get(&quot;pbar&quot;, None)</span>

<span class="sd">In addition to the `write_methods` dictionary, a dictionary mapping the</span>
<span class="sd">formats and their corresponding file extensions is also defined::</span>

<span class="sd">    format_ext = {&quot;ima2&quot;: &quot;.txt&quot;,</span>
<span class="sd">                  &quot;mcmctree&quot;: &quot;_mcmctree.phy&quot;,</span>
<span class="sd">                  &quot;phylip&quot;: &quot;.phy&quot;,</span>
<span class="sd">                  &quot;nexus&quot;: &quot;.nex&quot;,</span>
<span class="sd">                  &quot;fasta&quot;: &quot;.fas&quot;,</span>
<span class="sd">                  &quot;stockholm&quot;: &quot;.stockholm&quot;,</span>
<span class="sd">                  &quot;gphocs&quot;: &quot;.txt&quot;}</span>

<span class="sd">To add a new output format writer, simply create the method using the</span>
<span class="sd">`_write_&lt;format&gt;` notation and include it in the `write_methods` dictionary,</span>
<span class="sd">along with the extension in the `format_ext` variable.</span>

<span class="sd">`AlignmentList` class</span>
<span class="sd">---------------------</span>

<span class="sd">The `AlignmentList` is the main interface between the user and the alignment</span>
<span class="sd">data. It may contain one or more `Alignment` objects, which are considered</span>
<span class="sd">the building blocks of the total data set. These `Alignment` objects are bound</span>
<span class="sd">by the same sqlite database connection.</span>

<span class="sd">How to create an instance</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">An `AlignmentList` instance can be created with a single argument, which</span>
<span class="sd">is a list of paths to alignment files::</span>

<span class="sd">    aln_obj = AlignmentList([&quot;file1.fas&quot;, &quot;file2.fas&quot;])</span>

<span class="sd">Even when no information is provided, an sqlite database connection is</span>
<span class="sd">established automatically (generating a &quot;trifusion.sqlite3&quot; file in the</span>
<span class="sd">current working directory by default). However, it is possible and advisable</span>
<span class="sd">to specify the path to the sqlite database::</span>

<span class="sd">    aln_obj = AlignmentList([&quot;file1.fas&quot;, &quot;file2.fas&quot;], sql_path=&quot;.sql.db&quot;)</span>

<span class="sd">A connection to the database can also be provided when instantiating</span>
<span class="sd">the class, so it&#39;s perhaps more useful to see how the `__init__` works::</span>

<span class="sd">    def __init__(self, alignment_list, sql_db=None, db_cur=None, db_con=None,</span>
<span class="sd">                 pbar=None):</span>

<span class="sd">        if db_cur and db_con:</span>
<span class="sd">            self.con = db_con</span>
<span class="sd">            self.cur = db_cur</span>
<span class="sd">        elif sql_db:</span>
<span class="sd">            self.sql_path = sql_db</span>
<span class="sd">        else:</span>
<span class="sd">            self.sql_path = &quot;trifusion.sqlite3&quot;</span>

<span class="sd">As we can see, if a database connection is provided via the `db_cur` and</span>
<span class="sd">`db_con` arguments, the `sql_path` is ignored. If no sqlite information</span>
<span class="sd">is provided, the `sql_path` attribute defaults to &quot;trifusion.sqlite3&quot;.</span>

<span class="sd">Adding alignment data</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Alignment data can be loaded when initializing the class as above and/or</span>
<span class="sd">added later via the `add_alignments` and `add_alignment_files` methods</span>
<span class="sd">(in fact, the `add_alignment_files` method is the one called at `__init__`)::</span>

<span class="sd">    aln_obj.add_alignment_files([&quot;file3.fas&quot;, &quot;file4.fas&quot;])</span>

<span class="sd">The most important aspects when adding alignment data is how each alignment</span>
<span class="sd">is processed and how errors and exceptions are handled. Briefly, the flow is:</span>

<span class="sd">  1. Check for file duplications within the loaded file list. Duplications</span>
<span class="sd">     are stored in the `duplicate_alignments` attribute.</span>
<span class="sd">  2. Check for file duplications between the loaded files and any</span>
<span class="sd">     alignment already present. The duplications go to the same attribute</span>
<span class="sd">     as above.</span>
<span class="sd">  3. For each provided file, create an `Alignment` object. Errors that occur</span>
<span class="sd">     when creating an `Alignment` object are stored in its `e` attribute.</span>
<span class="sd">     This attribute is then checked before adding the alignment to the</span>
<span class="sd">     `AlignmentList` object.</span>
<span class="sd">      1. Check for `InputError` exception. These are malformated files</span>
<span class="sd">         and are stored in the `bad_alignments` attribute.</span>
<span class="sd">      2. Check for `AlignmentUnequalLength` exception. These are sequence</span>
<span class="sd">         sets of</span>
<span class="sd">         unequal length (not alignments) and are stored in the</span>
<span class="sd">         `non_alignments` attribute.</span>
<span class="sd">      3. Check for `EmptyAlignment` exception. These are empty alignments</span>
<span class="sd">         and are</span>
<span class="sd">         stored in the `bad_alignments` attribute.</span>
<span class="sd">      4. Check the sequence code of the `Alignment` object. For now,</span>
<span class="sd">         the `AlignmentList` object only accepts alignments of the same</span>
<span class="sd">         sequence type (e.g. DNA, protein).</span>
<span class="sd">      5. If all of the previous checks pass, add the `Alignment` object</span>
<span class="sd">         to the `alignments` attribute.</span>
<span class="sd">  4. Update the overall taxa names of the full data set after the</span>
<span class="sd">     inclusion of the new alignments.</span>

<span class="sd">Wrapping `Alignment` methods</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">The majority of the methods defined in the `Alignment` object can also be</span>
<span class="sd">accessed in the `AlignmentList` object. These are defined roughly with the</span>
<span class="sd">same arguments in both classes so that their behavior is the same.</span>
<span class="sd">These can be simple wrappers that call the respective `Alignment` method for</span>
<span class="sd">each alignment in the `alignments` attribute. For instance, the</span>
<span class="sd">`change_taxon_name` method is simply::</span>

<span class="sd">    def change_taxon_name(self, old_name, new_name):</span>

<span class="sd">        for alignment_obj in list(self.alignments.values()):</span>
<span class="sd">            alignment_obj.change_taxon_name(old_name, new_name)</span>

<span class="sd">        self.taxa_names = [new_name if x == old_name else x</span>
<span class="sd">                           for x in self.taxa_names]</span>

<span class="sd">To avoid duplicating the argument list, the wrapping method</span>
<span class="sd">can use `args` and `kwargs` to transfer arguments. This ensures that if</span>
<span class="sd">the argument list is modified in the `Alignment` method, it doesn&#39;t need</span>
<span class="sd">any modification in the wrapper method. For instance, the `write_to_file`</span>
<span class="sd">method of `Alignment` accepts a large number of positional and keyword</span>
<span class="sd">arguments, which would be an hassle to define an maintain in the wrapper</span>
<span class="sd">method of `AlignmentList`. So, the `write_to_file` method of</span>
<span class="sd">`AlignmentList` is simply defined as::</span>

<span class="sd">    def write_to_file(self, output_format, conversion_suffix=&quot;&quot;,</span>
<span class="sd">                      output_suffix=&quot;&quot;, *args, **kwargs):</span>

<span class="sd">The `output_format`, `conversion_suffix` and `output_suffix` are the only</span>
<span class="sd">positional arguments required when calling this wrapper. All the remaining</span>
<span class="sd">arguments are packed in the `args` and `kwargs` objects and used</span>
<span class="sd">normally by the wrapped method.</span>

<span class="sd">Exclusive `AlignmentList` methods</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Some methods are exclusive of `AlignmentList` because they only make sense</span>
<span class="sd">to be applied to lists of alignments (e.g. `concatenate`). These have </span>
<span class="sd">more freedom in how they are defined and called.</span>

<span class="sd">Active and inactive datasets</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Taxa and/or alignments can become &#39;inactive&#39;, that is, they</span>
<span class="sd">are temporarily removed from their respective attributes, `taxa_names`</span>
<span class="sd">and `alignments`. This means that these &#39;inactive&#39; elements are ignored</span>
<span class="sd">when performing most operations. To</span>
<span class="sd">change the &#39;active&#39; status of alignments, the `update_active_alignments`</span>
<span class="sd">and `update_active_alignment` methods are available. For taxa, the</span>
<span class="sd">`update_taxa_names` method can be used::</span>

<span class="sd">    # Set only two active alignments</span>
<span class="sd">    self.update_active_alignments([&quot;file1.fas&quot;, &quot;file2.fas&quot;])</span>

<span class="sd">    # Set only two active taxa</span>
<span class="sd">    self.update_taxa_names([&quot;taxonA&quot;, &quot;taxonB&quot;])</span>

<span class="sd">Note that all these modifications</span>
<span class="sd">are reversible. &#39;Inactive&#39; elements are stored in the `shelve_alignments`</span>
<span class="sd">attribute for alignments, and `shelved_taxa` for taxa.</span>

<span class="sd">Updating `Alignment` objects</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Some tasks perform changes to core attributes of `AlignmentList`, but they</span>
<span class="sd">may also be necessary on each `Alignment` object. For instance, the</span>
<span class="sd">`remove_taxa` method is used to remove a list of taxa from the</span>
<span class="sd">`AlignmentList` object. It is easy to change only the relevant `AlignmentList`</span>
<span class="sd">attributes, but this change also requires those particular taxa to be</span>
<span class="sd">removed in all `Alignment` objects. For this reason, such methods should</span>
<span class="sd">be defined in the same way in both classes. Using the `remove_taxa` example::</span>

<span class="sd">    def remove_taxa(self, taxa_list, mode=&quot;remove&quot;):</span>

<span class="sd">        # &lt;changes to AlignmentList&gt;</span>

<span class="sd">        for alignment_obj in list(self.alignments.values()):</span>
<span class="sd">            alignment_obj.remove_taxa(taxa_list, mode=mode)</span>

<span class="sd">As you can see, the usage is the same for both methods.</span>

<span class="sd">Plot data methods</span>
<span class="sd">~~~~~~~~~~~~~~~~~</span>

<span class="sd">The `AlignmentList` object contains all methods that generate data for plotting</span>
<span class="sd">in TriFusion&#39;s Statistics screen and TriStats CLI program. However, it&#39;s</span>
<span class="sd">important to establish a separation between the generation of plot data,</span>
<span class="sd">and the generation of the plot itself. The `AlignmentList` methods only</span>
<span class="sd">generate the data and relevant instructions necessary to</span>
<span class="sd">to draw the plot. This information is then passed on to the appropriate</span>
<span class="sd">plot generation functions, which are defined in `trifusion.base.plotter`.</span>
<span class="sd">The reason for this separation of tasks is that many different alignment</span>
<span class="sd">analyses are represented by the same plot.</span>

<span class="sd">The complete process of how new plots can be added to TriFusion is</span>
<span class="sd">described here_. In this section, we provide only a few guidelines on what</span>
<span class="sd">to expect from these methods.</span>

<span class="sd">All plot data methods must be decorated with the `check_data` function</span>
<span class="sd">and take at least a `Namespace` argument. In most cases, no more arguments</span>
<span class="sd">are required::</span>

<span class="sd">    @check_data</span>
<span class="sd">    def gene_occupancy(self, ns=None):</span>

<span class="sd">The `check_data` decorator is responsible for performing checks before and</span>
<span class="sd">after executing the method. The `Namespace` argument, `ns`, is used to allow</span>
<span class="sd">communication between the main and worker threads of TriFusion.</span>

<span class="sd">Additional keyword arguments may be defined, but in that case they must</span>
<span class="sd">be provided in TriFusion when the `trifusion.app.TriFusionApp.stats_show_plot`</span>
<span class="sd">method is called, using the `additional_args` argument. This object will</span>
<span class="sd">be passed to the `get_stats_data` function in</span>
<span class="sd">`trifusion.data.resources.background_tasks` and used when calling the</span>
<span class="sd">plot data methods::</span>

<span class="sd">    if additional_args:</span>
<span class="sd">        plot_data = methods[stats_idx](ns=ns, **additional_args)</span>
<span class="sd">    else:</span>
<span class="sd">        plot_data = methods[stats_idx](ns)</span>

<span class="sd">The other requirement of plot data methods, is that they must always return</span>
<span class="sd">a single dictionary object. This dictionary must contain at least one</span>
<span class="sd">`key:value` with the &quot;data&quot; key and a numpy.array with the plot data as the</span>
<span class="sd">value. The other entries in the dictionary are optional and refer to</span>
<span class="sd">instructions for plot generation. For example, the</span>
<span class="sd">`missing_data_distribution` method returns::</span>

<span class="sd">    return {&quot;data&quot;: data,</span>
<span class="sd">            &quot;title&quot;: &quot;Distribution of missing data&quot;,</span>
<span class="sd">            &quot;legend&quot;: legend,</span>
<span class="sd">            &quot;ax_names&quot;: [&quot;Proportion&quot;, &quot;Number of genes&quot;],</span>
<span class="sd">            &quot;table_header&quot;: [&quot;Bin&quot;] + legend}</span>

<span class="sd">The first entry is the mandatory &quot;data&quot; key with the numpy.array `data`.</span>
<span class="sd">The other instructions are &quot;title&quot;, which sets the title of the plot, &quot;legend&quot;,</span>
<span class="sd">which provides the labels for the plot legend, &quot;ax_names&quot;, which provides</span>
<span class="sd">the name of the `x` and `y` axis, and the &quot;table_header&quot;, which specified</span>
<span class="sd">the header for the table of that plot.</span>

<span class="sd">The allowed plot instructions depend on the plot function that will be used</span>
<span class="sd">and not all of them need to be specified.</span>

<span class="sd">.. _here: https://github.com/ODiogoSilva/TriFusion/wiki/</span>
<span class="sd">          Add-Statistics-plot-analysis</span>

<span class="sd">Logging progress</span>
<span class="sd">~~~~~~~~~~~~~~~~</span>

<span class="sd">The majority of `AlignmentList` methods support the setup and update of</span>
<span class="sd">progress indicators that can be used in TriFusion (GUI) and the CLI programs.</span>
<span class="sd">In the case of TriFusion, the progress indicator is provided via a</span>
<span class="sd">`multiprocessing.Namespace` object that transfers information between the</span>
<span class="sd">main thread (where the GUI changes take place) and the working thread.</span>
<span class="sd">In the case of the CLI programs, the indicator is provided via a `ProgressBar`</span>
<span class="sd">object. In either case, the setup, update and reseting of the progress</span>
<span class="sd">indicators is perfomed by the same methods.</span>

<span class="sd">At the beginning of an operation, the `_set_pipes` method is called::</span>

<span class="sd">    self._set_pipes(ns, pbar, total=len(self.alignments))</span>

<span class="sd">The `Namespace` object is defined as `ns` and the `ProgressBar` is defined</span>
<span class="sd">as `pbar`. Usually, only one of them is provided, depending on whether it</span>
<span class="sd">was called from TriFusion or from a CLI program. We also set the total</span>
<span class="sd">of the progress indicator. In this case it&#39;s the number of alignments. In case</span>
<span class="sd">the operation is called from TriFusion using the `Namespace` object, this</span>
<span class="sd">method also checks the number of active alignments. If there is only one</span>
<span class="sd">active alignment, it sets a Namespace attribute that will silence the</span>
<span class="sd">progress logging of the `AlignmentList` object and receive the information</span>
<span class="sd">from the `Alignment` object instead::</span>

<span class="sd">    if ns:</span>
<span class="sd">        if len(self.alignments) == 1:</span>
<span class="sd">            ns.sa = True</span>
<span class="sd">        else:</span>
<span class="sd">            ns.sa = False</span>

<span class="sd">Then, inside the task, we can update the progress within a loop using</span>
<span class="sd">the `_update_pipes` method::</span>

<span class="sd">    for p, alignment_obj in enumerate(list(self.alignments.values())):</span>
<span class="sd">        self._update_pipes(ns, pbar, value=p + 1,</span>
<span class="sd">                           msg=&quot;Some message&quot;)</span>

<span class="sd">Here, we provide the `Namespace` and `ProgressBar` objects as before.</span>
<span class="sd">In addition we provide the `value` associated with each iteration.</span>
<span class="sd">Optionally, the `msg` argument can be specified, which is used exclusively</span>
<span class="sd">by TriFusion to show a custom message.</span>

<span class="sd">At the end of a task its good pratice to reset the progress indicators</span>
<span class="sd">by using the `_reset_pipes` method::</span>

<span class="sd">    self._reset_pipes(ns)</span>

<span class="sd">Here, only the `Namespace` object is necessary, since the `ProgressBar`</span>
<span class="sd">indicator is automatically reset on `_set_pipes`.</span>

<span class="sd">Incorporating a kill switch</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">All time consuming methods of `AlignmentList` accept a `Namespace` object</span>
<span class="sd">when called from TriFusion, allowing communication between the main thread</span>
<span class="sd">and the worker thread. Since python `threads` cannot be forcibly terminated</span>
<span class="sd">like `processes`, most methods should listen to a kill switch</span>
<span class="sd">flag that is actually an attribute of the `Namespace` object.</span>
<span class="sd">This kill switch is already incorporated into the `_set_pipes`,</span>
<span class="sd">`_update_pipes` and `_reset_pipes` methods. ::</span>

<span class="sd">    if ns:</span>
<span class="sd">        if ns.stop:</span>
<span class="sd">            raise KillByUser(&quot;&quot;)</span>

<span class="sd">What it does is to listen to a kill signal from TriFusion&#39;s window (it can</span>
<span class="sd">be clicking on the &quot;Cancel&quot; button, for example). When this kill signal is</span>
<span class="sd">received in the main thread, it sets the `Namespace.stop` attribute to True</span>
<span class="sd">in both main and worker threads. In the worker thread, when the `stop`</span>
<span class="sd">attribute evaluates to True, a custom `KillByUser` exception is raised,</span>
<span class="sd">immediatelly stopping the task. The exeption is then handled in the</span>
<span class="sd">`trifusion.data.resources.background_tasks.process_execution` function</span>
<span class="sd">and transmitted to the main thread.</span>

<span class="sd">Using SQLite</span>
<span class="sd">------------</span>

<span class="sd">A great deal of the high performance and memory efficiency of the `sequence`</span>
<span class="sd">module comes from the</span>
<span class="sd">use of sqlite to store and manipulate alignment data on disk rather than RAM.</span>
<span class="sd">This means that parsing, modifying and writing alignment data must be</span>
<span class="sd">done with great care to ensure that only the essential data is loaded into</span>
<span class="sd">memory, while minimizing the number of (expensive) database queries. This</span>
<span class="sd">has some implications for the methods of both `Alignment` and `AlignmentList`</span>
<span class="sd">objects with respect to how parsing, alignment modification and output</span>
<span class="sd">writing is performed.</span>

<span class="sd">Implications for parsing</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">When writing or modifying parsing methods it is important to take into account</span>
<span class="sd">that alignment files can be very large (&gt; 1Gb) and loading the entire</span>
<span class="sd">data into memory should be avoided. Whenever possible, only the data of a</span>
<span class="sd">single taxon should be kept in memory before inserting it into the database</span>
<span class="sd">and then releasing that memory. For most formats,</span>
<span class="sd">particularly leave formats, it&#39;s</span>
<span class="sd">fairly straightforward to do this. However, interleave formats can fragment</span>
<span class="sd">the data from each taxon across the entire file. Since database insertions</span>
<span class="sd">and updates are expensive, loading the data in each line can greatly</span>
<span class="sd">decrease the performance in these formats. Therefore, it&#39;s preferable</span>
<span class="sd">to read the alignment file once per taxon, join the entire sequence</span>
<span class="sd">of that taxon, and then insert it into the database. This ensures that</span>
<span class="sd">only sequence data from one taxon is kept in memory at any given time and</span>
<span class="sd">only a minimal number of database insertions are performed. It will</span>
<span class="sd">also result in the same file being parsed N times, where N is the number of</span>
<span class="sd">taxa. However, the decrease in speed is marginal, since most lines are</span>
<span class="sd">actually skipped, whereas the efficiency increase is substantial.</span>

<span class="sd">Implications for fetching data</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">Retrieving data from an sqlite database is not as simple as accessing python</span>
<span class="sd">native data structure. Therefore, a set of methods and generators have</span>
<span class="sd">been defined in the `Alignment` object to facilitate the interface with</span>
<span class="sd">the data in the database (See `Data fetching`_). When some kind of data</span>
<span class="sd">is required from the database, it is preferable to modify or create a</span>
<span class="sd">dedicated method, instead of interacting with the database directly. This</span>
<span class="sd">creates a layer of abstraction between the database and `Alignment`/</span>
<span class="sd">`AlignmentList` methods that greatly facilitates future updates.</span>

<span class="sd">Implications for alignment modification</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">When performing modification to the alignment data it is important to take</span>
<span class="sd">into account that the original alignment data may need to be preserved for</span>
<span class="sd">future operations. These methods must be defined and called using</span>
<span class="sd">appropriate decorators and arguments that establish the name of the</span>
<span class="sd">database table from where information will be retrieved, and the name of the</span>
<span class="sd">table where the information will be written (See `Alignment modifiers`_).</span>

<span class="sd">Implications for writing files</span>
<span class="sd">~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</span>

<span class="sd">When writing alignment data into new output files, the same caution of the</span>
<span class="sd">alignment parsing is advised. It&#39;s quite easy to let the entire alignment</span>
<span class="sd">be loaded into RAM, particularly when writing in interleave format.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">Counter</span><span class="p">,</span> <span class="n">defaultdict</span><span class="p">,</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="k">import</span> <span class="n">join</span><span class="p">,</span> <span class="n">basename</span><span class="p">,</span> <span class="n">splitext</span><span class="p">,</span> <span class="n">exists</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">compress</span>
<span class="kn">from</span> <span class="nn">threading</span> <span class="k">import</span> <span class="n">Lock</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">import</span> <span class="nn">sqlite3</span>

<span class="c1"># TriFusion imports</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">process</span>
    <span class="kn">from</span> <span class="nn">process.base</span> <span class="k">import</span> <span class="n">dna_chars</span><span class="p">,</span> <span class="n">aminoacid_table</span><span class="p">,</span> <span class="n">iupac</span><span class="p">,</span> \
        <span class="n">iupac_rev</span><span class="p">,</span> <span class="n">iupac_conv</span><span class="p">,</span> <span class="n">Base</span>
    <span class="kn">from</span> <span class="nn">process.data</span> <span class="k">import</span> <span class="n">Partitions</span>
    <span class="kn">from</span> <span class="nn">process.data</span> <span class="k">import</span> <span class="n">PartitionException</span>
    <span class="kn">from</span> <span class="nn">process.error_handling</span> <span class="k">import</span> <span class="n">DuplicateTaxa</span><span class="p">,</span> <span class="n">KillByUser</span><span class="p">,</span> \
        <span class="n">InvalidSequenceType</span><span class="p">,</span> <span class="n">EmptyData</span><span class="p">,</span> <span class="n">InputError</span><span class="p">,</span> <span class="n">EmptyAlignment</span><span class="p">,</span> \
        <span class="n">MultipleSequenceTypes</span><span class="p">,</span> <span class="n">SingleAlignment</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">trifusion.process</span> <span class="k">as</span> <span class="nn">process</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.base</span> <span class="k">import</span> <span class="n">dna_chars</span><span class="p">,</span> <span class="n">aminoacid_table</span><span class="p">,</span> <span class="n">iupac</span><span class="p">,</span> \
        <span class="n">iupac_rev</span><span class="p">,</span> <span class="n">iupac_conv</span><span class="p">,</span> <span class="n">Base</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.data</span> <span class="k">import</span> <span class="n">Partitions</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.data</span> <span class="k">import</span> <span class="n">PartitionException</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.error_handling</span> <span class="k">import</span> <span class="n">DuplicateTaxa</span><span class="p">,</span> <span class="n">KillByUser</span><span class="p">,</span> \
        <span class="n">InvalidSequenceType</span><span class="p">,</span> <span class="n">EmptyData</span><span class="p">,</span> <span class="n">InputError</span><span class="p">,</span> <span class="n">EmptyAlignment</span><span class="p">,</span> \
        <span class="n">MultipleSequenceTypes</span><span class="p">,</span> <span class="n">SingleAlignment</span>

<span class="c1"># import pickle</span>
<span class="c1"># TODO: Create a SequenceSet class for sets of sequences that do not conform</span>
<span class="c1"># to an alignment, i.e. unequal length.This would eliminate the problems of</span>
<span class="c1"># applying methods designed for alignments to sets of sequences with unequal</span>
<span class="c1"># length would allows these sets of sequences to have methods of their own.</span>
<span class="c1"># However, the __init__ of the current Alignment object should apply to both</span>
<span class="c1"># SequenceSet and Alignment classes. So, I&#39;ll have to re-structure the code</span>
<span class="c1"># somehow.</span>
<span class="c1"># TODO After creating the SequenceSet class, an additional class should be</span>
<span class="c1"># used to make the triage of files to either the Alignment or SequenceSet</span>
<span class="c1"># classes</span>

<span class="c1"># Lock mechanism to prevent concurrent access to sqlite database</span>
<span class="n">lock</span> <span class="o">=</span> <span class="n">Lock</span><span class="p">()</span>


<div class="viewcode-block" id="LookupDatabase"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.LookupDatabase">[docs]</a><span class="k">class</span> <span class="nc">LookupDatabase</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Decorator handling hash lookup table with pre-calculated values.</span>
<span class="sd">    </span>
<span class="sd">    This decorator class is used to decorate class methods that calculate </span>
<span class="sd">    pairwise sequence similarity. To ensure proper functionality, the</span>
<span class="sd">    decorated method should be called first with a single &quot;connect&quot; argument</span>
<span class="sd">    and after finishing all calculations, with a single &quot;disconnect&quot; argument.</span>
<span class="sd">    These special method callings are necessary to setup and close the</span>
<span class="sd">    database storing the calculated values, respectively.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>
<span class="sd">    </span>
<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>
<span class="sd">    con : None or sqlite.Connection</span>
<span class="sd">        Connection object of the sqlite database</span>
<span class="sd">    c : None or sqlite.Cursor</span>
<span class="sd">        Cursor object of the sqlite database</span>
<span class="sd">    </span>
<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The `con` and `c` attributes are initialized when the decorated function</span>
<span class="sd">    is called with the single &quot;connect&quot; argument </span>
<span class="sd">    (e.g. decorated_func(&quot;connect&quot;)). When the decorated function is called</span>
<span class="sd">    with the single &quot;disconnect&quot; argument (e.g. decorated_func(&quot;disconnect&quot;)),</span>
<span class="sd">    the database changes are committed, the sqlite Connection is closed and</span>
<span class="sd">    the Cursor is reset.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">func</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">func</span> <span class="o">=</span> <span class="n">func</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">c</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="LookupDatabase.__call__"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.LookupDatabase.__call__">[docs]</a>    <span class="k">def</span> <span class="nf">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Wraps the call of `func`.</span>
<span class="sd">        </span>
<span class="sd">        When the decorated method is called, this code is wrapped around</span>
<span class="sd">        its execution. It accepts an arbitrary number of positional arguments</span>
<span class="sd">        but is is currently designed to decorate the `_get_similarity`</span>
<span class="sd">        method of the `AlignmentList` class. There are three expected calling</span>
<span class="sd">        modes:</span>
<span class="sd">        </span>
<span class="sd">            1. decorated_func(&quot;connect&quot;) : Initializes the Connection and</span>
<span class="sd">            Cursor objects of the database</span>
<span class="sd">            2. decorated_func(seq1, seq2, locus_lenght) : The typical main</span>
<span class="sd">            execution of the method, providing the sequence1 and sequence2</span>
<span class="sd">            strings along with the integer with their lenght</span>
<span class="sd">            3. decorated_func(&quot;disconnect&quot;) : Commits changes to database</span>
<span class="sd">            and closes Connection and Cursor objects.</span>
<span class="sd">        </span>
<span class="sd">        The path of this sqlite database is automatically obtained from </span>
<span class="sd">        the `AlignmentList` attribute `sql_path`. We use the path to the </span>
<span class="sd">        same directory, but with a different name for the database, &quot;pw.db&quot;.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        args : list</span>
<span class="sd">            List of positional arguments for decorated method</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        When the decorated method is called normally (i.e., not with a </span>
<span class="sd">        &quot;connect&quot; or &quot;disconnect&quot; argument&quot;), it first creates an hash</span>
<span class="sd">        of the argument list (excluding `self`). This hash is used to query</span>
<span class="sd">        the database to check if a value for that combination of sequence</span>
<span class="sd">        strings and sequence length has already been calculated. If yes, </span>
<span class="sd">        the result is immediately  returned without executing the decorated</span>
<span class="sd">        method. If no, the method is executed to perform calculations. The</span>
<span class="sd">        result of this calculation is then stored in the database and the </span>
<span class="sd">        result is returned.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">c_args</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>

        <span class="c1"># Initialize database connection.</span>
        <span class="k">if</span> <span class="n">c_args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;connect&quot;</span><span class="p">:</span>

            <span class="c1"># Get path to database based on the AlignmentList db</span>
            <span class="n">tmp_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sql_path</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">,</span> <span class="s2">&quot;pw.db&quot;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">c</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;PRAGMA synchronous = OFF&quot;</span><span class="p">)</span>

            <span class="c1"># Create table, if it does not yet exist</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&quot;</span>
                                  <span class="s2">&quot;&#39;table&#39; AND name=&#39;pw_table&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fetchall</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE pw_table (hash integer&quot;</span>
                               <span class="s2">&quot;, seq_len integer&quot;</span>
                               <span class="s2">&quot;, val float, &quot;</span>
                               <span class="s2">&quot;effective_len float,&quot;</span>
                               <span class="s2">&quot;PRIMARY KEY (hash, seq_len))&quot;</span><span class="p">)</span>

            <span class="k">return</span>

        <span class="k">elif</span> <span class="n">c_args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;disconnect&quot;</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">c</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">return</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="p">:</span>

            <span class="n">h</span> <span class="o">=</span> <span class="nb">hash</span><span class="p">(</span><span class="n">c_args</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;SELECT * FROM pw_table WHERE hash = ?&quot;</span>
                           <span class="s2">&quot; AND seq_len = ?&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">vals</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">vals</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">vals</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span>
                <span class="n">h</span> <span class="o">=</span> <span class="nb">hash</span><span class="p">(</span><span class="n">c_args</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO pw_table VALUES (?, ?, ?, ?)&quot;</span><span class="p">,</span>
                               <span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="n">args</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
                <span class="k">return</span> <span class="n">value</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Return the function&#39;s docstring. &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">func</span><span class="o">.</span><span class="vm">__doc__</span>

    <span class="k">def</span> <span class="nf">__get__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">obj</span><span class="p">,</span> <span class="n">objtype</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Support instance methods &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="fm">__call__</span><span class="p">,</span> <span class="n">obj</span><span class="p">)</span></div>


<div class="viewcode-block" id="check_data"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.check_data">[docs]</a><span class="k">def</span> <span class="nf">check_data</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Decorator handling the result from AlignmentList plotting methods.</span>
<span class="sd">    </span>
<span class="sd">    This should decorate all plotting methods from the `AlignmentList` class.</span>
<span class="sd">    It can be used to control warnings and handle exceptions when calling</span>
<span class="sd">    the plotting methods. Currently, it ensures that a list of methods are</span>
<span class="sd">    not executed when the `AlignmentList` instance contains a single </span>
<span class="sd">    alignment, and handles cases where the plotting methods return an</span>
<span class="sd">    empty array of data. Further control can be added here to prevent</span>
<span class="sd">    methods from being executed in certain conditions and handling </span>
<span class="sd">    certain outputs of the plotting methods.</span>
<span class="sd">    </span>
<span class="sd">    The only requirement is that, even when the plotting methods are not</span>
<span class="sd">    executed, this should always return a dictionary. In such cases,</span>
<span class="sd">    this dictionary should contain a single key:value with</span>
<span class="sd">    {&quot;exception&quot;:&lt;exception_type_string&gt;}. The &lt;exception_type_string&gt; should</span>
<span class="sd">    be a simple string with an informative name that will be handled in</span>
<span class="sd">    the `stats_write_plot` of the `TriFusionApp` class.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    res : dict</span>
<span class="sd">        The value returned by `func` is always a dictionary. If the</span>
<span class="sd">        decorator prevents the execution of `func` for some reason, ensure</span>
<span class="sd">        that a dictionary is always return with a single key:value</span>
<span class="sd">        {&quot;exception&quot;: &lt;exception_type_string&gt;}</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># functools.wraps(func) allows the correct generation of docstrings</span>
    <span class="c1"># by sphinx of the decorated objects.</span>
    <span class="nd">@functools</span><span class="o">.</span><span class="n">wraps</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">wrapper</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="c1"># Plot methods that should not be allowed to continue with only one</span>
        <span class="c1">#  active alignment</span>
        <span class="n">no_single_plot</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;outlier_missing_data&quot;</span><span class="p">,</span> <span class="s2">&quot;outlier_missing_data_sp&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;outlier_segregating&quot;</span><span class="p">,</span> <span class="s2">&quot;outlier_segregating_sp&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;outlier_sequence_size&quot;</span><span class="p">,</span> <span class="s2">&quot;outlier_sequence_size_sp&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;average_seqsize_per_species&quot;</span><span class="p">,</span> <span class="s2">&quot;average_seqsize&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;sequence_similarity&quot;</span><span class="p">,</span> <span class="s2">&quot;sequence_segregation&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;length_polymorphism_correlation&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;taxa_distribution&quot;</span><span class="p">,</span> <span class="s2">&quot;cumulative_missing_genes&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;gene_occupancy&quot;</span><span class="p">,</span> <span class="s2">&quot;missing_data_distribution&quot;</span><span class="p">,</span>
                          <span class="s2">&quot;missing_genes_average&quot;</span><span class="p">]</span>

        <span class="c1"># Calling outlier method with a single alignments should immediately</span>
        <span class="c1"># raise an exception</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">func</span><span class="o">.</span><span class="vm">__name__</span> <span class="ow">in</span> <span class="n">no_single_plot</span><span class="p">:</span>
                <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;exception&quot;</span><span class="p">:</span> <span class="s2">&quot;single_alignment&quot;</span><span class="p">}</span>

        <span class="n">res</span> <span class="o">=</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;data&quot;</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;data&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="k">return</span> <span class="n">res</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;exception&quot;</span><span class="p">:</span> <span class="s2">&quot;empty_data&quot;</span><span class="p">}</span>
        <span class="k">elif</span> <span class="s2">&quot;exception&quot;</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">res</span>

    <span class="k">return</span> <span class="n">wrapper</span></div>


<div class="viewcode-block" id="setup_database"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.setup_database">[docs]</a><span class="k">def</span> <span class="nf">setup_database</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Decorator handling the active database tables.</span>
<span class="sd">    </span>
<span class="sd">    Decorates methods from the `Alignment` object that use and</span>
<span class="sd">    perform modifications to the original alignment data. All methods </span>
<span class="sd">    decorated with this must have the keyword arguments `table_in` and</span>
<span class="sd">    `table_out` (and `use_main_table`, optionally). The values associated with </span>
<span class="sd">    these arguments will determine which tables will be used and created </span>
<span class="sd">    before the execution of the decorated method (See Notes for the </span>
<span class="sd">    rationale). The strings provided as arguments for `table_in` and </span>
<span class="sd">    `table_out` will serve as a suffix to the `Alignment` instance attribute</span>
<span class="sd">    `table_name`.</span>
<span class="sd">    </span>
<span class="sd">    If `use_main_table` is provided, and is set to True, both `table_in`</span>
<span class="sd">    and `table_out` will default to `Alignment.table_name`. Values provided</span>
<span class="sd">    for `table_in` and `table_out` at calling time will be ignored.</span>
<span class="sd">    </span>
<span class="sd">    The following cases assume `use_main_table` is not provided or set to </span>
<span class="sd">    False.</span>
<span class="sd">    </span>
<span class="sd">    If both `table_in` and `table_out` are provided, `table_out` is modified</span>
<span class="sd">    so that `table_out = Alignment.table_name + table_out`. If `table_out`</span>
<span class="sd">    is not provided, it defaults to `Alignment.table_name`.</span>
<span class="sd">    </span>
<span class="sd">    If the final `table_out` does not exist in the database, create it. </span>
<span class="sd">    In this case, if `table_in` will default to `Alignment.table_name`. If </span>
<span class="sd">    `table_out` already exists in the database, then `table_in=table_out`.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>
<span class="sd">    </span>
<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    In order to fully understand the mechanism behind the setup of the</span>
<span class="sd">    database tables, one must first know that when an `Alignment` instance</span>
<span class="sd">    is created, it generates a table in the database containing the</span>
<span class="sd">    original data from the alignment.</span>
<span class="sd">    In TriFusion (GUI), this original table MUST NOT be modified, since users</span>
<span class="sd">    may want to execute several methods on the same `Alignment` object. </span>
<span class="sd">    Therefore, when a particular method needs to modify the original alignment,</span>
<span class="sd">    a new temporary table is created to store the modified version</span>
<span class="sd">    until the end of the execution. If a second modification is requested,</span>
<span class="sd">    it will also be necessary to set the input table as the output table</span>
<span class="sd">    of the previous alignment modification.</span>
<span class="sd">    In the case of TriSeq (CLI), there is no such requirement, </span>
<span class="sd">    so it&#39;s much simpler to use the same table  for all modifications. </span>
<span class="sd">    </span>
<span class="sd">    This decorator greatly simplifies this process in the same way for </span>
<span class="sd">    all methods of the `Alignment` object that modify the original alignment</span>
<span class="sd">    data. To accomplish this, all decorated method must have the keyword</span>
<span class="sd">    arguments: `table_in` and `table_out` (and `use_main_table`, optionally).</span>
<span class="sd">    </span>
<span class="sd">    For methods called within the execution of TriFusion, the idea is simple.</span>
<span class="sd">    Since we don&#39;t know which methods will be used by the user, all chained</span>
<span class="sd">    methods that will create the same output file can be called with </span>
<span class="sd">    `table_in=new_table` and `table_out=new_table`. </span>
<span class="sd">    Note that bot arguments have the same value. Whatever is the first method</span>
<span class="sd">    being called, it will face the fact that &quot;new_table&quot; does not yet exist.</span>
<span class="sd">    In this case, the decorator will create a &quot;new_table&quot; in the database and </span>
<span class="sd">    reset `table_in` to the original `Alignment.table_name`. This ensures that</span>
<span class="sd">    the first method will still be able to fetch the alignment data. In the</span>
<span class="sd">    following methods, `table_in` and `table_out` will be used based on the</span>
<span class="sd">    original values, ensuring that the alignment data is being fetched</span>
<span class="sd">    from the last modification made and that the modification chain is </span>
<span class="sd">    maintained. In this way, the execution of `Alignment` methods can</span>
<span class="sd">    be the same, regardless of the order or number of operations requested</span>
<span class="sd">    by the user.</span>
<span class="sd">    </span>
<span class="sd">    In the case of TriSeq, the methods should set `use_main_table=True`.</span>
<span class="sd">    This will always set `table_in=table_out=Alignment.table_name`, which</span>
<span class="sd">    means that the main database will be used for all methods.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nd">@functools</span><span class="o">.</span><span class="n">wraps</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">wrapper</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="c1"># If use_main_table is set, the table_in will be overwritten by</span>
        <span class="c1"># table_out</span>
        <span class="k">if</span> <span class="s2">&quot;use_main_table&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;use_main_table&quot;</span><span class="p">]:</span>
                <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_out&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>
                <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_in&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>
                <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1"># Get sqlite database cursor and main table name</span>
        <span class="n">sql_cur</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cur</span>

        <span class="c1"># Define sqlite table name based on the main Alignment table name</span>
        <span class="c1"># and the provided table_name argument</span>
        <span class="k">if</span> <span class="s2">&quot;table_out&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">table_out</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_out&quot;</span><span class="p">]</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_out&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">table_out</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">table_out</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_out&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>

        <span class="c1"># Add table name to active table names</span>
        <span class="k">if</span> <span class="n">table_out</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>
            <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">table_out</span><span class="p">)</span>

        <span class="c1"># Check if the table already exists, if not create it.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">sql_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; AND&quot;</span>
                <span class="s2">&quot; name=&#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span><span class="o">.</span><span class="n">fetchall</span><span class="p">():</span>

            <span class="n">sql_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE [</span><span class="si">{}</span><span class="s2">](&quot;</span>
                             <span class="s2">&quot;txId INT,&quot;</span>
                             <span class="s2">&quot;taxon TEXT,&quot;</span>
                             <span class="s2">&quot;seq TEXT)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>
            <span class="c1"># Set the table_in argument (table where the sequences will be</span>
            <span class="c1"># fecthed) to the main Alignment table name</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_in&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Since the table_out already exists, sequences will be fetched</span>
            <span class="c1"># from this table instead of the main table name</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_in&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">table_out</span>

        <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">wrapper</span></div>


<div class="viewcode-block" id="setup_intable"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.setup_intable">[docs]</a><span class="k">def</span> <span class="nf">setup_intable</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Decorator handling active database table when fetching data.</span>
<span class="sd">    </span>
<span class="sd">    This class is mean to decorate methods of the `Alignment` class that</span>
<span class="sd">    retrieves alignment data from the database. The requirement is that</span>
<span class="sd">    these methods have the positional arguments `table_suffix`, </span>
<span class="sd">    `table_name` and `table`. These are all optional, and only `table_suffix`</span>
<span class="sd">    and `table_name` should be used when calling these methods. The values</span>
<span class="sd">    of these two will be used to define the value of `table`, so any</span>
<span class="sd">    value provided to this argument when calling the decorated method will</span>
<span class="sd">    be ignored. In the end, the `table` variable will contain the final</span>
<span class="sd">    table name and only this variable will be used to retrieve the alignment</span>
<span class="sd">    data.</span>
<span class="sd">    </span>
<span class="sd">    `table_suffix` is a suffix that is appended to `Alignment.table_name`.</span>
<span class="sd">    `table_name` defines a complete table name.</span>
<span class="sd">    </span>
<span class="sd">    If neither `table_suffix` nor `table_name` are provided, then `table`</span>
<span class="sd">    will default to `Alignment.table_name`.</span>
<span class="sd">    </span>
<span class="sd">    If `table_name` is provided, `table=table_name`. If `table_suffix`</span>
<span class="sd">    is provided, `table=Alignment.table_name + table_suffix`.</span>
<span class="sd">    </span>
<span class="sd">    If both are provided, `table_name` takes precedence over `table_suffix`</span>
<span class="sd">    and `table=table_name`.</span>
<span class="sd">    </span>
<span class="sd">    In any case, we test the existence of the final `table` value in the</span>
<span class="sd">    database. If it does not exist, `table` will revert to </span>
<span class="sd">    `Alignment.table_name` to prevent errors.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>
<span class="sd">    </span>
<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    func : function</span>
<span class="sd">        Decorated function</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># functools.wraps(func) allows the correct generation of docstrings</span>
    <span class="c1"># by sphinx of the decorated objects.</span>
    <span class="nd">@functools</span><span class="o">.</span><span class="n">wraps</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">wrapper</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="c1"># Get sqlite database cursor and main table name</span>
        <span class="n">sql_cur</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cur</span>

        <span class="k">if</span> <span class="s2">&quot;table_name&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s2">&quot;table_suffix&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>
            <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">tb_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_name&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">tb_name</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">tb_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">tb_suffix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">if</span> <span class="n">tb_name</span><span class="p">:</span>
            <span class="n">table</span> <span class="o">=</span> <span class="n">tb_name</span>
        <span class="k">elif</span> <span class="n">tb_suffix</span><span class="p">:</span>
            <span class="n">table</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span> <span class="o">+</span> <span class="n">tb_suffix</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">table</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>

        <span class="c1"># Check if the specified table exists. If not, revert to</span>
        <span class="c1"># self.table_name. When executing Process operations, the first valid</span>
        <span class="c1"># operation may be one that only has a &quot;table_in&quot; argument but several</span>
        <span class="c1"># other operations may be defined before with a different table suffix.</span>
        <span class="c1"># In such case, the table passed to &quot;table_in&quot; was not yet created and</span>
        <span class="c1"># populated in those prior methods. Therefore, here we account for</span>
        <span class="c1"># those cases by reverting the table to self.table_name if table_in</span>
        <span class="c1"># was not yet created.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">sql_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; AND&quot;</span>
                <span class="s2">&quot; name=&#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">))</span><span class="o">.</span><span class="n">fetchall</span><span class="p">():</span>
            <span class="n">table</span> <span class="o">=</span> <span class="n">args</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">table_name</span>

        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">table</span>

        <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">wrapper</span></div>


<div class="viewcode-block" id="AlignmentException"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentException">[docs]</a><span class="k">class</span> <span class="nc">AlignmentException</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Generic Alignment object exception. &quot;&quot;&quot;</span>
    <span class="k">pass</span></div>


<div class="viewcode-block" id="AlignmentUnequalLength"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentUnequalLength">[docs]</a><span class="k">class</span> <span class="nc">AlignmentUnequalLength</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Raised when sequences in alignment have unequal length. &quot;&quot;&quot;</span>
    <span class="k">pass</span></div>


<div class="viewcode-block" id="Alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment">[docs]</a><span class="k">class</span> <span class="nc">Alignment</span><span class="p">(</span><span class="n">Base</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Main interface for single alignment files.</span>
<span class="sd">    </span>
<span class="sd">    The `Alignment` class is the main interface for single alignment files,</span>
<span class="sd">    providing methods that parse, query, retrieve and modify alignment data.</span>
<span class="sd">    It is usually created within the context of `AlignmentList`,</span>
<span class="sd">    which handles the connection with the sqlite database. Nevertheless,</span>
<span class="sd">    `Alignment` instances can be created by providing the `input_alignment`</span>
<span class="sd">    argument, which can be one of the following:</span>

<span class="sd">    - path to alignment file. In this case, the file is parsed and</span>
<span class="sd">      information stored in a new sqlite table.</span>

<span class="sd">    - sqlite table name. In this case, the table name must be present</span>
<span class="sd">      in the database.</span>

<span class="sd">    In either case, the sqlite `Connection` and `Cursor` objects should</span>
<span class="sd">    be provided.</span>
<span class="sd">    </span>
<span class="sd">    When an `Alignment` object is instantiated, it first generates the </span>
<span class="sd">    `table_name` based on the `input_alignment` string, filtering all</span>
<span class="sd">    characters that are not alpha numeric. Then, it queries the database</span>
<span class="sd">    to check is a table already exists with that name. If yes, it is assumed</span>
<span class="sd">    that the alignment data is stored in the provided table name. If there</span>
<span class="sd">    is no table with that name, it is assumed that `input_alignment` is a </span>
<span class="sd">    path to the alignment file and the regular parsing ensues. An empty</span>
<span class="sd">    table is created, the sequence type, format and missing data symbol are</span>
<span class="sd">    automatically detected and the alignment is parsed according to the</span>
<span class="sd">    detected format.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input_alignment : str</span>
<span class="sd">        Can be either the path to an alignment file or the name of a table</span>
<span class="sd">        in the sqlite database.</span>
<span class="sd">    sql_cursor : sqlite3.Cursor</span>
<span class="sd">        Cursor object of the sqlite database.</span>
<span class="sd">    sql_con : sqlite3.Connection</span>
<span class="sd">        Connection object of the sqlite database.</span>
<span class="sd">    input_format : str, optional</span>
<span class="sd">        File format of `input_alignment`. If `input_alignment` is a</span>
<span class="sd">        file path, the format will be automatically detect from the file.</span>
<span class="sd">        The value provided with this argument overrides the automatic </span>
<span class="sd">        detection&#39;s result.</span>
<span class="sd">    partitions : `trifusion.process.data.Partitions`, optional</span>
<span class="sd">        If provided, it will set the `partitions` attribute. This should</span>
<span class="sd">        be used only when `input_alignment` is a database table name.</span>
<span class="sd">    locus_length : int, optional</span>
<span class="sd">        Sets the length of the current alignment, stored in the `locus_length`</span>
<span class="sd">        attribute. This option should only be used when `input_alignment`</span>
<span class="sd">        is a database table name. Otherwise, it is automatically set during</span>
<span class="sd">        alignment parsing.</span>
<span class="sd">    sequence_code : tuple, optional</span>
<span class="sd">        Sets the `sequence_code` attribute with the information on</span>
<span class="sd">        (&lt;sequence_type&gt;, &lt;missing data symbol&gt;). This option should only be</span>
<span class="sd">         used when `input_alignment` is a database table name. Otherwise, </span>
<span class="sd">         it is automatically set during alignment parsing.</span>
<span class="sd">    taxa_list : list, optional</span>
<span class="sd">        Sets the list attribute `taxa_list` with the names of the taxa</span>
<span class="sd">        present in the alignment. This option should only be</span>
<span class="sd">        used when `input_alignment` is a database table name. Otherwise,</span>
<span class="sd">        it is automatically set during alignment parsing.</span>
<span class="sd">    taxa_idx : dict, optional</span>
<span class="sd">        Sets the dictionary attribute `taxa_idx` that maps the taxon names</span>
<span class="sd">        to their index in the sqlite database table. This option should only</span>
<span class="sd">        be used when `input_alignment` is a database table name. Otherwise,</span>
<span class="sd">        it is automatically set during alignment parsing.</span>
<span class="sd">    </span>
<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    cur : sqlite3.Cursor</span>
<span class="sd">        Cursor object of the sqlite database.</span>
<span class="sd">    con : sqlite3.Connection</span>
<span class="sd">        Connection object of the sqlite database.</span>
<span class="sd">    table_name : str </span>
<span class="sd">        Name of the sqlite database&#39;s table storing the sequence</span>
<span class="sd">        data.</span>
<span class="sd">    tables : list</span>
<span class="sd">        Lists the sqlite database&#39;s tables created for the `Alignment`</span>
<span class="sd">        instance. The &quot;master&quot; table is always present and represents the</span>
<span class="sd">        original alignment.</span>
<span class="sd">    partitions : `trifusion.process.data.Partitions`</span>
<span class="sd">        Stores the partitions and substitution model definition for the</span>
<span class="sd">        alignment.</span>
<span class="sd">    locus_length : int</span>
<span class="sd">        Length of the alignment in base pairs or residues.</span>
<span class="sd">    restriction_range : str</span>
<span class="sd">        Only used when gaps are coded in a binary matrix. Stores a string</span>
<span class="sd">        with the range of the restriction-type data that will encode</span>
<span class="sd">        gaps and will only be used when nexus is in the output format.</span>
<span class="sd">    e : None or Exception</span>
<span class="sd">        Stores any exceptions that occur during the parsing of the </span>
<span class="sd">        alignment file. It remains None unless something wrong happens.</span>
<span class="sd">    taxa_list : list</span>
<span class="sd">        List with the active taxon names.</span>
<span class="sd">    taxa_idx : dict</span>
<span class="sd">        Maps the taxon names to their corresponding index in the sqlite </span>
<span class="sd">        database. The index is not retrieved from the position of the taxon</span>
<span class="sd">        in `taxa_list` to prevent messing up when taxa are removed from the</span>
<span class="sd">        `Alignment` object.</span>
<span class="sd">    shelved_taxa : list</span>
<span class="sd">        List of ignored taxon names. </span>
<span class="sd">    path : str</span>
<span class="sd">        Full path to alignment file.</span>
<span class="sd">    sname : str</span>
<span class="sd">        Basename of the alignment file without the extension.</span>
<span class="sd">    name : str</span>
<span class="sd">        Basename of the alignment file with the extension.</span>
<span class="sd">    sequence_code : tuple</span>
<span class="sd">        Contains information on (&lt;sequence type&gt;, &lt;missing data symbol&gt;),</span>
<span class="sd">        e.g. (&quot;Protein&quot;, &quot;x&quot;).</span>
<span class="sd">    interleave_data : bool</span>
<span class="sd">        Attribute that is set to True when interleave data has been</span>
<span class="sd">        created for the alignment data.</span>
<span class="sd">    </span>
<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    The `Alignment` class is not meant to be used directly (although it is</span>
<span class="sd">    quite possible to do so if you handle the database connections before).</span>
<span class="sd">    Instead, use the `AlignmentList` class, even if there is only one</span>
<span class="sd">    alignment. All `Alignment` methods are available from the `AlignmentList`</span>
<span class="sd">    and it is also possible to retrieve specific `Alignment` objects.</span>

<span class="sd">    The `Alignment` class was designed to be a lightweight, fast and</span>
<span class="sd">    powerful interface between alignment data and a set of manipulation</span>
<span class="sd">    and transformation methods.</span>
<span class="sd">    For performance and efficiency purposes, all alignment data is stored </span>
<span class="sd">    in a sqlite database that prevents the entire alignment from being</span>
<span class="sd">    loaded into memory. To facilitate the retrieval and iteration over the</span>
<span class="sd">    alignment data, several methods (`iter_columns`, &#39;iter_sequences`, etc)</span>
<span class="sd">    are available to handle the interface with the database and retrieve</span>
<span class="sd">    only the necessary information.</span>

<span class="sd">    See Also</span>
<span class="sd">    --------</span>
<span class="sd">    AlignmentList</span>
<span class="sd">    AlignmentList.add_alignment_files</span>
<span class="sd">    AlignmentList.retrieve_alignment</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_alignment</span><span class="p">,</span> <span class="n">input_format</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">partitions</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">locus_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sequence_code</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">taxa_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">taxa_idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sql_cursor</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sql_con</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="n">sql_cursor</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">sql_con</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">partitions</span><span class="p">,</span> <span class="n">Partitions</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">partitions</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">Partitions</span><span class="p">()</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Initializing a Partitions instance for the current alignment. By</span>
<span class="sd">            default, the Partitions instance will assume that the whole Alignment</span>
<span class="sd">            object is one single partition. However, if the current alignment is the</span>
<span class="sd">            result of a concatenation, or if a partitions file is provided, the</span>
<span class="sd">            partitions argument can be used. Partitions can be later changed using</span>
<span class="sd">            the set_partitions method. Substitution models objects are associated</span>
<span class="sd">            with each partition and they default to None</span>
<span class="sd">            &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">locus_length</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="n">locus_length</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            The length of the alignment object. Even if the current alignment</span>
<span class="sd">            object is partitioned, this will return the length of the entire</span>
<span class="sd">            alignment</span>
<span class="sd">            &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This option is only relevant when gaps are coded. This will store a</span>
<span class="sd">        string with the range of the restriction-type data that will encode</span>
<span class="sd">        gaps and will only be used when nexus is in the output format</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The `e` attribute will store any exceptions that occur during the</span>
<span class="sd">        parsing of the alignment object. It remains None unless something</span>
<span class="sd">        wrong happens.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute that will store the taxa names.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute that will store shelved taxa. When retrieving taxa from</span>
<span class="sd">        the database, this list will be checked and present taxa will</span>
<span class="sd">        be ignored</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute that will store the index of the taxa in the sql database.</span>
<span class="sd">        This index is stored in a dictionary instead of being retrieved</span>
<span class="sd">        from the index position of the taxa_list list because taxa may</span>
<span class="sd">        be removed (which would mess up the link between the two indexes)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="o">=</span> <span class="n">input_alignment</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute with full path to alignment file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sname</span> <span class="o">=</span> <span class="n">basename</span><span class="p">(</span><span class="n">splitext</span><span class="p">(</span><span class="n">input_alignment</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute with basename of alignment file without extension</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">basename</span><span class="p">(</span><span class="n">input_alignment</span><span class="p">)</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute with basename of alignment file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="n">sequence_code</span>
        <span class="k">if</span> <span class="n">taxa_list</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="n">taxa_list</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">taxa_idx</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span> <span class="o">=</span> <span class="n">taxa_idx</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">path</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">isalnum</span><span class="p">()])</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a database table for the current alignment. This will be the</span>
<span class="sd">        link attribute between the sqlite database and the remaining methods</span>
<span class="sd">        that require information on the sequence data. This also removes</span>
<span class="sd">        any non alpha numeric characters the table name might have and ensures</span>
<span class="sd">        that it starts with a aphabetic character to avoid a database error</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tables</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Lists the currently active tables for the Alignment object. The</span>
<span class="sd">        &#39;master&#39; table is always present and represents the original</span>
<span class="sd">        alignment. Additional tables may be added as needed, and then dropped</span>
<span class="sd">        when no longer necessary. When all tables, except the master, need to</span>
<span class="sd">        be removed, this attribute can be used to quickly drop all derived</span>
<span class="sd">        tables</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">interleave_data</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute that is set to True when interleave data has been </span>
<span class="sd">        created for the alignment data. Buiding the interleave matrix is a bit</span>
<span class="sd">        costly, so when it is requested by the user it is built once and</span>
<span class="sd">        stored in the database, and then further usages will use that table.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        NOTE ON POSSIBLE DUPLICATE TABLE NAMES: It is not the responsibility</span>
<span class="sd">        of the Alignment class to check on duplicate table names. If an</span>
<span class="sd">        exiting table name is found in the database, the Alignment class</span>
<span class="sd">        assumes that the alignment has already been parsed and inserted into</span>
<span class="sd">        the database. Therefore, it will not parse the alignment file again.</span>
<span class="sd">        This usually happens when creating concatenated or reverse concatenated</span>
<span class="sd">        alignments, where the sequence data is inserted into the database</span>
<span class="sd">        before instantiating the Alignment class. IT IS THE RESPONSABILITY</span>
<span class="sd">        OF THE AlignmentList CLASS AND OTHER WRAPPERS TO ASSESS THE VALIDITY</span>
<span class="sd">        OR NOT OF POTENTIAL DUPLICATE ALIGNMENT FILES.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check if table for current alignment file already exists. If it</span>
        <span class="c1"># does not exist, it is assumed that input_alignment is the path to</span>
        <span class="c1"># the alignment file</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">):</span>

            <span class="c1"># Get alignment format and code. Sequence code is a tuple of</span>
            <span class="c1"># (DNA, N) or (Protein, X)</span>
            <span class="n">finder_content</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">autofinder</span><span class="p">(</span><span class="n">input_alignment</span><span class="p">)</span>
            <span class="c1"># Handles the case where the input format is invalid and</span>
            <span class="c1"># finder_content is an Exception</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">finder_content</span><span class="p">,</span> <span class="ne">Exception</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">input_format</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">autofinder</span><span class="p">(</span>
                    <span class="n">input_alignment</span><span class="p">)</span>

                <span class="c1"># Create database table when there are no issues in the</span>
                <span class="c1"># recognition of the file format and sequence code</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">)</span>

                <span class="c1"># In case the input format is specified, overwrite the</span>
                <span class="c1"># attribute</span>
                <span class="k">if</span> <span class="n">input_format</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">input_format</span> <span class="o">=</span> <span class="n">input_format</span>

                <span class="c1"># parsing the alignment</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">read_alignment</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Setting the sequence code attribute for seq type checking</span>
                <span class="c1"># in AlignmentList</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">finder_content</span>

        <span class="c1"># In case there is a table for the provided input_alignment</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">input_format</span> <span class="o">=</span> <span class="n">input_format</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterator behavior for `Alignment`.</span>
<span class="sd">        </span>
<span class="sd">        Iterates over taxa and sequence data from the alignment. Taxon names</span>
<span class="sd">        in the `shelved_taxa` attribute are ignored. This will always</span>
<span class="sd">        retrieve the alignment data from the master database table, </span>
<span class="sd">        `table_name`.</span>
<span class="sd">        </span>
<span class="sd">        Yields</span>
<span class="sd">        ------</span>
<span class="sd">        tx : str</span>
<span class="sd">            taxon name.</span>
<span class="sd">        seq : str</span>
<span class="sd">            sequence string.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT taxon,seq from [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span>

<div class="viewcode-block" id="Alignment._create_table"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._create_table">[docs]</a>    <span class="k">def</span> <span class="nf">_create_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_name</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates a new table in the database.</span>
<span class="sd">        </span>
<span class="sd">        Convenience method that creates a new table in the sqlite database.</span>
<span class="sd">        It accepts a custom Cursor object, which overrides the `cur`</span>
<span class="sd">        attribute.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_name : str</span>
<span class="sd">            Name of the table to be created.</span>
<span class="sd">        cur : sqlite3.Cursor, optional</span>
<span class="sd">            Custom Cursor object used to query the database.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">cur</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span>

        <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE [</span><span class="si">{}</span><span class="s2">](&quot;</span>
                    <span class="s2">&quot;txId INT,&quot;</span>
                    <span class="s2">&quot;taxon TEXT,&quot;</span>
                    <span class="s2">&quot;seq TEXT)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span></div>

<div class="viewcode-block" id="Alignment._table_exists"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._table_exists">[docs]</a>    <span class="k">def</span> <span class="nf">_table_exists</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_name</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Checks if a table exists in the database.</span>
<span class="sd">        </span>
<span class="sd">        Convenience method that checks if a table exists in the database.</span>
<span class="sd">        It accepts a custom Cursor object, which overrides the `cur`</span>
<span class="sd">        attribute.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_name : str</span>
<span class="sd">            Name of the table.</span>
<span class="sd">        cur: sqlite3.Cursor, optional</span>
<span class="sd">            Custom Cursor object used to query the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        res : list</span>
<span class="sd">            List with the results of a query for &#39;table&#39; type with </span>
<span class="sd">            `table_name` name. Is empty when the table does not exist.</span>
<span class="sd">        </span>
<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This returns a list that will contain one item if the table exists</span>
<span class="sd">        in the database. If it doesn&#39;t exist, returns an empty list.</span>
<span class="sd">        Therefore, this can be used simply like::</span>
<span class="sd">        </span>
<span class="sd">            if self._table_exists(&quot;my_table&quot;):</span>
<span class="sd">                 # Do stuff`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">cur</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span>

        <span class="n">res</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
            <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; AND&quot;</span>
            <span class="s2">&quot; name=&#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">res</span></div>

<div class="viewcode-block" id="Alignment._set_format"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._set_format">[docs]</a>    <span class="k">def</span> <span class="nf">_set_format</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_format</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Manually set the input format of the `Alignment` object.</span>
<span class="sd">        </span>
<span class="sd">        Manually sets the input format associated with the `Alignment` object</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_format : str</span>
<span class="sd">            The input format.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">input_format</span> <span class="o">=</span> <span class="n">input_format</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Alignment._set_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._set_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">msg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Setup of progress indicators for both GUI and CLI task executions.</span>
<span class="sd">         </span>
<span class="sd">         This handles the setup of the objects responsible for updating</span>
<span class="sd">         the progress of task&#39;s execution of both TriFusion (GUI) and</span>
<span class="sd">         TriSeq (CLI). At the beginning of any given task, these objects</span>
<span class="sd">         can be initialized by providing either the Namespace object (`ns`)</span>
<span class="sd">         in the case of TriFusion, or the ProgressBar object (`pbar`), in</span>
<span class="sd">         the case of TriSeq. Along with one of these objects, the expected</span>
<span class="sd">         `total` of the progress should also be provided. The `ns` and</span>
<span class="sd">         `pbar` objects are updated at each iteration of a given task,</span>
<span class="sd">         and the `total` is used to get a measure of the progress.</span>

<span class="sd">         Optionally, a message can be also provided for the Namespace object</span>
<span class="sd">         that will be used by TriFusion.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        total : int</span>
<span class="sd">            Expected total of the task&#39;s progress.</span>
<span class="sd">        msg : str, optional</span>
<span class="sd">            A secondary message that appears in TriFusion&#39;s progress dialogs.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The progress of any given task can be provided by either an</span>
<span class="sd">        `Alignment` or `AlignmentList` instance. Generally, the tasks follow</span>
<span class="sd">        the progress of the `AlignmentList` instance, unless that instance</span>
<span class="sd">        contains only one `Alignment` object. In that case, the progress</span>
<span class="sd">        information is piped from the `Alignment` instance. For that reason,</span>
<span class="sd">        and for the Namespace (`ns`) object only, this method first checks</span>
<span class="sd">        if the `ns.sa` attribute exists. If it does, it means that the parent</span>
<span class="sd">        `AlignmentList` instance contains only one `Alignment` object and,</span>
<span class="sd">        therefore, the progress information shoud come from here.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        Start a progress counter  for a task that will make 100 iterations::</span>

<span class="sd">          self._set_pipes(ns=ns_obj, pbar=pbar_obj, total=100,</span>
<span class="sd">          msg=&quot;Some message&quot;)</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        _reset_pipes</span>
<span class="sd">        _update_pipes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check if the `sa` attribute exists in the Namespace object.</span>
        <span class="c1"># If yes, then the logging will be performed here. If not,</span>
        <span class="c1"># The progress attributes will not be set for `ns`.</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="c1"># Listen for kill switch</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="c1"># Set attributes only if parent AlignmentList contains a single</span>
            <span class="c1"># Alignment object</span>
            <span class="k">if</span> <span class="n">sa</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="n">total</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">msg</span>

        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">max_value</span> <span class="o">=</span> <span class="n">total</span>
            <span class="c1"># Reset ProgressBar object counter</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Alignment._update_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._update_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">msg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Update progress indicators for both GUI and CLI task executions.</span>

<span class="sd">        This method provides a single interface for updating the progress</span>
<span class="sd">        objects `ns` or `pbar`, which should have been initialized at the</span>
<span class="sd">        beginning of the task with the `_set_pipes` method.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        value : int</span>
<span class="sd">            Value of the current progress index</span>
<span class="sd">        msg : str, optional</span>
<span class="sd">            A secondary message that appears in TriFusion&#39;s progress dialogs.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The progress of any given task can be provided by either an</span>
<span class="sd">        `Alignment` or `AlignmentList` instance. Generally, the tasks follow</span>
<span class="sd">        the progress of the `AlignmentList` instance, unless that instance</span>
<span class="sd">        contains only one `Alignment` object. In that case, the progress</span>
<span class="sd">        information is piped from the `Alignment` instance. For that reason,</span>
<span class="sd">        and for the Namespace (`ns`) object only, this method first checks</span>
<span class="sd">        if the `ns.sa` attribute exists. If it does, it means that the parent</span>
<span class="sd">        `AlignmentList` instance contains only one `Alignment` object and,</span>
<span class="sd">        therefore, the progress information shoud come from here.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        Update the counter in an iteration of 100::</span>

<span class="sd">            for i in range(100):</span>
<span class="sd">                self._update_pipes(ns=ns_obj, pbar=pbar_obj, value=i,</span>
<span class="sd">                                   msg=&quot;Some string&quot;)</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        _set_pipes</span>
<span class="sd">        _reset_pipes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sa</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="n">value</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">msg</span>
        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">value</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Alignment._reset_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._reset_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reset progress indicators for both GUI and CLI task executions.</span>

<span class="sd">        This should be done at the end of any task that initialized the</span>
<span class="sd">        progress objects, but it only affects the Namespace object. It</span>
<span class="sd">        resets all Namespace attributes to None.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        _set_pipes</span>
<span class="sd">        _update_pipes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">sa</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sa</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">None</span></div>

    <span class="nd">@setup_intable</span>
<div class="viewcode-block" id="Alignment.iter_columns"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.iter_columns">[docs]</a>    <span class="k">def</span> <span class="nf">iter_columns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Generator over alignment columns.</span>

<span class="sd">        Generator that yields the alignment columns in a list of characters.</span>
<span class="sd">        Sequence data is retrieved from a database table specified either by</span>
<span class="sd">        the `table_name` or `table_suffix` arguments. `table_name` will always</span>
<span class="sd">        take precedence over `table_suffix` if both are provided. If none</span>
<span class="sd">        are provided, the default `Alignment.table_name` is used. If the</span>
<span class="sd">        table name provided by either `table_name` or `table_suffix` is</span>
<span class="sd">        invalid, the default `Alignment.table_name` is also used.</span>

<span class="sd">        The `table` variable will be automatically setup in the `SetupInTable`</span>
<span class="sd">        decorator according to `table_suffix` and `table_name`, and should</span>
<span class="sd">        not be used directly (values provided when calling it are actually</span>
<span class="sd">        ignored). The decorator will set this variable</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is appended to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        table : string</span>
<span class="sd">            This argument will be automatically setup by the `SetupInTable`</span>
<span class="sd">            decorator. Do not use directly.</span>

<span class="sd">        Yields</span>
<span class="sd">        ------</span>
<span class="sd">        i : list</span>
<span class="sd">            List with alignment column characters as elements</span>
<span class="sd">            ([&quot;A&quot;, &quot;A&quot;, &quot;C&quot;, &quot;A&quot;, &quot;A&quot;]).</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupInTable</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        To prevent the query from loading the entire alignment data into</span>
<span class="sd">        memory, two techniques are employed. First, the iteration over the</span>
<span class="sd">        query results is lazy, using `itertools.izip`. Second, a `range`</span>
<span class="sd">        iteration is performed to ensure that only blocks of 100k alignment</span>
<span class="sd">        sites are fetch at any given time. Several tests were performed to</span>
<span class="sd">        ensure that this block size provided a good trade-off between</span>
<span class="sd">        RAM usage and speed.</span>

<span class="sd">        All generators ignore data associated with taxa present in the</span>
<span class="sd">        `shelved_taxa` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Locking mechanism necessary to avoid concurrency issues when</span>
            <span class="c1"># accessing the database. This ensures that only one Cursor</span>
            <span class="c1"># object is querying the database at any given time</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">acquire</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="mi">100000</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">izip</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;SELECT taxon, substr(seq, </span><span class="si">{}</span><span class="s2">, 100000)&quot;</span>
                        <span class="s2">&quot; FROM [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">table</span><span class="p">))</span>
                               <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">)):</span>
                    <span class="k">yield</span> <span class="n">i</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">release</span><span class="p">()</span></div>

    <span class="nd">@setup_intable</span>
<div class="viewcode-block" id="Alignment.iter_columns_uniq"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.iter_columns_uniq">[docs]</a>    <span class="k">def</span> <span class="nf">iter_columns_uniq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                          <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Generator over unique characters of an alignment column.</span>

<span class="sd">        Generator that yields unique elements of alignment columns in a list.</span>
<span class="sd">        Sequence data is retrieved from a database table specified either by</span>
<span class="sd">        the `table_name` or `table_suffix` arguments. `table_name` will always</span>
<span class="sd">        take precedence over `table_suffix` if both are provided. If none</span>
<span class="sd">        are provided, the default `Alignment.table_name` is used. If the</span>
<span class="sd">        table name provided by either `table_name` or `table_suffix` is</span>
<span class="sd">        invalid, the default `Alignment.table_name` is also used.</span>

<span class="sd">        The `table` variable will be automatically setup in the `SetupInTable`</span>
<span class="sd">        decorator according to `table_suffix` and `table_name`, and should</span>
<span class="sd">        not be used directly (values provided when calling it are actually</span>
<span class="sd">        ignored). The decorator will set this variable</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is appended to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        table : string</span>
<span class="sd">            This argument will be automatically setup by the `SetupInTable`</span>
<span class="sd">            decorator. Do not use directly.</span>

<span class="sd">        Yields</span>
<span class="sd">        ------</span>
<span class="sd">        i : list</span>
<span class="sd">            List with alignment column characters as elements ([&quot;A&quot;, &quot;C&quot;]).</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupInTable</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        To prevent the query from loading the entire alignment data into</span>
<span class="sd">        memory, two techniques are employed. First, the iteration over the</span>
<span class="sd">        query results is lazy, using `itertools.izip`. Second, a `range`</span>
<span class="sd">        iteration is performed to ensure that only blocks of 100k alignment</span>
<span class="sd">        sites are fetch at any given time. Several tests were performed to</span>
<span class="sd">        ensure that this block size provided a good trade-off between</span>
<span class="sd">        RAM usage and speed.</span>

<span class="sd">        All generators ignore data associated with taxa present in the</span>
<span class="sd">        `shelved_taxa` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Locking mechanism necessary to avoid concurrency issues when</span>
            <span class="c1"># accessing the database. This ensures that only one Cursor</span>
            <span class="c1"># object is querying the database at any given time</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">acquire</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="mi">100000</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">izip</span><span class="p">(</span><span class="o">*</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;SELECT taxon, substr(seq, </span><span class="si">{}</span><span class="s2">, 100000)&quot;</span>
                        <span class="s2">&quot; FROM [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">table</span><span class="p">))</span>
                               <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">)):</span>
                    <span class="k">yield</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">release</span><span class="p">()</span></div>

    <span class="nd">@setup_intable</span>
<div class="viewcode-block" id="Alignment.iter_sequences"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.iter_sequences">[docs]</a>    <span class="k">def</span> <span class="nf">iter_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Generator over sequence strings in the alignment.</span>

<span class="sd">        Generator for sequence data of the `Alignment` object.</span>
<span class="sd">        Sequence data is retrieved from a database table specified either by</span>
<span class="sd">        the `table_name` or `table_suffix` arguments. `table_name` will always</span>
<span class="sd">        take precedence over `table_suffix` if both are provided. If none</span>
<span class="sd">        are provided, the default `Alignment.table_name` is used. If the</span>
<span class="sd">        table name provided by either `table_name` or `table_suffix` is</span>
<span class="sd">        invalid, the default `Alignment.table_name` is also used.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is appended to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        table : string</span>
<span class="sd">            This argument will be automatically setup by the `SetupInTable`</span>
<span class="sd">            decorator. Do not use directly.</span>

<span class="sd">        Yields</span>
<span class="sd">        ------</span>
<span class="sd">        seq : string</span>
<span class="sd">            Sequence string for a given taxon</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupInTable</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        All generators ignore data associated with taxa present in the</span>
<span class="sd">        `shelved_taxa` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Locking mechanism necessary to avoid concurrency issues when</span>
            <span class="c1"># accessing the database. This ensures that only one Cursor</span>
            <span class="c1"># object is querying the database at any given time</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">acquire</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;SELECT taxon,seq FROM [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">seq</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">release</span><span class="p">()</span></div>

    <span class="nd">@setup_intable</span>
<div class="viewcode-block" id="Alignment.iter_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.iter_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">iter_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Generator for (taxon, sequence) tuples.</span>

<span class="sd">        Generator that yields (taxon, sequence) tuples from the database.</span>
<span class="sd">        Sequence data is retrieved from a database table specified either by</span>
<span class="sd">        the `table_name` or `table_suffix` arguments. `table_name` will always</span>
<span class="sd">        take precedence over `table_suffix` if both are provided. If none</span>
<span class="sd">        are provided, the default `Alignment.table_name` is used. If the</span>
<span class="sd">        table name provided by either `table_name` or `table_suffix` is</span>
<span class="sd">        invalid, the default `Alignment.table_name` is also used.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        table : string</span>
<span class="sd">            This argument will be automatically setup by the `SetupInTable`</span>
<span class="sd">            decorator. Do not use directly.</span>

<span class="sd">        Yields</span>
<span class="sd">        ------</span>
<span class="sd">        tx : string</span>
<span class="sd">            Taxon name.</span>
<span class="sd">        seq : string</span>
<span class="sd">            Sequence string.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupInTable</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        All generators ignore data associated with taxa present in the</span>
<span class="sd">        `shelved_taxa` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Locking mechanism necessary to avoid concurrency issues when</span>
            <span class="c1"># accessing the database. This ensures that only one Cursor</span>
            <span class="c1"># object is querying the database at any given time</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">acquire</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;SELECT taxon, seq from [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">)):</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                    <span class="k">yield</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">release</span><span class="p">()</span></div>

    <span class="nd">@setup_intable</span>
<div class="viewcode-block" id="Alignment.get_sequence"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.get_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">get_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                       <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns the sequence string for a given taxon.</span>

<span class="sd">        Returns the sequence string of the corresponding `taxon`. If the</span>
<span class="sd">        taxon does not exist in the table, raises a KeyError.</span>
<span class="sd">        The sequence is retrieved from a database table specified either by</span>
<span class="sd">        the `table_name` or `table_suffix` arguments. `table_name` will always</span>
<span class="sd">        take precedence over `table_suffix` if both are provided. If none</span>
<span class="sd">        are provided, the default `Alignment.table_name` is used. If the</span>
<span class="sd">        table name provided by either `table_name` or `table_suffix` is</span>
<span class="sd">        invalid, the default `Alignment.table_name` is also used.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxon : str</span>
<span class="sd">            Name of the taxon.</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        table : string</span>
<span class="sd">            This argument will be automatically setup by the `SetupInTable`</span>
<span class="sd">            decorator. Do not use directly.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        seq : str</span>
<span class="sd">            Sequence string.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        KeyError</span>
<span class="sd">            If the taxon does not exist in the specified table.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Ignores data associated with taxa present in the `shelved_taxa`</span>
<span class="sd">        attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Locking mechanism necessary to avoid concurrency issues when</span>
            <span class="c1"># accessing the database. This ensures that only one Cursor</span>
            <span class="c1"># object is querying the database at any given time</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">acquire</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="ow">and</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;SELECT seq FROM [</span><span class="si">{}</span><span class="s2">] WHERE txId=?&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">),</span>
                    <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxon</span><span class="p">],))</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="n">seq</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">KeyError</span>
        <span class="k">finally</span><span class="p">:</span>
            <span class="n">lock</span><span class="o">.</span><span class="n">release</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment.remove_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.remove_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">remove_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Removes alignment data from the database. &quot;&quot;&quot;</span>

        <span class="c1"># Drop main alignment</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">))</span>

        <span class="c1"># If additional alignments were created, drop those tables as well</span>
        <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>
            <span class="k">except</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">OperationalError</span><span class="p">:</span>
                <span class="k">pass</span></div>

<div class="viewcode-block" id="Alignment.shelve_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.shelve_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">shelve_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">lst</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Shelves taxa from `Alignment` methods.</span>

<span class="sd">        The taxa provided in the `lst` argument will be ignored in</span>
<span class="sd">        subsequent operations. In practice, taxa from the `taxa_list`</span>
<span class="sd">        attribute that are present in `lst` will be moved to the</span>
<span class="sd">        `shelved_taxa` attribute and ignored in all methods of the class.</span>
<span class="sd">        To revert all taxa to the `taxa_list` attribute, simply call this</span>
<span class="sd">        method with an empty list.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        lst : list</span>
<span class="sd">            List with taxa names that should be ignored.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span> <span class="o">=</span> <span class="n">lst</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span>
                          <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">]</span></div>

<div class="viewcode-block" id="Alignment._read_interleave_phylip"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_interleave_phylip">[docs]</a>    <span class="k">def</span> <span class="nf">_read_interleave_phylip</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Alignment parser for interleave phylip format.</span>

<span class="sd">        Parses an interleave phylip alignment file and stores taxa and</span>
<span class="sd">        sequence data in the database. This method is only called from the</span>
<span class="sd">        `_read_phylip` method, when the regular phylip parser detects that the</span>
<span class="sd">        file is in interleave format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ntaxa : int</span>
<span class="sd">            Number of taxa contained in the alignment file.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        _read_phylip</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The phylip interleave format splits the alignment into blocks of</span>
<span class="sd">        a certain lenght (usually 90 characters) separated by blank lines.</span>
<span class="sd">        This means that to gather the complete sequence of a given taxon,</span>
<span class="sd">        the parser has to read the entire file. To prevent the entire</span>
<span class="sd">        alignment to be loaded into memory, we actually iterate over a range</span>
<span class="sd">        determined by the number of taxa in the alignment. In each iteration,</span>
<span class="sd">        we open a file handle and we retrieve only the sequence of</span>
<span class="sd">        a particular taxon. This ensures that only sequence data for</span>
<span class="sd">        a single taxon is store in memory at any given time. This also means</span>
<span class="sd">        that the alignment file has to be read N times, where N = number of</span>
<span class="sd">        taxa. However, since the vast majority of lines are actually skipped</span>
<span class="sd">        in each iteration, the decrease in speed is marginal, while the</span>
<span class="sd">        gains in memory efficient are much larger.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">):</span>

            <span class="c1"># Variable that will store the sequence for the current taxon.</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># Index identifier of the current taxon</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># When True, it means that the alignment line has the taxon</span>
            <span class="c1"># and sequence. When False, it means that the line contains</span>
            <span class="c1"># only sequence.</span>
            <span class="n">taxa_gather</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

            <span class="c1"># Skip header an any potential blank lines</span>
            <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">while</span> <span class="ow">not</span> <span class="n">header</span><span class="p">:</span>
                <span class="n">header</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>

                <span class="c1"># At blank lines, reset the idx and set taxa_gather to False.</span>
                <span class="c1"># All lines from this point on will only contain sequence.</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                    <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">taxa_gather</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># If the idx of the alignment block matches the taxon</span>
                    <span class="c1"># from the current iteration, add to sequence variable</span>
                    <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="n">i</span><span class="p">:</span>
                        <span class="c1"># Remove the taxon from the gathering</span>
                        <span class="k">if</span> <span class="n">taxa_gather</span><span class="p">:</span>
                            <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]))</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()))</span>
                    <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

            <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

            <span class="c1"># Insert data into database</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

            <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment._read_interleave_nexus"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_interleave_nexus">[docs]</a>    <span class="k">def</span> <span class="nf">_read_interleave_nexus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ntaxa</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Alignment parser for interleave nexus format.</span>

<span class="sd">        Parses an interleave nexus alignment file and stores taxa and</span>
<span class="sd">        sequence data in the database.This method is only called from the</span>
<span class="sd">        `_read_nexus` method, when the regular nexus parser detects that the</span>
<span class="sd">        file is in interleave format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ntaxa : int</span>
<span class="sd">            Number of taxa contained in the alignment file.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        _read_nexus</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The nexus interleave format splits the alignment into blocks of</span>
<span class="sd">        a certain lenght (usually 90 characters) separated by blank lines.</span>
<span class="sd">        This means that to gather the complete sequence of a given taxon,</span>
<span class="sd">        the parser has to read the entire file. To prevent the entire</span>
<span class="sd">        alignment to be loaded into memory, we actually iterate over a range</span>
<span class="sd">        determined by the number of taxa in the alignment. In each iteration,</span>
<span class="sd">        we open a file handle and we retrieve only the sequence of a</span>
<span class="sd">        particular taxon. This ensures that only sequence data for</span>
<span class="sd">        a single taxon is store in memory at any given time. This also means</span>
<span class="sd">        that the alignment file has to be read N times, where N = number of</span>
<span class="sd">        taxa. However, since the vast majority of lines are actually skipped</span>
<span class="sd">        in each iteration, the decrease in speed is marginal, while the</span>
<span class="sd">        gains in memory efficient are much larger.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">):</span>

            <span class="c1"># Variable that will store the sequence for the current taxon.</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="c1"># counter used to skip the Nexus header and footer</span>
            <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Index identifier of the current taxon</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;matrix&quot;</span> <span class="ow">and</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">counter</span> <span class="o">=</span> <span class="mi">1</span>

                <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;;&quot;</span> <span class="ow">and</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">counter</span> <span class="o">=</span> <span class="mi">2</span>

                <span class="k">elif</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>

                    <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                        <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="n">i</span><span class="p">:</span>
                            <span class="k">if</span> <span class="ow">not</span> <span class="n">taxa</span><span class="p">:</span>
                                <span class="n">taxa</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span>
                                                                       <span class="s2">&quot;&quot;</span><span class="p">)</span>
                                <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rm_illegal</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

                            <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]))</span>

                        <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxa</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

            <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

            <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._eval_missing_symbol"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._eval_missing_symbol">[docs]</a>    <span class="k">def</span> <span class="nf">_eval_missing_symbol</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Evaluates missing data symbol from sequence string.</span>

<span class="sd">        This method is performed when sequence data is being parsed from the</span>
<span class="sd">        alignment and only executes when the missing data symbol stored in</span>
<span class="sd">        `Alignment.sequence_code` is not defined. It attempts to count</span>
<span class="sd">        the regular characters used to denote missing data. First, tries to</span>
<span class="sd">        find &quot;?&quot; characters, which are set as the symbol if they exist.</span>
<span class="sd">        Then, it finds &quot;n&quot; characters and if sequence type is set to DNA,</span>
<span class="sd">        this character is set as the symbol.</span>
<span class="sd">        Finally, it finds &quot;x&quot; characters and if sequence type is set to</span>
<span class="sd">        Protein, this character is set as the symbol.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            Sequence string</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

            <span class="k">if</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;?&quot;</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;?&quot;</span>

            <span class="k">elif</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;n&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;n&quot;</span>

            <span class="k">elif</span> <span class="n">sequence</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Protein&quot;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;x&quot;</span></div>

<div class="viewcode-block" id="Alignment._read_phylip"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_phylip">[docs]</a>    <span class="k">def</span> <span class="nf">_read_phylip</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Alignment parser for phylip format.</span>

<span class="sd">        Parses a phylip alignment file and stored taxa and sequence data in</span>
<span class="sd">        the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        read_alignment</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Variable storing the lenght of each sequence</span>
        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Get the number of taxa and sequence length from the file header</span>
        <span class="n">header</span> <span class="o">=</span> <span class="n">fh</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">header</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_length</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>
        <span class="n">taxa_num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">header</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1"># These three following attributes allow the support for</span>
        <span class="c1"># interleave phylip files</span>
        <span class="c1"># Flags whether the current line should be parsed for taxon name</span>
        <span class="n">taxa_gather</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="c1"># Counter that makes the correspondence between the current line</span>
        <span class="c1"># and the appropriate taxon</span>
        <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>

            <span class="c1"># Ignore empty lines</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># The counter is reset when surpassing the number of</span>
            <span class="c1"># expected taxa.</span>
            <span class="k">if</span> <span class="n">c</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">&gt;</span> <span class="n">taxa_num</span><span class="p">:</span>
                <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">taxa_gather</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="k">try</span><span class="p">:</span>

                <span class="c1"># Here, assume that all lines with taxon names have</span>
                <span class="c1"># already been processed, since he taxa_pos variable</span>
                <span class="c1"># already has the same number as the expected taxa in the</span>
                <span class="c1">#  phylip header</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">taxa_gather</span><span class="p">:</span>

                    <span class="c1"># Oh boy, this seems like an interleave phylip file.</span>
                    <span class="c1"># Redirect parsing to appropriate method</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DELETE FROM [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">))</span>
                    <span class="n">size_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_interleave_phylip</span><span class="p">(</span><span class="n">taxa_num</span><span class="p">)</span>
                    <span class="k">break</span>

                <span class="c1"># To support interleave phylip, while the taxa_pos</span>
                <span class="c1"># variable has not reached the expected number of taxa</span>
                <span class="c1"># provided in header[0], treat the line as the first</span>
                <span class="c1"># lines of the phylip where the first field is the taxon</span>
                <span class="c1"># name</span>
                <span class="k">elif</span> <span class="n">taxa_gather</span><span class="p">:</span>

                    <span class="n">taxa</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rm_illegal</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1"># Joint multiple batches of sequence</span>
                        <span class="c1"># Remove any possible whitespace by splitting</span>
                        <span class="c1"># according to whitespace. This also ensures that</span>
                        <span class="c1"># sequence is always a list when added to</span>
                        <span class="c1"># sequence_data</span>
                        <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                        <span class="n">sequence</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxa</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span>

                    <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

                    <span class="c1"># Evaluate missing data symbol if undefined</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_eval_missing_symbol</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES&quot;</span>
                                     <span class="s2">&quot;(?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

                    <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

                    <span class="c1"># Add counter for interleave processing</span>
                    <span class="n">c</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># sequence_data.append((taxa, &quot;&quot;.join(sequence)))</span>

            <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="c1"># Updating partitions object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                      <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
        
        <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment._read_fasta"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_fasta">[docs]</a>    <span class="k">def</span> <span class="nf">_read_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Alignment parser for fasta format.</span>

<span class="sd">        Parses a fasta alignment file and stores taxa and sequence data in</span>
<span class="sd">        the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        read_alignment</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Variable storing the lenght of each sequence</span>
        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">sequence</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

                    <span class="c1"># Evaluate missing data symbol if undefined</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_eval_missing_symbol</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES&quot;</span>
                                     <span class="s2">&quot; (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxa</span><span class="p">]</span> <span class="o">=</span> <span class="n">idx</span>

                    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

                    <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

                    <span class="c1"># sequence_data.append((taxa, &quot;&quot;.join(sequence)))</span>
                    <span class="n">sequence</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">taxa</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rm_illegal</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">taxa</span><span class="p">:</span>
                <span class="n">sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span>
                                <span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">sequence</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

            <span class="c1"># Evaluate missing data symbol if undefined</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_eval_missing_symbol</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES&quot;</span>
                             <span class="s2">&quot; (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxa</span><span class="p">]</span> <span class="o">=</span> <span class="n">idx</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

        <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_length</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="c1"># Updating partitions object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                      <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment._read_loci"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_loci">[docs]</a>    <span class="k">def</span> <span class="nf">_read_loci</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Alignment parser for pyRAD and ipyrad loci format.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        read_alignment</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Variable storing the lenght of each sequence</span>
        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_loci_taxa</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Create empty dict</span>
        <span class="n">sequence_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">tx</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">])</span>

        <span class="c1"># Add a counter to name each locus</span>
        <span class="n">locus_c</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="c1"># This variable is used in comparison with taxa_list to check which</span>
        <span class="c1"># taxa are missing from the current partition. Missing taxa</span>
        <span class="c1"># are filled with missing data.</span>
        <span class="n">present_taxa</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;//&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">taxon</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">)</span>
                <span class="n">present_taxa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
                <span class="n">sequence_data</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>

            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;//&quot;</span><span class="p">):</span>

                <span class="n">locus_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">+=</span> <span class="n">locus_len</span>
                <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">locus_len</span><span class="p">)</span>

                <span class="c1"># Adding missing data</span>
                <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">present_taxa</span><span class="p">:</span>
                        <span class="n">sequence_data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span>
                                                 <span class="n">locus_len</span><span class="p">)</span>

                <span class="n">present_taxa</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span><span class="s2">&quot;locus_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">locus_c</span><span class="p">),</span>
                                              <span class="n">locus_len</span><span class="p">,</span>
                                              <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                <span class="n">locus_c</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">sequence_data</span> <span class="o">=</span> <span class="p">[(</span><span class="n">p</span><span class="p">,</span> <span class="n">tx</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span> <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span>
                         <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sequence_data</span><span class="o">.</span><span class="n">items</span><span class="p">()))]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="n">sequence_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_length</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment._read_nexus"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_nexus">[docs]</a>    <span class="k">def</span> <span class="nf">_read_nexus</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Alignment parser for nexus format.</span>

<span class="sd">        Parses a nexus alignment file and stores taxa and sequence data in</span>
<span class="sd">        the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        read_alignment</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Variable storing the lenght of each sequence</span>
        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ntaxa</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">interleave</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>

            <span class="c1"># Fetch the number of taxa from nexus header. This will be</span>
            <span class="c1"># necessary for efficient interleave parsing</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ntaxa</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">ntaxa</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*ntax=(.+?) &quot;</span><span class="p">,</span>
                                          <span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">InputError</span><span class="p">(</span><span class="s2">&quot;Could not recognize number&quot;</span>
                                        <span class="s2">&quot; of taxa from nexus header&quot;</span><span class="p">)</span>
                    <span class="k">return</span>
                <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                    <span class="k">pass</span>

            <span class="c1"># Fetch the number of sites from nexus header. This will be</span>
            <span class="c1"># necessary for efficient interleave parsing</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span>
                        <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*nchar=(.+?)[ ,;]&quot;</span><span class="p">,</span>
                                  <span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">InputError</span><span class="p">(</span><span class="s2">&quot;Could not recognize number&quot;</span>
                                        <span class="s2">&quot; of sites from nexus header&quot;</span><span class="p">)</span>
                    <span class="k">return</span>
                <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                    <span class="k">pass</span>

            <span class="c1"># Fetch the interleave value from the  nexus header. This</span>
            <span class="c1"># will be necessary for efficient interleave parsing</span>
            <span class="k">if</span> <span class="n">interleave</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">interleave</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;interleave=(.+?) &quot;</span><span class="p">,</span>
                                           <span class="n">line</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                    <span class="n">interleave</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">interleave</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;yes&quot;</span> \
                        <span class="k">else</span> <span class="kc">False</span>
                <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                    <span class="k">pass</span>

            <span class="c1"># Set the counter to beggin data parsing</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;matrix&quot;</span> <span class="ow">and</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="mi">1</span>

            <span class="c1"># Stop sequence parser here</span>
            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;;&quot;</span> <span class="ow">and</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="mi">2</span>

                <span class="c1"># self.locus_length = len(sequence_data[0][2])</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_length</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

            <span class="c1"># Start parsing here</span>
            <span class="k">elif</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>

                <span class="c1"># If neither ntaxa or interleave are defined, return</span>
                <span class="c1"># with an error in the &#39;e&#39; attribute</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">ntaxa</span> <span class="ow">or</span> <span class="n">interleave</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">InputError</span><span class="p">(</span><span class="s2">&quot;Could not recognize ntax or&quot;</span>
                                        <span class="s2">&quot; interleave parameters from&quot;</span>
                                        <span class="s2">&quot; nexus header&quot;</span><span class="p">)</span>
                    <span class="k">return</span>

                <span class="k">if</span> <span class="n">interleave</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>

                    <span class="c1"># In interleave, this marks the change in matrix blocks</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                        <span class="k">continue</span>

                    <span class="c1"># Get taxon name</span>
                    <span class="n">taxa</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                    <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rm_illegal</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

                    <span class="c1"># Update taxa_list and taxa_idx attributes</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxa</span><span class="p">]</span> <span class="o">=</span> <span class="n">idx</span>

                    <span class="c1"># Get sequence string</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>

                    <span class="c1"># Evaluate missing data symbol if undefined</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_eval_missing_symbol</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

                    <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>

                    <span class="c1"># Add sequence to sqlite database</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

                    <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_read_interleave_nexus</span><span class="p">(</span><span class="n">ntaxa</span><span class="p">)</span>
                    <span class="n">counter</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="c1"># If partitions are specified using the charset command, this</span>
            <span class="c1"># section will parse the partitions</span>
            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;charset&quot;</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">read_from_nexus_string</span><span class="p">(</span><span class="n">line</span><span class="p">,</span>
                                                       <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

            <span class="c1"># If substitution models are specified using the lset or prset</span>
            <span class="c1"># commands, this will parse the model parameters</span>
            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;lset&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">or</span> <span class="s2">&quot;prset&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span> <span class="ow">and</span> \
                            <span class="n">counter</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">parse_nexus_model</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

        <span class="c1"># If no partitions have been added during the parsing of the nexus</span>
        <span class="c1"># file, set a single partition</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span> <span class="o">==</span> <span class="n">OrderedDict</span><span class="p">():</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                          <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment._read_stockholm"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._read_stockholm">[docs]</a>    <span class="k">def</span> <span class="nf">_read_stockholm</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Alignment parser for stockholm format.</span>

<span class="sd">        Parses a stockholm alignment file and stores taxa and sequence data in</span>
<span class="sd">        the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        size_list : list</span>
<span class="sd">            List of sequence size (int) for each taxon in the alignment file.</span>
<span class="sd">            Used to check size consistency at the end of the alignment parsing</span>
<span class="sd">            in `read_alignment`.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        read_alignment</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fh</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="c1"># Variable storing the lenght of each sequence</span>
        <span class="n">size_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">sequence_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Skip first header line</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">fh</span><span class="p">:</span>
            <span class="c1"># Skip header and comments</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="k">pass</span>
            <span class="c1"># End of file</span>
            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;//&quot;</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="c1"># Parse sequence data</span>
            <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>

                <span class="n">taxa</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rm_illegal</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">sequence</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
                    <span class="c1"># Evaluate missing data symbol if undefined</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_eval_missing_symbol</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                    <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

                <span class="n">size_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span>

                <span class="n">sequence_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">taxa</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>

        <span class="n">sequence_data</span> <span class="o">=</span> <span class="p">[(</span><span class="n">p</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="ow">in</span>
                         <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">sequence_data</span><span class="p">))]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="n">sequence_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_data</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_length</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="c1"># Updating partitions object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                      <span class="n">file_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="n">fh</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">size_list</span></div>

<div class="viewcode-block" id="Alignment.read_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.read_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">read_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">size_check</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Main alignment parser method.</span>

<span class="sd">        This is the main alignment parsing method that is called when the</span>
<span class="sd">        `Alignment` object is instantiated with a file path as the argument.</span>
<span class="sd">        Given the alignment format automatically detected in `__init__`,</span>
<span class="sd">        it calls the specific method that parses that alignment format.</span>
<span class="sd">        After the execution of this method, all attributes of the class will</span>
<span class="sd">        be set and the full range of methods can be applied.</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        size_check : bool, optional</span>
<span class="sd">            If True, perform a check for sequence size consistency across</span>
<span class="sd">            taxa (default is True).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">parsing_methods</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_phylip</span><span class="p">,</span>
            <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_fasta</span><span class="p">,</span>
            <span class="s2">&quot;loci&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_loci</span><span class="p">,</span>
            <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_nexus</span><span class="p">,</span>
            <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_stockholm</span>
        <span class="p">}</span>

        <span class="n">size_list</span> <span class="o">=</span> <span class="n">parsing_methods</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">input_format</span><span class="p">]()</span>

        <span class="c1"># If the missing data symbol could not be evaluated during alignment</span>
        <span class="c1"># parsing, set the defaults</span>
        <span class="n">default_missing</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;DNA&quot;</span><span class="p">:</span> <span class="s2">&quot;n&quot;</span><span class="p">,</span> <span class="s2">&quot;Protein&quot;</span><span class="p">:</span> <span class="s2">&quot;x&quot;</span><span class="p">}</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">default_missing</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

        <span class="c1"># Checks the size consistency of the alignment</span>
        <span class="k">if</span> <span class="n">size_check</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">size_list</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">AlignmentUnequalLength</span><span class="p">()</span>
                <span class="k">return</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)):</span>
            <span class="n">duplicate_taxa</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_taxa</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">e</span> <span class="o">=</span> <span class="n">DuplicateTaxa</span><span class="p">(</span><span class="s2">&quot;The following taxa were duplicated in&quot;</span>
                                   <span class="s2">&quot; the alignment: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">duplicate_taxa</span><span class="p">)))</span></div>

<div class="viewcode-block" id="Alignment.remove_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.remove_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">remove_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list_file</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;remove&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Removes taxa from the `Alignment` object.</span>

<span class="sd">        Removes taxa based on the  `taxa_list_file` argument from the</span>
<span class="sd">        `Alignment` object.</span>

<span class="sd">        This method supports a list or path to CSV file as the</span>
<span class="sd">        `taxa_list_file` argument. In case of CSV path, it parses the CSV</span>
<span class="sd">        file into a list. The CSV file should be a simple text file containing</span>
<span class="sd">        taxon names in each line.</span>

<span class="sd">        This method also supports two removal modes.</span>
<span class="sd">            - remove: removes the specified taxa</span>
<span class="sd">            - inverse: removes all but the specified taxa</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxa_list_file : list or string</span>
<span class="sd">            List containing taxa names or string with path to a CSV file</span>
<span class="sd">            containig the taxa names.</span>
<span class="sd">        mode : {&quot;remove&quot;, &quot;inverse&quot;}, optional</span>
<span class="sd">            The removal mode (default is remove).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">remove</span><span class="p">(</span><span class="n">list_taxa</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">list_taxa</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;DELETE FROM [</span><span class="si">{}</span><span class="s2">] WHERE txId=?&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span>
                    <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],))</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span>

        <span class="k">def</span> <span class="nf">inverse</span><span class="p">(</span><span class="n">list_taxa</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">list_taxa</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;DELETE FROM [</span><span class="si">{}</span><span class="s2">] WHERE txId=?&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span>
                        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],))</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span>

        <span class="c1"># Checking if taxa_list is an input csv file:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">taxa_list_file</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_basic_csv</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>

        <span class="c1"># If not, then the method&#39;s argument is already the final list</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">IOError</span><span class="p">,</span> <span class="ne">IndexError</span><span class="p">):</span>
            <span class="n">taxa_list</span> <span class="o">=</span> <span class="n">taxa_list_file</span>

        <span class="c1"># Filter taxa_list for the taxa that are actually present in this</span>
        <span class="c1"># Alignment object</span>
        <span class="n">taxa_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">taxa_list</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;remove&quot;</span><span class="p">:</span>
            <span class="n">remove</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;inverse&quot;</span><span class="p">:</span>
            <span class="n">inverse</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span></div>

<div class="viewcode-block" id="Alignment.change_taxon_name"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.change_taxon_name">[docs]</a>    <span class="k">def</span> <span class="nf">change_taxon_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Changes the name of a particular taxon.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        old_name : str</span>
<span class="sd">            Original taxon name.</span>
<span class="sd">        new_name : str</span>
<span class="sd">            New taxon name.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Change in taxa_list</span>
        <span class="k">if</span> <span class="n">old_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">old_name</span><span class="p">)]</span> <span class="o">=</span> <span class="n">new_name</span>
            <span class="c1"># Change in the database</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;UPDATE [</span><span class="si">{}</span><span class="s2">] SET taxon=? WHERE txId=?&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">new_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">old_name</span><span class="p">]))</span>
            <span class="c1"># Change in taxa_index</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">new_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">old_name</span><span class="p">]</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">old_name</span><span class="p">]</span></div>

    <span class="nd">@setup_database</span>
<div class="viewcode-block" id="Alignment.collapse"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.collapse">[docs]</a>    <span class="k">def</span> <span class="nf">collapse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">write_haplotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">haplotypes_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">conversion_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">haplotype_name</span><span class="o">=</span><span class="s2">&quot;Hap&quot;</span><span class="p">,</span>
                 <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;collapsed&quot;</span><span class="p">,</span> <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Collapses equal sequences into unique haplotypes.</span>

<span class="sd">        Collapses equal sequences into unique haplotypes. This method fetches</span>
<span class="sd">        the sequences for the current alignment and creates a new database</span>
<span class="sd">        table with the collapsed haplotypes.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        write_haplotypes : bool, optional</span>
<span class="sd">            If True, a file mapping the taxa names name to their respective</span>
<span class="sd">            haplotype name will be generated (default is True).</span>
<span class="sd">        haplotypes_file : string, optional</span>
<span class="sd">            Name of the file mapping taxa names to their respective haplotype.</span>
<span class="sd">            Only used when `write_haplotypes` is True. If it not provided and</span>
<span class="sd">            `write_haplotypes` is True, the file name will be determined</span>
<span class="sd">            from `Alignment.name`.</span>
<span class="sd">        dest : string, optional</span>
<span class="sd">            Path of directory where `haplotypes_file` will be generated</span>
<span class="sd">            (default is &quot;.&quot;).</span>
<span class="sd">        conversion_suffix : string, optional</span>
<span class="sd">            Suffix appended to the haplotypes file. Only used when</span>
<span class="sd">            `write_haplotypes` is True and `haplotypes_file` is None.</span>
<span class="sd">        haplotype_name : string, optional</span>
<span class="sd">            Prefix of the haplotype string. The final haplotype name will be</span>
<span class="sd">            `haplotype_name` + &lt;int&gt; (default is &quot;Hap&quot;).</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string, optional</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;collapsed&quot;).</span>
<span class="sd">        use_main_table : bool, optional</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupDatabase</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        In order to collapse an alignment, all sequence would need to be</span>
<span class="sd">        compared and somehow stored until the last sequence is processed.</span>
<span class="sd">        This creates the immediate issue that, if all sequences are different,</span>
<span class="sd">        the entire alignment could have to be stored in memory. To increase</span>
<span class="sd">        memory efficiency and increase performance, sequences are converted</span>
<span class="sd">        into hashes, and these hashes are compared among each other instead</span>
<span class="sd">        of the actual sequences. This means that instead of comparing</span>
<span class="sd">        potentially very long string and maintaining them in memory, only a</span>
<span class="sd">        small hash string is maintained and compared. In this way, only a</span>
<span class="sd">        fraction of the information is stored in memory and the comparisons</span>
<span class="sd">        are much faster.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="c1"># Create temporary table for storing unique sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="s2">&quot;.collapsed&quot;</span><span class="p">)</span>

        <span class="c1"># Create a new cursor to edit the database while iterating over</span>
        <span class="c1"># table_in</span>
        <span class="n">col_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="c1"># Dict that will store the hashes of each sequence. It will be used to</span>
        <span class="c1"># check if the current sequence has already been processed. The value</span>
        <span class="c1"># will be the haplotype</span>
        <span class="n">hash_list</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># hap_dic will store the haplotype name as key and a list of the</span>
        <span class="c1"># taxa with the same sequence as a list value</span>
        <span class="n">hap_dic</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">hap_counter</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxa</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                <span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="s2">&quot;Collapsing taxon </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">taxa</span><span class="p">))</span>

            <span class="n">cur_hash</span> <span class="o">=</span> <span class="nb">hash</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

            <span class="c1"># If the current hash is unique, add it to the database</span>
            <span class="k">if</span> <span class="n">cur_hash</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hash_list</span><span class="p">:</span>

                <span class="c1"># Create name for new haplotype and update other haplotype</span>
                <span class="c1"># related variables</span>
                <span class="n">haplotype</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">haplotype_name</span><span class="p">,</span> <span class="n">hap_counter</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">hash_list</span><span class="p">[</span><span class="n">cur_hash</span><span class="p">]</span> <span class="o">=</span> <span class="n">haplotype</span>
                <span class="n">hap_dic</span><span class="p">[</span><span class="n">haplotype</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">taxa</span><span class="p">]</span>

                <span class="c1"># Add data to table</span>
                <span class="n">col_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.collapsed] &quot;</span>
                                <span class="s2">&quot;VALUES (?, ?, ?)&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">hap_counter</span><span class="p">,</span>
                                                     <span class="n">unicode</span><span class="p">(</span><span class="n">haplotype</span><span class="p">),</span>
                                                     <span class="n">seq</span><span class="p">))</span>

                <span class="n">hap_counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="n">hap_dic</span><span class="p">[</span><span class="n">hash_list</span><span class="p">[</span><span class="n">cur_hash</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>

        <span class="c1"># The collapse operation is special in the sense that the former taxon</span>
        <span class="c1"># names are no longer valid. Therefore, we drop the previous table and</span>
        <span class="c1"># populate a new one with the collapsed data</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">table_out</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">];&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>

        <span class="c1"># Replace table_out with collapsed table</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;ALTER TABLE [.collapsed] RENAME TO [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">table_out</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">write_haplotypes</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="c1"># If no output file for the haplotype correspondence is provided,</span>
            <span class="c1"># use the input alignment name as reference</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">haplotypes_file</span><span class="p">:</span>
                <span class="n">haplotypes_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">conversion_suffix</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">write_loci_correspondence</span><span class="p">(</span><span class="n">hap_dic</span><span class="p">,</span> <span class="n">haplotypes_file</span><span class="p">,</span>
                                           <span class="n">dest</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Alignment.write_loci_correspondence"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.write_loci_correspondence">[docs]</a>    <span class="k">def</span> <span class="nf">write_loci_correspondence</span><span class="p">(</span><span class="n">hap_dict</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes the file mapping taxa to unique haplotypes for `collapse`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        hap_dict : dict</span>
<span class="sd">            Dictionary mapping the haplotype (key) to a list of taxa (value)</span>
<span class="sd">            sharing the same haplotype.</span>
<span class="sd">        output_file : str</span>
<span class="sd">            Name of the haplotype correspondence file</span>
<span class="sd">        dest : str, optional</span>
<span class="sd">            Path to directory where the `output_file` will be written.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file</span> <span class="o">+</span> <span class="s2">&quot;.haplotypes&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">haplotype</span><span class="p">,</span> <span class="n">taxa_list</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">hap_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">haplotype</span><span class="p">,</span> <span class="s2">&quot;; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)))</span>

        <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

    <span class="nd">@setup_database</span>
<div class="viewcode-block" id="Alignment.consensus"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.consensus">[docs]</a>    <span class="k">def</span> <span class="nf">consensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">consensus_type</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;consensus&quot;</span><span class="p">,</span>
                  <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates a consensus sequence from the alignment.</span>

<span class="sd">        Converts the current `Alignment` alignment data into a single</span>
<span class="sd">        consensus sequence. The `consensus_type` argument determines how</span>
<span class="sd">        variation in the original alignment is handled for the generation</span>
<span class="sd">        of the consensus sequence.</span>

<span class="sd">        The options for handling sequence variation are:</span>

<span class="sd">            - IUPAC: Converts variable sites according to the corresponding</span>
<span class="sd">              IUPAC symbols (DNA sequence type only)</span>

<span class="sd">            - Soft mask: Converts variable sites into missing data</span>

<span class="sd">            - Remove: Removes variable sites</span>

<span class="sd">            - First sequence: Uses the first sequence in the dictionary</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        consensus_type : {&quot;IUPAC&quot;, &quot;Soft mask&quot;, &quot;Remove&quot;, &quot;First sequence&quot;}</span>
<span class="sd">            Type of variation handling. See summary above.</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string, optional</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;consensus&quot;).</span>
<span class="sd">        use_main_table : bool, optional</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupDatabase</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="c1"># Empty consensus sequence</span>
        <span class="n">consensus_seq</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># If sequence type is first sequence</span>
        <span class="k">if</span> <span class="n">consensus_type</span> <span class="o">==</span> <span class="s2">&quot;First sequence&quot;</span><span class="p">:</span>
            <span class="c1"># Grab first available sequence</span>
            <span class="n">consensus_seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">while</span> <span class="ow">not</span> <span class="n">consensus_seq</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">consensus_seq</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;SELECT seq from [</span><span class="si">{}</span><span class="s2">] WHERE txId=</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">table_in</span><span class="p">,</span> <span class="n">idx</span><span class="p">))</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span>
                <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                    <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">column</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">column</span><span class="p">))</span>

            <span class="c1"># If invariable, include in consensus and skip. Increases speed</span>
            <span class="c1"># when there are no missing or gap characters</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">column</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">continue</span>

            <span class="c1"># Remove missing data and gaps</span>
            <span class="n">column</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span>
                             <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">])</span>

            <span class="c1"># In case there is only missing/gap characters</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">column</span><span class="p">:</span>
                <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">continue</span>

            <span class="c1"># Check for invariable without missing data</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">column</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">consensus_type</span> <span class="o">==</span> <span class="s2">&quot;IUPAC&quot;</span><span class="p">:</span>
                <span class="c1"># Convert variation into IUPAC code</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">iupac</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">column</span><span class="p">)])</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">iupac_code</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">x</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">dna_chars</span> <span class="k">else</span> <span class="n">iupac_rev</span><span class="p">[</span><span class="n">x</span><span class="p">]</span>
                         <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span><span class="p">]))))</span>
                    <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">iupac</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">iupac_code</span><span class="p">)])</span>
                <span class="k">continue</span>

            <span class="k">elif</span> <span class="n">consensus_type</span> <span class="o">==</span> <span class="s2">&quot;Soft mask&quot;</span><span class="p">:</span>
                <span class="n">consensus_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">continue</span>

            <span class="k">elif</span> <span class="n">consensus_type</span> <span class="o">==</span> <span class="s2">&quot;Remove&quot;</span><span class="p">:</span>
                <span class="k">continue</span>

        <span class="c1"># The consensus operation is special in the sense that the former taxon</span>
        <span class="c1"># names are no longer valid. Therefore, we drop the previous table and</span>
        <span class="c1"># populate a new one with the consensus data</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">table_out</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DELETE FROM [</span><span class="si">{}</span><span class="s2">];&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>

        <span class="c1"># Insert into database</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES(?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">table_out</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;consensus&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">consensus_seq</span><span class="p">)))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;consensus&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="s2">&quot;consensus&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="Alignment._check_partitions"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._check_partitions">[docs]</a>    <span class="k">def</span> <span class="nf">_check_partitions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">partition_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Consistency check of the `partition_obj` for `Alignment` object.</span>

<span class="sd">        Performs checks to ensure that a `partition_obj` is consistent</span>
<span class="sd">        with the alignment data of the `Alignment` object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        partition_obj : trifusion.process.data.Partitions</span>
<span class="sd">            Partitions object.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        trifusion.process.data.Partitions</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Checks if total lenght of partitions matches the lenght of the</span>
        <span class="c1"># current alignment</span>

        <span class="k">if</span> <span class="n">partition_obj</span><span class="o">.</span><span class="n">counter</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">process</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">InvalidPartitionFile</span><span class="p">(</span><span class="s2">&quot;Partitions in partition&quot;</span>
                   <span class="s2">&quot; file are inconsistent with current alignment. Last &quot;</span>
                   <span class="s2">&quot;partition range: </span><span class="si">{}</span><span class="s2">; Alignmet length: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">partition_obj</span><span class="o">.</span><span class="n">counter</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span>
            <span class="p">))</span></div>

<div class="viewcode-block" id="Alignment.set_partitions"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.set_partitions">[docs]</a>    <span class="k">def</span> <span class="nf">set_partitions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">partitions</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Updates the `partition` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        partitions : trifusion.process.data.Partitions</span>
<span class="sd">            Partitions object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Checks partition&#39;s consistency</span>
        <span class="n">er</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_check_partitions</span><span class="p">(</span><span class="n">partitions</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">er</span><span class="p">,</span> <span class="n">process</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">InvalidPartitionFile</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">er</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">partitions</span></div>

<div class="viewcode-block" id="Alignment.reverse_concatenate"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.reverse_concatenate">[docs]</a>    <span class="k">def</span> <span class="nf">reverse_concatenate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">db_con</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reverses the alignment according to the `partitions` attribute.</span>

<span class="sd">        This method splits the current alignment according to the</span>
<span class="sd">        partitions set in the `partitions` attribute. Returns an</span>
<span class="sd">        `AlignmentList` object where each partition is an `Alignment` object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        db_con : sqlite3.Connection</span>
<span class="sd">            Database connection object that will be provided to each new</span>
<span class="sd">            `Alignment` object.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">add_alignment</span><span class="p">(</span><span class="n">part_name</span><span class="p">,</span> <span class="n">part_len</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">taxa_idx</span><span class="p">):</span>

            <span class="c1"># Create Alignment object</span>
            <span class="n">current_aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">part_name</span><span class="p">,</span> <span class="n">input_format</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">input_format</span><span class="p">,</span>
                                    <span class="n">sql_cursor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="p">,</span>
                                    <span class="n">locus_length</span><span class="o">=</span><span class="n">part_len</span><span class="p">,</span>
                                    <span class="n">sequence_code</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">,</span>
                                    <span class="n">taxa_list</span><span class="o">=</span><span class="n">taxa_list</span><span class="p">,</span>
                                    <span class="n">taxa_idx</span><span class="o">=</span><span class="n">taxa_idx</span><span class="p">)</span>

            <span class="n">alns</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_aln</span><span class="p">)</span>

        <span class="n">taxa_list_master</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="n">taxa_idx_master</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="n">rev_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">):</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">part_range</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">:</span>

                <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">isalnum</span><span class="p">()])</span>

                <span class="k">if</span> <span class="n">part_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>

                        <span class="n">part_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
                                       <span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">][</span><span class="n">i</span><span class="p">::</span><span class="mi">3</span><span class="p">]</span>

                        <span class="k">if</span> <span class="n">part_seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="k">continue</span>

                        <span class="n">cname</span> <span class="o">=</span> <span class="n">name</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                        <span class="n">taxa_list_master</span><span class="p">[</span><span class="n">cname</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
                        <span class="n">taxa_idx_master</span><span class="p">[</span><span class="n">cname</span><span class="p">][</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>

                        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">cname</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="n">rev_cur</span><span class="p">):</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="n">cname</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="n">rev_cur</span><span class="p">)</span>

                        <span class="n">rev_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                            <span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES&quot;</span>
                            <span class="s2">&quot;(?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cname</span><span class="p">),</span> <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">part_seq</span><span class="p">))</span>

                <span class="k">else</span><span class="p">:</span>

                    <span class="n">part_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span>
                                   <span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">part_seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="n">taxa_list_master</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
                    <span class="n">taxa_idx_master</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>

                    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="n">rev_cur</span><span class="p">):</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="n">rev_cur</span><span class="p">)</span>

                    <span class="n">rev_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                        <span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES&quot;</span>
                        <span class="s2">&quot;(?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">),</span> <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">part_seq</span><span class="p">))</span>

        <span class="n">concatenated_aln</span> <span class="o">=</span> <span class="n">AlignmentList</span><span class="p">([],</span> <span class="n">db_con</span><span class="o">=</span><span class="n">db_con</span><span class="p">,</span> <span class="n">db_cur</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="p">)</span>
        <span class="n">alns</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Create Alignment objects for each partition</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">part_range</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">:</span>

            <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">isalnum</span><span class="p">()])</span>

            <span class="k">if</span> <span class="n">part_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>

                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>

                    <span class="n">part_len</span> <span class="o">=</span> <span class="p">(</span><span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span>

                    <span class="n">taxa_list</span> <span class="o">=</span> <span class="n">taxa_list_master</span><span class="p">[</span><span class="n">name</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span>
                    <span class="n">taxa_idx</span> <span class="o">=</span> <span class="n">taxa_idx_master</span><span class="p">[</span><span class="n">name</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span>

                    <span class="n">add_alignment</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">part_len</span><span class="p">,</span>
                                  <span class="n">taxa_list</span><span class="o">=</span><span class="n">taxa_list</span><span class="p">,</span> <span class="n">taxa_idx</span><span class="o">=</span><span class="n">taxa_idx</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="n">part_len</span> <span class="o">=</span> <span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">part_range</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="n">taxa_list</span> <span class="o">=</span> <span class="n">taxa_list_master</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
                <span class="n">taxa_idx</span> <span class="o">=</span> <span class="n">taxa_idx_master</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>

                <span class="n">add_alignment</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">part_len</span><span class="p">,</span> <span class="n">taxa_list</span><span class="o">=</span><span class="n">taxa_list</span><span class="p">,</span>
                              <span class="n">taxa_idx</span><span class="o">=</span><span class="n">taxa_idx</span><span class="p">)</span>

        <span class="n">concatenated_aln</span><span class="o">.</span><span class="n">add_alignments</span><span class="p">(</span><span class="n">alns</span><span class="p">,</span> <span class="n">ignore_paths</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">concatenated_aln</span></div>

    <span class="nd">@setup_database</span>
<div class="viewcode-block" id="Alignment.filter_codon_positions"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.filter_codon_positions">[docs]</a>    <span class="k">def</span> <span class="nf">filter_codon_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">position_list</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;filter&quot;</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Filter alignment according to codon positions.</span>

<span class="sd">        Filters codon positions from `Alignment` object with DNA sequence</span>
<span class="sd">        type. The `position_list` argument determines which codon positions</span>
<span class="sd">        will be stored. It should be a three element list, corresponding to</span>
<span class="sd">        the three codon position, and the positions that are saved are the</span>
<span class="sd">        ones with a True value.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        position_list : list</span>
<span class="sd">            List of three bool elements that correspond to each codon position.</span>
<span class="sd">            Ex. [True, True, True] will save all positions while</span>
<span class="sd">            [True, True, False] will exclude the third codon position</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string, optional</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;filter&quot;).</span>
<span class="sd">        use_main_table : bool, optional</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupDatabase</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">index</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">pos</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            index generator</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="mi">3</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">pos</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">j</span><span class="p">:</span>
                        <span class="k">yield</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">yield</span> <span class="mi">0</span>

        <span class="c1"># Create temporary table for storing unique sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="s2">&quot;.codonfilter&quot;</span><span class="p">)</span>

        <span class="c1"># Create a new cursor to edit the database while iterating over</span>
        <span class="c1"># table_in</span>
        <span class="n">cod_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span>
                <span class="n">itertools</span><span class="o">.</span><span class="n">compress</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">index</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                              <span class="n">position_list</span><span class="p">))))</span>

            <span class="n">cod_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.codonfilter] VALUES &quot;</span>
                            <span class="s2">&quot;(?, ?, ?)&quot;</span><span class="p">,</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">taxon</span><span class="p">],</span>
                                         <span class="n">taxon</span><span class="p">,</span>
                                         <span class="n">seq</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">table_out</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>

        <span class="c1"># Replace table_out with collapsed table</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
            <span class="s2">&quot;ALTER TABLE [.codonfilter] RENAME TO [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">table_out</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span></div>

    <span class="nd">@setup_database</span>
<div class="viewcode-block" id="Alignment.code_gaps"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.code_gaps">[docs]</a>    <span class="k">def</span> <span class="nf">code_gaps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;gaps&quot;</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                  <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Code gap patterns as binary matrix at the end of alignment.</span>

<span class="sd">        This method codes gaps present in the alignment in binary format,</span>
<span class="sd">        according to the method of Simmons and Ochoterena (2000), to be read</span>
<span class="sd">        by phylogenetic programs such as MrBayes. The resultant alignment,</span>
<span class="sd">        however, can only be output in the Nexus format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string, optional</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;gaps&quot;).</span>
<span class="sd">        use_main_table : bool, optional</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        SetupDatabase</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="k">def</span> <span class="nf">gap_listing</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gap_symbol</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot; Function that parses a sequence string and returns the</span>
<span class="sd">            position of indel events. The returned list is composed of</span>
<span class="sd">            tuples with the span of each indel &quot;&quot;&quot;</span>
            <span class="n">gap</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">+&quot;</span> <span class="o">%</span> <span class="n">gap_symbol</span>
            <span class="n">span_regex</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">gap_list</span><span class="p">,</span> <span class="n">seq_start</span> <span class="o">=</span> <span class="p">[],</span> <span class="mi">0</span>
            <span class="k">while</span> <span class="n">span_regex</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">span_regex</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">gap</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">span_regex</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">seq_start</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">gap_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">())</span>
                    <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]</span>
                    <span class="n">seq_start</span> <span class="o">=</span> <span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">span_regex</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">seq_start</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">gap_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq_start</span><span class="p">,</span>
                                     <span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq_start</span><span class="p">))</span>
                    <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]</span>
                    <span class="n">seq_start</span> <span class="o">+=</span> <span class="n">span_regex</span><span class="o">.</span><span class="n">span</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">gap_list</span>

        <span class="k">def</span> <span class="nf">gap_binary_generator</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">gap_list</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot; This function contains the algorithm to construct the binary</span>
<span class="sd">             state block for the indel events &quot;&quot;&quot;</span>
            <span class="k">for</span> <span class="n">cur_gap</span> <span class="ow">in</span> <span class="n">gap_list</span><span class="p">:</span>
                <span class="n">cur_gap_start</span><span class="p">,</span> <span class="n">cur_gap_end</span> <span class="o">=</span> <span class="n">cur_gap</span>
                <span class="k">if</span> <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_start</span><span class="p">:</span><span class="n">cur_gap_end</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> \
                        <span class="p">(</span><span class="n">cur_gap_end</span> <span class="o">-</span> <span class="n">cur_gap_start</span><span class="p">)</span> <span class="ow">and</span> \
                        <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> \
                        <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_end</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                    <span class="n">sequence</span> <span class="o">+=</span> <span class="s2">&quot;1&quot;</span>

                <span class="k">elif</span> <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_start</span><span class="p">:</span><span class="n">cur_gap_end</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> \
                        <span class="p">(</span><span class="n">cur_gap_end</span> <span class="o">-</span> <span class="n">cur_gap_start</span><span class="p">):</span>

                    <span class="k">if</span> <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="ow">or</span> \
                            <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_end</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                        <span class="n">sequence</span> <span class="o">+=</span> <span class="s2">&quot;-&quot;</span>

                <span class="k">elif</span> <span class="n">sequence</span><span class="p">[</span><span class="n">cur_gap_start</span><span class="p">:</span><span class="n">cur_gap_end</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> \
                        <span class="p">(</span><span class="n">cur_gap_end</span> <span class="o">-</span> <span class="n">cur_gap_start</span><span class="p">):</span>
                    <span class="n">sequence</span> <span class="o">+=</span> <span class="s2">&quot;0&quot;</span>
            <span class="k">return</span> <span class="n">sequence</span>

        <span class="n">complete_gap_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">sequence_data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Get the complete list of unique gap positions in the alignment</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Fetching gaps for </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tx</span><span class="p">))</span>

            <span class="n">current_list</span> <span class="o">=</span> <span class="n">gap_listing</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">complete_gap_list</span> <span class="o">+=</span> <span class="p">[</span><span class="n">gap</span> <span class="k">for</span> <span class="n">gap</span> <span class="ow">in</span> <span class="n">current_list</span> <span class="k">if</span> <span class="n">gap</span> <span class="ow">not</span> <span class="ow">in</span>
                                  <span class="n">complete_gap_list</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="c1"># This will add the binary matrix of the unique gaps listed at the</span>
        <span class="c1"># end of each alignment sequence</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Adding gaps for </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tx</span><span class="p">))</span>

            <span class="n">final_seq</span> <span class="o">=</span> <span class="n">gap_binary_generator</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">complete_gap_list</span><span class="p">)</span>

            <span class="c1"># Check if input and output tables are the same. If they are,</span>
            <span class="c1"># it means that the output table already exists and is being</span>
            <span class="c1"># updated</span>
            <span class="k">if</span> <span class="n">table_in</span> <span class="o">==</span> <span class="n">table_out</span><span class="p">:</span>
                <span class="n">sequence_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">final_seq</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">]))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sequence_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="n">final_seq</span><span class="p">))</span>

        <span class="c1"># Populate/modify table</span>
        <span class="k">if</span> <span class="n">table_in</span> <span class="o">==</span> <span class="n">table_out</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span><span class="s2">&quot;UPDATE [</span><span class="si">{}</span><span class="s2">] SET seq=? WHERE txId=?&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">table_out</span><span class="p">),</span> <span class="n">sequence_data</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">table_out</span><span class="p">),</span> <span class="n">sequence_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">-</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">),</span>
                                            <span class="nb">len</span><span class="p">(</span><span class="n">complete_gap_list</span><span class="p">)</span> <span class="o">+</span>
                                            <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="Alignment._filter_terminals"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._filter_terminals">[docs]</a>    <span class="k">def</span> <span class="nf">_filter_terminals</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_in</span><span class="p">,</span> <span class="n">table_out</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Replace gaps at alignment&#39;s extremities with missing data.</span>
<span class="sd">        </span>
<span class="sd">        This method should not be called directly. Instead it is used by the</span>
<span class="sd">        `filter_missing_data` method.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;collapsed&quot;).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        filter_missing_data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="s2">&quot;.filterterminals&quot;</span><span class="p">)</span>

        <span class="n">filt_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering terminals&quot;</span><span class="p">)</span>

            <span class="c1"># Condition where the sequence only has gaps</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">seq</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">):</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
                <span class="n">filt_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.filterterminals] &quot;</span>
                                 <span class="s2">&quot;VALUES (?, ?, ?)&quot;</span><span class="p">,</span>
                                 <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>
                <span class="k">continue</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">counter</span><span class="p">,</span> <span class="n">reverse_counter</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span>

            <span class="k">while</span> <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">while</span> <span class="n">seq</span><span class="p">[</span><span class="n">reverse_counter</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                <span class="n">seq</span><span class="p">[</span><span class="n">reverse_counter</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">reverse_counter</span> <span class="o">-=</span> <span class="mi">1</span>

            <span class="n">filt_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.filterterminals] &quot;</span>
                             <span class="s2">&quot;VALUES (?, ?, ?)&quot;</span><span class="p">,</span>
                             <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)))</span>

        <span class="c1"># Check if input and output tables are the same. If they are,</span>
        <span class="c1"># drop the old table and replace with this new one</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">table_out</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;ALTER TABLE [.filterterminals] &quot;</span>
                         <span class="s2">&quot;RENAME TO [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span></div>

<div class="viewcode-block" id="Alignment._filter_columns"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._filter_columns">[docs]</a>    <span class="k">def</span> <span class="nf">_filter_columns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gap_threshold</span><span class="p">,</span> <span class="n">missing_threshold</span><span class="p">,</span> <span class="n">table_in</span><span class="p">,</span>
                        <span class="n">table_out</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters alignment columns based on missing data content.</span>
<span class="sd">        </span>
<span class="sd">        Filters alignment columns based on the amount of missing data and</span>
<span class="sd">        gap content. If the calculated content is higher than a defined</span>
<span class="sd">        threshold, the column is removed from the final alignment. Setting</span>
<span class="sd">        `gap_threshold` and `missing_threshold` to 0 results in an alignment</span>
<span class="sd">        with no missing data, while setting them to 100 will not remove any</span>
<span class="sd">        column.</span>

<span class="sd">        This method should not be called directly. Instead it is used by the</span>
<span class="sd">        `filter_missing_data` method.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gap_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap percentage is removed.</span>
<span class="sd">        missing_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap+missing percentage is removed.</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;collapsed&quot;).</span>
<span class="sd">        ns : multiprocessing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        filter_missing_data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="n">taxa_number</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="n">filtered_cols</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="s2">&quot;.filtercolumns&quot;</span><span class="p">)</span>

        <span class="n">filt_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="c1"># Creating the column list variable</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering columns&quot;</span><span class="p">)</span>

            <span class="n">cadd</span> <span class="o">=</span> <span class="n">column</span><span class="o">.</span><span class="n">count</span>

            <span class="c1"># Calculating metrics</span>
            <span class="n">gap_proportion</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">cadd</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">))</span> <span class="o">/</span>
                              <span class="n">taxa_number</span><span class="p">)</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
            <span class="n">missing_proportion</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">cadd</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span> <span class="o">/</span>
                                  <span class="n">taxa_number</span><span class="p">)</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
            <span class="n">total_missing_proportion</span> <span class="o">=</span> <span class="n">gap_proportion</span> <span class="o">+</span> <span class="n">missing_proportion</span>

            <span class="k">if</span> <span class="n">gap_proportion</span> <span class="o">&lt;=</span> <span class="n">gap_threshold</span> <span class="ow">and</span> \
                    <span class="n">total_missing_proportion</span> <span class="o">&lt;=</span> <span class="n">missing_threshold</span><span class="p">:</span>

                <span class="n">filtered_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">filtered_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">compress</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">filtered_cols</span><span class="p">))</span>

            <span class="n">filt_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.filtercolumns] &quot;</span>
                             <span class="s2">&quot;VALUES (?, ?, ?)&quot;</span><span class="p">,</span>
                             <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="c1"># Check if input and output tables are the same. If they are,</span>
        <span class="c1"># it means that the output table already exists and is being</span>
        <span class="c1"># updated</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="n">table_out</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">];&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_out</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;ALTER TABLE [.filtercolumns] RENAME TO [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">table_out</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span></div>

    <span class="nd">@setup_database</span>
<div class="viewcode-block" id="Alignment.filter_missing_data"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.filter_missing_data">[docs]</a>    <span class="k">def</span> <span class="nf">filter_missing_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gap_threshold</span><span class="p">,</span> <span class="n">missing_threshold</span><span class="p">,</span>
                            <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;filter&quot;</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters alignment according to missing data.</span>

<span class="sd">        Filters gaps and true missing data from the alignment using tolerance</span>
<span class="sd">        thresholds for each type of missing data. Both thresholds are maximum</span>
<span class="sd">        percentages of sites in an alignment column containing the type of</span>
<span class="sd">        missing data. If gap_threshold=50, for example, alignment columns with</span>
<span class="sd">        more than 50% of sites with gaps are removed.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gap_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap percentage is removed.</span>
<span class="sd">        missing_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap+missing percentage is removed.</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string, optional</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;filter&quot;).</span>
<span class="sd">        use_main_table : bool, optional</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_terminals</span><span class="p">(</span><span class="n">table_in</span><span class="o">=</span><span class="n">table_in</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="n">table_out</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="n">ns</span><span class="p">)</span>

        <span class="c1"># Update output table. This will make the _filter_columns method modify</span>
        <span class="c1"># the alignment produced in the _filter_terminals method.</span>
        <span class="n">table_in</span> <span class="o">=</span> <span class="n">table_out</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_columns</span><span class="p">(</span><span class="n">gap_threshold</span><span class="p">,</span> <span class="n">missing_threshold</span><span class="p">,</span>
                             <span class="n">table_in</span><span class="o">=</span><span class="n">table_in</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="n">table_out</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="n">ns</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Alignment._test_range"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._test_range">[docs]</a>    <span class="k">def</span> <span class="nf">_test_range</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Test if a given `s` integer in inside a specified range.</span>

<span class="sd">        Wrapper for the tests that determine whether a certain value</span>
<span class="sd">        (`s`) is within the range provided by `min_val` and `max_val`.</span>
<span class="sd">        Both of these arguments can be None, in which case there are not</span>
<span class="sd">        lower and/or upper boundaries, respectively.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        s : int</span>
<span class="sd">            Test value</span>
<span class="sd">        min_val : int</span>
<span class="sd">            Minimum range allowed for the test value.</span>
<span class="sd">        max_val : int</span>
<span class="sd">            Maximum range allowed for the test value.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : bool</span>
<span class="sd">            True if `s` within boundaries. False if not.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># If both values were specified, check if s is within range</span>
        <span class="k">if</span> <span class="n">max_val</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">min_val</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> \
                <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;=</span> <span class="n">min_val</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&lt;=</span> <span class="n">max_val</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="c1"># If only min_val was specified, check if s is higher</span>
        <span class="k">elif</span> <span class="n">max_val</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;=</span> <span class="n">min_val</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="c1"># If only max_val was specified, check if s is lower</span>
        <span class="k">elif</span> <span class="n">min_val</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&lt;=</span> <span class="n">max_val</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="Alignment.filter_segregating_sites"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.filter_segregating_sites">[docs]</a>    <span class="k">def</span> <span class="nf">filter_segregating_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                    <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Checks if the segregating sites from alignment are within range.</span>

<span class="sd">        Evaluates the number of segregating sites of the current alignment</span>
<span class="sd">        and returns True if it falls between the range specified by the</span>
<span class="sd">        `min_val` and `max_val` arguments.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_val : int</span>
<span class="sd">            Minimum number of segregating sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no lower bound.</span>
<span class="sd">        max_val : int</span>
<span class="sd">            Maximum number of segregating sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no upper bound.</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : bool</span>
<span class="sd">            True if the number of segregating sites are within the specified</span>
<span class="sd">            range. False if not.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        filter_informative_sites</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">max_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span>

        <span class="c1"># Counter for segregating sites</span>
        <span class="n">s</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Creating the column list variable</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_columns_uniq</span><span class="p">(</span>
                <span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
                <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">v</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span>
                     <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;-&quot;</span><span class="p">]])</span>

            <span class="k">if</span> <span class="n">v</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">v</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">s</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Add these tests so that the method may exit earlier if the</span>
            <span class="c1"># conditions are met, precluding the analysis of the entire</span>
            <span class="c1"># alignment</span>
            <span class="k">if</span> <span class="n">min_val</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;=</span> <span class="n">min_val</span> <span class="ow">and</span> <span class="n">max_val</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">True</span>

            <span class="k">if</span> <span class="n">max_val</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;</span> <span class="n">max_val</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_test_range</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">)</span></div>

<div class="viewcode-block" id="Alignment.filter_informative_sites"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.filter_informative_sites">[docs]</a>    <span class="k">def</span> <span class="nf">filter_informative_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                    <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Checks if the variable sites from alignment are within range.</span>

<span class="sd">        Similar to `filter_segregating_sites` method, but only considers</span>
<span class="sd">        informative sites (variable sites present in more than 2 taxa).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_val : int</span>
<span class="sd">            Minimum number of informative sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no lower bound.</span>
<span class="sd">        max_val : int</span>
<span class="sd">            Maximum number of informative sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no upper bound.</span>
<span class="sd">        table_in : string, optional</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : bool</span>
<span class="sd">            True if the number of informative sites are within the specified</span>
<span class="sd">            range. False if not.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        filter_segregating_sites</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">max_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span>

        <span class="c1"># Counter for informative sites</span>
        <span class="n">s</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Creating the column list variable</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="n">table_in</span><span class="p">)):</span>

            <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
                <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="n">column</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span>
                              <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;-&quot;</span><span class="p">]])</span>

            <span class="k">if</span> <span class="n">column</span><span class="p">:</span>

                <span class="c1"># Skip if column has no variation</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Delete most common</span>
                <span class="k">del</span> <span class="n">column</span><span class="p">[</span><span class="n">column</span><span class="o">.</span><span class="n">most_common</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>

                <span class="c1"># If any of the remaining sites is present in more than two</span>
                <span class="c1"># taxa score the site as informative</span>
                <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
                    <span class="n">s</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="c1"># Add these tests so that the method may exit earlier if the</span>
                <span class="c1"># conditions are met, precluding the analysis of the entire</span>
                <span class="c1"># alignment</span>
                <span class="k">if</span> <span class="n">min_val</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;=</span> <span class="n">min_val</span> <span class="ow">and</span> <span class="n">max_val</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">return</span> <span class="kc">True</span>

                <span class="k">if</span> <span class="n">max_val</span> <span class="ow">and</span> <span class="n">s</span> <span class="o">&gt;</span> <span class="n">max_val</span><span class="p">:</span>
                    <span class="k">return</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_test_range</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">)</span></div>

<div class="viewcode-block" id="Alignment._get_interleave_data"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._get_interleave_data">[docs]</a>    <span class="k">def</span> <span class="nf">_get_interleave_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                             <span class="n">ns_pipe</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Builds an interleave matrix on the database.</span>

<span class="sd">        When the user requests an interleave output format, this method</span>
<span class="sd">        rearranges the data matrix (which is leave in the database) in</span>
<span class="sd">        interleave format in a new table. This should be performed only once,</span>
<span class="sd">        even when multiple formats are specified in interleave. The first one</span>
<span class="sd">        calls this class, and the subsequent ones use the this table.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; AND&quot;</span>
                <span class="s2">&quot; name=&#39;.interleavedata&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fetchall</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DELETE FROM [.interleavedata]&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE [.interleavedata](&quot;</span>
                             <span class="s2">&quot;taxon TEXT,&quot;</span>
                             <span class="s2">&quot;slice INT,&quot;</span>
                             <span class="s2">&quot;seq TEXT)&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE INDEX interindex ON &quot;</span>
                            <span class="s2">&quot;[.interleavedata](slice)&quot;</span><span class="p">)</span>

        <span class="n">temp_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                               <span class="n">value</span><span class="o">=</span><span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">90</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="mi">90</span><span class="p">):</span>

                <span class="n">temp_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.interleavedata] VALUES&quot;</span>
                                 <span class="s2">&quot; (?, ?, ?)&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">:</span><span class="n">i</span><span class="p">]))</span>
                <span class="c1"># temp_storage[p].append(seq[counter:i])</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="n">i</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">%</span> <span class="mi">90</span><span class="p">:</span>
                    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">temp_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.interleavedata] VALUES&quot;</span>
                                     <span class="s2">&quot; (?, ?, ?)&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">:]))</span>
                    <span class="c1"># temp_storage[p].append(seq[c:],)</span>
            <span class="c1"># This likely means that the alignment is less than</span>
            <span class="c1"># 90 characters long. In this case, append to the storage&#39;s</span>
            <span class="c1"># 0 index</span>
            <span class="k">except</span> <span class="ne">UnboundLocalError</span><span class="p">:</span>
                <span class="n">temp_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [.interleavedata] VALUES &quot;</span>
                                 <span class="s2">&quot;(?, ?, ?)&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="k">return</span> <span class="kc">True</span></div>

<div class="viewcode-block" id="Alignment._write_fasta"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_fasta">[docs]</a>    <span class="k">def</span> <span class="nf">_write_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into fasta output format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_file : str</span>
<span class="sd">            Name of the output file. If using `ns_pipe` in TriFusion,</span>
<span class="sd">            it will prompt the user if the output file already exists.</span>
<span class="sd">        interleave : bool, optional</span>
<span class="sd">            Determines whether the output alignment</span>
<span class="sd">            will be in leave (False) or interleave (True) format. Not all</span>
<span class="sd">            output formats support this option.</span>
<span class="sd">        ld_hat : bool, optional</span>
<span class="sd">            If True, the fasta output format will include a first line</span>
<span class="sd">            compliant with the format of LD Hat and will truncate sequence</span>
<span class="sd">            names and sequence length per line accordingly.</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get relevant keyword arguments</span>
        <span class="n">ld_hat</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ld_hat&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">interleave</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.fas&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="c1"># If LD HAT sub format has been specificed, write the first line</span>
        <span class="c1"># containing the number of sequences, sites and genotype phase</span>
        <span class="k">if</span> <span class="n">ld_hat</span><span class="p">:</span>
            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                        <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing fasta output&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">ld_hat</span><span class="p">:</span>
                <span class="c1"># Truncate sequence name to 30 characters</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">[:</span><span class="mi">30</span><span class="p">]))</span>
                <span class="c1"># Limit each sequence line to 2000 characters</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2000</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">),</span> <span class="mi">2000</span><span class="p">):</span>
                        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="mi">2000</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
            <span class="k">elif</span> <span class="n">interleave</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">taxon</span><span class="p">)</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">90</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="mi">90</span><span class="p">):</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">:</span><span class="n">i</span><span class="p">])</span>
                    <span class="n">counter</span> <span class="o">=</span> <span class="n">i</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">seq</span><span class="p">[</span><span class="n">counter</span><span class="p">:])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._write_phylip"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_phylip">[docs]</a>    <span class="k">def</span> <span class="nf">_write_phylip</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into phylip format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_file : str</span>
<span class="sd">            Name of the output file. If using `ns_pipe` in TriFusion,</span>
<span class="sd">            it will prompt the user if the output file already exists.</span>
<span class="sd">        tx_space_phy : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in phylip</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        cut_space_phy : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            phylip format. Longer  names are truncated (default is 39).</span>
<span class="sd">        interleave : bool, optional</span>
<span class="sd">            Determines whether the output alignment</span>
<span class="sd">            will be in leave (False) or interleave (True) format. Not all</span>
<span class="sd">            output formats support this option.</span>
<span class="sd">        phy_truncate_names : bool, optional</span>
<span class="sd">            If True, taxon names in phylip format are truncated to a maximum</span>
<span class="sd">            of 10 characters (default is False).</span>
<span class="sd">        partition_file : bool, optional</span>
<span class="sd">            If True, the auxiliary partitions file will be written (default</span>
<span class="sd">            is True).</span>
<span class="sd">        model_phylip : str, optional</span>
<span class="sd">            Substitution model for the auxiliary partition file of phylip</span>
<span class="sd">            format, compliant with RAxML.</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get relevant keyword arguments</span>
        <span class="n">interleave</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">tx_space_phy</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tx_space_phy&quot;</span><span class="p">,</span> <span class="mi">40</span><span class="p">)</span>
        <span class="n">cut_space_phy</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cut_space_phy&quot;</span><span class="p">,</span> <span class="mi">39</span><span class="p">)</span>
        <span class="n">phy_truncate_names</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;phy_truncate_names&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">partition_file</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;partition_file&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">model_phylip</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;model_phylip&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="c1"># Change taxa space if phy_truncate_names option is set to True</span>
        <span class="k">if</span> <span class="n">phy_truncate_names</span><span class="p">:</span>
            <span class="n">cut_space_phy</span> <span class="o">=</span> <span class="mi">10</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.phy&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                        <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">interleave</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing phylip output&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">interleave_data</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">interleave_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_interleave_data</span><span class="p">(</span>
                    <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">,</span>
                    <span class="n">ns_pipe</span><span class="o">=</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span>
                <span class="p">)</span>

            <span class="n">write_tx</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">prev</span> <span class="o">=</span> <span class="mi">90</span>
            <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;SELECT taxon, slice, seq FROM [.interleavedata] &quot;</span>
                    <span class="s2">&quot;ORDER BY slice&quot;</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">p</span> <span class="o">!=</span> <span class="n">prev</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">prev</span> <span class="o">=</span> <span class="n">p</span>
                    <span class="n">write_tx</span> <span class="o">=</span> <span class="kc">False</span>

                <span class="k">if</span> <span class="n">write_tx</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">tx</span><span class="p">[:</span><span class="n">cut_space_phy</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span>
                            <span class="n">tx_space_phy</span><span class="p">),</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing phylip output&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                    <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                                   <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_phy</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">tx_space_phy</span><span class="p">),</span>
                    <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

        <span class="c1"># In case there is a concatenated alignment being written</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">()</span> <span class="ow">and</span> <span class="n">partition_file</span><span class="p">:</span>
            <span class="n">partition_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file</span> <span class="o">+</span> <span class="s2">&quot;_part.File&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">lrange</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">:</span>
                <span class="c1"># Get model from app if it exists and there are no codon</span>
                <span class="c1"># positions</span>
                <span class="n">model</span> <span class="o">=</span> <span class="n">model_phylip</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">==</span> \
                                        <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span> \
                    <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">partition_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2"> = </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="n">model</span> <span class="k">if</span> <span class="n">model</span> <span class="k">else</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">name</span><span class="p">,</span>
                    <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                              <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]])))</span>
            <span class="n">partition_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._write_nexus"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_nexus">[docs]</a>    <span class="k">def</span> <span class="nf">_write_nexus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into nexus file format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_format : list</span>
<span class="sd">            List with the output formats to generate. Options are:</span>
<span class="sd">            {&quot;fasta&quot;, &quot;phylip&quot;, &quot;nexus&quot;, &quot;stockholm&quot;, &quot;gphocs&quot;, &quot;ima2&quot;}.</span>
<span class="sd">        tx_space_nex : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in nexus</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        cut_space_nex : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            nexus format. Longer  names are truncated (default is 39).</span>
<span class="sd">        interleave : bool, optional</span>
<span class="sd">            Determines whether the output alignment</span>
<span class="sd">            will be in leave (False) or interleave (True) format. Not all</span>
<span class="sd">            output formats support this option.</span>
<span class="sd">        gap : str, optional</span>
<span class="sd">            Symbol for alignment gaps (default is &#39;-&#39;).</span>
<span class="sd">        use_charset : bool, optional</span>
<span class="sd">            If True, partitions from the `Partitions` object will be written</span>
<span class="sd">            in the nexus output format (default is True).</span>
<span class="sd">        use_nexus_models : bool, optional</span>
<span class="sd">            If True, writes the partitions charset block in nexus format.</span>
<span class="sd">        outgroup_list : list, optional</span>
<span class="sd">            Specify list of taxon names that will be defined as the outgroup</span>
<span class="sd">            in the nexus output format. This may be useful for analyses with</span>
<span class="sd">            MrBayes or other software that may require outgroups.</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get relevant keyword arguments</span>
        <span class="n">interleave</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">tx_space_nex</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tx_space_nex&quot;</span><span class="p">,</span> <span class="mi">40</span><span class="p">)</span>
        <span class="n">cut_space_nex</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cut_space_nex&quot;</span><span class="p">,</span> <span class="mi">39</span><span class="p">)</span>
        <span class="n">gap</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;gap&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="n">use_charset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;use_charset&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">use_nexus_models</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;use_nexus_models&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">outgroup_list</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;outgroup_list&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.nex&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="c1"># This writes the output in interleave format</span>
        <span class="k">if</span> <span class="n">interleave</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing nexus output&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#NEXUS</span><span class="se">\n\n</span><span class="s2">Begin data;</span><span class="se">\n\t</span><span class="s2">dimensions &quot;</span>
                               <span class="s2">&quot;ntax=</span><span class="si">%s</span><span class="s2"> nchar=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">format datatype=&quot;</span>
                               <span class="s2">&quot;mixed(</span><span class="si">%s</span><span class="s2">:1-</span><span class="si">%s</span><span class="s2">, restriction:</span><span class="si">%s</span><span class="s2">) interleave=&quot;</span>
                               <span class="s2">&quot;yes gap=</span><span class="si">%s</span><span class="s2"> missing=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">matrix</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                               <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span><span class="p">,</span>
                                <span class="n">gap</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#NEXUS</span><span class="se">\n\n</span><span class="s2">Begin data;</span><span class="se">\n\t</span><span class="s2">dimensions &quot;</span>
                               <span class="s2">&quot;ntax=</span><span class="si">%s</span><span class="s2"> nchar=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">format datatype=</span><span class="si">%s</span><span class="s2"> &quot;</span>
                               <span class="s2">&quot;interleave=yes gap=</span><span class="si">%s</span><span class="s2"> missing=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">&quot;</span>
                               <span class="s2">&quot;matrix</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                               <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
                                <span class="n">gap</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">interleave_data</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">interleave_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_interleave_data</span><span class="p">(</span>
                    <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                    <span class="n">ns_pipe</span><span class="o">=</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span>
                <span class="p">)</span>

            <span class="c1"># Iterate over interleave data</span>
            <span class="n">prev</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                    <span class="s2">&quot;SELECT taxon, slice, seq FROM [.interleavedata] &quot;</span>
                    <span class="s2">&quot;ORDER BY slice&quot;</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">p</span> <span class="o">!=</span> <span class="n">prev</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">prev</span> <span class="o">=</span> <span class="n">p</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">tx</span><span class="p">[:</span><span class="n">cut_space_nex</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">tx_space_nex</span><span class="p">),</span>
                    <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\t</span><span class="s2">end;&quot;</span><span class="p">)</span>

        <span class="c1"># This writes the output in leave format (default)</span>
        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing nexus output&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#NEXUS</span><span class="se">\n\n</span><span class="s2">Begin data;</span><span class="se">\n\t</span><span class="s2">dimensions &quot;</span>
                               <span class="s2">&quot;ntax=</span><span class="si">%s</span><span class="s2"> nchar=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">format datatype=mixed&quot;</span>
                               <span class="s2">&quot;(</span><span class="si">%s</span><span class="s2">:1-</span><span class="si">%s</span><span class="s2">, restriction:</span><span class="si">%s</span><span class="s2">) interleave=no &quot;</span>
                               <span class="s2">&quot;gap=</span><span class="si">%s</span><span class="s2"> missing=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">matrix</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                               <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span><span class="p">,</span>
                                <span class="n">gap</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;#NEXUS</span><span class="se">\n\n</span><span class="s2">Begin data;</span><span class="se">\n\t</span><span class="s2">dimensions ntax=</span><span class="si">%s</span><span class="s2">&quot;</span>
                               <span class="s2">&quot; nchar=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">format datatype=</span><span class="si">%s</span><span class="s2"> &quot;</span>
                               <span class="s2">&quot;interleave=no gap=</span><span class="si">%s</span><span class="s2"> missing=</span><span class="si">%s</span><span class="s2"> ;</span><span class="se">\n\t</span><span class="s2">&quot;</span>
                               <span class="s2">&quot;matrix</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                                   <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                   <span class="n">gap</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                    <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_nex</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span>
                    <span class="n">tx_space_nex</span><span class="p">),</span> <span class="n">seq</span><span class="p">))</span>

            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;;</span><span class="se">\n\t</span><span class="s2">end;&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">use_charset</span><span class="p">:</span>
            <span class="c1"># Writing partitions, if any</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">begin mrbayes;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">p</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="c1"># Full gene partitions</span>
                <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">lrange</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">:</span>
                    <span class="c1"># If there are codon partitions, write those</span>
                    <span class="k">if</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">charset </span><span class="si">%s</span><span class="s2">_</span><span class="si">%s</span><span class="s2"> = </span><span class="si">%s</span><span class="s2">-</span><span class="si">%s</span><span class="se">\\</span><span class="s2">3;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                                           <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                            <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
                            <span class="n">p</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">charset </span><span class="si">%s</span><span class="s2"> = </span><span class="si">%s</span><span class="s2">-</span><span class="si">%s</span><span class="s2">;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                                       <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                        <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
                        <span class="n">p</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">partition part = </span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2">;</span><span class="se">\n\t</span><span class="s2">set &quot;</span>
                               <span class="s2">&quot;partition=part;</span><span class="se">\n</span><span class="s2">end;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                               <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">name</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span>
                                              <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">get_partition_names</span><span class="p">()])))</span>

        <span class="c1"># Write models, if any</span>
        <span class="k">if</span> <span class="n">use_nexus_models</span><span class="p">:</span>
            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">begin mrbayes;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">models</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">models</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">models</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="n">m</span><span class="p">:</span>
                            <span class="n">m_str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">_models</span><span class="p">[</span><span class="s2">&quot;mrbayes&quot;</span><span class="p">][</span><span class="n">m</span><span class="p">]</span>
                            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>
                                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;applyto=(</span><span class="si">{}</span><span class="s2">) &quot;</span>
                                               <span class="s2">&quot;lset &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span>
                                               <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">m_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;lset &quot;</span> <span class="o">+</span>
                                               <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">m_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;end;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># In case outgroup taxa are specified</span>
        <span class="k">if</span> <span class="n">outgroup_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="c1"># This assures that only the outgroups present in the current</span>
            <span class="c1">#  file are written</span>
            <span class="n">compliant_outgroups</span> <span class="o">=</span> <span class="p">[</span><span class="n">taxon</span> <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">outgroup_list</span>
                                   <span class="k">if</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">compliant_outgroups</span> <span class="ow">is</span> <span class="ow">not</span> <span class="p">[]:</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">begin mrbayes;</span><span class="se">\n\t</span><span class="s2">outgroup </span><span class="si">%s</span><span class="se">\n</span><span class="s2">end;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                               <span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">compliant_outgroups</span><span class="p">)))</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._write_stockholm"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_stockholm">[docs]</a>    <span class="k">def</span> <span class="nf">_write_stockholm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into stockholm file format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                        <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing stockholm output&quot;</span><span class="p">)</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.stockholm&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;# STOCKHOLM V1.0</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;//</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._write_gphocs"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_gphocs">[docs]</a>    <span class="k">def</span> <span class="nf">_write_gphocs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into gphocs file format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="c1"># Write number of loci</span>
        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="p">)))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing gphocs output&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">lrange</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">lrange</span> <span class="o">=</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                               <span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

                <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                        <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                        <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">):</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span>
                                                 <span class="n">seq</span><span class="p">[</span><span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]]))</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing gphocs output&quot;</span><span class="p">)</span>

            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sname</span><span class="p">,</span>
                                           <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>

            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                    <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                    <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment._write_ima2"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_ima2">[docs]</a>    <span class="k">def</span> <span class="nf">_write_ima2</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into ima2 file format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        tx_space_ima2 : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in ima2</span>
<span class="sd">            format (default is 10).</span>
<span class="sd">        cut_space_ima2 : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            ima2 format. Longer  names are truncated (default is 8).</span>
<span class="sd">        ima2_params : list</span>
<span class="sd">            A list with the additional information required for the ima2</span>
<span class="sd">            output format. The list should contains the following information:</span>

<span class="sd">               1. (str) path to file containing the species and populations.</span>
<span class="sd">               2. (str) Population tree in newick format, e.g. (0,1):2.</span>
<span class="sd">               3. (str) Mutational model for all alignments.</span>
<span class="sd">               4. (str) inheritance scalar.</span>

<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">tx_space_ima2</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tx_space_ima2&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
        <span class="n">cut_space_ima2</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cut_space_ima2&quot;</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
        <span class="n">ima2_params</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ima2_params&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">population_file</span> <span class="o">=</span> <span class="n">ima2_params</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">population_tree</span> <span class="o">=</span> <span class="n">ima2_params</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">mutational_model</span> <span class="o">=</span> <span class="n">ima2_params</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">inheritance_scalar</span> <span class="o">=</span> <span class="n">ima2_params</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

        <span class="c1"># Get information on which species belong to each population from</span>
        <span class="c1">#  the populations file</span>
        <span class="n">population_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">population_file</span><span class="p">)</span>
        <span class="n">population_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">population_handle</span><span class="p">:</span>
            <span class="n">taxon</span><span class="p">,</span> <span class="n">population</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">population_storage</span><span class="p">[</span><span class="n">population</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">population_storage</span><span class="p">[</span><span class="n">population</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span> <span class="o">=</span> <span class="p">[</span><span class="n">taxon</span><span class="p">]</span>

        <span class="c1"># Write the general header of the IMa2 input file</span>
        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="c1"># First line with general description</span>
        <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Input file for IMa2 using </span><span class="si">%s</span><span class="s2"> alignments</span><span class="se">\n</span><span class="s2">&quot;</span>
                       <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span>  <span class="c1"># Line with number of loci</span>
                       <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span>  <span class="c1"># Line with name of populations</span>
                       <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span>  <span class="c1"># Line with population string</span>
                       <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="p">),</span>
                          <span class="nb">len</span><span class="p">(</span><span class="n">population_storage</span><span class="p">),</span>
                          <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">population_storage</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                          <span class="n">population_tree</span><span class="p">))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing IMa2 output&quot;</span><span class="p">)</span>

            <span class="c1"># Write each locus</span>
            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">partition</span><span class="p">,</span> <span class="n">lrange</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">):</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="c1"># Retrieving taxon names and sequence data. This step is</span>
                <span class="c1"># the first because it will enable the removal of species</span>
                <span class="c1"># containing only missing data.</span>
                <span class="n">new_alignment</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="c1"># This temporary ordered dictionary is created so that</span>
                <span class="c1"># the number of taxa per populations is corrected in</span>
                <span class="c1"># each locus</span>
                <span class="n">current_locus_populations</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">population_storage</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">population</span><span class="p">,</span> <span class="n">taxa_list</span> <span class="ow">in</span> <span class="n">population_storage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">:</span>
                        <span class="c1"># This try statement catches common errors, such as</span>
                        <span class="c1">#  providing a species in the mapping file that</span>
                        <span class="c1"># does not exist in the alignment</span>
                        <span class="k">try</span><span class="p">:</span>
                            <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span>
                                <span class="n">taxon</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                                <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)[</span>
                                  <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]:</span><span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Taxon </span><span class="si">%s</span><span class="s2"> provided in auxiliary &quot;</span>
                                  <span class="s2">&quot;population mapping file is not found &quot;</span>
                                  <span class="s2">&quot;in the alignment&quot;</span><span class="p">)</span>
                            <span class="k">raise</span> <span class="ne">SystemExit</span>

                        <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="n">new_alignment</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_ima2</span><span class="p">]</span>
                                                  <span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">tx_space_ima2</span><span class="p">),</span>
                                                  <span class="n">seq</span><span class="p">))</span>

                            <span class="n">current_locus_populations</span><span class="p">[</span><span class="n">population</span><span class="p">]</span> \
                                <span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>

                <span class="c1"># Write the header of each partition</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="n">partition</span><span class="p">,</span>
                    <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                              <span class="nb">list</span><span class="p">(</span><span class="n">current_locus_populations</span><span class="o">.</span><span class="n">values</span><span class="p">())]),</span>
                    <span class="p">(</span><span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">mutational_model</span><span class="p">,</span>
                    <span class="n">inheritance_scalar</span><span class="p">))</span>

                <span class="c1"># Write sequence data according to the order of the</span>
                <span class="c1"># population mapping file</span>
                <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">new_alignment</span><span class="p">:</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">population_storage</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing IMa2 output&quot;</span><span class="p">)</span>

            <span class="c1"># Write the header for the single</span>
            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                <span class="n">partition</span><span class="p">,</span>
                <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">population_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                <span class="n">mutational_model</span><span class="p">,</span>
                <span class="n">inheritance_scalar</span><span class="p">))</span>

            <span class="c1"># Write sequence data</span>
            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">population</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                    <span class="n">population_storage</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                                   <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span>
                        <span class="n">taxon</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span>
                        <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span>
                                   <span class="p">(</span>
                                   <span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_ima2</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">tx_space_ima2</span><span class="p">),</span>
                                   <span class="n">seq</span><span class="p">))</span></div>

<div class="viewcode-block" id="Alignment._write_mcmctree"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment._write_mcmctree">[docs]</a>    <span class="k">def</span> <span class="nf">_write_mcmctree</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` object into mcmctree file format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        tx_space_phy : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in phylip</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        cut_space_phy : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            phylip format. Longer  names are truncated (default is 39).</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">tx_space_phy</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;tx_space_phy&quot;</span><span class="p">,</span> <span class="mi">40</span><span class="p">)</span>
        <span class="n">cut_space_phy</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;cut_space_phy&quot;</span><span class="p">,</span> <span class="mi">39</span><span class="p">)</span>
        <span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">out_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_mcmctree.phy&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">taxa_number</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">()</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span>
                            <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing mcmctree output&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">lrange</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">lrange</span> <span class="o">=</span> <span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxa_number</span><span class="p">,</span>
                                            <span class="p">(</span><span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="p">(</span><span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]))))</span>

                <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                        <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">):</span>
                    <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">  </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                        <span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_phy</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span>
                            <span class="n">tx_space_phy</span><span class="p">),</span>
                        <span class="n">seq</span><span class="p">[</span><span class="n">lrange</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">lrange</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>
        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writing mcmctree output&quot;</span><span class="p">)</span>

            <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">taxa_number</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span>
                    <span class="n">table_name</span><span class="o">=</span><span class="n">table_name</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_suffix</span><span class="p">)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

                <span class="n">out_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">  </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="n">taxon</span><span class="p">[:</span><span class="n">cut_space_phy</span><span class="p">]</span><span class="o">.</span><span class="n">ljust</span><span class="p">(</span><span class="n">tx_space_phy</span><span class="p">),</span>
                    <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()))</span>

        <span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Alignment.write_to_file"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.Alignment.write_to_file">[docs]</a>    <span class="k">def</span> <span class="nf">write_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_format</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes alignment data into an output file.</span>

<span class="sd">        Writes the alignment object into a specified output file,</span>
<span class="sd">        automatically adding the extension, according to the specified</span>
<span class="sd">        output formats.</span>

<span class="sd">        This function supports the writing of both converted (no partitions)</span>
<span class="sd">        and concatenated (partitioned files). The choice between these modes</span>
<span class="sd">        is determined by the Partitions object associated with the Alignment</span>
<span class="sd">        object. If it contains multiple partitions, it will produce a</span>
<span class="sd">        concatenated alignment and the auxiliary partition files where</span>
<span class="sd">        necessary. Otherwise it will treat the alignment as a single partition.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_format : list</span>
<span class="sd">            List with the output formats to generate. Options are:</span>
<span class="sd">            {&quot;fasta&quot;, &quot;phylip&quot;, &quot;nexus&quot;, &quot;stockholm&quot;, &quot;gphocs&quot;, &quot;ima2&quot;}.</span>
<span class="sd">        output_file : str</span>
<span class="sd">            Name of the output file. If using `ns_pipe` in TriFusion,</span>
<span class="sd">            it will prompt the user if the output file already exists.</span>
<span class="sd">        tx_space_nex : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in nexus</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        tx_space_phy : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in phylip</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        tx_space_ima2 : int, optional</span>
<span class="sd">            Space (in characters) provided for the taxon name in ima2</span>
<span class="sd">            format (default is 10).</span>
<span class="sd">        cut_space_nex : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            nexus format. Longer  names are truncated (default is 39).</span>
<span class="sd">        cut_space_phy : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            phylip format. Longer  names are truncated (default is 39).</span>
<span class="sd">        cut_space_ima2 : int, optional</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            ima2 format. Longer  names are truncated (default is 8).</span>
<span class="sd">        interleave : bool, optional</span>
<span class="sd">            Determines whether the output alignment</span>
<span class="sd">            will be in leave (False) or interleave (True) format. Not all</span>
<span class="sd">            output formats support this option.</span>
<span class="sd">        gap : str, optional</span>
<span class="sd">            Symbol for alignment gaps (default is &#39;-&#39;).</span>
<span class="sd">        model_phylip : str, optional</span>
<span class="sd">            Substitution model for the auxiliary partition file of phylip</span>
<span class="sd">            format, compliant with RAxML.</span>
<span class="sd">        outgroup_list : list, optional</span>
<span class="sd">            Specify list of taxon names that will be defined as the outgroup</span>
<span class="sd">            in the nexus output format. This may be useful for analyses with</span>
<span class="sd">            MrBayes or other software that may require outgroups.</span>
<span class="sd">        ima2_params : list, optional</span>
<span class="sd">            A list with the additional information required for the ima2</span>
<span class="sd">            output format. The list should contains the following information:</span>

<span class="sd">               1. (str) path to file containing the species and populations.</span>
<span class="sd">               2. (str) Population tree in newick format, e.g. (0,1):2.</span>
<span class="sd">               3. (str) Mutational model for all alignments.</span>
<span class="sd">               4. (str) inheritance scalar.</span>

<span class="sd">        use_charset : bool, optional</span>
<span class="sd">            If True, partitions from the `Partitions` object will be written</span>
<span class="sd">            in the nexus output format (default is True).</span>
<span class="sd">        partition_file : bool, optional</span>
<span class="sd">            If True, the auxiliary partitions file will be written (default</span>
<span class="sd">            is True).</span>
<span class="sd">        output_dir : str, optional</span>
<span class="sd">            If provided, the output file will be written on the specified</span>
<span class="sd">            directory.</span>
<span class="sd">        phy_truncate_names : bool, optional</span>
<span class="sd">            If True, taxon names in phylip format are truncated to a maximum</span>
<span class="sd">            of 10 characters (default is False).</span>
<span class="sd">        ld_hat : bool, optional</span>
<span class="sd">            If True, the fasta output format will include a first line</span>
<span class="sd">            compliant with the format of LD Hat and will truncate sequence</span>
<span class="sd">            names and sequence length per line accordingly.</span>
<span class="sd">        use_nexus_models : bool, optional</span>
<span class="sd">            If True, writes the partitions charset block in nexus format.</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace, optional</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar, optional</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string, optional</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string, optional</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Get relevant keyword arguments</span>
        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">model_phylip</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;model_phylip&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">write_methods</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_fasta</span><span class="p">,</span>
            <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_phylip</span><span class="p">,</span>
            <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_nexus</span><span class="p">,</span>
            <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_stockholm</span><span class="p">,</span>
            <span class="s2">&quot;gphocs&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_gphocs</span><span class="p">,</span>
            <span class="s2">&quot;ima2&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_ima2</span><span class="p">,</span>
            <span class="s2">&quot;mcmctree&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_mcmctree</span>
        <span class="p">}</span>

        <span class="c1"># If a specific output directory is provided, the output file will be</span>
        <span class="c1"># written there</span>
        <span class="n">output_dir</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;output_dir&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_dir</span><span class="p">:</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_file</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">exists</span><span class="p">(</span><span class="n">output_dir</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

        <span class="c1"># Stores suffixes for each format</span>
        <span class="n">format_ext</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ima2&quot;</span><span class="p">:</span> <span class="s2">&quot;.txt&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;mcmctree&quot;</span><span class="p">:</span> <span class="s2">&quot;_mcmctree.phy&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="s2">&quot;.phy&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="s2">&quot;.nex&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="s2">&quot;.fas&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="s2">&quot;.stockholm&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;gphocs&quot;</span><span class="p">:</span> <span class="s2">&quot;.txt&quot;</span><span class="p">}</span>

        <span class="c1"># Check if any output file already exist. If so, issue a warning</span>
        <span class="c1"># through the app or terminal pipes</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">output_format</span><span class="p">:</span>
            <span class="n">fname</span> <span class="o">=</span> <span class="n">output_file</span> <span class="o">+</span> <span class="n">format_ext</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">exists</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>

                <span class="c1"># File exists issue warning through appropriate pipe</span>
                <span class="k">if</span> <span class="n">ns_pipe</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">ns_pipe</span><span class="o">.</span><span class="n">apply_all</span><span class="p">:</span>
                        <span class="n">ns_pipe</span><span class="o">.</span><span class="n">file_dialog</span> <span class="o">=</span> <span class="n">fname</span>
                        <span class="k">while</span> <span class="ow">not</span> <span class="n">ns_pipe</span><span class="o">.</span><span class="n">status</span><span class="p">:</span>
                            <span class="k">pass</span>

                <span class="c1"># When the dialog has been close, check if file is to be</span>
                <span class="c1"># skipped or overwritten</span>
                <span class="k">if</span> <span class="n">ns_pipe</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ns_pipe</span><span class="o">.</span><span class="n">status</span> <span class="o">==</span> <span class="s2">&quot;skip&quot;</span><span class="p">:</span>
                        <span class="c1"># Skip file</span>
                        <span class="k">return</span>

        <span class="c1"># Reset pipes, if any</span>
        <span class="k">if</span> <span class="n">ns_pipe</span><span class="p">:</span>
            <span class="n">ns_pipe</span><span class="o">.</span><span class="n">status</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># This will determine the default model value. GTR for nucleotides</span>
        <span class="c1"># and LG for proteins</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">model_phylip</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
                <span class="n">model_phylip</span> <span class="o">=</span> <span class="s2">&quot;GTR&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">model_phylip</span> <span class="o">=</span> <span class="s2">&quot;LG&quot;</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;model_phylip&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_phylip</span>

        <span class="c1"># Checks if there is any other format besides Nexus if the</span>
        <span class="c1"># alignment&#39;s gap have been coded</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">restriction_range</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">output_format</span> <span class="o">!=</span> <span class="p">[</span><span class="s2">&quot;nexus&quot;</span><span class="p">]:</span>
                <span class="k">return</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="k">for</span> <span class="n">fmt</span> <span class="ow">in</span> <span class="n">output_format</span><span class="p">:</span>
            <span class="n">write_methods</span><span class="p">[</span><span class="n">fmt</span><span class="p">](</span><span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">)</span>

        <span class="c1"># Remove temporary interleave data, if present</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="s2">&quot;.interleavedata&quot;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [.interleavedata]&quot;</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="AlignmentList"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList">[docs]</a><span class="k">class</span> <span class="nc">AlignmentList</span><span class="p">(</span><span class="n">Base</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Main interface for groups of `Alignment` objects.</span>

<span class="sd">    The `AlignmentList` class can be seen as a group of `Alignment` objects</span>
<span class="sd">    that is treated as a single cohesive data set. It is the main class</span>
<span class="sd">    used to perform alignment operations on TriFusion and TriSeq programs,</span>
<span class="sd">    even when there is only a single alignment.</span>

<span class="sd">    An instance can be created by simply providing a list of paths to</span>
<span class="sd">    alignment files, which can be empty (and alignments added later). It</span>
<span class="sd">    is recommended that the path to the sqlite database be specified</span>
<span class="sd">    via the `sql_db` argument, but this is optional.</span>

<span class="sd">    Individual alignment files can be provided when instantiating the class</span>
<span class="sd">    or added later, and each is processed and stored as an `Alignment` object.</span>
<span class="sd">    The `AlignmentList` object provides attributes that are relative to the</span>
<span class="sd">    global data set. For instance, each `Alignment` object may have their</span>
<span class="sd">    individual and unique list of taxa, but the corresponding attribute in the</span>
<span class="sd">    `AlignmentList` object contains the list of all taxa that are found</span>
<span class="sd">    in the `Alignment` object list.</span>
<span class="sd">    </span>
<span class="sd">    Several methods of the `Alignment` class are also present in this class</span>
<span class="sd">    with the same usage (e.g. `collapse`). These are basically wrappers that</span>
<span class="sd">    apply the method to all `Alignment` objects and may have some</span>
<span class="sd">    modifications adapted to multiple alignments.</span>
<span class="sd">    </span>
<span class="sd">    Other methods are exclusive and desgined to deal with multiple alignments,</span>
<span class="sd">    such as concatenation (`concatenate`), filter by taxa (`filter_by_taxa`),</span>
<span class="sd">    etc.</span>

<span class="sd">    Plotting methods are exclusive to `AlignmentList`, even the</span>
<span class="sd">    ones that are focused on single alignments. In this case, the</span>
<span class="sd">    `Alignment` object can be specified and processed individually directly</span>
<span class="sd">    from this class.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    alignment_list : list</span>
<span class="sd">        List with paths to alignment files. Can be empty.</span>
<span class="sd">    sql_db : str, optional</span>
<span class="sd">        Path where the sqlite3 database will be created.</span>
<span class="sd">    db_cur : sqlite3.Cursor, optional</span>
<span class="sd">        Provide a database Cursor object in conjunction with the Connection</span>
<span class="sd">        object (`db_con`) to connect to an existing database.</span>
<span class="sd">    db_con : sqlite3.Connection, optional</span>
<span class="sd">        Provide a database Connection object in conjunction with the Cursor</span>
<span class="sd">        object (`db_cur`) to connect to an existing database.</span>
<span class="sd">    pbar : ProgressBar, optional</span>
<span class="sd">        A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    con : sqlite3.Connection</span>
<span class="sd">        Connection object of the sqlite database.</span>
<span class="sd">    cur : sqlite3.Cursor</span>
<span class="sd">        Cursor object of the sqlite database.</span>
<span class="sd">    sql_path : str</span>
<span class="sd">        Path to sqlite3 database file.</span>
<span class="sd">    alignments : collections.OrderedDict</span>
<span class="sd">        Stores the &#39;active&#39; `Alignment objects` as values, with the</span>
<span class="sd">        corresponding key being `Alignment.path`.</span>
<span class="sd">    shelve_alignments : collections.OrderedDict</span>
<span class="sd">        Stores the &#39;inactive&#39; or &#39;shelved&#39; `Alignment` objects.</span>
<span class="sd">        `AlignmentList` methods will operate only on the `alignments`</span>
<span class="sd">        attribute, unless explicitly stated otherwise. The key:value is the</span>
<span class="sd">        same as in the `alignments` attribute.</span>
<span class="sd">    bad_alignments : list</span>
<span class="sd">        Stores the `Alignment.name` attribute of badly formatted alignments.</span>
<span class="sd">    duplicate_alignments : list</span>
<span class="sd">        Stores the `Alignment.name` attribute of duplicated alignment paths.</span>
<span class="sd">    non_alignments : list</span>
<span class="sd">        Stores the `Alignment.name` attribute of sequence sets of unequal</span>
<span class="sd">        length.</span>
<span class="sd">    taxa_names : list</span>
<span class="sd">        Stores the name of all taxa across the `Alignment` object list.</span>
<span class="sd">    shelved_taxa : list</span>
<span class="sd">        Stores the &#39;inactive&#39; or &#39;shelved&#39; taxa. `AlignmentList` methods</span>
<span class="sd">        will operate only on the `taxa_names` attribute, unless explicitly</span>
<span class="sd">        stated otherwise.</span>
<span class="sd">    path_list : list</span>
<span class="sd">        List of `Alignment.paths` for all `Alignment` objects.</span>
<span class="sd">    filtered_alignments : collections.OrderedDict</span>
<span class="sd">        Ordered dictionary of four key:values used to count the number</span>
<span class="sd">        of `Alignment` objects that were filtered by four operations. The</span>
<span class="sd">        keys are {&quot;By minimum taxa&quot;, &quot;By taxa&quot;, &quot;By variable sites&quot;,</span>
<span class="sd">        &quot;By informative sites&quot;}.</span>
<span class="sd">    sequence_code : tuple</span>
<span class="sd">        Contains information on (&lt;sequence type&gt;, &lt;missing data symbol&gt;),</span>
<span class="sd">        e.g. (&quot;Protein&quot;, &quot;x&quot;).</span>
<span class="sd">    gap_symbol : str</span>
<span class="sd">        Symbol used to denote gaps in the alignment (default is &#39;-&#39;).</span>
<span class="sd">    summary_stats : dict</span>
<span class="sd">        Dictionary that stores several summary statistics that are calculated</span>
<span class="sd">        once the `get_summary_stats` method is executed.</span>
<span class="sd">    summary_gene_table : pandas.DataFrame</span>
<span class="sd">        DataFrame containing the summary statistics for each `Alignment`</span>
<span class="sd">        object. Also populated when the `get_summary_stats` method is</span>
<span class="sd">        executed.</span>
<span class="sd">    active_tables : list</span>
<span class="sd">        List of active database tables associated with the `AlignmentList`</span>
<span class="sd">        instance.</span>
<span class="sd">    partitions : trifusion.process.data.Partitions</span>
<span class="sd">        Partitions object that refers to the total `AlignmentList`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alignment_list</span><span class="p">,</span> <span class="n">sql_db</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">db_cur</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">db_con</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="c1"># Create connection and cursor for sqlite database</span>
        <span class="c1"># If `db_cur` and `db_con` are both provided, setup the database</span>
        <span class="c1"># connection and ignore `sql_db`</span>
        <span class="k">if</span> <span class="n">db_cur</span> <span class="ow">and</span> <span class="n">db_con</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">db_con</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Database Connection object</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="n">db_cur</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Database Cursor object</span>
<span class="sd">            &quot;&quot;&quot;</span>
        <span class="c1"># If only `sql_db` is provided, set it as the attribute</span>
        <span class="k">elif</span> <span class="n">sql_db</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span> <span class="o">=</span> <span class="n">sql_db</span>
        <span class="c1"># If no database information is provided, set a temporary database</span>
        <span class="c1"># file in the current working directory</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span> <span class="o">=</span> <span class="s2">&quot;trifusion.sqlite3&quot;</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Path to sqlite3 database file </span>
<span class="sd">            &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">db_cur</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">db_con</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span><span class="p">,</span> <span class="n">check_same_thread</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;PRAGMA synchronous = OFF&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Stores the &quot;active&quot; `Alignment` objects for the current </span>
<span class="sd">        `AlignmentList`. Keys will be the `Alignment.path` for quick lookup of </span>
<span class="sd">        `Alignment` object values</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Stores the &quot;inactive&quot; or &quot;shelved&quot; `Alignment` objects. All </span>
<span class="sd">        `AlignmentList` methods will operate only on the alignments </span>
<span class="sd">        attribute, unless explicitly stated otherwise. </span>
<span class="sd">        Key-value is the same as the `alignments` attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">bad_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute list that stores the Alignment.name attribute of badly</span>
<span class="sd">        formatted alignments</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute list that stores duplicate Alignment.name.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">non_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attribute list that stores sequence sets of unequal lenght</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        List with the name of the taxa included in the AlignmentList object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        List with non active taxa</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">path_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        List of `Alignment.path`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="s2">&quot;By minimum taxa&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                               <span class="p">(</span><span class="s2">&quot;By taxa&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                               <span class="p">(</span><span class="s2">&quot;By variable sites&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                               <span class="p">(</span><span class="s2">&quot;By informative sites&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)])</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Records the number of filtered alignments by each individual filter type</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Tuple with the AlignmentList sequence code. Either (&quot;DNA&quot;, &quot;n&quot;) or</span>
<span class="sd">        (&quot;Protein&quot;, &quot;x&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Dictionary with summary statistics for the active alignments</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;seq_len&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;gaps&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_gaps&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;missing&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_missing&quot;</span><span class="p">:</span> <span class="p">[],</span>
                              <span class="s2">&quot;variable&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_var&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;informative&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_inf&quot;</span><span class="p">:</span> <span class="p">[]}</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Dictionary with summary statistics for each alignment.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">,</span> <span class="s2">&quot;nsites&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;var&quot;</span><span class="p">,</span> <span class="s2">&quot;inf&quot;</span><span class="p">,</span> <span class="s2">&quot;gap&quot;</span><span class="p">,</span> <span class="s2">&quot;missing&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Lists the currently active tables. This is mainly used for the</span>
<span class="sd">        conversion of the consensus alignments into a single Alignment object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">active_tables</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Set partitions object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">Partitions</span><span class="p">()</span>

        <span class="c1"># if type(alignment_list[0]) is str:</span>
        <span class="k">if</span> <span class="n">alignment_list</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">add_alignment_files</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Iterator behavior for `AlignmentList`</span>

<span class="sd">        Returns an iterator over the &#39;active&#39; `Alignment` objects.</span>

<span class="sd">        Returns</span>

<span class="sd">        _ : iter</span>
<span class="sd">            Iterable over `alignments.values()`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

<div class="viewcode-block" id="AlignmentList._create_table"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._create_table">[docs]</a>    <span class="k">def</span> <span class="nf">_create_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_name</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates a new table in the database.</span>

<span class="sd">        Convenience method that creates a new table in the sqlite database.</span>
<span class="sd">        It accepts a custom Cursor object, which overrides the `cur`</span>
<span class="sd">        attribute. Optionally, an index can also be specified.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_name : str</span>
<span class="sd">            Name of the table to be created.</span>
<span class="sd">        index : list, optional</span>
<span class="sd">            Provide a list with [&lt;name of index&gt;, &lt;columns to be indexed&gt;].</span>
<span class="sd">            (e.g., [&quot;concindex&quot;, &quot;txId&quot;]).</span>
<span class="sd">        cur : sqlite3.Cursor, optional</span>
<span class="sd">            Custom Cursor object used to query the database.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">cur</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span>

        <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE [</span><span class="si">{}</span><span class="s2">](&quot;</span>
                    <span class="s2">&quot;txId INT,&quot;</span>
                    <span class="s2">&quot;taxon TEXT,&quot;</span>
                    <span class="s2">&quot;seq TEXT)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">index</span><span class="p">:</span>
            <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE INDEX </span><span class="si">{}</span><span class="s2"> ON [</span><span class="si">{}</span><span class="s2">](</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">table_name</span><span class="p">,</span> <span class="n">index</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span></div>

<div class="viewcode-block" id="AlignmentList._table_exists"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._table_exists">[docs]</a>    <span class="k">def</span> <span class="nf">_table_exists</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_name</span><span class="p">,</span> <span class="n">cur</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Checks if a table exists in the database.</span>

<span class="sd">        Convenience method that checks if a table exists in the database.</span>
<span class="sd">        It accepts a custom Cursor object, which overrides the `cur`</span>
<span class="sd">        attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_name : str</span>
<span class="sd">            Name of the table.</span>
<span class="sd">        cur: sqlite3.Cursor, optional</span>
<span class="sd">            Custom Cursor object used to query the database.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        res : list</span>
<span class="sd">            List with the results of a query for &#39;table&#39; type with</span>
<span class="sd">            `table_name` name. Is empty when the table does not exist.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This returns a list that will contain one item if the table exists</span>
<span class="sd">        in the database. If it doesn&#39;t exist, returns an empty list.</span>
<span class="sd">        Therefore, this can be used simply like::</span>

<span class="sd">            if self._table_exists(&quot;my_table&quot;):</span>
<span class="sd">                 # Do stuff</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">cur</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span>

        <span class="k">return</span> <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
            <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; AND&quot;</span>
            <span class="s2">&quot; name=&#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList._set_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._set_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_set_pipes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">msg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Setup of progress indicators for both GUI and CLI task executions.</span>

<span class="sd">        This handles the setup of the objects responsible for updating</span>
<span class="sd">        the progress of task&#39;s execution of both TriFusion (GUI) and</span>
<span class="sd">        TriSeq (CLI). At the beginning of any given task, these objects</span>
<span class="sd">        can be initialized by providing either the Namespace object (`ns`)</span>
<span class="sd">        in the case of TriFusion, or the ProgressBar object (`pbar`), in</span>
<span class="sd">        the case of TriSeq. Along with one of these objects, the expected</span>
<span class="sd">        `total` of the progress should also be provided. The `ns` and</span>
<span class="sd">        `pbar` objects are updated at each iteration of a given task,</span>
<span class="sd">        and the `total` is used to get a measure of the progress.</span>

<span class="sd">        Optionally, a message can be also provided for the Namespace object</span>
<span class="sd">        that will be used by TriFusion.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        total : int</span>
<span class="sd">            Expected total of the task&#39;s progress.</span>
<span class="sd">        msg : str, optional</span>
<span class="sd">            A secondary message that appears in TriFusion&#39;s progress dialogs.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The progress of any given task can be provided by either an</span>
<span class="sd">        `Alignment` or `AlignmentList` instance. Generally, the tasks follow</span>
<span class="sd">        the progress of the `AlignmentList` instance, unless that instance</span>
<span class="sd">        contains only one `Alignment` object. In that case, the progress</span>
<span class="sd">        information is piped from the `Alignment` instance. For that reason,</span>
<span class="sd">        and for the Namespace (`ns`) object only, this method checks if</span>
<span class="sd">        there is only one &#39;active&#39; `Alignment` object. If yes, set the</span>
<span class="sd">        `Namepsace.sa` flag to True, so that the progress indicators</span>
<span class="sd">        are piped from the `Alignment` object instead.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        Start a progress counter  for a task that will make 100 iterations::</span>

<span class="sd">            self._set_pipes(ns=ns_obj, pbar=pbar_obj, total=100,</span>
<span class="sd">                            msg=&quot;Some message&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="n">total</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">max_value</span> <span class="o">=</span> <span class="n">total</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="AlignmentList._update_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._update_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">msg</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Update progress indicators for both GUI and CLI task executions.</span>

<span class="sd">        This method provides a single interface for updating the progress</span>
<span class="sd">        objects `ns` or `pbar`, which should have been initialized at the</span>
<span class="sd">        beginning of the task with the `_set_pipes` method.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        value : int</span>
<span class="sd">            Value of the current progress index</span>
<span class="sd">        msg : str, optional</span>
<span class="sd">            A secondary message that appears in TriFusion&#39;s progress dialogs.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        The progress of any given task can be provided by either an</span>
<span class="sd">        `Alignment` or `AlignmentList` instance. Generally, the tasks follow</span>
<span class="sd">        the progress of the `AlignmentList` instance, unless that instance</span>
<span class="sd">        contains only one `Alignment` object. In that case, the progress</span>
<span class="sd">        information is piped from the `Alignment` instance. For that reason,</span>
<span class="sd">        and for the Namespace (`ns`) object only, if the `Namespace.sa`</span>
<span class="sd">        attribute is set to True, then to not update the progress. Progress</span>
<span class="sd">        is being piped directly from the `Alignment` object.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        Update the counter in an iteration of 100::</span>

<span class="sd">            for i in range(100):</span>
<span class="sd">                self._update_pipes(ns=ns_obj, pbar=pbar_obj, value=i,</span>
<span class="sd">                                   msg=&quot;Some string&quot;)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">msg</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="n">value</span>
        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">value</span><span class="p">)</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="AlignmentList._reset_pipes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._reset_pipes">[docs]</a>    <span class="k">def</span> <span class="nf">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reset progress indicators for both GUI and CLI task executions.</span>

<span class="sd">        This should be done at the end of any task that initialized the</span>
<span class="sd">        progress objects, but it only affects the Namespace object. It</span>
<span class="sd">        resets all Namespace attributes to None.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="AlignmentList.aln_names"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.aln_names">[docs]</a>    <span class="k">def</span> <span class="nf">aln_names</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns list of basenames of `alignments` file paths.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">basename</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">])</span></div>

<div class="viewcode-block" id="AlignmentList.resume_database"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.resume_database">[docs]</a>    <span class="k">def</span> <span class="nf">resume_database</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reconnects to the sqlite database.</span>

<span class="sd">        The connection to the database can be closes or even severed when</span>
<span class="sd">        passing objects between threads. This methods allows the</span>
<span class="sd">        reconnection of the database for the `AlignmentList` and</span>
<span class="sd">        *all* (even the shelved ones) `Alignment` objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span><span class="p">,</span> <span class="n">check_same_thread</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">aln</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span>
            <span class="n">aln</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span></div>

<div class="viewcode-block" id="AlignmentList.set_database_connections"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.set_database_connections">[docs]</a>    <span class="k">def</span> <span class="nf">set_database_connections</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cur</span><span class="p">,</span> <span class="n">con</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Provides Connection and Cursor to `Alignment` objects.</span>

<span class="sd">        Sets the database connections manually for *all* (even the</span>
<span class="sd">        shelved ones) `Alignment` objects.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        cur : sqlite3.Cursor</span>
<span class="sd">            Provide a database Cursor object.</span>
<span class="sd">        con : sqlite3.Connection</span>
<span class="sd">            Provide a database Connection object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="n">cur</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">con</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">aln</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="n">cur</span>
            <span class="n">aln</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">con</span></div>

<div class="viewcode-block" id="AlignmentList.get_tables"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.get_tables">[docs]</a>    <span class="k">def</span> <span class="nf">get_tables</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return list with `table_name` of *all* `Alignment` objects.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : list</span>
<span class="sd">            List with `Alignment.table_name`.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">table_name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span></div>

<div class="viewcode-block" id="AlignmentList.remove_tables"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.remove_tables">[docs]</a>    <span class="k">def</span> <span class="nf">remove_tables</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">preserve_tables</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">trash_tables</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Drops tables from the database.</span>

<span class="sd">        This method can be used to drop tables from the database.</span>

<span class="sd">        If no arguments are provided it will remove *all*  database tables</span>
<span class="sd">        associated with the `AlignmentList` object.  If the</span>
<span class="sd">        `preserve_tables` argument is provided, the table names present in</span>
<span class="sd">        this list will not be dropped. If `trash_tables` is provided,</span>
<span class="sd">        only the tables names specified in that list are dropped. If</span>
<span class="sd">        both are provided, `trash_tables` will take precedence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        preserve_tables : list, optional</span>
<span class="sd">            If provided, the table names in this list will NOT be dropped.</span>
<span class="sd">        trash_tables : list, optional</span>
<span class="sd">            If provided, only the table names in this list will be dropped.</span>
<span class="sd">            Takes precedence over `preserve_tables` if both are provided.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">preserve_tables</span><span class="p">:</span>
            <span class="n">preserve_tables</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">trash_tables</span><span class="p">:</span>
            <span class="n">trash_tables</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">tables</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
            <span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39;;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">trash_tables</span><span class="p">:</span>
            <span class="n">tables_delete</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tables</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">trash_tables</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tables_delete</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tables</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span>
                             <span class="n">preserve_tables</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">tb</span> <span class="ow">in</span> <span class="n">tables_delete</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tb</span><span class="p">))</span></div>

<div class="viewcode-block" id="AlignmentList.clear_alignments"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.clear_alignments">[docs]</a>    <span class="k">def</span> <span class="nf">clear_alignments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Clears all attributes and data from the `AlignmentList` object.&quot;&quot;&quot;</span>

        <span class="c1"># Drop all database tables related to the current Alignments</span>
        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">aln</span><span class="o">.</span><span class="n">remove_alignment</span><span class="p">()</span>

        <span class="c1"># Drop active databases of the AlignmentList instance, namely consensus</span>
        <span class="c1"># and concatenation, if they exist</span>
        <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">active_tables</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bad_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">non_alignments</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">path_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">([(</span><span class="s2">&quot;By minimum taxa&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                                <span class="p">(</span><span class="s2">&quot;By taxa&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                                <span class="p">(</span><span class="s2">&quot;By variable sites&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                                                <span class="p">(</span><span class="s2">&quot;By informative sites&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;seq_len&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;gaps&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_gaps&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;missing&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_missing&quot;</span><span class="p">:</span> <span class="p">[],</span>
                              <span class="s2">&quot;variable&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_var&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;informative&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_inf&quot;</span><span class="p">:</span> <span class="p">[]}</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">,</span> <span class="s2">&quot;nsites&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;var&quot;</span><span class="p">,</span> <span class="s2">&quot;inf&quot;</span><span class="p">,</span> <span class="s2">&quot;gap&quot;</span><span class="p">,</span> <span class="s2">&quot;missing&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">active_tables</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">Partitions</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList._reset_summary_stats"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._reset_summary_stats">[docs]</a>    <span class="k">def</span> <span class="nf">_reset_summary_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Resets the `summary_stats` attribute.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;genes&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;seq_len&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;gaps&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_gaps&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;missing&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_missing&quot;</span><span class="p">:</span> <span class="p">[],</span>
                              <span class="s2">&quot;variable&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;avg_var&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;informative&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                              <span class="s2">&quot;avg_inf&quot;</span><span class="p">:</span> <span class="p">[]}</span></div>

<div class="viewcode-block" id="AlignmentList.update_active_alignments"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.update_active_alignments">[docs]</a>    <span class="k">def</span> <span class="nf">update_active_alignments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aln_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">all_files</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Sets the active alignments.</span>

<span class="sd">        Sets the &#39;active&#39; alignments stored in the `alignments` attribute</span>
<span class="sd">        and the &#39;inactive&#39; alignments stored in the `shelve_alignments`</span>
<span class="sd">        attribute, based on the `aln_list` argument. Regardless of the</span>
<span class="sd">        current &#39;active&#39; and &#39;inactive&#39; alignments, calling this method</span>
<span class="sd">        with the `aln_list` argument will set only those as the &#39;active&#39;</span>
<span class="sd">        alignments.</span>

<span class="sd">        Alternatively, the `all_files` bool argument can be provided to</span>
<span class="sd">        set all alignments as &#39;active&#39;.</span>

<span class="sd">        After updating the `alignments`and `shelve_alignments` attributes,</span>
<span class="sd">        the `taxa_names` attribute is also updated.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        aln_list : list</span>
<span class="sd">            List of `Alignment.name` that will be set as &#39;active&#39;.</span>
<span class="sd">        all_files : bool</span>
<span class="sd">            If True, all alignments become active. Ignores `aln_list`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">all_files</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span>

        <span class="k">elif</span> <span class="n">aln_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

                <span class="k">if</span> <span class="n">aln</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aln_list</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span>
                        <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span>
                        <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln</span><span class="p">]</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="k">pass</span>

        <span class="c1"># Update taxa names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.update_active_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.update_active_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">update_active_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aln_name</span><span class="p">,</span> <span class="n">direction</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Updates the &#39;active&#39; status of a single alignment.</span>

<span class="sd">        Changes the &#39;active&#39; status of a single alignment in a specified</span>
<span class="sd">        direction.</span>

<span class="sd">        After updating the `alignments`and `shelve_alignments` attributes,</span>
<span class="sd">        the `taxa_names` attribute is also updated.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        aln_name : str</span>
<span class="sd">            `Alignment.name` to update.</span>
<span class="sd">        direction : str</span>
<span class="sd">            Direction where `aln_name` will be moved. Can be</span>
<span class="sd">            {&quot;shelve&quot;, &quot;active&quot;}.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">direction</span> <span class="o">==</span> <span class="s2">&quot;shelve&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">aln_name</span><span class="p">]</span>

        <span class="c1"># Update taxa names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.update_taxa_names"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.update_taxa_names">[docs]</a>    <span class="k">def</span> <span class="nf">update_taxa_names</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">all_taxa</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Sets the active taxa.</span>

<span class="sd">        Sets the &#39;active&#39; taxa stored in the `taxa_names` attribute and the</span>
<span class="sd">        &#39;inactive&#39; taxa stored in the `shelved_taxa` attribute, based on</span>
<span class="sd">        the `taxa_list` argument.</span>

<span class="sd">        Alternatively, the `all_taxa` bool argument can be provided to</span>
<span class="sd">        set all taxa as &#39;active&#39;.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxa_list : list</span>
<span class="sd">            List of taxon names that will be set as &#39;active&#39;.</span>
<span class="sd">        all_taxa : bool</span>
<span class="sd">            If True, all alignments become active. Ignores `taxa_list`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Activate all taxa</span>
        <span class="k">if</span> <span class="n">all_taxa</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Activate only taxa specified by taxa_list</span>
        <span class="k">elif</span> <span class="n">taxa_list</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxa_list</span> <span class="ow">and</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="k">pass</span>

                <span class="k">elif</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">taxa_list</span> <span class="ow">and</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="k">pass</span>

        <span class="c1"># Update individual Alignment objects</span>
        <span class="k">for</span> <span class="n">aln_obj</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">aln_obj</span><span class="o">.</span><span class="n">shelve_taxa</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shelved_taxa</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.format_list"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.format_list">[docs]</a>    <span class="k">def</span> <span class="nf">format_list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns list of unique sequence types from `Alignment` objects.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : list</span>
<span class="sd">            List with unique sequence types for `Alignment` objects as</span>
<span class="sd">            strings.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                         <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="p">]))</span></div>

<div class="viewcode-block" id="AlignmentList._get_taxa_list"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._get_taxa_list">[docs]</a>    <span class="k">def</span> <span class="nf">_get_taxa_list</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">only_active</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the global taxa list from all alignments.</span>

<span class="sd">        If called with no arguments, it returns the taxa list from *all*</span>
<span class="sd">        alignments, including the &#39;inactive&#39; ones. The `only_active`</span>
<span class="sd">        argument can be set to True to return only the taxa list from the</span>
<span class="sd">        &#39;active&#39; files.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        only_active : bool</span>
<span class="sd">            If True, only returns the taxa list from the &#39;active&#39; alignments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        full_taxa : list</span>
<span class="sd">            List with taxon names as strings.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">only_active</span><span class="p">:</span>
            <span class="n">full_taxa</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">()</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">taxa_list</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()]))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">full_taxa</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">()</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">taxa_list</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">+</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()]))</span>

        <span class="k">return</span> <span class="n">full_taxa</span></div>

<div class="viewcode-block" id="AlignmentList._get_filename_list"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._get_filename_list">[docs]</a>    <span class="k">def</span> <span class="nf">_get_filename_list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns list with the `Alignment.name` of alignments.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : list</span>
<span class="sd">            List of `Alignment.name` of alignments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">alignment</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span></div>

<div class="viewcode-block" id="AlignmentList.set_partition_from_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.set_partition_from_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">set_partition_from_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Updates `partitions` object with the provided `Alignment` object.</span>

<span class="sd">        Uses the `partitions` of an `Alignment` object to set the `partitions`</span>
<span class="sd">        attribute of `AlignmentList`</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        alignment_obj : trifusion.process.sequence.Alignment</span>
<span class="sd">            `Alignment` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Update partitions object</span>
        <span class="c1"># NOTE: The use_counter argument is set to False here so that when</span>
        <span class="c1"># the locus_range is provided, the counter does not interfere with the</span>
        <span class="c1"># ranges.</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">is_single</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">partitions</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span>
                    <span class="n">k</span><span class="p">,</span> <span class="n">locus_range</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">codon</span><span class="o">=</span><span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="n">use_counter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                    <span class="n">model_cls</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">add_partition</span><span class="p">(</span>
                <span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">use_counter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">length</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                <span class="n">model_cls</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">])</span></div>

<div class="viewcode-block" id="AlignmentList.add_alignments"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.add_alignments">[docs]</a>    <span class="k">def</span> <span class="nf">add_alignments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alignment_obj_list</span><span class="p">,</span> <span class="n">ignore_paths</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add a list of `Alignment` objects to the current `AlignmentList`.</span>

<span class="sd">        Adds a list of alignments, already as `Alignment` objects, to the</span>
<span class="sd">        current `AlignmentList`. This method performs several checks</span>
<span class="sd">        to assure that the `Alignment` objects being added are correct. If</span>
<span class="sd">        all checks out, updates all attributes of the class with the new</span>
<span class="sd">        alignments.</span>

<span class="sd">        The `ignore_paths` argument can be used to prevent `Alignment`</span>
<span class="sd">        objects with short paths (only basename, for instance) from being</span>
<span class="sd">        added if they have the same basename and another in the `alignments`</span>
<span class="sd">        dict.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        alignment_obj_list : list</span>
<span class="sd">            List with `Alignment` objects.</span>
<span class="sd">        ignore_paths : bool</span>
<span class="sd">            If True, the `Alignment.path` of the new alignments are not</span>
<span class="sd">            checked with the loaded alignments.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="n">alignment_obj_list</span><span class="p">:</span>

            <span class="c1"># Check for badly formatted alignments</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">e</span><span class="p">,</span> <span class="n">InputError</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bad_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">e</span><span class="p">,</span>
                            <span class="n">AlignmentUnequalLength</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">non_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

            <span class="c1"># if not ignore_paths:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ignore_paths</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">path</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                          <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Get seq code</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">sequence_code</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">]</span> <span class="o">=</span> <span class="n">alignment_obj</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">set_partition_from_alignment</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Get seq code</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">sequence_code</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">alignment_obj</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">set_partition_from_alignment</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>
                <span class="c1"># self.filename_list.append(alignment_obj.name)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.add_alignment_files"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.add_alignment_files">[docs]</a>    <span class="k">def</span> <span class="nf">add_alignment_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_name_list</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Adds a list of alignment files to the current `AlignmentList`.</span>

<span class="sd">        Adds a list of alignment paths to the current `AlignmentList`. Each</span>
<span class="sd">        path in the list is used to instantiate an `Alignment` object and</span>
<span class="sd">        several checks are performed to ensure that the alignments are</span>
<span class="sd">        correct and compliant with the other members of the `AlignmentList`</span>
<span class="sd">        object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file_name_list : list</span>
<span class="sd">            List with paths to sequence alignment files.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check for duplicates among current file list</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">Counter</span><span class="p">(</span><span class="n">file_name_list</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="n">file_name_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

        <span class="c1"># Check for duplicates between current file list and previous file list</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path_list</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">file_name_list</span><span class="p">)):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">file_name_list</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
            <span class="n">pbar</span><span class="o">.</span><span class="n">max_value</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_name_list</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">aln_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">file_name_list</span><span class="p">):</span>

            <span class="c1"># Progress bar update for command line version</span>
            <span class="k">if</span> <span class="n">pbar</span><span class="p">:</span>
                <span class="n">pbar</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">m</span> <span class="o">=</span> <span class="s2">&quot;Processing file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">aln_path</span><span class="p">))</span>

                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;Child thread killed by user&quot;</span><span class="p">)</span>

            <span class="n">aln_obj</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">aln_path</span><span class="p">,</span> <span class="n">sql_cursor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="p">,</span>
                                <span class="n">sql_con</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">e</span><span class="p">,</span> <span class="n">InputError</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bad_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">e</span><span class="p">,</span> <span class="n">AlignmentUnequalLength</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">non_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">e</span><span class="p">,</span> <span class="n">EmptyAlignment</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bad_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Get seq code</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span> <span class="o">=</span> <span class="n">aln_obj</span><span class="o">.</span><span class="n">sequence_code</span>
                <span class="c1"># Check for multiple sequence types. If True,</span>
                <span class="c1"># raise Exception</span>
                <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">aln_obj</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="k">raise</span> <span class="n">MultipleSequenceTypes</span><span class="p">(</span><span class="s2">&quot;Multiple sequence &quot;</span>
                        <span class="s2">&quot;types detected: </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">aln_obj</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">]</span> <span class="o">=</span> <span class="n">aln_obj</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">set_partition_from_alignment</span><span class="p">(</span><span class="n">aln_obj</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">path_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.retrieve_alignment"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.retrieve_alignment">[docs]</a>    <span class="k">def</span> <span class="nf">retrieve_alignment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return `Alignment` object with a given `name`.</span>

<span class="sd">        Retrieves the `Alignment` object with a matching `name` attribute</span>
<span class="sd">        from `alignments` or `shelve_alignments`. Returns None if no match</span>
<span class="sd">        is found.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        name : str</span>
<span class="sd">            `name` attribute of an `Alignment` object.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : trifusion.process.sequence.Alignment</span>
<span class="sd">            `Alignment` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="AlignmentList.iter_alignment_files"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.iter_alignment_files">[docs]</a>    <span class="k">def</span> <span class="nf">iter_alignment_files</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns an iterable on `Alignment.path` of each alignment.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : iter</span>
<span class="sd">            Iterable over the `Alignment.path` attribute of each alignment.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">path</span> <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">())</span></div>

<div class="viewcode-block" id="AlignmentList.write_taxa_to_file"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.write_taxa_to_file">[docs]</a>    <span class="k">def</span> <span class="nf">write_taxa_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_name</span><span class="o">=</span><span class="s2">&quot;Taxa_list.csv&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes the taxa in `taxa_list` to a file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file_name : string</span>
<span class="sd">            Path to the generated file (default is &#39;Taxa_list.csv&#39;).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">:</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">taxon</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.concatenate"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.concatenate">[docs]</a>    <span class="k">def</span> <span class="nf">concatenate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Concatenates alignments into a single `Alignment` object.</span>

<span class="sd">        Concatenates multiple sequence alignments creating a single `Alignment`</span>
<span class="sd">        object and the auxiliary Partitions object defining the partitions</span>
<span class="sd">        of the concatenated alignment</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        concatenated_alignment : trifusion.process.sequence.Alignment</span>
<span class="sd">            The `Alignment` object with the concatenated data and partitions.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        Since sqlite queries are quite expensive, the previous approach</span>
<span class="sd">        of querying the sequence for each taxa AND alignment table was</span>
<span class="sd">        dropped. In this approach, sqlite will do all the heavy lifting.</span>
<span class="sd">        First, a temporary table with the same definition as all alignment</span>
<span class="sd">        tables is populated with data from all taxa and alignments. It is</span>
<span class="sd">        important that and index is created on the new txId column, which</span>
<span class="sd">        will redifine the txId of individual alignments according to the</span>
<span class="sd">        global `taxa_names` attribute. In the first procedure, there</span>
<span class="sd">        will be only one query per alignment. When the temporary table is</span>
<span class="sd">        complete, some sql operations are used to group sequences from each</span>
<span class="sd">        taxon for all alignments and then concatenated them, all in a single</span>
<span class="sd">        query. This query returns the concatenated sequences, corrected txId</span>
<span class="sd">        and taxon information, ready to populate the final concatenation</span>
<span class="sd">        table. This approach is an order o magnitude faster than the</span>
<span class="sd">        previous one where thousands of queries could be performed.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="c1"># Create table temporary table and create an index</span>
        <span class="n">temp_table</span> <span class="o">=</span> <span class="s2">&quot;concatenation_temp&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="n">temp_table</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;concidex&quot;</span><span class="p">,</span> <span class="s2">&quot;txId&quot;</span><span class="p">])</span>

        <span class="c1"># Variables that will store the taxa_list and taxa_idx that will be</span>
        <span class="c1"># provided when instantiating an Alignment object</span>
        <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span>
        <span class="n">taxa_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">tx</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">tx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))</span>

        <span class="c1"># Create new cursor to insert data into database while the main</span>
        <span class="c1"># cursor returns the query</span>
        <span class="n">conc_cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">aln</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="c1"># Update progress information</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Processing alignment </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>

            <span class="c1"># For each taxon present in the current Alignment object,</span>
            <span class="c1"># update the txId in the new table using the taxa_idx local</span>
            <span class="c1"># variable.</span>
            <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_alignment</span><span class="p">(</span><span class="n">table_suffix</span><span class="o">=</span><span class="n">table_in</span><span class="p">):</span>

                <span class="n">conc_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">temp_table</span><span class="p">),</span> <span class="p">(</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

            <span class="c1"># Get all the taxa missing from the current Alignment object</span>
            <span class="n">missing_tx</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span>

            <span class="c1"># Insert missing data for the missing taxa in the teporary database</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">missing_tx</span><span class="p">:</span>

                <span class="n">conc_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">temp_table</span><span class="p">),</span>
                    <span class="p">(</span><span class="n">taxa_idx</span><span class="p">[</span><span class="n">tx</span><span class="p">],</span> <span class="n">tx</span><span class="p">,</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>

        <span class="c1"># Set the name of the final concatenated table</span>
        <span class="n">table</span> <span class="o">=</span> <span class="s2">&quot;concatenation&quot;</span>

        <span class="c1"># Create the table</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">active_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>

        <span class="c1"># Reset progress information for next loop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))</span>

        <span class="c1"># Here is where sqlite will do the heavy lifting of grouping sequences</span>
        <span class="c1"># from the same taxon across all alignments and then concatenating</span>
        <span class="c1"># them. It is important that the column that will be grouped has</span>
        <span class="c1"># an index (for perfomance and memory reasons). While seqs are</span>
        <span class="c1"># grouped by txId, the GROUP_CONCAT() method is used to return</span>
        <span class="c1"># concatenated strings.</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
            <span class="s2">&quot;SELECT txId, taxon, GROUP_CONCAT(seq, &#39;&#39;) FROM [</span><span class="si">{}</span><span class="s2">] &quot;</span>
            <span class="s2">&quot;GROUP BY txId&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">temp_table</span><span class="p">))):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Concatenating taxon </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tx</span><span class="p">))</span>

            <span class="n">conc_cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">table</span><span class="p">),</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">tx</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>

        <span class="c1"># Variable that will store the length of the concatenated alignment</span>
        <span class="c1"># and provided it when initializing the Alignment object</span>
        <span class="n">locus_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

        <span class="c1"># DROP the temporary table</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;DROP TABLE [</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">temp_table</span><span class="p">))</span>

        <span class="c1"># Removes partitions that are currently in the shelve</span>
        <span class="k">for</span> <span class="n">aln_obj</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span><span class="n">file_name</span><span class="o">=</span><span class="n">aln_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">PartitionException</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="c1"># Create the concatenated file in an Alignment object</span>
        <span class="n">concatenated_alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">table</span><span class="p">,</span>
                                           <span class="n">partitions</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="p">,</span>
                                           <span class="n">locus_length</span><span class="o">=</span><span class="n">locus_length</span><span class="p">,</span>
                                           <span class="n">sql_cursor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="p">,</span>
                                           <span class="n">sql_con</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="p">,</span>
                                           <span class="n">sequence_code</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">,</span>
                                           <span class="n">taxa_list</span><span class="o">=</span><span class="n">taxa_list</span><span class="p">,</span>
                                           <span class="n">taxa_idx</span><span class="o">=</span><span class="n">taxa_idx</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">concatenated_alignment</span></div>

<div class="viewcode-block" id="AlignmentList.filter_min_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_min_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">filter_min_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_taxa</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters `alignments` by minimum taxa proportion.</span>

<span class="sd">        Filters `Alignment` objects based on a minimum taxa representation</span>
<span class="sd">        threshold. Alignments with less that the specified minimum taxa</span>
<span class="sd">        percentage will be moved to the `filtered_alignments` attribute.</span>

<span class="sd">        NOTE: Since this filtering is meant to be performed when executing</span>
<span class="sd">        the process operations, it will permanently change the AlignmentList</span>
<span class="sd">        object, which means both self.alignments and self.partitions. Not doing</span>
<span class="sd">        so and removing/adding the partitions would create a great deal of</span>
<span class="sd">        conflicts that can be easily avoided by simply copying the</span>
<span class="sd">        AlignmentList object and modifying this object for the process</span>
<span class="sd">        execution</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_taxa : int</span>
<span class="sd">            Percentage of minimum allowed taxa representation (e.g. 50).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By minimum taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Evaluating file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">&lt;</span> \
                    <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">min_taxa</span><span class="p">)</span> <span class="o">/</span> <span class="mf">100.</span><span class="p">)</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignment</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;shelve&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span><span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By minimum taxa&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.filter_by_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_by_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">filter_by_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">filter_mode</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters `alignments` if they contain or exclude certain taxa.</span>

<span class="sd">        Filters the `alignments` attribute by a given taxa list.</span>
<span class="sd">        The filtering may be performed to filter alignments</span>
<span class="sd">        depending on whether they include or exclude certain taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxa_list : list</span>
<span class="sd">            List of taxa names.</span>
<span class="sd">        filter_mode : str</span>
<span class="sd">            Filter mode. Can be &quot;contain&quot; or &quot;exclude&quot;.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Support automatic file detection if taxa_list is a string</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span>
                <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_basic_csv</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
                <span class="k">pass</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_basic_csv</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="c1"># Filter alignments that do not contain at least all taxa in</span>
            <span class="c1"># taxa_list</span>
            <span class="k">if</span> <span class="n">filter_mode</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;contain&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">():</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignment</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;shelve&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span>
                        <span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By taxa&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Filter alignments that contain the taxa in taxa list</span>
            <span class="k">if</span> <span class="n">filter_mode</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;exclude&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">any</span><span class="p">((</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">taxa_list</span>
                        <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)):</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignment</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;shelve&quot;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span>
                        <span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By taxa&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">(</span><span class="n">only_active</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># If the resulting alignment is empty, raise an Exception</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span> <span class="o">==</span> <span class="p">{}:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">exception</span> <span class="o">=</span> <span class="s2">&quot;EmptyAlignment&quot;</span>
            <span class="k">raise</span> <span class="n">EmptyAlignment</span><span class="p">(</span><span class="s2">&quot;Alignment is empty after taxa filter&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.filter_codon_positions"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_codon_positions">[docs]</a>    <span class="k">def</span> <span class="nf">filter_codon_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters codon positions in each `Alignment` object.</span>

<span class="sd">        This wraps the execution of `filter_codon_positions` method for</span>
<span class="sd">        each `Alignment` object in the `alignments` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        position_list : list</span>
<span class="sd">            List of three bool elements that correspond to each codon position.</span>
<span class="sd">            Ex. [True, True, True] will save all positions while</span>
<span class="sd">            [True, True, False] will exclude the third codon position</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;filter&quot;).</span>
<span class="sd">        use_main_table : bool</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.filter_codon_positions</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="c1"># Reset partitions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">Partitions</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">filter_codon_positions</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">set_partition_from_alignment</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.filter_missing_data"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_missing_data">[docs]</a>    <span class="k">def</span> <span class="nf">filter_missing_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters missing data in each `Alignment` object.</span>

<span class="sd">        This wraps the execution of `filter_missing_data` method for</span>
<span class="sd">        each `Alignment` object in the `alignments` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gap_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap percentage is removed.</span>
<span class="sd">        missing_threshold : int</span>
<span class="sd">            Integer between 0 and 100 defining the percentage above which</span>
<span class="sd">            a column with that gap+missing percentage is removed.</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;filter&quot;).</span>
<span class="sd">        use_main_table : bool</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.filter_missing_data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="c1"># Reset partitions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span> <span class="o">=</span> <span class="n">Partitions</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">filter_missing_data</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">set_partition_from_alignment</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.filter_segregating_sites"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_segregating_sites">[docs]</a>    <span class="k">def</span> <span class="nf">filter_segregating_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters `Alignment` objects according to segregating sites number.</span>

<span class="sd">        Filters `alignments` according to whether or not their number</span>
<span class="sd">        of segregating sites is within a specified range. `Alignment` objects</span>
<span class="sd">        with a number of segregating sites outside the range are move to</span>
<span class="sd">        the `shelve_alignments` and the counter of `filtered_alignments` is</span>
<span class="sd">        updated.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_val : int</span>
<span class="sd">            Minimum number of segregating sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no lower bound.</span>
<span class="sd">        max_val : int</span>
<span class="sd">            Maximum number of segregating sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no upper bound.</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.filter_segregating_sites</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By variable sites&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">filter_segregating_sites</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignment</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;shelve&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span><span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By variable sites&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.filter_informative_sites"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.filter_informative_sites">[docs]</a>    <span class="k">def</span> <span class="nf">filter_informative_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters `Alignment` objects according to informative sites number.</span>

<span class="sd">        Filters `alignments` according to whether or not their number</span>
<span class="sd">        of informative sites is within a specified range. `Alignment` objects</span>
<span class="sd">        with a number of informative sites outside the range are move to</span>
<span class="sd">        the `shelve_alignments` and the counter of `filtered_alignments` is</span>
<span class="sd">        updated.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        min_val : int</span>
<span class="sd">            Minimum number of informative sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no lower bound.</span>
<span class="sd">        max_val : int</span>
<span class="sd">            Maximum number of informative sites for the alignment to pass.</span>
<span class="sd">            Can be None, in which case there is no upper bound.</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.filter_informative_sites</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By informative sites&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Filtering file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">filter_informative_sites</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignment</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s2">&quot;shelve&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span><span class="n">file_name</span><span class="o">=</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filtered_alignments</span><span class="p">[</span><span class="s2">&quot;By informative sites&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.remove_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.remove_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">remove_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;remove&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Removes the specified taxa.</span>

<span class="sd">        Removes the specified taxa from the global `AlignmentList` attributes</span>
<span class="sd">        and from each `Alignment` object.</span>

<span class="sd">        This method supports a list or path to CSV file as the</span>
<span class="sd">        `taxa_list_file` argument. In case of CSV path, it parses the CSV</span>
<span class="sd">        file into a list. The CSV file should be a simple text file containing</span>
<span class="sd">        taxon names in each line.</span>

<span class="sd">        This method also supports two removal modes.</span>
<span class="sd">            - remove: removes the specified taxa</span>
<span class="sd">            - inverse: removes all but the specified taxa</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxa_list_file : list or string</span>
<span class="sd">            List containing taxa names or string with path to a CSV file</span>
<span class="sd">            containig the taxa names.</span>
<span class="sd">        mode : {&quot;remove&quot;, &quot;inverse&quot;}</span>
<span class="sd">            The removal mode (default is remove).</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.remove_taxa</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Checking if taxa_list is an input csv file:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_basic_csv</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>

        <span class="c1"># If not, then the method&#39;s argument is already the final list</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">IOError</span><span class="p">,</span> <span class="ne">IndexError</span><span class="p">):</span>
            <span class="k">pass</span>

        <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">)</span>

        <span class="c1"># Updates taxa names</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;remove&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="c1"># TODO: log a warning</span>
                    <span class="k">pass</span>
        <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;inverse&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">tx</span> <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">taxa_list</span> <span class="k">if</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">]</span></div>

<div class="viewcode-block" id="AlignmentList.change_taxon_name"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.change_taxon_name">[docs]</a>    <span class="k">def</span> <span class="nf">change_taxon_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Changes the name of a taxon. &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">change_taxon_name</span><span class="p">(</span><span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">)</span>

        <span class="c1"># update taxa names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_name</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">old_name</span> <span class="k">else</span> <span class="n">x</span>
                           <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">]</span></div>

<div class="viewcode-block" id="AlignmentList.remove_file"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.remove_file">[docs]</a>    <span class="k">def</span> <span class="nf">remove_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Removes alignments.</span>

<span class="sd">        Removes `Alignment` objects from `alignments` and `shelve_alignments`</span>
<span class="sd">        based on their `name` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        filename_list : list</span>
<span class="sd">            List of `Alignment.name` to be removed.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># If filename_list corresponds to all files in the current alignment</span>
        <span class="c1"># list, dispatch the clear_alignments methods</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">path_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">filename_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">([]):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">clear_alignments</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">nm</span> <span class="ow">in</span> <span class="n">filename_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">nm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">nm</span><span class="p">]</span><span class="o">.</span><span class="n">remove_alignment</span><span class="p">()</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">[</span><span class="n">nm</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">nm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">nm</span><span class="p">]</span><span class="o">.</span><span class="n">remove_alignment</span><span class="p">()</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">shelve_alignments</span><span class="p">[</span><span class="n">nm</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">remove_partition</span><span class="p">(</span><span class="n">file_name</span><span class="o">=</span><span class="n">nm</span><span class="p">)</span>
            <span class="k">except</span> <span class="n">PartitionException</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="c1"># Updates taxa names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_taxa_list</span><span class="p">()</span></div>

<div class="viewcode-block" id="AlignmentList.select_by_taxa"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.select_by_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">select_by_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;strict&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Selects a list of `Alignment` objects by taxa.</span>

<span class="sd">        This method is used to select `Alignments` based on whether they</span>
<span class="sd">        contain or exclude certain taxa. The selected alignments are</span>
<span class="sd">        returned in a list.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        taxa_list : list</span>
<span class="sd">            List of taxa names.</span>
<span class="sd">        mode : str</span>
<span class="sd">            Selection mode. Can be:</span>

<span class="sd">                - &quot;strict&quot;: Selects alignments containing all and only the</span>
<span class="sd">                provide taxa.</span>
<span class="sd">                - &quot;inclusive&quot;: Selects alignments containing the provided</span>
<span class="sd">                taxa.</span>
<span class="sd">                - &quot;relaxed&quot;: Selects alignments if they contain at least</span>
<span class="sd">                one of the provided taxa.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        selected_alignments : list</span>
<span class="sd">            List of `Alignment` objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">selected_alignments</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># taxa_list may be a file name (string) or a list containing the name</span>
        <span class="c1"># of the taxa. If taxa_list is a file name this code will parse the</span>
        <span class="c1"># csv file and return a list of the taxa. Otherwise, the taxa_list</span>
        <span class="c1"># variable remains the same.</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">))</span>
            <span class="n">taxa_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_basic_csv</span><span class="p">(</span><span class="n">file_handle</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">EnvironmentError</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="n">alignment_taxa</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">taxa_list</span>

            <span class="c1"># Selected only the alignments with the exact same taxa</span>
            <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;strict&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">alignment_taxa</span><span class="p">):</span>
                    <span class="n">selected_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>

            <span class="c1"># Selected alignments that include the specified taxa</span>
            <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;inclusive&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">alignment_taxa</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">():</span>
                    <span class="n">selected_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;relaxed&quot;</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">alignment_taxa</span><span class="p">:</span>
                        <span class="n">selected_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">)</span>
                        <span class="k">continue</span>

        <span class="k">return</span> <span class="n">selected_alignments</span></div>

<div class="viewcode-block" id="AlignmentList.code_gaps"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.code_gaps">[docs]</a>    <span class="k">def</span> <span class="nf">code_gaps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Code gaps in each `Alignment` object.</span>

<span class="sd">        This wraps the execution of `code_gaps` method for</span>
<span class="sd">        each `Alignment` object in the `alignments` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;gaps&quot;).</span>
<span class="sd">        use_main_table : bool</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.code_gaps</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Coding file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">code_gaps</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.collapse"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.collapse">[docs]</a>    <span class="k">def</span> <span class="nf">collapse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Collapses equal sequences for each `Alignment` object.</span>

<span class="sd">        This wraps the execution of `collapse` method for</span>
<span class="sd">        each `Alignment` object in the `alignments` attribute.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        write_haplotypes : bool</span>
<span class="sd">            If True, a file mapping the taxa names name to their respective</span>
<span class="sd">            haplotype name will be generated (default is True).</span>
<span class="sd">        haplotypes_file : string</span>
<span class="sd">            Name of the file mapping taxa names to their respective haplotype.</span>
<span class="sd">            Only used when `write_haplotypes` is True. If it not provided and</span>
<span class="sd">            `write_haplotypes` is True, the file name will be determined</span>
<span class="sd">            from `Alignment.name`.</span>
<span class="sd">        dest : string</span>
<span class="sd">            Path of directory where `haplotypes_file` will be generated</span>
<span class="sd">            (default is &quot;.&quot;).</span>
<span class="sd">        conversion_suffix : string</span>
<span class="sd">            Suffix appended to the haplotypes file. Only used when</span>
<span class="sd">            `write_haplotypes` is True and `haplotypes_file` is None.</span>
<span class="sd">        haplotype_name : string</span>
<span class="sd">            Prefix of the haplotype string. The final haplotype name will be</span>
<span class="sd">            `haplotype_name` + &lt;int&gt; (default is &quot;Hap&quot;).</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;collapsed&quot;).</span>
<span class="sd">        use_main_table : bool</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.collapse</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="n">write_haplotypes</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;write_haplotypes&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">conversion_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;conversion_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">haplotypes_file</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;haplotypes_file&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Collapsing file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="k">if</span> <span class="n">write_haplotypes</span><span class="p">:</span>
                <span class="c1"># Set name for haplotypes file</span>
                <span class="n">output_file</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                              <span class="n">conversion_suffix</span> <span class="o">+</span> <span class="n">haplotypes_file</span>
                <span class="n">alignment_obj</span><span class="o">.</span><span class="n">collapse</span><span class="p">(</span><span class="n">haplotypes_file</span><span class="o">=</span><span class="n">output_file</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span>
                                       <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">alignment_obj</span><span class="o">.</span><span class="n">collapse</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.consensus"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.consensus">[docs]</a>    <span class="k">def</span> <span class="nf">consensus</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates consensus sequences for each alignment.</span>

<span class="sd">        This wraps the execution of `consensus` method for</span>
<span class="sd">        each `Alignment` object in the `alignments` attribute. It has an</span>
<span class="sd">        additional argument, `single_file`, which can be used to merge</span>
<span class="sd">        the consensus sequence from each alignment into a new `Alignment`</span>
<span class="sd">        object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        consensus_type : {&quot;IUPAC&quot;, &quot;Soft mask&quot;, &quot;Remove&quot;, &quot;First sequence&quot;}</span>
<span class="sd">            Type of variation handling. See summary above.</span>
<span class="sd">        single_file : bool</span>
<span class="sd">            If True, the consensus sequence of each `Alignment` object will</span>
<span class="sd">            be merged into a single `Alignment` object.</span>
<span class="sd">        table_in : string</span>
<span class="sd">            Name of database table containing the alignment data that is</span>
<span class="sd">            used for this operation.</span>
<span class="sd">        table_out : string</span>
<span class="sd">            Name of database table where the final alignment will be inserted</span>
<span class="sd">            (default is &quot;consensus&quot;).</span>
<span class="sd">        use_main_table : bool</span>
<span class="sd">            If True, both `table_in` and `table_out` are ignore and the main</span>
<span class="sd">            table `Alignment.table_name` is used as the input and output</span>
<span class="sd">            table (default is False).</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        consensus_aln : trifusion.process.sequence.Alignment</span>
<span class="sd">            Alignment object with all the consensus sequences merged. It&#39;s</span>
<span class="sd">            only returned when the `single_file` parameter is set to True.</span>
<span class="sd">        </span>
<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        Alignment.consensus</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">single_file</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;single_file&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;consensus&quot;</span><span class="p">]</span>

        <span class="c1"># Variables that will store the taxa_list and taxa_idx to provide</span>
        <span class="c1"># to the Alignment object</span>
        <span class="n">taxa_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">taxa_idx</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">if</span> <span class="n">single_file</span><span class="p">:</span>
            <span class="c1"># Create a table that will harbor the consensus sequences of all</span>
            <span class="c1"># Alignment objects</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_create_table</span><span class="p">(</span><span class="s2">&quot;consensus&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">active_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;consensus&quot;</span><span class="p">)</span>
            <span class="n">consensus_data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Processing file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">consensus</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">single_file</span><span class="p">:</span>
                <span class="n">table_out</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_out&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                <span class="n">sequence</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="s2">&quot;consensus&quot;</span><span class="p">,</span>
                                                      <span class="n">table_suffix</span><span class="o">=</span><span class="n">table_out</span><span class="p">)</span>
                <span class="n">consensus_data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">sname</span><span class="p">,</span> <span class="n">sequence</span><span class="p">))</span>
                <span class="n">taxa_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">sname</span><span class="p">)</span>
                <span class="n">taxa_idx</span><span class="p">[</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">sname</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">single_file</span><span class="p">:</span>

            <span class="n">use_main_table</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;use_main_table&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">use_main_table</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">remove_tables</span><span class="p">(</span><span class="n">preserve_tables</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;consensus&quot;</span><span class="p">])</span>

            <span class="c1"># Populate database table</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="o">.</span><span class="n">executemany</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="s2">&quot;consensus&quot;</span><span class="p">),</span> <span class="n">consensus_data</span><span class="p">)</span>

            <span class="c1"># Create Alignment object</span>
            <span class="n">consensus_aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="s2">&quot;consensus&quot;</span><span class="p">,</span> <span class="n">sql_cursor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">cur</span><span class="p">,</span>
                                      <span class="n">sql_con</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="p">,</span>
                                      <span class="n">sequence_code</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">,</span>
                                      <span class="n">taxa_list</span><span class="o">=</span><span class="n">taxa_list</span><span class="p">,</span>
                                      <span class="n">taxa_idx</span><span class="o">=</span><span class="n">taxa_idx</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">consensus_aln</span></div>

<div class="viewcode-block" id="AlignmentList.reverse_concatenate"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.reverse_concatenate">[docs]</a>    <span class="k">def</span> <span class="nf">reverse_concatenate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Reverse a concatenated file according to the partitions.</span>

<span class="sd">        This is basically a wrapper of the `reverse_concatenate` method</span>
<span class="sd">        of `Alignment` and is rarely useful. It is preferable to retrieve</span>
<span class="sd">        an `Alignment` object, provide some partitions and then</span>
<span class="sd">        perform the reverse concatenation. Since the reverse concatenation</span>
<span class="sd">        operation can only be applied to `Alignment` objects, this method</span>
<span class="sd">        actually creates a concatenated `Alignment` .</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        reverted_alns : AlignmentList</span>
<span class="sd">            AlignmentList with the partitions as `Alignment` objects.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">concatenated_aln</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">alignment_name</span><span class="o">=</span><span class="s2">&quot;concatenated&quot;</span><span class="p">)</span>

        <span class="n">reverted_alns</span> <span class="o">=</span> <span class="n">concatenated_aln</span><span class="o">.</span><span class="n">reverse_concatenate</span><span class="p">(</span><span class="n">db_con</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">con</span><span class="p">,</span>
                                                             <span class="n">ns</span><span class="o">=</span><span class="n">ns</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">reverted_alns</span></div>

<div class="viewcode-block" id="AlignmentList.write_to_file"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.write_to_file">[docs]</a>    <span class="k">def</span> <span class="nf">write_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_format</span><span class="p">,</span> <span class="n">conversion_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                      <span class="n">output_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes `Alignment` objects into files.</span>

<span class="sd">        Wrapper of the `write_to_file` method of the `Alignment` object.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_format : list</span>
<span class="sd">            List with the output formats to generate. Options are:</span>
<span class="sd">            {&quot;fasta&quot;, &quot;phylip&quot;, &quot;nexus&quot;, &quot;stockholm&quot;, &quot;gphocs&quot;, &quot;ima2&quot;}.</span>
<span class="sd">        output_file : str</span>
<span class="sd">            Name of the output file. If using `ns_pipe` in TriFusion,</span>
<span class="sd">            it will prompt the user if the output file already exists.</span>
<span class="sd">        tx_space_nex : int</span>
<span class="sd">            Space (in characters) provided for the taxon name in nexus</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        tx_space_phy : int</span>
<span class="sd">            Space (in characters) provided for the taxon name in phylip</span>
<span class="sd">            format (default is 40).</span>
<span class="sd">        tx_space_ima2 : int</span>
<span class="sd">            Space (in characters) provided for the taxon name in ima2</span>
<span class="sd">            format (default is 10).</span>
<span class="sd">        cut_space_nex : int</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            nexus format. Longer  names are truncated (default is 39).</span>
<span class="sd">        cut_space_phy : int</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            phylip format. Longer  names are truncated (default is 39).</span>
<span class="sd">        cut_space_ima2 : int</span>
<span class="sd">            Set the maximum allowed character length for taxon names in</span>
<span class="sd">            ima2 format. Longer  names are truncated (default is 8).</span>
<span class="sd">        interleave : bool</span>
<span class="sd">            Determines whether the output alignment</span>
<span class="sd">            will be in leave (False) or interleave (True) format. Not all</span>
<span class="sd">            output formats support this option.</span>
<span class="sd">        gap : str</span>
<span class="sd">            Symbol for alignment gaps (default is &#39;-&#39;).</span>
<span class="sd">        model_phylip : str</span>
<span class="sd">            Substitution model for the auxiliary partition file of phylip</span>
<span class="sd">            format, compliant with RAxML.</span>
<span class="sd">        outgroup_list : list</span>
<span class="sd">            Specify list of taxon names that will be defined as the outgroup</span>
<span class="sd">            in the nexus output format. This may be useful for analyses with</span>
<span class="sd">            MrBayes or other software that may require outgroups.</span>
<span class="sd">        ima2_params : list</span>
<span class="sd">            A list with the additional information required for the ima2</span>
<span class="sd">            output format. The list should contains the following information:</span>

<span class="sd">               1. (str) path to file containing the species and populations.</span>
<span class="sd">               2. (str) Population tree in newick format, e.g. (0,1):2.</span>
<span class="sd">               3. (str) Mutational model for all alignments.</span>
<span class="sd">               4. (str) inheritance scalar.</span>

<span class="sd">        use_charset : bool</span>
<span class="sd">            If True, partitions from the `Partitions` object will be written</span>
<span class="sd">            in the nexus output format (default is True).</span>
<span class="sd">        partition_file : bool</span>
<span class="sd">            If True, the auxiliary partitions file will be written (default</span>
<span class="sd">            is True).</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            If provided, the output file will be written on the specified</span>
<span class="sd">            directory.</span>
<span class="sd">        phy_truncate_names : bool</span>
<span class="sd">            If True, taxon names in phylip format are truncated to a maximum</span>
<span class="sd">            of 10 characters (default is False).</span>
<span class="sd">        ld_hat : bool</span>
<span class="sd">            If True, the fasta output format will include a first line</span>
<span class="sd">            compliant with the format of LD Hat and will truncate sequence</span>
<span class="sd">            names and sequence length per line accordingly.</span>
<span class="sd">        use_nexus_models : bool</span>
<span class="sd">            If True, writes the partitions charset block in nexus format.</span>
<span class="sd">        ns_pipe : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>
<span class="sd">        pbar : ProgressBar</span>
<span class="sd">            A ProgressBar object used to log the progress of TriSeq execution.</span>
<span class="sd">        table_suffix : string</span>
<span class="sd">            Suffix of the table from where the sequence data is fetched.</span>
<span class="sd">            The suffix is append to `Alignment.table_name`.</span>
<span class="sd">        table_name : string</span>
<span class="sd">            Name of the table from where the sequence data is fetched.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Writting file </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                   <span class="n">basename</span><span class="p">(</span><span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)))</span>

            <span class="n">output_file_name</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                <span class="n">conversion_suffix</span> <span class="o">+</span> <span class="n">output_suffix</span>

            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;output_file&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">output_file_name</span>

            <span class="c1"># Get partition name for current alignment object</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">part_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span>
                             <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions_alignments</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span>
                             <span class="n">alignment_obj</span><span class="o">.</span><span class="n">path</span> <span class="ow">in</span> <span class="n">y</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">part_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span>
                                 <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">partitions_alignments</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                                 <span class="k">if</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">y</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                    <span class="n">part_name</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># Get model from partitions</span>
            <span class="k">if</span> <span class="n">part_name</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="n">part_name</span><span class="p">]</span>
                <span class="n">alignment_obj</span><span class="o">.</span><span class="n">partitions</span><span class="o">.</span><span class="n">set_model</span><span class="p">(</span><span class="n">part_name</span><span class="p">,</span> <span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

            <span class="n">alignment_obj</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">output_format</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns_pipe</span><span class="p">)</span></div>

<div class="viewcode-block" id="AlignmentList.get_gene_table_stats"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.get_gene_table_stats">[docs]</a>    <span class="k">def</span> <span class="nf">get_gene_table_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">active_alignments</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sortby</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                             <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns summary statistics for each `Alignment`.</span>

<span class="sd">        Gets summary statistics for each individual gene for the gene table</span>
<span class="sd">        view of TriFusion. Summary statistics have already been calculated in</span>
<span class="sd">        get_summary_stats, so here we only get the active alignments for</span>
<span class="sd">        showing. Therefore, this function should only be called after</span>
<span class="sd">        get_summary_stats.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        active_alignments : list</span>
<span class="sd">            List of `Alignment.name` objects to show</span>
<span class="sd">        sortby : str</span>
<span class="sd">            Table header used for sorting. Can be {&quot;nsites&quot;, &quot;taxa&quot;, &quot;var&quot;,</span>
<span class="sd">            &quot;inf&quot;, &quot;gap&quot;, &quot;missing&quot;}.</span>
<span class="sd">        ascending : bool</span>
<span class="sd">            If True, the `sortby` column will be sorted in ascending order</span>
<span class="sd">            (default is True).</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        summary_gene_table : pandas.DataFrame</span>
<span class="sd">            DataFrame with summary statistics for each alignment.</span>
<span class="sd">        table : list</span>
<span class="sd">            List with the data for creating .csv tables.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Set table header</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">[[</span><span class="s2">&quot;Gene name&quot;</span><span class="p">,</span> <span class="s2">&quot;Number of sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Number of taxa&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;Variable sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Informative sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Gaps&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;Missing data&quot;</span><span class="p">]]</span>
        <span class="c1"># Table line keys matching summary_gene_table dict values</span>
        <span class="n">tl</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;nsites&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;var&quot;</span><span class="p">,</span> <span class="s2">&quot;inf&quot;</span><span class="p">,</span> <span class="s2">&quot;gap&quot;</span><span class="p">,</span> <span class="s2">&quot;missing&quot;</span><span class="p">]</span>

        <span class="c1"># Update the active alignments</span>
        <span class="k">if</span> <span class="n">active_alignments</span> <span class="ow">and</span> <span class="n">active_alignments</span> <span class="o">!=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignments</span><span class="p">(</span><span class="n">active_alignments</span><span class="p">)</span>

        <span class="c1"># Filter table with active alignments</span>
        <span class="n">summary_gene_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aln_names</span><span class="p">())]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">sortby</span><span class="p">:</span>
            <span class="n">sortby</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">]</span>

        <span class="n">summary_gene_table</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">sortby</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="n">ascending</span><span class="p">,</span>
                                       <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Populate table information</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="o">.</span><span class="n">itertuples</span><span class="p">():</span>
            <span class="c1"># Add table line</span>
            <span class="n">table</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">k</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>

        <span class="k">return</span> <span class="n">summary_gene_table</span><span class="p">,</span> <span class="n">table</span></div>

    <span class="c1"># Stats methods</span>
<div class="viewcode-block" id="AlignmentList.get_summary_stats"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.get_summary_stats">[docs]</a>    <span class="k">def</span> <span class="nf">get_summary_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">active_alignments</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Calculates summary statistics for the &#39;active&#39; alignments.</span>

<span class="sd">        Creates/Updates summary statistics for the active alignments.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        active_alignments : list</span>
<span class="sd">            List of `Alignment` objects for which the summary statistics</span>
<span class="sd">            will be calculated</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        summary_stats : dict</span>
<span class="sd">            Dictionary with the overall summary statistics</span>
<span class="sd">        table : list</span>
<span class="sd">            List with overall summary statistics for creating .csv tables.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Update active alignments if they changed since last update</span>
        <span class="k">if</span> <span class="n">active_alignments</span> <span class="ow">and</span> \
                <span class="n">active_alignments</span> <span class="o">!=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignments</span><span class="p">(</span><span class="n">active_alignments</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="c1"># Set table header for summary_stats</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">[[</span><span class="s2">&quot;Genes&quot;</span><span class="p">,</span> <span class="s2">&quot;Taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Alignment length&quot;</span><span class="p">,</span> <span class="s2">&quot;Gaps&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;Gaps per gene&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing data&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing data per gene&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;Variable sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Variable sites per gene&quot;</span><span class="p">,</span>
                  <span class="s2">&quot;Informative sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Informative sites per gene&quot;</span><span class="p">]]</span>
        <span class="c1"># Table line keys matching summary_stats for table completion</span>
        <span class="n">tl</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">,</span> <span class="s2">&quot;taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;seq_len&quot;</span><span class="p">,</span> <span class="s2">&quot;gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;missing&quot;</span><span class="p">,</span>
              <span class="s2">&quot;avg_missing&quot;</span><span class="p">,</span> <span class="s2">&quot;variable&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_var&quot;</span><span class="p">,</span> <span class="s2">&quot;informative&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_inf&quot;</span><span class="p">]</span>

        <span class="c1"># Reset summary_stats</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_summary_stats</span><span class="p">()</span>

        <span class="c1"># Get number of alignments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="c1"># Get number of taxa</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;taxa&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="c1"># Get statistics that require iteration over alignments</span>
        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">aln</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;Child thread killed by user&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get alignment size info</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;seq_len&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span>

            <span class="n">cur_gap</span><span class="p">,</span> <span class="n">cur_missing</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
            <span class="n">cur_var</span><span class="p">,</span> <span class="n">cur_inf</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

                <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="n">col</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>

                <span class="c1"># Get missing data and gaps</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">col</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;missing&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">cur_missing</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span> <span class="ow">in</span> <span class="n">col</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;gaps&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">cur_gap</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="c1"># Get variability information</span>
                <span class="c1"># Filter missing data</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">]:</span>
                    <span class="k">del</span> <span class="n">col</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

                <span class="c1"># If it&#39;s only missing data, ignore</span>
                <span class="k">if</span> <span class="n">col</span><span class="p">:</span>
                    <span class="c1"># Get variable sites</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">col</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;variable&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">cur_var</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="c1"># If any of the remaining sites is present in more than two</span>
                    <span class="c1"># taxa score the site as informative</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">col</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;informative&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">cur_inf</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get values for current alignment for average calculations</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;avg_gaps&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_gap</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;avg_missing&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_missing</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;avg_var&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_var</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;avg_inf&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_inf</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">aln</span><span class="o">.</span><span class="n">name</span><span class="p">):</span>
                <span class="c1"># Get row information for current gene in dict format</span>
                <span class="n">row_info</span> <span class="o">=</span> <span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                            <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span>
                            <span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">),</span>
                            <span class="n">cur_var</span><span class="p">,</span>
                            <span class="n">cur_inf</span><span class="p">,</span>
                            <span class="n">cur_gap</span><span class="p">,</span>
                            <span class="n">cur_missing</span><span class="p">)</span>
                <span class="c1"># Create DataFrame from dict</span>
                <span class="n">row_dt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">row_info</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="n">p</span><span class="p">],</span>
                                      <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
                <span class="c1"># Add row DF to main DF</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_gene_table</span><span class="p">,</span>
                                                     <span class="n">row_dt</span><span class="p">])</span>

        <span class="c1"># Get average values</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;avg_gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_missing&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_var&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_inf&quot;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span>

        <span class="c1"># Get percentage values for missing data</span>
        <span class="n">total</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;seq_len&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;taxa&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;missing&quot;</span><span class="p">]:</span>
            <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="n">percentage</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="nb">float</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">total</span><span class="p">))</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0:,}</span><span class="s2"> (</span><span class="si">{1}</span><span class="s2">%)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">percentage</span><span class="p">)</span>

        <span class="c1"># Get percentage values for variation</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;variable&quot;</span><span class="p">,</span> <span class="s2">&quot;informative&quot;</span><span class="p">]:</span>
            <span class="n">n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="n">percentage</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="nb">float</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">/</span>
                                <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="s2">&quot;seq_len&quot;</span><span class="p">]))</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0:,}</span><span class="s2"> (</span><span class="si">{1}</span><span class="s2">%)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">percentage</span><span class="p">)</span>

        <span class="c1"># Complete table information</span>
        <span class="n">table</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tl</span><span class="p">])</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_stats</span><span class="p">),</span> <span class="n">table</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.gene_occupancy"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.gene_occupancy">[docs]</a>    <span class="k">def</span> <span class="nf">gene_occupancy</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for gene occupancy plot.</span>

<span class="sd">        Creates data for an interpolation plot to visualize the amount of</span>
<span class="sd">        missing genes in a alignment dataset.</span>

<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">()</span>

            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="mi">1</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">taxa_list</span> <span class="k">else</span> <span class="mi">0</span>
                         <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">])</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Genes&quot;</span><span class="p">,</span> <span class="s2">&quot;Taxa&quot;</span><span class="p">],</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene occupancy&quot;</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.missing_data_distribution"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.missing_data_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">missing_data_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for overall distribution of missing data plot.</span>

<span class="sd">        Creates data for overall distribution of missing data. This will</span>
<span class="sd">        calculate the amount of gaps, missing and actual data.</span>

<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;legend&quot;: mpl Legend object,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">legend</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing data&quot;</span><span class="p">,</span> <span class="s2">&quot;Data&quot;</span><span class="p">]</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">legend</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch to interrupt worker</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="c1"># Add to progress counter</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">gaps_g</span><span class="p">,</span> <span class="n">missing_g</span><span class="p">,</span> <span class="n">data_g</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="c1"># Get gaps</span>
                    <span class="n">gaps</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">))</span>
                    <span class="n">gaps_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gaps</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>
                    <span class="c1"># Get missing data</span>
                    <span class="n">missing</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
                    <span class="n">missing_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">missing</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>
                    <span class="c1"># Get actual data</span>
                    <span class="n">actual_data</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span> <span class="o">-</span> <span class="n">gaps</span> <span class="o">-</span> <span class="n">missing</span><span class="p">)</span> <span class="o">/</span> \
                        <span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>
                    <span class="n">data_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_data</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">gaps_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">missing_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                    <span class="n">data_g</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

            <span class="n">data_storage</span><span class="p">[</span><span class="s2">&quot;Gaps&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">gaps_g</span><span class="p">))</span>
            <span class="n">data_storage</span><span class="p">[</span><span class="s2">&quot;Missing data&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">missing_g</span><span class="p">))</span>
            <span class="n">data_storage</span><span class="p">[</span><span class="s2">&quot;Data&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">data_g</span><span class="p">))</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of missing data&quot;</span><span class="p">,</span>
                <span class="s2">&quot;legend&quot;</span><span class="p">:</span> <span class="n">legend</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion&quot;</span><span class="p">,</span> <span class="s2">&quot;Number of genes&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Bin&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">legend</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.missing_data_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.missing_data_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">missing_data_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for missing data proportion per species plot.</span>

<span class="sd">        For each taxon in `taxon_list` calculate the proportion of gap,</span>
<span class="sd">        missing and actual data.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;legend&quot;: mpl Legend object,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table,</span>

<span class="sd">            &quot;normalize&quot;: bool, whether data values should be normalized,</span>

<span class="sd">            &quot;normalize_factor&quot;: int, factor to use in normalization</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Data for a stacked bar plot. First element for gaps, second for</span>
        <span class="c1"># missing, third for actual data</span>
        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">taxon</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span>
                                   <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>
        <span class="n">total_len</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">legend</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing&quot;</span><span class="p">,</span> <span class="s2">&quot;Data&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch to interrupt worker</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="c1"># Add to progress counter</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">total_len</span> <span class="o">+=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span>
            <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">data_storage</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
                    <span class="c1"># Get gaps</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>
                    <span class="n">gaps</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
                    <span class="n">data_storage</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="n">gaps</span>
                    <span class="c1"># Get missing</span>
                    <span class="n">missing</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                    <span class="n">data_storage</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">missing</span>
                    <span class="c1"># Get actual data</span>
                    <span class="n">actual_data</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">-</span> <span class="n">gaps</span> <span class="o">-</span> <span class="n">missing</span>
                    <span class="n">data_storage</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">+=</span> <span class="n">actual_data</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">data_storage</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
                                          <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                                          <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                          <span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()],</span>
                         <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                          <span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()],</span>
                         <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                          <span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()]])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of missing data per species&quot;</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                <span class="s2">&quot;legend&quot;</span><span class="p">:</span> <span class="n">legend</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Taxon&quot;</span><span class="p">,</span> <span class="s2">&quot;Gaps&quot;</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing&quot;</span><span class="p">,</span> <span class="s2">&quot;%&quot;</span><span class="p">,</span> <span class="s2">&quot;Data&quot;</span><span class="p">,</span>
                                 <span class="s2">&quot;%&quot;</span><span class="p">],</span>
                <span class="s2">&quot;normalize&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                <span class="s2">&quot;normalize_factor&quot;</span><span class="p">:</span> <span class="n">total_len</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.missing_genes_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.missing_genes_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">missing_genes_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data with distribution of missing genes per species plot.</span>

<span class="sd">        Calculates, for each taxon, the number of alignments where it is</span>
<span class="sd">        present.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">taxon</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch to interrupt worker</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="c1"># Add to progress counter</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">data_storage</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
                    <span class="n">data_storage</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># Sort data in descending order of missing genes</span>
        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                          <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">())],</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of missing genes per species&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Taxon&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing genes&quot;</span><span class="p">]</span>
                <span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.missing_genes_average"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.missing_genes_average">[docs]</a>    <span class="k">def</span> <span class="nf">missing_genes_average</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates histogram data with average missing taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of missing taxa&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of missing taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of missing taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.average_seqsize_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.average_seqsize_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">average_seqsize_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for average sequence size per taxon.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">taxon</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">sp</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="p">:</span>
                <span class="n">data_storage</span><span class="p">[</span><span class="n">sp</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span>
                                        <span class="n">replace</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)))</span>

        <span class="c1"># Adapt y-axis label according to sequence code</span>
        <span class="n">seq_code</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ax_ylabel</span> <span class="o">=</span> <span class="s2">&quot;Size (bp)&quot;</span> <span class="k">if</span> <span class="n">seq_code</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> <span class="k">else</span> <span class="s2">&quot;Size (residues)&quot;</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                   <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence size distribution per species&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax_ylabel</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.average_seqsize"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.average_seqsize">[docs]</a>    <span class="k">def</span> <span class="nf">average_seqsize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for the average sequence size for the entire data set.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">data_storage</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="c1"># Adapt y-axis label according to sequence code</span>
        <span class="n">seq_code</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ax_xlabel</span> <span class="o">=</span> <span class="s2">&quot;Size (bp)&quot;</span> <span class="k">if</span> <span class="n">seq_code</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> <span class="k">else</span> <span class="s2">&quot;Size (residues)&quot;</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_storage</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Average sequence size distribution&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ax_xlabel</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ax_xlabel</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.characters_proportion"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.characters_proportion">[docs]</a>    <span class="k">def</span> <span class="nf">characters_proportion</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for the proportion of nt/aa for the data set.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table,</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1">#Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">():</span>
                <span class="n">data_storage</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span>
                                        <span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

        <span class="c1"># Determine total number of characters</span>
        <span class="n">chars</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="c1"># Valid characters list</span>
        <span class="n">valid_chars</span> <span class="o">=</span> <span class="n">dna_chars</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> <span class="k">else</span> \
            <span class="nb">list</span><span class="p">(</span><span class="n">aminoacid_table</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="n">data</span><span class="p">,</span> <span class="n">xlabels</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="p">[(</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">/</span> <span class="n">chars</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span> <span class="k">for</span> <span class="n">y</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span>
                              <span class="n">data_storage</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">valid_chars</span><span class="p">])</span>

        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Nucleotide proportions&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> \
            <span class="k">else</span> <span class="s2">&quot;Amino acid proportions&quot;</span>
        <span class="n">ax_xlabel</span> <span class="o">=</span> <span class="s2">&quot;Nucleotide&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> \
            <span class="k">else</span> <span class="s2">&quot;Amino acid&quot;</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">data</span><span class="p">],</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">xlabels</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="n">title</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ax_xlabel</span><span class="p">,</span> <span class="s2">&quot;Proportion&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">ax_xlabel</span><span class="p">,</span> <span class="s2">&quot;Proportion&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.characters_proportion_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.characters_proportion_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">characters_proportion_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for the proportion of nt/aa per species.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;legend&quot;: mpl Legend object,</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">Counter</span><span class="p">())</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">sp</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="p">:</span>
                <span class="n">data_storage</span><span class="p">[</span><span class="n">sp</span><span class="p">]</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span>
                                            <span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

        <span class="n">legend</span> <span class="o">=</span> <span class="n">dna_chars</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> <span class="k">else</span> \
            <span class="nb">list</span><span class="p">(</span><span class="n">aminoacid_table</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">legend</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">char</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">legend</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">data_storage</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">chars</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">y</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">legend</span><span class="p">]))</span>
                <span class="n">data</span><span class="p">[</span><span class="n">p</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">c</span><span class="p">[</span><span class="n">char</span><span class="p">])</span> <span class="o">/</span> <span class="n">chars</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Nucleotide proportions&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> \
            <span class="k">else</span> <span class="s2">&quot;Amino acid proportions&quot;</span>

        <span class="n">ax_ylabel</span> <span class="o">=</span> <span class="s2">&quot;Nucleotide&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span> \
            <span class="k">else</span> <span class="s2">&quot;Amino acid&quot;</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="n">title</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_storage</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                <span class="s2">&quot;legend&quot;</span><span class="p">:</span> <span class="n">legend</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Taxa&quot;</span><span class="p">,</span> <span class="n">ax_ylabel</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Taxon&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">legend</span><span class="p">}</span></div>

    <span class="nd">@LookupDatabase</span>
    <span class="k">def</span> <span class="nf">_get_similarity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">aln_len</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets the similarity between two strings and the effective length.</span>

<span class="sd">        Compares two sequences and calculates the number of similarities.</span>
<span class="sd">        This comparison ignores columns with missing data, so the final</span>
<span class="sd">        sequence length that is actually compared has to be adjusted.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        seq1, seq2 : str</span>
<span class="sd">            Sequence strings.</span>
<span class="sd">        aln_len : int</span>
<span class="sd">            Original lenght of `seq1` and `seq2`.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sim : float</span>
<span class="sd">            Number of pairwise similarities.</span>
<span class="sd">        ef_len : float</span>
<span class="sd">            Effective sequence length, without missing data.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        To avoid performing the calculations between identical string pairs,</span>
<span class="sd">        this method is decorated with a class that stores the calculation</span>
<span class="sd">        result of new string pairs in a database, and fetches that result</span>
<span class="sd">        when the same string pair appears later on.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">seq1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seq1</span><span class="p">))</span>
        <span class="n">seq2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seq2</span><span class="p">))</span>

        <span class="n">ef_len_</span> <span class="o">=</span> <span class="p">(</span><span class="n">seq1</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&amp;</span> \
            <span class="p">(</span><span class="n">seq1</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">)</span> <span class="o">&amp;</span> \
            <span class="p">(</span><span class="n">seq2</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&amp;</span> \
            <span class="p">(</span><span class="n">seq2</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">)</span>

        <span class="n">sim_</span> <span class="o">=</span> <span class="p">(</span><span class="n">seq1</span> <span class="o">==</span> <span class="n">seq2</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">ef_len_</span>

        <span class="n">sim</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">sim_</span><span class="p">)</span>
        <span class="n">ef_len</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">ef_len_</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">sim</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="n">ef_len</span><span class="p">)</span>

<div class="viewcode-block" id="AlignmentList._get_informative_sites"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._get_informative_sites">[docs]</a>    <span class="k">def</span> <span class="nf">_get_informative_sites</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">aln</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Calculates number of informative sites in an `Alignment`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        aln : trifusion.process.sequence.Alignment</span>
<span class="sd">            `Alignment` object.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        informative_sites : int</span>
<span class="sd">            Number of informative sites for `Alignment` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">informative_sites</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

            <span class="n">column</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span>
                              <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">])</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Remove most common and check the length of the remaining</span>
                <span class="k">del</span> <span class="n">column</span><span class="p">[</span><span class="n">column</span><span class="o">.</span><span class="n">most_common</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
                <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">column</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">informative_sites</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="n">informative_sites</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_similarity"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_similarity">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_similarity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for average sequence similarity plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y]</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;connect&quot;</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="c1"># self._get_reference_data(list(aln.sequences()))</span>
            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">aln_similarities</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="n">sim</span><span class="p">,</span> <span class="n">total_len</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span>
                                                      <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">total_len</span><span class="p">:</span>
                    <span class="n">aln_similarities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sim</span> <span class="o">/</span> <span class="n">total_len</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">aln_similarities</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">aln_similarities</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;disconnect&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Similarity (%)&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_similarity_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_similarity_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_similarity_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for sequence similarity per species pair plot.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;color_label&quot;: str, label for colorbar</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;connect&quot;</span><span class="p">)</span>

        <span class="c1"># Create matrix for parwise comparisons</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">),</span> <span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span>
                <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))]</span>

        <span class="n">taxa_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">tx1</span><span class="p">,</span> <span class="n">tx2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">taxa_pos</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                    <span class="c1"># Kill switch</span>
                    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx1</span><span class="p">),</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx2</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">sim</span><span class="p">,</span> <span class="n">l</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">l</span><span class="p">:</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">taxa_pos</span><span class="p">[</span><span class="n">tx1</span><span class="p">]][</span><span class="n">taxa_pos</span><span class="p">[</span><span class="n">tx2</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sim</span> <span class="o">/</span> <span class="n">l</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="k">if</span> <span class="n">y</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">])</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tri</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;disconnect&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;color_label&quot;</span><span class="p">:</span> <span class="s2">&quot;Pairwise sequence similarity&quot;</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">taxa_pos</span><span class="p">)}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_similarity_gene"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_similarity_gene">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_similarity_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for sliding window sequence similarity for alignment.</span>

<span class="sd">        Retrieves an alignment using `gene_name` and calculates the</span>
<span class="sd">        average pair-wise sequence similarity along the alignment length</span>
<span class="sd">        using a sliding window approach.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gene_name : str</span>
<span class="sd">            `Alignment.name` from an alignment</span>
<span class="sd">        window_size : int</span>
<span class="sd">            Size of sliding window.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;window_size&quot;: int, size of sliding window,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aln_obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">retrieve_alignment</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">aln_obj</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="n">window_size</span><span class="p">):</span>

            <span class="n">window_similarities</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="n">seqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">]]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                             <span class="n">aln_obj</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">()])</span>

            <span class="k">for</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>

                <span class="n">s</span><span class="p">,</span> <span class="n">t</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq1</span><span class="p">),</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq2</span><span class="p">),</span>
                                            <span class="n">window_size</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">t</span><span class="p">:</span>
                    <span class="n">sim</span> <span class="o">=</span> <span class="n">s</span> <span class="o">/</span> <span class="n">t</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">sim</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="n">window_similarities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sim</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">window_similarities</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">window_similarities</span><span class="p">))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence similarity sliding window for gene</span><span class="se">\n</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span>
                         <span class="o">%</span> <span class="n">basename</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span>
                <span class="s2">&quot;window_size&quot;</span><span class="p">:</span> <span class="n">window_size</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence (bp)&quot;</span><span class="p">,</span> <span class="s2">&quot;Similarity (%)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence (bp)&quot;</span><span class="p">,</span> <span class="s2">&quot;Similarity (%)&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_segregation"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_segregation">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_segregation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">proportions</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for overall distribution of segregating sites.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        proportions : bool</span>
<span class="sd">            If True, use proportions instead of absolute values</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table,</span>

<span class="sd">            &quot;real_bin_num&quot;:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="n">segregating_sites</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

                <span class="c1"># Remove gaps and missing characters</span>
                <span class="n">column</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                              <span class="ow">and</span> <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">])</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">segregating_sites</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">proportions</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="nb">float</span><span class="p">(</span><span class="n">segregating_sites</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)))</span>
                <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Segregating sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Percentage&quot;</span><span class="p">]</span>
                <span class="n">real_bin</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">segregating_sites</span><span class="p">)</span>
                <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Segregating sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]</span>
                <span class="n">real_bin</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of segregating sites&quot;</span><span class="p">,</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">,</span>
                <span class="s2">&quot;real_bin_num&quot;</span><span class="p">:</span> <span class="n">real_bin</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_segregation_per_species"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_segregation_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_segregation_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for a triangular matrix of sequence segregation for</span>
<span class="sd">        pairs of taxa.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;color_label&quot;: str, label for colorbar</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;connect&quot;</span><span class="p">)</span>

        <span class="c1"># Create matrix for parwise comparisons</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">),</span> <span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span>
                <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))]</span>

        <span class="n">taxa_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">y</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">tx1</span><span class="p">,</span> <span class="n">tx2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">taxa_pos</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx1</span><span class="p">),</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx2</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">s</span><span class="p">,</span> <span class="n">t</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">t</span><span class="p">:</span>
                    <span class="n">aln_diff</span> <span class="o">=</span> <span class="n">t</span> <span class="o">-</span> <span class="n">s</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">aln_diff</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="n">data</span><span class="p">[</span><span class="n">taxa_pos</span><span class="p">[</span><span class="n">tx1</span><span class="p">]][</span><span class="n">taxa_pos</span><span class="p">[</span><span class="n">tx2</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln_diff</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">y</span><span class="p">)</span> <span class="k">if</span> <span class="n">y</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">])</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tri</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">k</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="n">mask</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;disconnect&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">taxa_pos</span><span class="p">),</span>
                <span class="s2">&quot;color_label&quot;</span><span class="p">:</span> <span class="s2">&quot;Segregating sites&quot;</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.sequence_segregation_gene"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.sequence_segregation_gene">[docs]</a>    <span class="k">def</span> <span class="nf">sequence_segregation_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for a sliding window analysis of segregating sites.</span>

<span class="sd">        Retrieves an alignment using `gene_name` and calculates the</span>
<span class="sd">        number of segregating sites along the alignment length</span>
<span class="sd">        using a sliding window approach.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gene_name : str</span>
<span class="sd">            `Alignment.name` from an alignment</span>
<span class="sd">        window_size : int</span>
<span class="sd">            Size of sliding window.</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;window_size&quot;: int, size of sliding window,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">aln_obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">retrieve_alignment</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">aln_obj</span><span class="o">.</span><span class="n">locus_length</span><span class="p">,</span> <span class="n">window_size</span><span class="p">):</span>

            <span class="n">segregating_sites</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="n">seqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">]]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                             <span class="n">aln_obj</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">()])</span>

            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">seqs</span><span class="p">):</span>
                <span class="n">column</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">aln_obj</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                              <span class="ow">and</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">])</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">segregating_sites</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">segregating_sites</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Number of segregating sites sliding window for &quot;</span>
                         <span class="s2">&quot;gene</span><span class="se">\n</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">basename</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span>
                <span class="s2">&quot;window_size&quot;</span><span class="p">:</span> <span class="n">window_size</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence (bp)&quot;</span><span class="p">,</span> <span class="s2">&quot;Segregating sites&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence (bp)&quot;</span><span class="p">,</span> <span class="s2">&quot;Segregating sites&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.length_polymorphism_correlation"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.length_polymorphism_correlation">[docs]</a>    <span class="k">def</span> <span class="nf">length_polymorphism_correlation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for correlation between alignment length and</span>
<span class="sd">        informative sites.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;correlation&quot;: bool, whether to perform spearman correlation,</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data_length</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_inf</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get informative sites for alignment</span>
            <span class="n">inf_sites</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_informative_sites</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
            <span class="n">data_inf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">inf_sites</span><span class="p">)</span>

            <span class="c1"># Get size</span>
            <span class="n">data_length</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">data_length</span><span class="p">,</span> <span class="n">data_inf</span><span class="p">],</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Correlation between alignment length and number of &quot;</span>
                         <span class="s2">&quot;variable sites&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Alignment length&quot;</span><span class="p">,</span> <span class="s2">&quot;Informative sites&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Alignment length&quot;</span><span class="p">,</span> <span class="s2">&quot;Informative sites&quot;</span><span class="p">],</span>
                <span class="s2">&quot;correlation&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.allele_frequency_spectrum"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum">[docs]</a>    <span class="k">def</span> <span class="nf">allele_frequency_spectrum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">proportions</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for the allele frequency spectrum.</span>

<span class="sd">        Calculates the allele frequency spectrum of the entire</span>
<span class="sd">        alignment data set. Here, multiple alignments are effectively treated</span>
<span class="sd">        as a single one. This method is exclusive of DNA sequence type and</span>
<span class="sd">        supports IUPAC ambiguity codes.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        proportions : bool</span>
<span class="sd">            If True, use proportions instead of absolute values</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make check for sequence type consistency here for TriStats.py. In</span>
        <span class="c1"># TriFusion the check is made before calling this method</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;exception&quot;</span><span class="p">:</span> <span class="n">InvalidSequenceType</span><span class="p">}</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

                <span class="n">col</span> <span class="o">=</span> <span class="p">[</span><span class="n">iupac_conv</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span>
                                  <span class="n">x</span> <span class="o">!=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span>
                                  <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">]</span>
                <span class="n">col_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>

                <span class="c1"># Remove gaps and missing characters</span>
                <span class="n">column</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>

                <span class="c1"># Consider only bi-allelic SNPs</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="c1"># Remove most common and check the length of the remaining</span>
                    <span class="k">del</span> <span class="n">column</span><span class="p">[</span><span class="n">column</span><span class="o">.</span><span class="n">most_common</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
                    <span class="c1"># Append number of derived alleles</span>
                    <span class="k">if</span> <span class="n">proportions</span><span class="p">:</span>
                        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">column</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span> <span class="o">/</span>
                                    <span class="nb">float</span><span class="p">(</span><span class="n">col_len</span><span class="p">))</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">column</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele frequency spectrum&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Derived allele frequency&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Derived allele frequency&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;real_bin_num&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.allele_frequency_spectrum_gene"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum_gene">[docs]</a>    <span class="k">def</span> <span class="nf">allele_frequency_spectrum_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for the allele frequency spectrum of an `Alignment`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gene_name : str</span>
<span class="sd">            `Alignment.name` of an alignment.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table,</span>

<span class="sd">            &quot;real_bin_num&quot;:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Make check for sequence type consistency here for TriStats.py. In</span>
        <span class="c1"># TriFusion the check is made before calling this method</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;exception&quot;</span><span class="p">:</span> <span class="n">InvalidSequenceType</span><span class="p">}</span>

        <span class="n">aln</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">retrieve_alignment</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

            <span class="c1"># Remove gaps and missing characters</span>
            <span class="n">column</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">iupac_conv</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span>
                              <span class="n">x</span> <span class="o">!=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span>
                              <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">])</span>

            <span class="c1"># Consider only bi-allelic SNPs</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="c1"># Remove most common and check the length of the remaining</span>
                <span class="k">del</span> <span class="n">column</span><span class="p">[</span><span class="n">column</span><span class="o">.</span><span class="n">most_common</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
                <span class="c1"># Append number of derived alleles</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">column</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Allele frequency spectrum&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Derived allele frequency&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Derived allele frequency&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;real_bin_num&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.taxa_distribution"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.taxa_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">taxa_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for a distribution of taxa frequency across alignments.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;color_label&quot;: str, label for colorbar,</span>

<span class="sd">            &quot;real_bin_num&quot;:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get number of taxa</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of taxa frequency&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;real_bin_num&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.cumulative_missing_genes"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.cumulative_missing_genes">[docs]</a>    <span class="k">def</span> <span class="nf">cumulative_missing_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for a distribution of the maximum number of genes</span>
<span class="sd">        available for consecutive thresholds of missing data.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;labels&quot;: list, label for xticks,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;table_header&quot;: list with headers of table}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">size_storage</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="c1"># total number of taxa in data set</span>
        <span class="n">taxa</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="c1"># Kill switch</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get number of taxa</span>
            <span class="n">size_storage</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">))</span> <span class="o">/</span> <span class="n">taxa</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>

        <span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">105</span><span class="p">,</span> <span class="mi">5</span><span class="p">):</span>

            <span class="c1"># Get percentage</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">size_storage</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">]))</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">data</span><span class="p">],</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene frequency for decreasing values of missing &quot;</span>
                         <span class="s2">&quot;data&quot;</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Minimum taxa representation&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene frequency&quot;</span><span class="p">],</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Percentage&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="AlignmentList._mad_based_outlier"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList._mad_based_outlier">[docs]</a>    <span class="k">def</span> <span class="nf">_mad_based_outlier</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">3.5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Calculate MAD based outliers.</span>

<span class="sd">        An outlier detection method based on median absolute deviation. This</span>
<span class="sd">        code was adapted from http://stackoverflow.com/a/22357811/1990165.</span>
<span class="sd">        The usage of the median is much less biased that the mean and is</span>
<span class="sd">        robust to smaller data sets.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        p : numpy.array</span>
<span class="sd">            Array with data observations.</span>
<span class="sd">        threshold : float</span>
<span class="sd">            Modified Z-score to use as a threshold.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">p</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span>
        <span class="c1"># Get sample median.</span>
        <span class="n">median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">p</span> <span class="o">-</span> <span class="n">median</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>
        <span class="n">med_abs_deviation</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span>

        <span class="c1"># The 0.6745is a constant that makes values roughly equivalent in</span>
        <span class="c1"># units to standard deviations.</span>
        <span class="n">z_score</span> <span class="o">=</span> <span class="mf">0.6745</span> <span class="o">*</span> <span class="n">diff</span> <span class="o">/</span> <span class="n">med_abs_deviation</span>

        <span class="k">return</span> <span class="n">z_score</span> <span class="o">&gt;</span> <span class="n">threshold</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_missing_data"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_missing_data">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_missing_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for missing data alignment outliers.</span>

<span class="sd">        Get data for outlier detection of genes based on the distribution of</span>
<span class="sd">        average missing data per gene. Data points will be based on the</span>
<span class="sd">        proportion of missing data symbols out of the possible total. For</span>
<span class="sd">        example, in an alignment with three taxa, each with 100 sites,</span>
<span class="sd">        the total possible missing data is 300 (100 * 3). Here, missing data</span>
<span class="sd">        will be gathered from all taxa and a proportion will be calculated</span>
<span class="sd">        based n the total possible</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">gn</span><span class="p">,</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">total_len</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">)</span>
            <span class="n">gn_data</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">():</span>
                <span class="n">gn_data</span> <span class="o">+=</span> <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> \
                          <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">)</span>

            <span class="n">m_data</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gn_data</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">total_len</span><span class="p">)</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">m_data</span><span class="p">)</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gn</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outlier labels</span>
        <span class="n">outliers_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span>
            <span class="n">data_points</span><span class="p">)])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Missing data outlier gene detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outliers_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion of missing data&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_missing_data_sp"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_missing_data_sp">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_missing_data_sp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for missing data taxa outliers.</span>

<span class="sd">        Gets data for outlier detection of species based on missing data. For</span>
<span class="sd">        this analysis, genes for which a taxa is completely absent will be</span>
<span class="sd">        ignored for the calculations of that taxa. The reason for this,</span>
<span class="sd">        is that including genes where the taxon is absent would bias the</span>
<span class="sd">        outlier detection towards taxa that have low prevalence in the data</span>
<span class="sd">        set, even if they have low missing data in the alignments where they</span>
<span class="sd">        are present.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">tx</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">total_len</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span>

            <span class="c1"># Get missing data for every taxon</span>
            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="n">m_data</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span>
                                   <span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">))</span> <span class="o">/</span> \
                        <span class="nb">float</span><span class="p">(</span><span class="n">total_len</span><span class="p">)</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">m_data</span><span class="p">)</span>

        <span class="c1"># Get average for each taxon</span>
        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>

        <span class="c1"># Prepara data for plotting</span>
        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outlier taxa</span>
        <span class="n">outlier_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Missing data outlier taxa detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outlier_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion of missing data&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_segregating"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_segregating">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_segregating</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for segregating site alignment outliers.</span>

<span class="sd">        Generates data for the outlier detection of genes based on</span>
<span class="sd">        segregating sites. The data will be based on the number of alignments</span>
<span class="sd">        columns with a variable number of sites, excluding gaps and missing</span>
<span class="sd">        data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">segregating_sites</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_columns</span><span class="p">():</span>

                <span class="c1"># Remove gaps and missing characters</span>
                <span class="n">column</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">column</span> <span class="k">if</span> <span class="n">x</span> <span class="o">!=</span> <span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span>
                              <span class="n">x</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">])</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">segregating_sites</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># Get proportion of segregating sites for current alignment</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">segregating_sites</span><span class="p">)</span> <span class="o">/</span>
                               <span class="nb">float</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">))</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outlier taxa</span>
        <span class="n">outlier_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence variation outlier gene detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outlier_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion of segregating sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_segregating_sp"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_segregating_sp">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_segregating_sp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for segregating sites taxa outliers.</span>

<span class="sd">        Generates data for the outlier detection of species based on their</span>
<span class="sd">        average pair-wise proportion of segregating sites. Comparisons</span>
<span class="sd">        with gaps or missing data are ignored</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;connect&quot;</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">((</span><span class="n">tx</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">tx1</span><span class="p">,</span> <span class="n">tx2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx1</span><span class="p">),</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx2</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">s</span><span class="p">,</span> <span class="n">t_len</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">aln</span><span class="o">.</span><span class="n">locus_length</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">t_len</span><span class="p">:</span>
                    <span class="n">s_data</span> <span class="o">=</span> <span class="p">(</span><span class="n">t_len</span> <span class="o">-</span> <span class="n">s</span><span class="p">)</span> <span class="o">/</span> <span class="n">t_len</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">s_data</span> <span class="o">=</span> <span class="mi">0</span>

                <span class="n">data</span><span class="p">[</span><span class="n">tx1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s_data</span><span class="p">)</span>
                <span class="n">data</span><span class="p">[</span><span class="n">tx2</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s_data</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>

        <span class="c1"># Prepara data for plotting</span>
        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outlier taxa</span>
        <span class="n">outlier_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_get_similarity</span><span class="p">(</span><span class="s2">&quot;disconnect&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence variation outlier taxa detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outlier_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion of segregating sites&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_sequence_size"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_sequence_size">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_sequence_size</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Creates data for sequence size alignment outliers.</span>

<span class="sd">        Generates data for the outlier detection of genes based on their</span>
<span class="sd">        sequence length (excluding missing data).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">gn</span><span class="p">,</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">gn_l</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">():</span>
                <span class="n">gn_l</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span>
                              <span class="n">replace</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span>
                              <span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)))</span>

            <span class="n">gn_avg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">gn_l</span><span class="p">)</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gn_avg</span><span class="p">)</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gn</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outliers labels</span>
        <span class="n">outliers_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span>
            <span class="n">data_points</span><span class="p">)])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence size outlier gene detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outliers_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence size&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div>

    <span class="nd">@check_data</span>
<div class="viewcode-block" id="AlignmentList.outlier_sequence_size_sp"><a class="viewcode-back" href="../../../trifusion.process.sequence.html#trifusion.process.sequence.AlignmentList.outlier_sequence_size_sp">[docs]</a>    <span class="k">def</span> <span class="nf">outlier_sequence_size_sp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create data for sequence size taxa outliers.</span>

<span class="sd">        Generates data for the outlier detection of species based on their</span>
<span class="sd">        sequence length (excluding missing data)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ns : multiprocesssing.Manager.Namespace</span>
<span class="sd">            A Namespace object used to communicate with the main thread</span>
<span class="sd">            in TriFusion.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        _ : dict</span>
<span class="sd">            &quot;data&quot;: `numpy.array` with data for plotting,</span>

<span class="sd">            &quot;ax_names&quot;: 2 element list with axis labels [x, y],</span>

<span class="sd">            &quot;title&quot;: str with title,</span>

<span class="sd">            &quot;outliers&quot;: list of outlier data points,</span>

<span class="sd">            &quot;outliers_labels&quot;: list of outlier labels</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">tx</span><span class="p">,</span> <span class="p">[])</span> <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">tx</span> <span class="ow">in</span> <span class="n">aln</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>
                    <span class="n">s_data</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span>
                                 <span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gap_symbol</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
                    <span class="n">data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s_data</span><span class="p">)</span>

        <span class="c1"># Get average for each taxon</span>
        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data</span><span class="p">[</span><span class="n">tx</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>

        <span class="c1"># Preparing data for plotting</span>
        <span class="n">data_points</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tx</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">data_points</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
            <span class="n">data_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx</span><span class="p">)</span>

        <span class="n">data_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_points</span><span class="p">)</span>
        <span class="n">data_labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">data_labels</span><span class="p">)</span>

        <span class="c1"># Get outliers</span>
        <span class="n">outliers_points</span> <span class="o">=</span> <span class="n">data_points</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)]</span>
        <span class="c1"># Get outlier taxa</span>
        <span class="n">outlier_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data_labels</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mad_based_outlier</span><span class="p">(</span><span class="n">data_points</span><span class="p">)])</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data_points</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence size outlier taxa detection&quot;</span><span class="p">,</span>
                <span class="s2">&quot;outliers&quot;</span><span class="p">:</span> <span class="n">outliers_points</span><span class="p">,</span>
                <span class="s2">&quot;outliers_labels&quot;</span><span class="p">:</span> <span class="n">outlier_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Sequence size&quot;</span><span class="p">,</span> <span class="s2">&quot;Frequency&quot;</span><span class="p">]}</span></div></div>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Diogo N. Silva&quot;</span>
<span class="n">__credits__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Diogo N. Silva&quot;</span><span class="p">,</span> <span class="s2">&quot;Tiago F. Jesus&quot;</span><span class="p">]</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Diogo N. Silva.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.5.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>