

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>trifusion.ortho.OrthomclToolbox &mdash; TriFusion 0.5.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="TriFusion 0.5.5 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> TriFusion
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../trifusion.html">TriFusion modules and API reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">TriFusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>trifusion.ortho.OrthomclToolbox</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for trifusion.ortho.OrthomclToolbox</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1">#</span>
<span class="c1">#  Copyright 2012 Unknown &lt;diogo@arch&gt;</span>
<span class="c1">#  </span>
<span class="c1">#  This program is free software; you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation; either version 2 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#  </span>
<span class="c1">#  This program is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#  </span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with this program; if not, write to the Free Software</span>
<span class="c1">#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,</span>
<span class="c1">#  MA 02110-1301, USA.</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">process.sequence</span> <span class="k">import</span> <span class="n">Alignment</span>
    <span class="kn">from</span> <span class="nn">base.plotter</span> <span class="k">import</span> <span class="n">bar_plot</span><span class="p">,</span> <span class="n">multi_bar_plot</span>
    <span class="kn">from</span> <span class="nn">process.error_handling</span> <span class="k">import</span> <span class="n">KillByUser</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.sequence</span> <span class="k">import</span> <span class="n">Alignment</span>
    <span class="kn">from</span> <span class="nn">trifusion.base.plotter</span> <span class="k">import</span> <span class="n">bar_plot</span><span class="p">,</span> <span class="n">multi_bar_plot</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.error_handling</span> <span class="k">import</span> <span class="n">KillByUser</span>

<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span><span class="p">,</span> <span class="n">Counter</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sqlite3</span>
<span class="kn">from</span> <span class="nn">os.path</span> <span class="k">import</span> <span class="n">join</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">string</span>
<span class="kn">import</span> <span class="nn">copy</span>


<div class="viewcode-block" id="Cluster"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Cluster">[docs]</a><span class="k">class</span> <span class="nc">Cluster</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Object for clusters of the OrthoMCL groups file. It is useful to set a</span>
<span class="sd">     number of attributes that will make subsequent filtration and</span>
<span class="sd">     processing much easier &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        To initialize a Cluster object, only a string compliant with the</span>
<span class="sd">        format of a cluster in an OrthoMCL groups file has to be provided.</span>
<span class="sd">        This line should contain the name of the group, a colon, and the</span>
<span class="sd">        sequences belonging to that group separated by whitespace</span>
<span class="sd">        :param line_string: String of a cluster</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initializing attributes for parse_string</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Initializing attributes for apply filter</span>
        <span class="c1"># If the value is different than None, this will inform downstream</span>
        <span class="c1"># objects of whether this cluster is compliant with the specified</span>
        <span class="c1"># gene_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_compliant</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># If the value is different than None, this will inform downstream</span>
        <span class="c1"># objects of whether this cluster is compliant with the specified</span>
        <span class="c1"># species_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_compliant</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">parse_string</span><span class="p">(</span><span class="n">line_string</span><span class="p">)</span>

<div class="viewcode-block" id="Cluster.parse_string"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Cluster.parse_string">[docs]</a>    <span class="k">def</span> <span class="nf">parse_string</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cluster_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the string and sets the group name and sequence list attributes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fields</span> <span class="o">=</span> <span class="n">cluster_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
        <span class="c1"># Setting the name and sequence list of the clusters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>

        <span class="c1"># Setting the gene frequency for each species in the cluster</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">field</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span>
                                          <span class="bp">self</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span></div>

<div class="viewcode-block" id="Cluster.remove_taxa"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Cluster.remove_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">remove_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Removes the taxa contained in taxa_list from self.sequences and</span>
<span class="sd">        self.species_frequency</span>
<span class="sd">        :param taxa_list: list, each element should be a taxon name</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequences</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                          <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">taxon</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span> <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span>
                                      <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                                      <span class="k">if</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">)</span></div>

<div class="viewcode-block" id="Cluster.apply_filter"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Cluster.apply_filter">[docs]</a>    <span class="k">def</span> <span class="nf">apply_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_threshold</span><span class="p">,</span> <span class="n">species_threshold</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method will update two Cluster attributes, self.gene_flag and</span>
<span class="sd">        self.species_flag, which will inform downstream objects if this</span>
<span class="sd">        cluster respects the gene and species threshold</span>
<span class="sd">        :param gene_threshold: Integer for the maximum number of gene copies</span>
<span class="sd">        per species</span>
<span class="sd">        :param species_threshold: Integer for the minimum number of species</span>
<span class="sd">        present</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check whether cluster is compliant with species_threshold</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">species_threshold</span> <span class="ow">and</span> \
                <span class="n">species_threshold</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">species_compliant</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">species_compliant</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># Check whether cluster is compliant with gene_threshold</span>
        <span class="k">if</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">&lt;=</span> <span class="n">gene_threshold</span> <span class="ow">and</span> \
                <span class="n">gene_threshold</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_compliant</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_compliant</span> <span class="o">=</span> <span class="kc">False</span></div></div>


<div class="viewcode-block" id="OrthoGroupException"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.OrthoGroupException">[docs]</a><span class="k">class</span> <span class="nc">OrthoGroupException</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="k">pass</span></div>


<div class="viewcode-block" id="GroupLight"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight">[docs]</a><span class="k">class</span> <span class="nc">GroupLight</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analogous to Group object but with several changes to reduce memory usage</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">groups_file</span><span class="p">,</span> <span class="n">gene_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">species_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="n">gene_threshold</span> <span class="k">if</span> <span class="n">gene_threshold</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="n">species_threshold</span> <span class="k">if</span> <span class="n">species_threshold</span> \
            <span class="k">else</span> <span class="kc">None</span>

        <span class="c1"># Attribute containing the list of included species</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Attribute that will contain taxa to be excluded from analyses</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Attributes that will store the number (int) of cluster after gene and</span>
        <span class="c1"># species filter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_clusters</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Attribute containing the total number of sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Attribute containing the maximum number of extra copies found in the</span>
        <span class="c1"># clusters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Attribute with name of the group file, which will be an ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">groups_file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">table</span> <span class="o">=</span> <span class="n">groups_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">sep</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Initialize atribute containing the groups filtered using the gene and</span>
        <span class="c1"># species threshold. This attribute can be updated at any time using</span>
        <span class="c1"># the update_filtered_group method</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_parse_groups</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">float</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_sp_proportion</span><span class="p">()</span>

<div class="viewcode-block" id="GroupLight.groups"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.groups">[docs]</a>    <span class="k">def</span> <span class="nf">groups</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generator for group file. This replaces the self.groups attribute of</span>
<span class="sd">        the original Group Object. Instead of loading the whole file into</span>
<span class="sd">        memory, a generator is created to iterate over its contents. It may</span>
<span class="sd">        run a bit slower but its a lot more memory efficient.</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_handle</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">yield</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span></div>

<div class="viewcode-block" id="GroupLight.iter_species_frequency"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.iter_species_frequency">[docs]</a>    <span class="k">def</span> <span class="nf">iter_species_frequency</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        In order to prevent permanent changes to the species_frequency</span>
<span class="sd">        attribute due to the filtering of taxa, this iterable should be used</span>
<span class="sd">        instead of the said variable. This creates a temporary deepcopy of</span>
<span class="sd">        species_frequency which will be iterated over and eventually modified.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Since the items of species_frequency are mutable, this ensures</span>
        <span class="c1"># that even those objects are correctly cloned</span>
        <span class="n">sp_freq</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">sp_freq</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">cl</span></div>

    <span class="k">def</span> <span class="nf">_remove_tx</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a group line, remove all references to the excluded taxa</span>
<span class="sd">        :param line: raw group file line</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">new_line</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

        <span class="n">tx_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="k">if</span>
                           <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">new_line</span> <span class="o">+</span> <span class="n">tx_str</span>

    <span class="k">def</span> <span class="nf">_apply_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cl</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets or updates the basic group statistics, such as the number of</span>
<span class="sd">        orthologs compliant with the gene copy and minimum taxa filters.</span>

<span class="sd">        :param cl: dictionary. Contains the number of occurrences for each</span>
<span class="sd">         taxon present in the ortholog cluster</span>
<span class="sd">         (e.g. {&quot;taxonA&quot;: 2, &quot;taxonB&quot;: 1).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># First, remove excluded taxa from the cl object since this will</span>
        <span class="c1"># impact all other filters</span>
        <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">:</span>
            <span class="n">cl</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cl</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">all_clusters</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">extra_copies</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">extra_copies</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="n">extra_copies</span>

            <span class="k">if</span> <span class="n">extra_copies</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span>\
                <span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span>  \
                    <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">elif</span> <span class="p">(</span><span class="n">extra_copies</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">)</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span> \
                    <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="nf">_get_compliance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cl</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determines whether an ortholog cluster is compliant with the specified</span>
<span class="sd">        ortholog filters.</span>
<span class="sd">        :param ccl: dictionary. Contains the number of occurrences for each</span>
<span class="sd">         taxon present in the ortholog cluster</span>
<span class="sd">         (e.g. {&quot;taxonA&quot;: 2, &quot;taxonB&quot;: 1).</span>

<span class="sd">        :return: tuple. The first element refers to the gene copy filter</span>
<span class="sd">        while the second refers to the minimum taxa filter. Values of 1</span>
<span class="sd">        indicate that the ortholg cluster is compliant.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">tx</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">:</span>
            <span class="n">cl</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">tx</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cl</span><span class="p">:</span>

            <span class="n">cp</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">cp</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span>\
                <span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span>  \
                    <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>

            <span class="k">elif</span> <span class="p">(</span><span class="n">cp</span> <span class="o">&lt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">)</span> <span class="ow">or</span> \
                    <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span>

            <span class="k">elif</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>

    <span class="k">def</span> <span class="nf">_reset_counter</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">all_clusters</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">def</span> <span class="nf">_parse_groups</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">():</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="c1"># Retrieve the field containing the ortholog sequences</span>
            <span class="n">sequence_field</span> <span class="o">=</span> <span class="n">cl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>

            <span class="c1"># Update species frequency list</span>
            <span class="n">sp_freq</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">((</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                              <span class="n">sequence_field</span><span class="o">.</span><span class="n">split</span><span class="p">()))</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sp_freq</span><span class="p">)</span>

            <span class="c1"># Update number of sequences</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_field</span><span class="p">)</span>

            <span class="c1"># Update max number of extra copies</span>
            <span class="n">extra_copies</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">sp_freq</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">extra_copies</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">sp_freq</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sp_freq</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span>
                                      <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">])</span>

            <span class="c1"># Apply filters, if any</span>
            <span class="c1"># gene filter</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">sp_freq</span><span class="p">)</span>

<div class="viewcode-block" id="GroupLight.exclude_taxa"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.exclude_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">exclude_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">update_stats</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Updates the excluded_taxa attribute and updates group statistics if</span>
<span class="sd">        update_stats is True. This does not change the Group object data</span>
<span class="sd">        permanently, only sets an attribute that will be taken into account</span>
<span class="sd">        when plotting and exporting data.</span>
<span class="sd">        :param taxa_list: list. List of taxa that should be excluded from</span>
<span class="sd">        downstream operations</span>
<span class="sd">        :param update_stats: boolean. If True, it will update the group</span>
<span class="sd">        statistics</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># IF the taxa_list is the same as the excluded_taxa attribute,</span>
        <span class="c1"># there is nothing to do</span>
        <span class="k">if</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span> <span class="o">==</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">):</span>
            <span class="k">return</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span>
                             <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxa_list</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span> <span class="o">=</span> <span class="n">taxa_list</span>

        <span class="k">if</span> <span class="n">update_stats</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reset_counter</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span></div>

<div class="viewcode-block" id="GroupLight.basic_group_statistics"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.basic_group_statistics">[docs]</a>    <span class="k">def</span> <span class="nf">basic_group_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">update_stats</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">update_stats</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reset_counter</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span>

        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span></div>

    <span class="k">def</span> <span class="nf">_get_sp_proportion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        When the species filter is a float value between 0 and 1, convert</span>
<span class="sd">        this proportion into absolute values (rounded up), since filters were</span>
<span class="sd">        already designed for absolutes.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">*</span>
                                     <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>

<div class="viewcode-block" id="GroupLight.update_filters"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.update_filters">[docs]</a>    <span class="k">def</span> <span class="nf">update_filters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">,</span> <span class="n">update_stats</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Updates the group filter attributes and group summary stats if</span>
<span class="sd">        update_stats is True. This method does not change the</span>
<span class="sd">        data of the Group object, only sets attributes that will be taken into</span>
<span class="sd">        account when plotting or exporting data</span>
<span class="sd">        :param gn_filter: integer. Maximum number of gene copies allowed in an</span>
<span class="sd">        ortholog cluster</span>
<span class="sd">        :param sp_filter: integer/float. Minimum number/proportion of taxa</span>
<span class="sd">        representation</span>
<span class="sd">        :param update_stats: boolean. If True it will update the group summary</span>
<span class="sd">        statistics</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># If the provided filters are the same as the current group attributes</span>
        <span class="c1"># there is nothing to do</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span>
                                      <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">):</span>
            <span class="k">return</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="n">gn_filter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="n">sp_filter</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">float</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_sp_proportion</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">update_stats</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reset_counter</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">():</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span></div>

<div class="viewcode-block" id="GroupLight.retrieve_sequences"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.retrieve_sequences">[docs]</a>    <span class="k">def</span> <span class="nf">retrieve_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sqldb</span><span class="p">,</span> <span class="n">protein_db</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span>
                             <span class="n">shared_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param sqldb: srting. Path to sqlite database file</span>
<span class="sd">        :param protein_db: string. Path to protein database file</span>
<span class="sd">        :param dest: string. Directory where sequences will be exported</span>
<span class="sd">        :param shared_namespace: Namespace object to communicate with</span>
<span class="sd">        TriFusion&#39;s main process</span>
<span class="sd">        :param outfile: If set, all sequeces will be instead saved in a</span>
<span class="sd">        single output file. This is used for the nucleotide sequence export</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dest</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">outfile</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">dest</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Creating database&quot;</span>
            <span class="c1"># Stores sequences that could not be retrieved</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">missed</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Get number of lines of protein database</span>
            <span class="n">p</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">protein_db</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fh</span><span class="p">):</span>
                    <span class="k">pass</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">max_pb</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="c1"># Connect to database</span>
        <span class="n">conn</span> <span class="o">=</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">sqldb</span><span class="p">,</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">100000</span><span class="p">)</span>
        <span class="n">c</span> <span class="o">=</span> <span class="n">conn</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
        <span class="n">table_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protein_db</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">isalnum</span><span class="p">()])</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span>
            <span class="s2">&quot;utf8&quot;</span><span class="p">)</span>

        <span class="c1"># Create table if it does not exist</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;SELECT name FROM sqlite_master WHERE type=&#39;table&#39; &quot;</span>
                         <span class="s2">&quot;AND name=&#39;</span><span class="si">{}</span><span class="s2">&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span><span class="o">.</span><span class="n">fetchall</span><span class="p">():</span>

            <span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE </span><span class="si">{}</span><span class="s2"> (seq_id text PRIMARY KEY, seq text)&quot;</span><span class="o">.</span>
                      <span class="nb">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span>

            <span class="c1"># Populate database</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">protein_db</span><span class="p">)</span> <span class="k">as</span> <span class="n">ph</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ph</span><span class="p">:</span>

                    <span class="c1"># Kill switch</span>
                    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                            <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
                        <span class="n">seq_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span>
                        <span class="k">if</span> <span class="n">seq</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO </span><span class="si">{}</span><span class="s2"> VALUES (?, ?)&quot;</span><span class="o">.</span>
                                      <span class="nb">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>
                        <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

            <span class="n">conn</span><span class="o">.</span><span class="n">commit</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Fetching sequences&quot;</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">good</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">max_pb</span> <span class="o">=</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">total</span> <span class="o">=</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span>

        <span class="c1"># Set single output file, if option is set</span>
        <span class="k">if</span> <span class="n">outfile</span><span class="p">:</span>
            <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">outfile</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="c1"># Fetching sequences</span>
        <span class="k">for</span> <span class="n">line</span><span class="p">,</span> <span class="n">cl</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">()):</span>

            <span class="c1"># Kill switch</span>
            <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="c1"># Filter sequences</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                    <span class="n">shared_namespace</span><span class="o">.</span><span class="n">good</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">shared_namespace</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="c1"># Retrieve sequences from current cluster</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">:</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_remove_tx</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>

                <span class="c1"># Open file</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">outfile</span><span class="p">:</span>
                    <span class="n">cl_name</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">cl_name</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.fas&quot;</span><span class="p">,</span>
                                         <span class="s2">&quot;w&quot;</span><span class="p">)</span>

                <span class="n">seqs</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">:</span>
                    <span class="c1"># Query database</span>
                    <span class="n">c</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{}</span><span class="s2"> WHERE seq_id = ?&quot;</span><span class="o">.</span>
                              <span class="nb">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">),</span> <span class="p">(</span><span class="n">i</span><span class="p">,))</span>
                    <span class="n">vals</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">fetchone</span><span class="p">()</span>
                    <span class="c1"># Handles cases where the sequence could not be retrieved</span>
                    <span class="c1"># If outfile is set, output_handle will be a single file</span>
                    <span class="c1"># for all groups. If not, it will represent an individual</span>
                    <span class="c1"># group file</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">{}</span><span class="se">\n</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">vals</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                                               <span class="n">vals</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
                    <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
                        <span class="k">pass</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">outfile</span><span class="p">:</span>
                    <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">outfile</span><span class="p">:</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="n">conn</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="GroupLight.export_filtered_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.export_filtered_group">[docs]</a>    <span class="k">def</span> <span class="nf">export_filtered_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;filtered_groups&quot;</span><span class="p">,</span>
                                 <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">shared_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Exporting filtered orthologs&quot;</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">missed</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">good</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">cl</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">(),</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">())):</span>

            <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="n">p</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                    <span class="n">shared_namespace</span><span class="o">.</span><span class="n">good</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">:</span>
                    <span class="n">l</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_remove_tx</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">l</span> <span class="o">=</span> <span class="n">line</span>
                <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">l</span><span class="p">))</span>

        <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="GroupLight.bar_species_distribution"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.bar_species_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">bar_species_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">()</span> <span class="k">if</span>
                           <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">))</span>

        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="p">)]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># When data is empty, return an exception</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>

        <span class="c1"># Sort lists</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span> <span class="n">data</span><span class="p">)))][</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Convert label to strings</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_labels</span><span class="p">]</span>

        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Taxa frequency distribution&quot;</span>
        <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Number of taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">data</span><span class="p">],</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="n">title</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">x_labels</span><span class="p">,</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of species&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]}</span></div>

<div class="viewcode-block" id="GroupLight.bar_genecopy_distribution"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.bar_genecopy_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">bar_genecopy_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a bar plot with the distribution of gene copies across</span>
<span class="sd">        clusters</span>
<span class="sd">        :param filt: Boolean, whether or not to use the filtered groups.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">((</span><span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">()</span> <span class="k">if</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">((</span><span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_frequency</span>
                           <span class="k">if</span> <span class="n">cl</span><span class="p">))</span>

        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="p">)]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># When data is empty, return an exception</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>

        <span class="n">x_labels</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span> <span class="n">data</span><span class="p">))))</span>

        <span class="c1"># Convert label to strings</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_labels</span><span class="p">]</span>

        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Gene copy distribution&quot;</span>
        <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Number of gene copies&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">data</span><span class="p">],</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">x_labels</span><span class="p">,</span>
                <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="n">title</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">,</span>
                <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Number of gene copies&quot;</span><span class="p">,</span>
                                 <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]}</span></div>

<div class="viewcode-block" id="GroupLight.bar_species_coverage"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.bar_species_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">bar_species_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a stacked bar plot with the proportion of</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_counter</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">cl</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span>
                           <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">cl</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">most_common</span><span class="p">()</span>

        <span class="c1"># When data is empty, return an exception</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>

        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">],</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">all_clusters</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span>
                                      <span class="n">filt</span> <span class="k">else</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                                      <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]]</span>

        <span class="n">lgd_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Available data&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing data&quot;</span><span class="p">]</span>
        <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">x_labels</span><span class="p">,</span>
                <span class="s2">&quot;lgd_list&quot;</span><span class="p">:</span> <span class="n">lgd_list</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">}</span></div>

<div class="viewcode-block" id="GroupLight.bar_genecopy_per_species"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.GroupLight.bar_genecopy_per_species">[docs]</a>    <span class="k">def</span> <span class="nf">bar_genecopy_per_species</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_reset_counter</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_apply_filter</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">cl</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span>
                           <span class="bp">self</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">cl</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">))</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">most_common</span><span class="p">()</span>

        <span class="c1"># When data is empty, return an exception</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>

        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">data</span><span class="p">]]</span>
        <span class="n">ax_names</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Gene copies&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
                <span class="s2">&quot;labels&quot;</span><span class="p">:</span> <span class="n">x_labels</span><span class="p">,</span>
                <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="n">ax_names</span><span class="p">}</span></div></div>


<div class="viewcode-block" id="Group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group">[docs]</a><span class="k">class</span> <span class="nc">Group</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This represents the main object of the orthomcl toolbox module. It is</span>
<span class="sd">     initialized with a file name of a orthomcl groups file and provides</span>
<span class="sd">     several methods that act on that group file. To process multiple Group</span>
<span class="sd">     objects, see MultiGroups object &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">groups_file</span><span class="p">,</span> <span class="n">gene_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">species_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">project_prefix</span><span class="o">=</span><span class="s2">&quot;MyGroups&quot;</span><span class="p">):</span>

        <span class="c1"># Initializing thresholds. These may be set from the start, or using</span>
        <span class="c1">#  some method that uses them as arguments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="n">gene_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="n">species_threshold</span>

        <span class="c1"># Attribute containing the list of included species</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Attribute that will contain taxa to be excluded from analyses</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Attributes that will store the number (int) of cluster after gene and</span>
        <span class="c1"># species filter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Attribute containing the total number of sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Attribute containing the maximum number of extra copies found in the</span>
        <span class="c1"># clusters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="c1"># Attribute with name of the group file, which will be an ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">group_name</span> <span class="o">=</span> <span class="n">groups_file</span>
        <span class="c1"># Initialize the project prefix for possible output files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">project_prefix</span>
        <span class="c1"># Initialize attribute containing the original groups</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">groups</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Initialize atribute containing the groups filtered using the gene and</span>
        <span class="c1"># species threshold. This attribute can be updated at any time using</span>
        <span class="c1"># the update_filtered_group method</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># Parse groups file and populate groups attribute</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__parse_groups</span><span class="p">(</span><span class="n">groups_file</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__parse_groups</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">groups_file</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the ortholog clusters in the groups file and populates the</span>
<span class="sd">         self.groups list with Cluster objects for each line in the groups file.</span>
<span class="sd">        :param groups_file: File name for the orthomcl groups file</span>
<span class="sd">        :return: populates the groups attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">groups_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">groups_file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">groups_file</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">groups_file_handle</span><span class="p">:</span>
            <span class="n">cluster_object</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

            <span class="c1"># Add cluster to general group list</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster_object</span><span class="p">)</span>

            <span class="c1"># Update total sequence counter</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster_object</span><span class="o">.</span><span class="n">sequences</span><span class="p">)</span>

            <span class="c1"># Update maximum number of extra copies, if needed</span>
            <span class="k">if</span> <span class="nb">max</span><span class="p">(</span><span class="n">cluster_object</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">&gt;</span> \
                    <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> \
                    <span class="nb">max</span><span class="p">(</span><span class="n">cluster_object</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

            <span class="c1"># Update species_list attribute</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">cluster_object</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>

            <span class="c1"># If thresholds have been specified, update self.filtered_groups</span>
            <span class="c1"># attribute</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">:</span>
                <span class="n">cluster_object</span><span class="o">.</span><span class="n">apply_filter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span>
                                            <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">cluster_object</span><span class="o">.</span><span class="n">species_compliant</span> <span class="ow">and</span> \
                        <span class="n">cluster_object</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                    <span class="c1"># Add cluster to the filtered group list</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster_object</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">all_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="c1"># Update num_species_compliant attribute</span>
                <span class="k">if</span> <span class="n">cluster_object</span><span class="o">.</span><span class="n">species_compliant</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Update num_gene_compliant attribute</span>
                <span class="k">if</span> <span class="n">cluster_object</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

<div class="viewcode-block" id="Group.exclude_taxa"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.exclude_taxa">[docs]</a>    <span class="k">def</span> <span class="nf">exclude_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds a taxon_name to the excluded_taxa list and updates the</span>
<span class="sd">        filtered_groups list</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span>

        <span class="c1"># Storage variable for new filtered groups</span>
        <span class="n">filtered_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Reset max_extra_copy attribute</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
            <span class="n">cl</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">cl</span><span class="o">.</span><span class="n">iter_sequences</span> <span class="ow">and</span> <span class="n">cl</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">:</span>
                <span class="n">filtered_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span>

                <span class="c1"># Update maximum number of extra copies, if needed</span>
                <span class="k">if</span> <span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Update species_list</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span> <span class="o">-</span>
                                        <span class="nb">set</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">)))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span> <span class="o">=</span> <span class="n">filtered_groups</span></div>

<div class="viewcode-block" id="Group.get_filters"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.get_filters">[docs]</a>    <span class="k">def</span> <span class="nf">get_filters</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a tuple with the thresholds for max gene copies and min species</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span></div>

<div class="viewcode-block" id="Group.basic_group_statistics"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.basic_group_statistics">[docs]</a>    <span class="k">def</span> <span class="nf">basic_group_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates a basic table in list format containing basic</span>
<span class="sd">        information of the groups file (total number of clusters, total number</span>
<span class="sd">        of sequences, number of clusters below the gene threshold, number of</span>
<span class="sd">        clusters below the species threshold and number of clusters below the</span>
<span class="sd">        gene AND species threshold)</span>
<span class="sd">        :return: List containing number of</span>

<span class="sd">        [total clusters,</span>
<span class="sd">         total sequences,</span>
<span class="sd">         clusters above gene threshold,</span>
<span class="sd">         clusters above species threshold,</span>
<span class="sd">         clusters above gene and species threshold]</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Total number of clusters</span>
        <span class="n">total_cluster_num</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">)</span>

        <span class="c1"># Total number of sequenes</span>
        <span class="n">total_sequence_num</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_seqs</span>

        <span class="c1"># Gene compliant clusters</span>
        <span class="n">clusters_gene_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span>

        <span class="c1"># Species compliant clusters</span>
        <span class="n">clusters_species_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span>

        <span class="n">clusters_all_threshold</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="p">)</span>

        <span class="n">statistics</span> <span class="o">=</span> <span class="p">[</span><span class="n">total_cluster_num</span><span class="p">,</span> <span class="n">total_sequence_num</span><span class="p">,</span>
                      <span class="n">clusters_gene_threshold</span><span class="p">,</span> <span class="n">clusters_species_threshold</span><span class="p">,</span>
                      <span class="n">clusters_all_threshold</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">statistics</span></div>

<div class="viewcode-block" id="Group.paralog_per_species_statistic"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.paralog_per_species_statistic">[docs]</a>    <span class="k">def</span> <span class="nf">paralog_per_species_statistic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span>
                                      <span class="s2">&quot;Paralog_per_species.csv&quot;</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates a CSV table with information on the number of</span>
<span class="sd">        paralog clusters per species</span>
<span class="sd">        :param output_file_name: string. Name of the output csv file</span>
<span class="sd">        :param filt: Boolean. Whether to use the filtered groups (True) or</span>
<span class="sd">        total groups (False)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Setting which clusters to use</span>
        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="n">paralog_count</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">species</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">species</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">species</span> <span class="ow">in</span> <span class="n">paralog_count</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">species_frequency</span><span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">paralog_count</span><span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="c1"># Writing table</span>
        <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Species; Clusters with paralogs</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">paralog_count</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">; </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">species</span><span class="p">,</span> <span class="n">val</span><span class="p">))</span>

        <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="Group.export_filtered_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.export_filtered_group">[docs]</a>    <span class="k">def</span> <span class="nf">export_filtered_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;filtered_groups&quot;</span><span class="p">,</span>
                              <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">get_stats</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                              <span class="n">shared_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Export the filtered groups into a new file.</span>
<span class="sd">        :param output_file_name: string, name of the filtered groups file</span>
<span class="sd">        :param dest: string, path to directory where the filtered groups file</span>
<span class="sd">        will be created</span>
<span class="sd">        :param get_stats: Boolean, whether to return the basic count stats or</span>
<span class="sd">        not</span>
<span class="sd">        :param shared_namespace: Namespace object, for communicating with</span>
<span class="sd">        main process.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Exporting filtered orthologs&quot;</span>

            <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">get_stats</span><span class="p">:</span>
                <span class="n">all_orthologs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">)</span>
                <span class="n">sp_compliant</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">gene_compliant</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">final_orthologs</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                    <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> \
                        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">species_compliant</span> <span class="ow">and</span> <span class="n">cluster</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                    <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                                    <span class="n">cluster</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">)))</span>
                    <span class="k">if</span> <span class="n">get_stats</span><span class="p">:</span>
                        <span class="n">final_orthologs</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">get_stats</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">species_compliant</span><span class="p">:</span>
                        <span class="n">sp_compliant</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                        <span class="n">gene_compliant</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">get_stats</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">all_orthologs</span><span class="p">,</span> <span class="n">sp_compliant</span><span class="p">,</span> <span class="n">gene_compliant</span><span class="p">,</span>\
                           <span class="n">final_orthologs</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">OrthoGroupException</span><span class="p">(</span><span class="s2">&quot;The groups object must be filtered &quot;</span>
                                      <span class="s2">&quot;before using the export_filtered_group&quot;</span>
                                      <span class="s2">&quot;method&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Group.update_filters"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.update_filters">[docs]</a>    <span class="k">def</span> <span class="nf">update_filters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets new values for the self.species_threshold and self.gene_threshold</span>
<span class="sd">        and updates the filtered_group</span>
<span class="sd">        :param gn_filter: int. Maximum value for gene copies in cluster</span>
<span class="sd">        :param sp_filter:  int. Minimum value for species in cluster</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sp_filter</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gn_filter</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">update_filtered_group</span><span class="p">()</span></div>

<div class="viewcode-block" id="Group.update_filtered_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.update_filtered_group">[docs]</a>    <span class="k">def</span> <span class="nf">update_filtered_group</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This method creates a new filtered group variable, like</span>
<span class="sd">        export_filtered_group, but instead of writing into a new file, it</span>
<span class="sd">        replaces the self.filtered_groups variable</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Reset gene and species compliant counters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">apply_filter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">species_compliant</span> <span class="ow">and</span> <span class="n">cluster</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>

            <span class="c1"># Update num_species_compliant attribute</span>
            <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">species_compliant</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_species_compliant</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># Update num_gene_compliant attribute</span>
            <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">gene_compliant</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">num_gene_compliant</span> <span class="o">+=</span> <span class="mi">1</span></div>

<div class="viewcode-block" id="Group.retrieve_sequences"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.retrieve_sequences">[docs]</a>    <span class="k">def</span> <span class="nf">retrieve_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span>
                             <span class="n">filt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shared_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        When provided with a database in Fasta format, this will use the</span>
<span class="sd">        Alignment object to retrieve sequences</span>
<span class="sd">        :param database: String. Fasta file</span>
<span class="sd">        :param dest: directory where files will be save</span>
<span class="sd">        :param mode: string, whether to retrieve sequences to a file (&#39;fasta&#39;),</span>
<span class="sd">        or a dictionary (&#39;dict&#39;)</span>
<span class="sd">        :param filt: Boolean. Whether to use the filtered groups (True) or</span>
<span class="sd">        total groups (False)</span>
<span class="sd">        :param shared_namespace: Namespace object. This argument is meant for</span>
<span class="sd">        when fast are retrieved in a background process, where there is a need</span>
<span class="sd">        to update the main process of the changes in this method</span>
<span class="sd">        :param dest: string. Path to directory where the retrieved sequences</span>
<span class="sd">        will be created.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;dict&quot;</span><span class="p">:</span>
            <span class="n">seq_storage</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s2">&quot;Orthologs&quot;</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;Orthologs&quot;</span><span class="p">)</span>

        <span class="c1"># Update method progress</span>
        <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Creating database&quot;</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating db&quot;</span><span class="p">)</span>
        <span class="c1"># Check what type of database was provided</span>
        <span class="c1">#TODO: Add exception handling if file is not parsed with Aligment</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">db_aln</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
            <span class="k">except</span> <span class="p">(</span><span class="ne">EnvironmentError</span><span class="p">,</span> <span class="n">pickle</span><span class="o">.</span><span class="n">UnpicklingError</span><span class="p">):</span>
                <span class="n">db_aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">database</span><span class="p">)</span>
                <span class="n">db_aln</span> <span class="o">=</span> <span class="n">db_aln</span><span class="o">.</span><span class="n">alignment</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">database</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="n">db_aln</span> <span class="o">=</span> <span class="n">database</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">OrthoGroupException</span><span class="p">(</span><span class="s2">&quot;The input database is neither a string&quot;</span>
                                      <span class="s2">&quot;nor a dictionary object&quot;</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Retrieving seqs&quot;</span><span class="p">)</span>
        <span class="c1"># Update method progress</span>
        <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
            <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Retrieving sequences&quot;</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
                <span class="n">shared_namespace</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;dict&quot;</span><span class="p">:</span>
                <span class="n">seq_storage</span><span class="p">[</span><span class="n">cluster</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;.fas&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">sequence_id</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="n">db_aln</span><span class="p">[</span><span class="n">sequence_id</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;fasta&quot;</span><span class="p">:</span>
                    <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">seq</span><span class="p">))</span>
                <span class="k">elif</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;dict&quot;</span><span class="p">:</span>
                    <span class="n">seq_storage</span><span class="p">[</span><span class="n">cluster</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">sequence_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                                                      <span class="n">seq</span><span class="p">])</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;dict&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">seq_storage</span></div>

<div class="viewcode-block" id="Group.bar_species_distribution"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.bar_species_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">bar_species_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                 <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;Species_distribution&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a bar plot with the distribution of species numbers across</span>
<span class="sd">        clusters</span>
<span class="sd">        :param dest: string, destination directory</span>
<span class="sd">        :param filt: Boolean, whether or not to use the filtered groups.</span>
<span class="sd">        :param output_file_name: string, name of the output file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Determine which groups to use</span>
        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                             <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]))</span>

        <span class="c1"># Transform data into histogram-like</span>
        <span class="n">transform_data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">transform_data</span><span class="p">)]</span>
        <span class="n">y_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transform_data</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Sort lists</span>
        <span class="n">x_labels</span><span class="p">,</span> <span class="n">y_vals</span> <span class="o">=</span> <span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span>
                                                             <span class="n">y_vals</span><span class="p">))))</span>
        <span class="c1"># Convert label to strings</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_labels</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="c1"># Create plot</span>
        <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">bar_plot</span><span class="p">([</span><span class="n">y_vals</span><span class="p">],</span> <span class="n">x_labels</span><span class="p">,</span>
                                 <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Taxa frequency distribution&quot;</span><span class="p">,</span>
                                 <span class="n">ax_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Number of taxa&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">])</span>
        <span class="n">b_plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
                      <span class="n">dpi</span><span class="o">=</span><span class="mi">400</span><span class="p">)</span>

        <span class="c1"># Create table</span>
        <span class="n">table_list</span> <span class="o">=</span> <span class="p">[[</span><span class="s2">&quot;Number of species&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span> <span class="n">y_vals</span><span class="p">):</span>
            <span class="n">table_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">table_list</span></div>

<div class="viewcode-block" id="Group.bar_genecopy_distribution"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.bar_genecopy_distribution">[docs]</a>    <span class="k">def</span> <span class="nf">bar_genecopy_distribution</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;Gene_copy_distribution.png&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a bar plot with the distribution of gene copies across</span>
<span class="sd">        clusters</span>
<span class="sd">        :param dest: string, destination directory</span>
<span class="sd">        :param filt: Boolean, whether or not to use the filtered groups.</span>
<span class="sd">        :param output_file_name: string, name of the output file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Determin which groups to use</span>
        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="c1"># Get max number of copies</span>
            <span class="n">max_copies</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">cl</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">max_copies</span><span class="p">)</span>

        <span class="c1"># Transform data into histogram-like</span>
        <span class="n">transform_data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">transform_data</span><span class="p">)]</span>
        <span class="n">y_vals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transform_data</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Sort lists</span>
        <span class="n">x_labels</span><span class="p">,</span> <span class="n">y_vals</span> <span class="o">=</span> <span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span>
                                                             <span class="n">y_vals</span><span class="p">))))</span>
        <span class="c1"># Convert label to strings</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_labels</span><span class="p">]</span>

        <span class="c1"># Create plot</span>
        <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">bar_plot</span><span class="p">([</span><span class="n">y_vals</span><span class="p">],</span> <span class="n">x_labels</span><span class="p">,</span>
                    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Gene copy distribution&quot;</span><span class="p">,</span>
                    <span class="n">ax_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Number of gene copies&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">],</span>
                    <span class="n">reverse_x</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">b_plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
                      <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">x_labels</span><span class="p">)</span> <span class="o">/</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

        <span class="c1"># Create table</span>
        <span class="n">table_list</span> <span class="o">=</span> <span class="p">[[</span><span class="s2">&quot;Number of gene copies&quot;</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">x_labels</span><span class="p">,</span> <span class="n">y_vals</span><span class="p">):</span>
            <span class="n">table_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">table_list</span></div>

<div class="viewcode-block" id="Group.bar_species_coverage"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.Group.bar_species_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">bar_species_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">filt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                             <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;Species_coverage&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a stacked bar plot with the proportion of</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Determine which groups to use</span>
        <span class="k">if</span> <span class="n">filt</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filtered_groups</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">+=</span> <span class="n">Counter</span><span class="p">(</span><span class="nb">dict</span><span class="p">((</span><span class="n">x</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">cl</span><span class="o">.</span><span class="n">species_frequency</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                            <span class="k">if</span> <span class="n">y</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>

        <span class="n">xlabels</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span> <span class="o">-</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                      <span class="n">data</span><span class="o">.</span><span class="n">values</span><span class="p">()]]</span>

        <span class="n">lgd_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Available data&quot;</span><span class="p">,</span> <span class="s2">&quot;Missing data&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">bar_plot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">xlabels</span><span class="p">,</span> <span class="n">lgd_list</span><span class="o">=</span><span class="n">lgd_list</span><span class="p">,</span>
                              <span class="n">ax_names</span><span class="o">=</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Ortholog frequency&quot;</span><span class="p">])</span>
        <span class="n">b_plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span>
                      <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="s2">&quot;&quot;</span></div></div>


<div class="viewcode-block" id="MultiGroups"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups">[docs]</a><span class="k">class</span> <span class="nc">MultiGroups</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Creates an object composed of multiple Group objects &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">groups_files</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">species_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">project_prefix</span><span class="o">=</span><span class="s2">&quot;MyGroups&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param groups_files: A list containing the file names of the multiple</span>
<span class="sd">        group files</span>
<span class="sd">        :return: Populates the self.multiple_groups attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># If a MultiGroups is initialized with duplicate Group objects, these</span>
        <span class="c1"># will be stored in a list. If all Group objects are unique, the list</span>
        <span class="c1"># will remain empty</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Initializing thresholds. These may be set from the start, or using</span>
        <span class="c1"># some method that uses them as arguments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="n">gene_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="n">species_threshold</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">project_prefix</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filters</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">if</span> <span class="n">groups_files</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">group_file</span> <span class="ow">in</span> <span class="n">groups_files</span><span class="p">:</span>

                <span class="c1"># If group_file is already a Group object, just add it</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">group_file</span><span class="p">,</span> <span class="n">Group</span><span class="p">):</span>
                    <span class="c1"># Check for duplicate group files</span>
                    <span class="n">group_object</span> <span class="o">=</span> <span class="n">Group</span><span class="p">(</span><span class="n">group_file</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span>
                                         <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">group_object</span> <span class="o">=</span> <span class="n">group_file</span>

                <span class="k">if</span> <span class="n">group_object</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group_object</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="n">group_object</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_object</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">[</span><span class="n">group_object</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span>
                                                <span class="nb">len</span><span class="p">(</span><span class="n">group_object</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">)</span>

<div class="viewcode-block" id="MultiGroups.iter_gnames"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.iter_gnames">[docs]</a>    <span class="k">def</span> <span class="nf">iter_gnames</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">)</span></div>

<div class="viewcode-block" id="MultiGroups.get_gnames"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.get_gnames">[docs]</a>    <span class="k">def</span> <span class="nf">get_gnames</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">]</span></div>

<div class="viewcode-block" id="MultiGroups.add_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.add_group">[docs]</a>    <span class="k">def</span> <span class="nf">add_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds a group object</span>
<span class="sd">        :param group_obj: Group object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check for duplicate groups</span>
        <span class="k">if</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_obj</span></div>

<div class="viewcode-block" id="MultiGroups.remove_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.remove_group">[docs]</a>    <span class="k">def</span> <span class="nf">remove_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Removes a group object according to its name</span>
<span class="sd">        :param group_id: string, name matching a Group object name attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">group_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">:</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="n">group_id</span><span class="p">]</span></div>

<div class="viewcode-block" id="MultiGroups.get_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.get_group">[docs]</a>    <span class="k">def</span> <span class="nf">get_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a group object based on its name. If the name does not match</span>
<span class="sd">        any group object, returns None</span>
<span class="sd">        :param group_id: string. Name of group object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="n">group_id</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">return</span></div>

<div class="viewcode-block" id="MultiGroups.add_multigroups"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.add_multigroups">[docs]</a>    <span class="k">def</span> <span class="nf">add_multigroups</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">multigroup_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merges a MultiGroup object</span>
<span class="sd">        :param multigroup_obj: MultiGroup object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">group_obj</span> <span class="ow">in</span> <span class="n">multigroup_obj</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_group</span><span class="p">(</span><span class="n">group_obj</span><span class="p">)</span></div>

<div class="viewcode-block" id="MultiGroups.update_filters"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.update_filters">[docs]</a>    <span class="k">def</span> <span class="nf">update_filters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">,</span> <span class="n">group_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                       <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This will not change the Group object themselves, only the filter</span>
<span class="sd">        mapping. The filter is only applied when the Group object is retrieved</span>
<span class="sd">        to reduce computations</span>
<span class="sd">        :param gn_filter: int, filter for max gene copies</span>
<span class="sd">        :param sp_filter: int, filter for min species</span>
<span class="sd">        :param group_names: list, with names of group objects</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">group_names</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">group_name</span> <span class="ow">in</span> <span class="n">group_names</span><span class="p">:</span>
                <span class="c1"># Get group object</span>
                <span class="n">group_obj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span>
                <span class="c1"># Define filters</span>
                <span class="n">gn_filter</span> <span class="o">=</span> <span class="n">gn_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> <span class="mi">1</span>
                <span class="n">sp_filter</span> <span class="o">=</span> <span class="n">sp_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> \
                    <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span>
                <span class="c1"># Update Group object with new filters</span>
                <span class="n">group_obj</span><span class="o">.</span><span class="n">update_filters</span><span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span>
                <span class="c1"># Update filter map</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">group_name</span><span class="p">,</span> <span class="n">group_obj</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="c1"># Define filters</span>
                <span class="n">gn_filter</span> <span class="o">=</span> <span class="n">gn_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> <span class="mi">1</span>
                <span class="n">sp_filter</span> <span class="o">=</span> <span class="n">sp_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> \
                    <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span>
                <span class="c1"># Update Group object with new filters</span>
                <span class="n">group_obj</span><span class="o">.</span><span class="n">update_filters</span><span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span>
                <span class="c1"># Update filter map</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span></div>

<div class="viewcode-block" id="MultiGroups.basic_multigroup_statistics"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.basic_multigroup_statistics">[docs]</a>    <span class="k">def</span> <span class="nf">basic_multigroup_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span>
                                    <span class="s2">&quot;multigroup_base_statistics.csv&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param output_file_name:</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Creates the storage for the statistics of the several files</span>
        <span class="n">statistics_storage</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">:</span>
            <span class="n">group_statistics</span> <span class="o">=</span> <span class="n">group</span><span class="o">.</span><span class="n">basic_group_statistics</span><span class="p">()</span>
            <span class="n">statistics_storage</span><span class="p">[</span><span class="n">group</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_statistics</span>

        <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.&quot;</span> <span class="o">+</span> <span class="n">output_file_name</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;Group file; Total clusters; Total sequences; &quot;</span>
                            <span class="s2">&quot;Clusters below gene threshold; Clusters above &quot;</span>
                            <span class="s2">&quot;species threshold; Clusters below gene and above&quot;</span>
                            <span class="s2">&quot; species thresholds</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">vals</span> <span class="ow">in</span> <span class="n">statistics_storage</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">; </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span>
                                                               <span class="ow">in</span> <span class="n">vals</span><span class="p">])))</span>

        <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="MultiGroups.bar_orthologs"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.bar_orthologs">[docs]</a>    <span class="k">def</span> <span class="nf">bar_orthologs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;Final_orthologs&quot;</span><span class="p">,</span>
                             <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">stats</span><span class="o">=</span><span class="s2">&quot;total&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a bar plot with the final ortholog values for each group file</span>
<span class="sd">        :param output_file_name: string. Name of output file</span>
<span class="sd">        :param dest: string. output directory</span>
<span class="sd">        :param stats: string. The statistics that should be used to generate</span>
<span class="sd">        the bar plot. Options are:</span>
<span class="sd">            ..: &quot;1&quot;: Total orthologs</span>
<span class="sd">            ..: &quot;2&quot;: Species compliant orthologs</span>
<span class="sd">            ..: &quot;3&quot;: Gene compliant orthologs</span>
<span class="sd">            ..: &quot;4&quot;: Final orthologs</span>
<span class="sd">            ..: &quot;all&quot;: All of the above</span>
<span class="sd">            Multiple combinations can be provided, for instance: &quot;123&quot; will</span>
<span class="sd">            display bars for total, species compliant and gene compliant stats</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Stores the x-axis labels</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Stores final ortholog values for all 4 possible data sets</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]]</span>
        <span class="n">lgd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Total orthologs&quot;</span><span class="p">,</span> <span class="s2">&quot;After species filter&quot;</span><span class="p">,</span> <span class="s2">&quot;After gene filter&quot;</span><span class="p">,</span>
               <span class="s2">&quot;Final orthologs&quot;</span><span class="p">]</span>

        <span class="c1"># Get final ortholog values</span>
        <span class="k">for</span> <span class="n">g_obj</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">:</span>

            <span class="n">x_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_obj</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">sep</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="c1"># Populate total orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;1&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">g_obj</span><span class="o">.</span><span class="n">groups</span><span class="p">))</span>
            <span class="c1"># Populate species compliant orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;2&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_obj</span><span class="o">.</span><span class="n">num_species_compliant</span><span class="p">)</span>
            <span class="c1"># Populate gene compliant orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;3&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_obj</span><span class="o">.</span><span class="n">num_gene_compliant</span><span class="p">)</span>
            <span class="c1"># Populate final orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;4&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">g_obj</span><span class="o">.</span><span class="n">filtered_groups</span><span class="p">))</span>

        <span class="c1"># Filter valid data sets</span>
        <span class="n">lgd_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lgd</span> <span class="k">if</span> <span class="n">vals</span><span class="p">[</span><span class="n">lgd</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">)]]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">l</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">vals</span> <span class="k">if</span> <span class="n">l</span><span class="p">]</span>

        <span class="c1"># Create plot</span>
        <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span> <span class="o">=</span> <span class="n">multi_bar_plot</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">x_labels</span><span class="p">,</span> <span class="n">lgd_list</span><span class="o">=</span><span class="n">lgd_list</span><span class="p">)</span>
        <span class="n">b_plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span>
                      <span class="n">bbox_extra_artists</span><span class="o">=</span><span class="p">(</span><span class="n">lgd</span><span class="p">,),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>

        <span class="c1"># Create table list object</span>
        <span class="n">table_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Create header</span>
        <span class="n">table_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s2">&quot;&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">x_labels</span><span class="p">)</span>
        <span class="c1"># Create content</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vals</span><span class="p">)):</span>
            <span class="n">table_list</span> <span class="o">+=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[[</span><span class="n">lgd_list</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">+</span> <span class="n">vals</span><span class="p">[</span><span class="n">i</span><span class="p">]]]</span>

        <span class="k">return</span> <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">table_list</span></div>

<div class="viewcode-block" id="MultiGroups.group_overlap"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroups.group_overlap">[docs]</a>    <span class="k">def</span> <span class="nf">group_overlap</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This will find the overlap of orthologs between two group files.</span>
<span class="sd">        THIS METHOD IS TEMPORARY AND EXPERIMENTAL</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span> <span class="nf">parse_groups</span><span class="p">(</span><span class="n">group_obj</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Returns a list with the sorted ortholog clusters</span>
<span class="sd">            &quot;&quot;&quot;</span>

            <span class="n">storage</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
                <span class="n">storage</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">))</span>

            <span class="k">return</span> <span class="n">storage</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">SystemExit</span><span class="p">(</span><span class="s2">&quot;This method can only be used with two group &quot;</span>
                             <span class="s2">&quot;files&quot;</span><span class="p">)</span>

        <span class="n">group1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">group2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">multiple_groups</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">group1_list</span> <span class="o">=</span> <span class="n">parse_groups</span><span class="p">(</span><span class="n">group1</span><span class="p">)</span>
        <span class="n">group2_list</span> <span class="o">=</span> <span class="n">parse_groups</span><span class="p">(</span><span class="n">group2</span><span class="p">)</span>

        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">group1_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">group2_list</span><span class="p">:</span>
                <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="MultiGroupsLight"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight">[docs]</a><span class="k">class</span> <span class="nc">MultiGroupsLight</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates an object composed of multiple Group objects like MultiGroups.</span>
<span class="sd">    However, instead of storing the groups in memory, these are shelved in</span>
<span class="sd">    the disk</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># The report calls available</span>
    <span class="n">calls</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;bar_genecopy_distribution&#39;</span><span class="p">,</span>
             <span class="s1">&#39;bar_species_distribution&#39;</span><span class="p">,</span>
             <span class="s1">&#39;bar_species_coverage&#39;</span><span class="p">,</span>
             <span class="s1">&#39;bar_genecopy_per_species&#39;</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">db_path</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">species_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">project_prefix</span><span class="o">=</span><span class="s2">&quot;MyGroups&quot;</span><span class="p">,</span>
                 <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param groups: A list containing the file names of the multiple</span>
<span class="sd">        group files</span>
<span class="sd">        :return: Populates the self.multiple_groups attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">db_path</span> <span class="o">=</span> <span class="n">db_path</span>

        <span class="c1"># If a MultiGroups is initialized with duplicate Group objects, their</span>
        <span class="c1"># names will be stored in a list. If all Group objects are unique, the</span>
        <span class="c1"># list will remain empty</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">groups</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">groups_stats</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Attribute that will store the paths of badly formated group files</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bad_groups</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Initializing thresholds. These may be set from the start, or using</span>
        <span class="c1"># some method that uses them as arguments</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="n">gene_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="n">species_threshold</span>

        <span class="c1"># Initializing mapping of group filters to their names. Should be</span>
        <span class="c1"># something like {&quot;groupA&quot;: (1, 10)}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filters</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># This attribute will contain a dictionary with the maximum extra copies</span>
        <span class="c1"># for each group object</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># This attribute will contain a list with the number of species for</span>
        <span class="c1"># each group object, excluding replicates. If a MultiGroupLight object</span>
        <span class="c1"># contains Group objects with different taxa numbers, this attribute</span>
        <span class="c1"># can be used to issue a warning</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_number</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">project_prefix</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">group_file</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                <span class="c1"># If group_file is already a Group object, just add it</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">group_file</span><span class="p">,</span> <span class="n">GroupLight</span><span class="p">):</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="n">ns</span><span class="o">.</span><span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">group_object</span> <span class="o">=</span> <span class="n">GroupLight</span><span class="p">(</span><span class="n">group_file</span><span class="p">,</span>
                                                  <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span><span class="p">,</span>
                                                  <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">,</span>
                                                  <span class="n">ns</span><span class="o">=</span><span class="n">ns</span><span class="p">)</span>
                    <span class="k">except</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">bad_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group_file</span><span class="p">)</span>
                        <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">group_object</span> <span class="o">=</span> <span class="n">group_file</span>

                <span class="c1"># Check for duplicate group files</span>
                <span class="k">if</span> <span class="n">group_object</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group_file</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">add_group</span><span class="p">(</span><span class="n">group_object</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">yield</span> <span class="n">k</span><span class="p">,</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>

<div class="viewcode-block" id="MultiGroupsLight.clear_groups"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.clear_groups">[docs]</a>    <span class="k">def</span> <span class="nf">clear_groups</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Clears the current MultiGroupsLight object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">groups</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">groups_stats</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filters</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species_number</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_threshold</span> <span class="o">=</span> <span class="mi">0</span></div>

<div class="viewcode-block" id="MultiGroupsLight.add_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.add_group">[docs]</a>    <span class="k">def</span> <span class="nf">add_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds a group object</span>
<span class="sd">        :param group_obj: Group object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check for duplicate groups</span>
        <span class="k">if</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
            <span class="n">gpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db_path</span><span class="p">,</span>
                    <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">string</span><span class="o">.</span><span class="n">ascii_uppercase</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span>
                            <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">)))</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">group_obj</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">gpath</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">gpath</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">[</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">),</span> <span class="p">[])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">[</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">max_extra_copy</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_number</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">species_number</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">duplicate_groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>

<div class="viewcode-block" id="MultiGroupsLight.remove_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.remove_group">[docs]</a>    <span class="k">def</span> <span class="nf">remove_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Removes a group object according to its name</span>
<span class="sd">        :param group_id: string, name matching a Group object name attribute</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">group_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_id</span><span class="p">])</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_id</span><span class="p">]</span></div>

<div class="viewcode-block" id="MultiGroupsLight.get_group"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.get_group">[docs]</a>    <span class="k">def</span> <span class="nf">get_group</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a group object based on its name. If the name does not match</span>
<span class="sd">        any group object, returns None</span>
<span class="sd">        :param group_id: string. Name of group object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_id</span><span class="p">],</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">return</span></div>

<div class="viewcode-block" id="MultiGroupsLight.add_multigroups"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.add_multigroups">[docs]</a>    <span class="k">def</span> <span class="nf">add_multigroups</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">multigroup_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merges a MultiGroup object</span>
<span class="sd">        :param multigroup_obj: MultiGroup object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">group_obj</span> <span class="ow">in</span> <span class="n">multigroup_obj</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_group</span><span class="p">(</span><span class="n">group_obj</span><span class="p">)</span></div>

<div class="viewcode-block" id="MultiGroupsLight.update_filters"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.update_filters">[docs]</a>    <span class="k">def</span> <span class="nf">update_filters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">,</span> <span class="n">excluded_taxa</span><span class="p">,</span>
                       <span class="n">group_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This will not change the Group object themselves, only the filter</span>
<span class="sd">        mapping. The filter is only applied when the Group object is retrieved</span>
<span class="sd">        to reduce computations</span>

<span class="sd">        :param gn_filter: int, filter for max gene copies</span>
<span class="sd">        :param sp_filter: int, filter for min species</span>
<span class="sd">        :param group_names: list, with names of group objects</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># There are no groups to update</span>
        <span class="k">if</span> <span class="n">group_names</span> <span class="o">==</span> <span class="p">[]:</span>
            <span class="k">return</span>

        <span class="k">if</span> <span class="n">group_names</span><span class="p">:</span>
            <span class="n">glist</span> <span class="o">=</span> <span class="n">group_names</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">glist</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups</span>

        <span class="k">for</span> <span class="n">group_name</span> <span class="ow">in</span> <span class="n">glist</span><span class="p">:</span>
            <span class="c1"># Get group object</span>
            <span class="n">group_obj</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_name</span><span class="p">],</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>

            <span class="c1"># Define excluded taxa</span>
            <span class="n">group_obj</span><span class="o">.</span><span class="n">exclude_taxa</span><span class="p">(</span><span class="n">excluded_taxa</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Define filters</span>
            <span class="n">gn_filter</span> <span class="o">=</span> <span class="n">gn_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> <span class="mi">1</span>
            <span class="n">sp_filter</span> <span class="o">=</span> <span class="n">sp_filter</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">default</span> <span class="k">else</span> \
                <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span>

            <span class="c1"># Correct maximum filter values after excluding taxa</span>
            <span class="n">gn_filter</span> <span class="o">=</span> <span class="n">gn_filter</span> <span class="k">if</span> <span class="n">gn_filter</span> <span class="o">&lt;=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">max_extra_copy</span> \
                <span class="k">else</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">max_extra_copy</span>
            <span class="n">sp_filter</span> <span class="o">=</span> <span class="n">sp_filter</span> <span class="k">if</span> <span class="n">sp_filter</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span> \
                <span class="k">else</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">)</span>

            <span class="c1"># Update Group object with new filters</span>
            <span class="n">group_obj</span><span class="o">.</span><span class="n">update_filters</span><span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">sp_filter</span><span class="p">)</span>
            <span class="c1"># Update group stats</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">get_multigroup_statistics</span><span class="p">(</span><span class="n">group_obj</span><span class="p">)</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">group_obj</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">group_name</span><span class="p">],</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
            <span class="c1"># Update filter map</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">filters</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">gn_filter</span><span class="p">,</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">species_threshold</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">taxa_list</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">excluded_taxa</span><span class="p">[</span><span class="n">group_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">excluded_taxa</span></div>

<div class="viewcode-block" id="MultiGroupsLight.get_multigroup_statistics"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.get_multigroup_statistics">[docs]</a>    <span class="k">def</span> <span class="nf">get_multigroup_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_obj</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">basic_group_statistics</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">groups_stats</span><span class="p">[</span><span class="n">group_obj</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="n">stats</span><span class="p">,</span>
                                        <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">species_list</span><span class="p">,</span>
                                        <span class="s2">&quot;max_copies&quot;</span><span class="p">:</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">max_extra_copy</span><span class="p">}</span></div>

<div class="viewcode-block" id="MultiGroupsLight.bar_orthologs"><a class="viewcode-back" href="../../../trifusion.ortho.OrthomclToolbox.html#trifusion.ortho.OrthomclToolbox.MultiGroupsLight.bar_orthologs">[docs]</a>    <span class="k">def</span> <span class="nf">bar_orthologs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">group_names</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_file_name</span><span class="o">=</span><span class="s2">&quot;Final_orthologs&quot;</span><span class="p">,</span>
                             <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">stats</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a bar plot with the final ortholog values for each group file</span>
<span class="sd">        :param group_names: list. If None, all groups in self.group_stats will</span>
<span class="sd">        be used to generate the plot. Else, only the groups with the names in</span>
<span class="sd">        the list will be plotted.</span>
<span class="sd">        :param output_file_name: string. Name of output file</span>
<span class="sd">        :param dest: string. output directory</span>
<span class="sd">        :param stats: string. The statistics that should be used to generate</span>
<span class="sd">        the bar plot. Options are:</span>
<span class="sd">            ..: &quot;1&quot;: Total orthologs</span>
<span class="sd">            ..: &quot;2&quot;: Species compliant orthologs</span>
<span class="sd">            ..: &quot;3&quot;: Gene compliant orthologs</span>
<span class="sd">            ..: &quot;4&quot;: Final orthologs</span>
<span class="sd">            ..: &quot;all&quot;: All of the above</span>
<span class="sd">            Multiple combinations can be provided, for instance: &quot;123&quot; will</span>
<span class="sd">            display bars for total, species compliant and gene compliant stats</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Stores the x-axis labels</span>
        <span class="n">x_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Stores final ortholog values for all 4 possible data sets</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]]</span>
        <span class="n">lgd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Total orthologs&quot;</span><span class="p">,</span> <span class="s2">&quot;After species filter&quot;</span><span class="p">,</span> <span class="s2">&quot;After gene filter&quot;</span><span class="p">,</span>
               <span class="s2">&quot;Final orthologs&quot;</span><span class="p">]</span>

        <span class="c1"># Determine which groups will be plotted</span>
        <span class="k">if</span> <span class="n">group_names</span><span class="p">:</span>
            <span class="n">groups_lst</span> <span class="o">=</span> <span class="n">group_names</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">groups_lst</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups_stats</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">gname</span> <span class="ow">in</span> <span class="n">groups_lst</span><span class="p">:</span>

            <span class="n">gstats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">groups_stats</span><span class="p">[</span><span class="n">gname</span><span class="p">]</span>

            <span class="n">x_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">sep</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
            <span class="c1"># Populate total orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;1&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gstats</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="c1"># Populate species compliant orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;2&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gstats</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">][</span><span class="mi">3</span><span class="p">])</span>
            <span class="c1"># Populate gene compliant orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;3&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gstats</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">][</span><span class="mi">2</span><span class="p">])</span>
            <span class="c1"># Populate final orthologs</span>
            <span class="k">if</span> <span class="s2">&quot;4&quot;</span> <span class="ow">in</span> <span class="n">stats</span> <span class="ow">or</span> <span class="n">stats</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
                <span class="n">vals</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gstats</span><span class="p">[</span><span class="s2">&quot;stats&quot;</span><span class="p">][</span><span class="mi">4</span><span class="p">])</span>

        <span class="c1"># Filter valid data sets</span>
        <span class="n">lgd_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lgd</span> <span class="k">if</span> <span class="n">vals</span><span class="p">[</span><span class="n">lgd</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">x</span><span class="p">)]]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">[</span><span class="n">l</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">vals</span> <span class="k">if</span> <span class="n">l</span><span class="p">]</span>

        <span class="c1"># Create plot</span>
        <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span> <span class="o">=</span> <span class="n">multi_bar_plot</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">x_labels</span><span class="p">,</span> <span class="n">lgd_list</span><span class="o">=</span><span class="n">lgd_list</span><span class="p">)</span>
        <span class="n">b_plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">),</span>
                      <span class="n">bbox_extra_artists</span><span class="o">=</span><span class="p">(</span><span class="n">lgd</span><span class="p">,),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

        <span class="c1"># Create table list object</span>
        <span class="n">table_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># Create header</span>
        <span class="n">table_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s2">&quot;&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">x_labels</span><span class="p">)</span>
        <span class="c1"># Create content</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">vals</span><span class="p">)):</span>
            <span class="n">table_list</span> <span class="o">+=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[[</span><span class="n">lgd_list</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="o">+</span> <span class="n">vals</span><span class="p">[</span><span class="n">i</span><span class="p">]]]</span>

        <span class="k">return</span> <span class="n">b_plt</span><span class="p">,</span> <span class="n">lgd</span><span class="p">,</span> <span class="n">table_list</span></div></div>


<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Diogo N. Silva&quot;</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Diogo N. Silva.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.5.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>