

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>trifusion.ortho.protein2dna &mdash; TriFusion 0.5.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="TriFusion 0.5.5 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> TriFusion
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../trifusion.html">TriFusion modules and API reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">TriFusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>trifusion.ortho.protein2dna</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for trifusion.ortho.protein2dna</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1">#</span>
<span class="c1">#  Copyright 2012 Unknown &lt;diogo@arch&gt;</span>
<span class="c1">#</span>
<span class="c1">#  This program is free software; you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation; either version 2 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1">#  This program is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with this program; if not, write to the Free Software</span>
<span class="c1">#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,</span>
<span class="c1">#  MA 02110-1301, USA.</span>
<span class="c1">#</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module deals with the conversion of protein sequences into their</span>
<span class="sd">corresponding nucleotide sequences. Since the conversion from protein to DNA</span>
<span class="sd">cannot be made without knowing the nucleotide sequence, this module contains</span>
<span class="sd">functions that compile and store DNA sequences, convert them into amino acid</span>
<span class="sd">sequences and then tries to match them to the original protein sequences.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">process.error_handling</span> <span class="k">import</span> <span class="n">KillByUser</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">trifusion.process.error_handling</span> <span class="k">import</span> <span class="n">KillByUser</span>

<span class="kn">from</span> <span class="nn">os.path</span> <span class="k">import</span> <span class="n">join</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="n">dna_map</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;ATA&#39;</span><span class="p">:</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATC&#39;</span><span class="p">:</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATT&#39;</span><span class="p">:</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ACA&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACC&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACG&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;ACT&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span>
    <span class="s1">&#39;AAC&#39;</span><span class="p">:</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;AAT&#39;</span><span class="p">:</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;AAA&#39;</span><span class="p">:</span> <span class="s1">&#39;K&#39;</span><span class="p">,</span> <span class="s1">&#39;AAG&#39;</span><span class="p">:</span> <span class="s1">&#39;K&#39;</span><span class="p">,</span>
    <span class="s1">&#39;AGC&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;AGT&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;AGA&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;AGG&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CTA&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTC&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTG&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;CTT&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CCA&#39;</span><span class="p">:</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCC&#39;</span><span class="p">:</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCG&#39;</span><span class="p">:</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;CCT&#39;</span><span class="p">:</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CAC&#39;</span><span class="p">:</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;CAT&#39;</span><span class="p">:</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;CAA&#39;</span><span class="p">:</span> <span class="s1">&#39;Q&#39;</span><span class="p">,</span> <span class="s1">&#39;CAG&#39;</span><span class="p">:</span> <span class="s1">&#39;Q&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CGA&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGC&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGG&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;CGT&#39;</span><span class="p">:</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span>
    <span class="s1">&#39;GTA&#39;</span><span class="p">:</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTC&#39;</span><span class="p">:</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTG&#39;</span><span class="p">:</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;GTT&#39;</span><span class="p">:</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span>
    <span class="s1">&#39;GCA&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCC&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCG&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;GCT&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span>
    <span class="s1">&#39;GAC&#39;</span><span class="p">:</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;GAT&#39;</span><span class="p">:</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;GAA&#39;</span><span class="p">:</span> <span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;GAG&#39;</span><span class="p">:</span> <span class="s1">&#39;E&#39;</span><span class="p">,</span>
    <span class="s1">&#39;GGA&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGC&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;GGT&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TCA&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCC&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCG&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;TCT&#39;</span><span class="p">:</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TTC&#39;</span><span class="p">:</span> <span class="s1">&#39;F&#39;</span><span class="p">,</span> <span class="s1">&#39;TTT&#39;</span><span class="p">:</span> <span class="s1">&#39;F&#39;</span><span class="p">,</span> <span class="s1">&#39;TTA&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;TTG&#39;</span><span class="p">:</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TAC&#39;</span><span class="p">:</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;TAT&#39;</span><span class="p">:</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;TAG&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span>
    <span class="s1">&#39;TGC&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;TGT&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">:</span> <span class="s1">&#39;W&#39;</span><span class="p">}</span>


<div class="viewcode-block" id="translate"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.translate">[docs]</a><span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Translates a DNA string into an amino acid sequence</span>
<span class="sd">    :param sequence: string. DNA sequence</span>
<span class="sd">    :return: String. Protein sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">aa_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span> <span class="mi">3</span><span class="p">):</span>
        <span class="n">codon</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>

        <span class="c1"># Check if codon is multiple of 3. If not, place an X (missing data)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="c1"># Check if there is missing data on the codon. If so, place an X</span>
            <span class="k">if</span> <span class="s2">&quot;n&quot;</span> <span class="ow">in</span> <span class="n">codon</span><span class="p">:</span>
                <span class="n">aa</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">aa</span> <span class="o">=</span> <span class="n">dna_map</span><span class="p">[</span><span class="n">codon</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">aa</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">aa</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="n">aa_sequence</span> <span class="o">+=</span> <span class="n">aa</span>

    <span class="k">return</span> <span class="n">aa_sequence</span></div>


<div class="viewcode-block" id="create_db"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.create_db">[docs]</a><span class="k">def</span> <span class="nf">create_db</span><span class="p">(</span><span class="n">f_list</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a fasta database file containing the translated protein sequences</span>
<span class="sd">    from the cds files. The final transcripts.fas file will be use</span>
<span class="sd">    by USEARCH to get matches between the original protein sequences and their</span>
<span class="sd">    nucleotide counterparts. A dictionary database will also be created where</span>
<span class="sd">    the transcript headers will be associated with the original DNA sequence,</span>
<span class="sd">    so that they will be later retrieved</span>
<span class="sd">    :param f_list. List, containing the file names of the transcript files</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">output_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;transcripts.fas&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">id_dic</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ns</span><span class="o">.</span><span class="n">max_pb</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">f_list</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">f_list</span><span class="p">:</span>
        <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">header</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">handle</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">seq</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="n">aa_seq</span> <span class="o">=</span> <span class="n">translate</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
                    <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="n">aa_seq</span><span class="p">))</span>
                    <span class="n">id_dic</span><span class="p">[</span><span class="n">header</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq</span>

                <span class="n">header</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;;;&quot;</span><span class="p">)</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="n">output_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">id_dic</span></div>


<div class="viewcode-block" id="create_query"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.create_query">[docs]</a><span class="k">def</span> <span class="nf">create_query</span><span class="p">(</span><span class="n">input_list</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    To speed things up, all sequences in the input protein files will be</span>
<span class="sd">    concatenated into a single file, which will be used as query in USEARCH.</span>
<span class="sd">    :param input_list: List, with file names of the protein files to convert</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">f_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;query.fas&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">query_db</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">input_list</span><span class="p">:</span>
        <span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">handle</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">):</span>
                <span class="n">query_db</span><span class="p">[</span><span class="n">f</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;;;&quot;</span><span class="p">))</span>
                <span class="n">f_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;;;&quot;</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">f_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

        <span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">f_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">query_db</span></div>


<div class="viewcode-block" id="create_query_from_dict"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.create_query_from_dict">[docs]</a><span class="k">def</span> <span class="nf">create_query_from_dict</span><span class="p">(</span><span class="n">protein_dict</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analogous to create_query, but begins from a dictionary provided by</span>
<span class="sd">    processed group files</span>
<span class="sd">    :param protein_dict: dictionary</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">query_db</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="p">[]}</span>
    <span class="n">f_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">protein_dict</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">seq_id</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">cl</span><span class="p">:</span>
            <span class="n">query_db</span><span class="p">[</span><span class="n">cl</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;;;&quot;</span><span class="p">))</span>
            <span class="n">f_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">seq_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;;;&quot;</span><span class="p">),</span> <span class="n">seq</span><span class="p">))</span>

    <span class="n">f_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">query_db</span></div>


<div class="viewcode-block" id="pair_search"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.pair_search">[docs]</a><span class="k">def</span> <span class="nf">pair_search</span><span class="p">(</span><span class="n">usearch_bin</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This will use USEARCH to search for translated transcript sequences</span>
<span class="sd">    identical to the original protein files</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">query_path</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;query.fas&quot;</span><span class="p">)</span>
    <span class="n">db_path</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;transcripts.fas&quot;</span><span class="p">)</span>
    <span class="n">out_path</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;pairs.out&quot;</span><span class="p">)</span>

    <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="n">usearch_bin</span><span class="p">,</span>
                      <span class="s2">&quot;-usearch_global&quot;</span><span class="p">,</span>
                      <span class="n">query_path</span><span class="p">,</span>
                      <span class="s2">&quot;-db&quot;</span><span class="p">,</span>
                      <span class="n">db_path</span><span class="p">,</span>
                      <span class="s2">&quot;-id&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;1&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;-maxaccepts&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;.9&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;-blast6out&quot;</span><span class="p">,</span>
                      <span class="n">out_path</span><span class="p">])</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span></div>


<div class="viewcode-block" id="get_pairs"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.get_pairs">[docs]</a><span class="k">def</span> <span class="nf">get_pairs</span><span class="p">(</span><span class="n">dest</span><span class="o">=</span><span class="s2">&quot;./&quot;</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parses the output of USEARCH and creates a dictionary with the header</span>
<span class="sd">    pairs between original protein and transcripts</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">file_h</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;pairs.out&quot;</span><span class="p">))</span>
    <span class="n">pair_db</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;pairs.out&quot;</span><span class="p">))</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
                <span class="k">pass</span>
        <span class="n">ns</span><span class="o">.</span><span class="n">max_pb</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">file_h</span><span class="p">:</span>

        <span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">ns</span><span class="o">.</span><span class="n">progress</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">fields</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">pair_db</span><span class="p">[</span><span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="n">file_h</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">pair_db</span></div>


<div class="viewcode-block" id="convert_protein_file"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.convert_protein_file">[docs]</a><span class="k">def</span> <span class="nf">convert_protein_file</span><span class="p">(</span><span class="n">pairs</span><span class="p">,</span> <span class="n">group_obj</span><span class="p">,</span> <span class="n">id_db</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">shared_ns</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A given protein file will be converted into their corresponding nucleotide</span>
<span class="sd">    sequences using a previously set database using the create_db function</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Create handle for file storing bad sequence headers.</span>
    <span class="n">bad_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;missed_sequences.log&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">line</span><span class="p">,</span> <span class="n">cl</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">group_obj</span><span class="o">.</span><span class="n">groups</span><span class="p">(),</span>
                        <span class="n">group_obj</span><span class="o">.</span><span class="n">iter_species_frequency</span><span class="p">()):</span>

        <span class="k">if</span> <span class="n">shared_ns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">shared_ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">_get_compliance</span><span class="p">(</span><span class="n">cl</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>

            <span class="n">line</span> <span class="o">=</span> <span class="n">group_obj</span><span class="o">.</span><span class="n">_remove_tx</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
            <span class="n">orto_name</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">seq_headers</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>

            <span class="n">f_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">orto_name</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.fas&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">seq_headers</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">pairs</span><span class="p">:</span>
                    <span class="n">seq</span> <span class="o">=</span> <span class="n">id_db</span><span class="p">[</span><span class="n">pairs</span><span class="p">[</span><span class="n">h</span><span class="p">]]</span>
                    <span class="n">shared_ns</span><span class="o">.</span><span class="n">good</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">f_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;</span><span class="si">%s</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;;;&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">),</span> <span class="n">seq</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">shared_ns</span><span class="o">.</span><span class="n">missed</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">bad_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="se">\t</span><span class="si">{}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">orto_name</span><span class="p">,</span> <span class="n">h</span><span class="p">))</span></div>


<div class="viewcode-block" id="convert_group"><a class="viewcode-back" href="../../../trifusion.ortho.protein2dna.html#trifusion.ortho.protein2dna.convert_group">[docs]</a><span class="k">def</span> <span class="nf">convert_group</span><span class="p">(</span><span class="n">sqldb</span><span class="p">,</span> <span class="n">cds_file_list</span><span class="p">,</span> <span class="n">protein_db</span><span class="p">,</span> <span class="n">group_sequences</span><span class="p">,</span>
                <span class="n">usearch_bin</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">shared_namespace</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convenience function that wraps all required operations to convert protein</span>
<span class="sd">    to nucleotide files from a Group object</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Creating database&quot;</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">missed</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">good</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># Create database</span>
    <span class="n">id_db</span> <span class="o">=</span> <span class="n">create_db</span><span class="p">(</span><span class="n">cds_file_list</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">shared_namespace</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Creating query&quot;</span>

        <span class="c1"># Kill switch</span>
        <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># Create query for USEARCH</span>
    <span class="n">group_sequences</span><span class="o">.</span><span class="n">retrieve_sequences</span><span class="p">(</span><span class="n">sqldb</span><span class="p">,</span> <span class="n">protein_db</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span>
                                       <span class="n">outfile</span><span class="o">=</span><span class="s2">&quot;query.fas&quot;</span><span class="p">,</span>
                                       <span class="n">shared_namespace</span><span class="o">=</span><span class="n">shared_namespace</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="c1"># Kill switch</span>
        <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="c1"># Execute search</span>
    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Performing search&quot;</span>
    <span class="n">pair_search</span><span class="p">(</span><span class="n">usearch_bin</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="c1"># Kill switch</span>
        <span class="k">if</span> <span class="n">shared_namespace</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

    <span class="n">pair_db</span> <span class="o">=</span> <span class="n">get_pairs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="n">shared_namespace</span><span class="p">)</span>
    <span class="c1"># Convert files</span>

    <span class="k">if</span> <span class="n">shared_namespace</span><span class="p">:</span>
        <span class="n">shared_namespace</span><span class="o">.</span><span class="n">act</span> <span class="o">=</span> <span class="s2">&quot;Converting to nucleotide&quot;</span>
    <span class="n">convert_protein_file</span><span class="p">(</span><span class="n">pair_db</span><span class="p">,</span> <span class="n">group_sequences</span><span class="p">,</span> <span class="n">id_db</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span>
                         <span class="n">shared_namespace</span><span class="p">)</span>

    <span class="c1"># Remove temporary files</span>
    <span class="n">temp_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;query.fas&quot;</span><span class="p">),</span>
                  <span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;transcripts.fas&quot;</span><span class="p">),</span>
                  <span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;pairs.out&quot;</span><span class="p">)]</span>

    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">temp_files</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></div>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Diogo N. Silva&quot;</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Diogo N. Silva.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.5.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>