

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>trifusion.TriSeq &mdash; TriFusion 0.5.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="TriFusion 0.5.5 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> TriFusion
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../trifusion.html">TriFusion modules and API reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">TriFusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>trifusion.TriSeq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for trifusion.TriSeq</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python2</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="c1">#</span>
<span class="c1">#  Copyright 2012 Unknown &lt;diogo@arch&gt;</span>
<span class="c1">#  </span>
<span class="c1">#  This program is free software; you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation; either version 2 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#  </span>
<span class="c1">#  This program is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#  </span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with this program; if not, write to the Free Software</span>
<span class="c1">#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,</span>
<span class="c1">#  MA 02110-1301, USA.</span>

<span class="kn">import</span> <span class="nn">warnings</span>

<span class="c1"># Suppress import warnings</span>
<span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="kn">import</span> <span class="nn">shutil</span>
    <span class="kn">import</span> <span class="nn">time</span>
    <span class="kn">import</span> <span class="nn">argparse</span>
    <span class="kn">from</span> <span class="nn">glob</span> <span class="k">import</span> <span class="n">glob</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">process.base</span> <span class="k">import</span> <span class="n">print_col</span><span class="p">,</span> <span class="n">RED</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">YELLOW</span><span class="p">,</span> <span class="n">CleanUp</span>
        <span class="kn">from</span> <span class="nn">process</span> <span class="k">import</span> <span class="n">sequence</span> <span class="k">as</span> <span class="n">seqset</span>
        <span class="kn">from</span> <span class="nn">process</span> <span class="k">import</span> <span class="n">data</span>
        <span class="kn">from</span> <span class="nn">process.error_handling</span> <span class="k">import</span> <span class="o">*</span>
        <span class="kn">from</span> <span class="nn">base.sanity</span> <span class="k">import</span> <span class="n">triseq_arg_check</span><span class="p">,</span> <span class="n">mfilters</span><span class="p">,</span> <span class="n">post_aln_checks</span><span class="p">,</span> \
            <span class="n">check_infile_list</span>
        <span class="kn">from</span> <span class="nn">progressbar</span> <span class="k">import</span> <span class="n">ProgressBar</span><span class="p">,</span> <span class="n">Timer</span><span class="p">,</span> <span class="n">Bar</span><span class="p">,</span> <span class="n">Percentage</span><span class="p">,</span> \
            <span class="n">SimpleProgress</span>
    <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">trifusion.process.base</span> <span class="k">import</span> <span class="n">print_col</span><span class="p">,</span> <span class="n">RED</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">YELLOW</span><span class="p">,</span>\
            <span class="n">CleanUp</span>
        <span class="kn">from</span> <span class="nn">trifusion.process</span> <span class="k">import</span> <span class="n">sequence</span> <span class="k">as</span> <span class="n">seqset</span>
        <span class="kn">from</span> <span class="nn">trifusion.process</span> <span class="k">import</span> <span class="n">data</span>
        <span class="kn">from</span> <span class="nn">trifusion.process.error_handling</span> <span class="k">import</span> <span class="o">*</span>
        <span class="kn">from</span> <span class="nn">trifusion.base.sanity</span> <span class="k">import</span> <span class="n">triseq_arg_check</span><span class="p">,</span> <span class="n">mfilters</span><span class="p">,</span> \
            <span class="n">post_aln_checks</span><span class="p">,</span> <span class="n">check_infile_list</span>
        <span class="kn">from</span> <span class="nn">trifusion.progressbar</span> <span class="k">import</span> <span class="n">ProgressBar</span><span class="p">,</span> <span class="n">Timer</span><span class="p">,</span> <span class="n">Bar</span><span class="p">,</span>\
            <span class="n">Percentage</span><span class="p">,</span> <span class="n">SimpleProgress</span>


<div class="viewcode-block" id="gen_wgt"><a class="viewcode-back" href="../../trifusion.TriSeq.html#trifusion.TriSeq.gen_wgt">[docs]</a><span class="k">def</span> <span class="nf">gen_wgt</span><span class="p">(</span><span class="n">msg</span><span class="p">):</span>

    <span class="n">bar_wdg</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;( &quot;</span><span class="p">,</span> <span class="n">SimpleProgress</span><span class="p">(),</span> <span class="s2">&quot; ) &quot;</span><span class="p">,</span>
        <span class="n">Bar</span><span class="p">(),</span>
        <span class="n">Percentage</span><span class="p">(),</span>
        <span class="s2">&quot; [&quot;</span><span class="p">,</span> <span class="n">Timer</span><span class="p">(),</span> <span class="s2">&quot;] &quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">bar_wdg</span></div>

<span class="nd">@CleanUp</span>
<span class="k">def</span> <span class="nf">main_parser</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="n">alignment_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Function with the main operations of TriSeq &quot;&quot;&quot;</span>

    <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Executing TriSeq module at </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">/%m/%Y&quot;</span><span class="p">),</span> <span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%I:%M:%S&quot;</span><span class="p">)),</span> <span class="n">GREEN</span><span class="p">,</span>
              <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>

    <span class="c1"># Create temp directory</span>
    <span class="n">tmp_dir</span> <span class="o">=</span> <span class="s2">&quot;.trifusion-temp&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">)</span>

    <span class="c1"># Set path to temporary sqlite database</span>
    <span class="n">sql_db</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">,</span> <span class="s2">&quot;trifusion.db&quot;</span><span class="p">)</span>

    <span class="c1"># If database already exists, erase it. Make sure we start fresh.</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">sql_db</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">sql_db</span><span class="p">)</span>

    <span class="c1"># Defining main variables</span>
    <span class="n">conversion</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">conversion</span>
    <span class="n">output_format</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">output_format</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">outfile</span>
    <span class="n">interleave</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">interleave</span>
    <span class="n">model_phy</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">model_phy</span>
    <span class="c1"># outgroup_taxa = arg.outgroup_taxa</span>

    <span class="c1"># Defining output file name</span>
    <span class="k">if</span> <span class="n">conversion</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">arg</span><span class="o">.</span><span class="n">outfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">conversion</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">arg</span><span class="o">.</span><span class="n">outfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">outfile</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus</span> <span class="ow">and</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus_single</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">arg</span><span class="o">.</span><span class="n">outfile</span><span class="p">:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;consensus&quot;</span>
    <span class="k">elif</span> <span class="n">conversion</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">arg</span><span class="o">.</span><span class="n">outfile</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># The input file at this stage is not necessary</span>
    <span class="c1"># If just converting the partition file format do this and exit</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">partition_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Initializing Partitions instance and reading partitions file</span>
        <span class="n">partition</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">Partitions</span><span class="p">()</span>
        <span class="n">partition</span><span class="o">.</span><span class="n">read_from_file</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">partition_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">partition</span><span class="o">.</span><span class="n">partition_format</span> <span class="o">==</span> <span class="s2">&quot;nexus&quot;</span><span class="p">:</span>
            <span class="n">partition</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="s2">&quot;raxml&quot;</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">model_phy</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">partition</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">0</span>

    <span class="c1"># Support wildcars as arguments for windows</span>
    <span class="n">fl</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;win32&quot;</span><span class="p">,</span> <span class="s2">&quot;cygwin&quot;</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">alignment_list</span><span class="p">:</span>
            <span class="n">fl</span> <span class="o">+=</span> <span class="n">glob</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="n">alignment_list</span> <span class="o">=</span> <span class="n">fl</span>

    <span class="c1"># Check input files for directories</span>
    <span class="n">alignment_list</span><span class="p">,</span> <span class="n">dirs</span><span class="p">,</span> <span class="n">lost</span> <span class="o">=</span> <span class="n">check_infile_list</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">dirs</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Ignoring input files pointing to a directory: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirs</span><span class="p">)),</span> <span class="n">YELLOW</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">lost</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Ignoring input files that do not exist: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">lost</span><span class="p">)),</span> <span class="n">YELLOW</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_list</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;No valid input files have been provided. Terminating...&quot;</span><span class="p">,</span>
                  <span class="n">RED</span><span class="p">)</span>

    <span class="c1"># Input alignments are mandatory from now on</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">:</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="n">ProgressBar</span><span class="p">(</span><span class="n">max_value</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">),</span> <span class="n">widgets</span><span class="o">=</span><span class="n">gen_wgt</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pbar</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Parsing </span><span class="si">%s</span><span class="s2"> alignments&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">),</span> <span class="n">GREEN</span><span class="p">,</span>
              <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
    <span class="n">alignments</span> <span class="o">=</span> <span class="n">seqset</span><span class="o">.</span><span class="n">AlignmentList</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">,</span> <span class="n">sql_db</span><span class="o">=</span><span class="n">sql_db</span><span class="p">,</span>
                                      <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="n">post_aln_checks</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="n">alignments</span><span class="p">)</span>

    <span class="c1"># ################################ Utilities ##############################</span>
    <span class="c1"># Return a file with taxa list and exit</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">get_taxa</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Writing taxa to new file&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">write_taxa_to_file</span><span class="p">()</span>
        <span class="k">return</span> <span class="mi">0</span>

    <span class="c1"># Remove taxa</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">remove</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Removing taxa&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">remove</span><span class="p">)</span>

    <span class="c1"># Grep taxa</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">grep</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Grepping taxa&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">grep</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;inverse&quot;</span><span class="p">)</span>

    <span class="c1"># Select alignments</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">select</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Selecting alignments&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s2">&quot;Taxa_selection&quot;</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s2">&quot;Taxa_selection&quot;</span><span class="p">)</span>

        <span class="c1"># Check if any of the provided taxa is absent from the alignments</span>
        <span class="n">absent_taxa</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">arg</span><span class="o">.</span><span class="n">select</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">alignments</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">absent_taxa</span><span class="p">:</span>
            <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;The following taxa were not found in any alignment and&quot;</span>
                      <span class="s2">&quot; will be ignored: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">absent_taxa</span><span class="p">)),</span>
                      <span class="n">YELLOW</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>

        <span class="n">selected_alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">select_by_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">select</span><span class="p">,</span>
                                                        <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;relaxed&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="n">selected_alignments</span><span class="p">:</span>
            <span class="n">alignment_file</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">path</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">,</span> <span class="s2">&quot;Taxa_selection&quot;</span><span class="p">)</span>

        <span class="k">return</span>

    <span class="c1"># ############################# Main operations ###########################</span>
    <span class="c1"># Reverse concatenation</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Reverse concatenating&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ArgumentError</span><span class="p">(</span><span class="s2">&quot;Only one input file allowed for reverse &quot;</span>
                                <span class="s2">&quot;concatenation&quot;</span><span class="p">)</span>
        <span class="n">aln</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1"># Initializing and reading partition file</span>
        <span class="n">partition</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">Partitions</span><span class="p">()</span>
        <span class="n">partition</span><span class="o">.</span><span class="n">read_from_file</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">reverse</span><span class="p">)</span>
        <span class="c1"># Updating alignment partitions</span>
        <span class="n">aln</span><span class="o">.</span><span class="n">set_partitions</span><span class="p">(</span><span class="n">partition</span><span class="p">)</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">reverse_concatenate</span><span class="p">(</span><span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filtering</span>
    <span class="c1"># Filter by minimum taxa</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">min_taxa</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering by minimum taxa&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_min_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">min_taxa</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filter by alignments that contain taxa</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">contain_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering alignment(s) including a taxa group&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span>
                  <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_by_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">contain_filter</span><span class="p">,</span> <span class="s2">&quot;Contain&quot;</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filter by alignments that exclude taxa</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">exclude_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering alignments excluding a taxa group&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span>
                  <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_by_taxa</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">exclude_filter</span><span class="p">,</span> <span class="s2">&quot;Exclude&quot;</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filter by codon position</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">codon_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering by codon positions&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">alignments</span><span class="o">.</span><span class="n">sequence_code</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;DNA&quot;</span><span class="p">:</span>
            <span class="n">codon_settings</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="n">arg</span><span class="o">.</span><span class="n">codon_filter</span> <span class="k">else</span> <span class="kc">False</span>
                              <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)]</span>
            <span class="n">alignments</span><span class="o">.</span><span class="n">filter_codon_positions</span><span class="p">(</span><span class="n">codon_settings</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filter by missing data</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">m_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering by missing data&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_missing_data</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">m_filter</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">arg</span><span class="o">.</span><span class="n">m_filter</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                       <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">,</span> <span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Filtering by variable sites</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">var_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering by variable sites&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_segregating_sites</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">var_filter</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                            <span class="n">arg</span><span class="o">.</span><span class="n">var_filter</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                            <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Filtering by informative sites</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">inf_filter</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Filtering by variable sites&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">filter_informative_sites</span><span class="p">(</span><span class="n">arg</span><span class="o">.</span><span class="n">inf_filter</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                            <span class="n">arg</span><span class="o">.</span><span class="n">inf_filter</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                            <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Concatenation</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">arg</span><span class="o">.</span><span class="n">conversion</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">arg</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Concatenating&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

        <span class="c1"># Concatenate zorro files</span>
        <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">zorro</span><span class="p">:</span>
            <span class="n">zorro</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">Zorro</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">,</span> <span class="n">arg</span><span class="o">.</span><span class="n">zorro</span><span class="p">)</span>
            <span class="n">zorro</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">outfile</span><span class="p">)</span>

    <span class="c1"># Collapsing</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">collapse</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Collapsing&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">collapse</span><span class="p">(</span><span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">,</span>
                            <span class="n">haplotypes_file</span><span class="o">=</span><span class="n">outfile</span><span class="p">)</span>

    <span class="c1"># Gcoder</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">gcoder</span><span class="p">:</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Coding gaps&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_format</span> <span class="o">==</span> <span class="p">[</span><span class="s2">&quot;nexus&quot;</span><span class="p">]:</span>
            <span class="n">alignments</span><span class="o">.</span><span class="n">code_gaps</span><span class="p">(</span><span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>

    <span class="c1"># Consensus</span>
    <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus</span><span class="p">:</span>
        <span class="n">consensus_type</span> <span class="o">=</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Creating consensus sequences&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">arg</span><span class="o">.</span><span class="n">consensus_single</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">seqset</span><span class="o">.</span><span class="n">AlignmentList</span><span class="p">):</span>
                <span class="n">alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">consensus</span><span class="p">(</span>
                    <span class="n">consensus_type</span><span class="p">,</span> <span class="n">single_file</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">consensus_single</span><span class="p">,</span>
                    <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">,</span> <span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">alignments</span><span class="o">.</span><span class="n">consensus</span><span class="p">(</span><span class="n">consensus_type</span><span class="o">=</span><span class="n">consensus_type</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">,</span>
                                     <span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">alignments</span><span class="o">.</span><span class="n">consensus</span><span class="p">(</span><span class="n">consensus_type</span><span class="o">=</span><span class="n">consensus_type</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">,</span>
                                 <span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Write output</span>
    <span class="n">print_col</span><span class="p">(</span><span class="s2">&quot;Writing output&quot;</span><span class="p">,</span> <span class="n">GREEN</span><span class="p">,</span> <span class="n">quiet</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">quiet</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">seqset</span><span class="o">.</span><span class="n">Alignment</span><span class="p">):</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">output_format</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span>
                                 <span class="n">interleave</span><span class="o">=</span><span class="n">interleave</span><span class="p">,</span>
                                 <span class="n">partition_file</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                 <span class="n">use_charset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                 <span class="n">ima2_params</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">ima2_params</span><span class="p">,</span>
                                 <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">seqset</span><span class="o">.</span><span class="n">AlignmentList</span><span class="p">):</span>
        <span class="n">alignments</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">output_format</span><span class="p">,</span>
                                 <span class="n">interleave</span><span class="o">=</span><span class="n">interleave</span><span class="p">,</span>
                                 <span class="n">ima2_params</span><span class="o">=</span><span class="n">arg</span><span class="o">.</span><span class="n">ima2_params</span><span class="p">,</span>
                                 <span class="n">pbar</span><span class="o">=</span><span class="n">pbar</span><span class="p">)</span>


<div class="viewcode-block" id="get_args"><a class="viewcode-back" href="../../trifusion.TriSeq.html#trifusion.TriSeq.get_args">[docs]</a><span class="k">def</span> <span class="nf">get_args</span><span class="p">(</span><span class="n">arg_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">unittest</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

    <span class="c1"># The inclusion of the argument definition in main, makes it possible to</span>
    <span class="c1"># import this file as a module and not triggering argparse. The</span>
    <span class="c1"># alternative of using a if __name__ == &quot;__main__&quot; statement does not</span>
    <span class="c1"># work well with the entry_points parameter of setup.py, since they call</span>
    <span class="c1"># the main function but do nothing inside said statement.</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;Command line interface for &quot;</span>
                                                 <span class="s2">&quot;TriFusion Process module&quot;</span><span class="p">)</span>

    <span class="c1"># Main execution</span>
    <span class="n">main_exec</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Main execution&quot;</span><span class="p">)</span>
    <span class="n">main_exec</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-in&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;infile&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide the &quot;</span>
                           <span class="s2">&quot;input file name. If multiple files are provided&quot;</span>
                           <span class="s2">&quot;, please separated the names with spaces&quot;</span><span class="p">)</span>
    <span class="n">main_exec</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-of&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;output_format&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
                           <span class="n">default</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;nexus&quot;</span><span class="p">],</span>
                           <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;nexus&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;mcmctree&quot;</span><span class="p">,</span>
                                    <span class="s2">&quot;ima2&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">,</span> <span class="s2">&quot;gphocs&quot;</span><span class="p">],</span>
                           <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Format of the output file(s). You may select &quot;</span>
                           <span class="s2">&quot;multiple output formats simultaneously (default is&quot;</span>
                           <span class="s2">&quot; &#39;</span><span class="si">%(default)s</span><span class="s2">&#39;)&quot;</span><span class="p">)</span>
    <span class="n">main_exec</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;outfile&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Name of the output file&quot;</span><span class="p">)</span>

    <span class="c1"># Alternative modes</span>
    <span class="n">alternative</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Alternative execution modes&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;conversion&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
                             <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Used for conversion of the &quot;</span>
                             <span class="s2">&quot;input files passed as arguments with the -in &quot;</span>
                             <span class="s2">&quot;option. This flag precludes the usage of the -o &quot;</span>
                             <span class="s2">&quot;option, as the output file name is automatically &quot;</span>
                             <span class="s2">&quot;generated based on the input file name&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-r&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;reverse&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Reverse a concatenated&quot;</span>
                             <span class="s2">&quot; file into its original single locus alignments.&quot;</span>
                             <span class="s2">&quot; A partition file similar to the one read by &quot;</span>
                             <span class="s2">&quot;RAxML must be provided&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-z&quot;</span><span class="p">,</span> <span class="s2">&quot;--zorro-suffix&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;zorro&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                             <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use this option if you wish to concatenate &quot;</span>
                             <span class="s2">&quot;auxiliary Zorro files associated with each &quot;</span>
                             <span class="s2">&quot;alignment. Provide the sufix for the concatenated&quot;</span>
                             <span class="s2">&quot; zorro file&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-p&quot;</span><span class="p">,</span> <span class="s2">&quot;--partition-file&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;partition_file&quot;</span><span class="p">,</span>
                             <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Using this option and providing &quot;</span>
                             <span class="s2">&quot;the partition file will convert it between a &quot;</span>
                             <span class="s2">&quot;RAxML or Nexus format&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-s&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;select&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Selects &quot;</span>
                             <span class="s2">&quot;alignments containing the provided taxa &quot;</span>
                             <span class="s2">&quot;(separate multiple taxa with whitespace)&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--collapse&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;collapse&quot;</span><span class="p">,</span>
                             <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
                             <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use this flag if &quot;</span>
                             <span class="s2">&quot;you would like to collapse the input alignment(s)&quot;</span>
                             <span class="s2">&quot; into unique haplotypes&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--code-gaps&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;gcoder&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
                             <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use this flag to &quot;</span>
                             <span class="s2">&quot;code the gaps of the alignment into a binary &quot;</span>
                             <span class="s2">&quot;state  matrix that is appended to the end of &quot;</span>
                             <span class="s2">&quot;the alignment&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--consensus&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;consensus&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                             <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;First sequence&quot;</span><span class="p">,</span> <span class="s2">&quot;IUPAC&quot;</span><span class="p">,</span> <span class="s2">&quot;Soft mask&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;Remove&quot;</span><span class="p">],</span>
                             <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Creates a consensus of the final alignments&quot;</span>
                             <span class="s2">&quot; specifying how variation is handled&quot;</span><span class="p">)</span>
    <span class="n">alternative</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--consensus-single-file&quot;</span><span class="p">,</span>
                             <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;consensus_single&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
                             <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Merges &quot;</span>
                             <span class="s2">&quot;consensus sequences in a single file&quot;</span><span class="p">)</span>

    <span class="n">filter_g</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Filter options&quot;</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--missing-filter&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;m_filter&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Use this  option if you wish to filter the&quot;</span>
                          <span class="s2">&quot; alignment&#39;s missing data. Along with this &quot;</span>
                          <span class="s2">&quot;option provide the threshold percentages for &quot;</span>
                          <span class="s2">&quot;gap and missing data, respectively (e.g. -filter &quot;</span>
                          <span class="s2">&quot;50 75 - filters alignments columns with more &quot;</span>
                          <span class="s2">&quot;than 50</span><span class="si">%%</span><span class="s2"> of gap+missing data and columns with &quot;</span>
                          <span class="s2">&quot;more than 75</span><span class="si">%%</span><span class="s2"> of true missing data)&quot;</span><span class="p">,</span>
                          <span class="nb">type</span><span class="o">=</span><span class="n">mfilters</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--min-taxa&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;min_taxa&quot;</span><span class="p">,</span>
                          <span class="nb">type</span><span class="o">=</span><span class="n">mfilters</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Set the minimum percentage &quot;</span>
                          <span class="s2">&quot;of taxa that needs to be present in an alignment&quot;</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--contain-taxa&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;contain_filter&quot;</span><span class="p">,</span>
                          <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Only &quot;</span>
                          <span class="s2">&quot;processes alignments that contain the specified &quot;</span>
                          <span class="s2">&quot;taxa&quot;</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--exclude-taxa&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;exclude_filter&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Only &quot;</span>
                          <span class="s2">&quot;process alignments that do NOT contain the &quot;</span>
                          <span class="s2">&quot;specified taxa&quot;</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--codon-filter&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;codon_filter&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span>
                          <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">],</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Include only the &quot;</span>
                          <span class="s2">&quot;codon positions specified by this option (DNA only)&quot;</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--variable-filter&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;var_filter&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide minimum and maximum values of &quot;</span>
                          <span class="s2">&quot;variable sites for each alignment. Filters &quot;</span>
                          <span class="s2">&quot;alignments with a number of variable sites outside &quot;</span>
                          <span class="s2">&quot;the specified range&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">filter_g</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--informative-filter&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;inf_filter&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide minimum and maximum values of &quot;</span>
                          <span class="s2">&quot;informative sites for each alignment. Filters &quot;</span>
                          <span class="s2">&quot;alignments with a number of informative sites &quot;</span>
                          <span class="s2">&quot;outside the specified range&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1"># Formatting options</span>
    <span class="n">formatting</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Formatting options&quot;</span><span class="p">)</span>
    <span class="n">formatting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--model&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;model_phy&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;LG&quot;</span><span class="p">,</span>
                            <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;DAYHOFF&quot;</span><span class="p">,</span> <span class="s2">&quot;DCMUT&quot;</span><span class="p">,</span> <span class="s2">&quot;JTT&quot;</span><span class="p">,</span> <span class="s2">&quot;MTREV&quot;</span><span class="p">,</span> <span class="s2">&quot;WAG&quot;</span><span class="p">,</span>
                                     <span class="s2">&quot;RTREV&quot;</span><span class="p">,</span> <span class="s2">&quot;CPREV&quot;</span><span class="p">,</span> <span class="s2">&quot;VT&quot;</span><span class="p">,</span> <span class="s2">&quot;BLOSUM62&quot;</span><span class="p">,</span>
                                     <span class="s2">&quot;MTMAM&quot;</span><span class="p">,</span> <span class="s2">&quot;LG&quot;</span><span class="p">],</span>
                            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;This option only applies for the &quot;</span>
                            <span class="s2">&quot;concatenation  of protein data into phylip &quot;</span>
                            <span class="s2">&quot;format. Specify the model for all partitions &quot;</span>
                            <span class="s2">&quot;defined in the partition file  (default is &#39;%(&quot;</span>
                            <span class="s2">&quot;default)s&#39;)&quot;</span><span class="p">)</span>
    <span class="n">formatting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--interleave&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span>
                            <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span> <span class="n">const</span><span class="o">=</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span>
                            <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Specify this  option to &quot;</span>
                            <span class="s2">&quot;write output files in interleave format (currently&quot;</span>
                            <span class="s2">&quot; only supported for nexus files&quot;</span><span class="p">)</span>
    <span class="n">formatting</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--ima2-params&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;ima2_params&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                            <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Provide 4 additional arguments needed to &quot;</span>
                            <span class="s2">&quot;write the output in a format compliant with &quot;</span>
                            <span class="s2">&quot;IMa2. The order of the required arguments &quot;</span>
                            <span class="s2">&quot;(separated by whitespace is as follows: &quot;</span>
                            <span class="s2">&quot;[(1) File name of population mapping]&quot;</span>
                            <span class="s2">&quot;[(2) Population tree]&quot;</span>
                            <span class="s2">&quot;[(3) Mutational model]&quot;</span>
                            <span class="s2">&quot;[(4) Inheritance Scalar]. &quot;</span>
                            <span class="s2">&quot;Additional notes: (1) The  population mapping &quot;</span>
                            <span class="s2">&quot;file is a simple .csv file  containing two &quot;</span>
                            <span class="s2">&quot;columns separated by a semi-colon, &quot;</span>
                            <span class="s2">&quot;in which the first column contains the taxon &quot;</span>
                            <span class="s2">&quot;name and the second column contains the &quot;</span>
                            <span class="s2">&quot;corresponding population name; (2) The order of &quot;</span>
                            <span class="s2">&quot;the population names in the population tree &quot;</span>
                            <span class="s2">&quot;must be the same as the order in the file with &quot;</span>
                            <span class="s2">&quot;the population mapping&quot;</span><span class="p">)</span>

    <span class="c1"># Data manipulation</span>
    <span class="n">manipulation</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Data manipulation&quot;</span><span class="p">)</span>
    <span class="n">manipulation</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-rm&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;remove&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Removes &quot;</span>
                              <span class="s2">&quot;the  specified taxa from the final alignment. &quot;</span>
                              <span class="s2">&quot;Unwanted taxa my be provided in a csv file &quot;</span>
                              <span class="s2">&quot;containing 1 column with a species name in &quot;</span>
                              <span class="s2">&quot;each line or they may be specified in the &quot;</span>
                              <span class="s2">&quot;command line and separated by whitespace&quot;</span><span class="p">)</span>
    <span class="n">manipulation</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-grep&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;grep&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The &quot;</span>
                              <span class="s2">&quot;inverse  of the -rm command. It removes all &quot;</span>
                              <span class="s2">&quot;taxa from the alignment except for the ones &quot;</span>
                              <span class="s2">&quot;specified with this option. Taxa names may be &quot;</span>
                              <span class="s2">&quot;specified in a csv file containing 1 column &quot;</span>
                              <span class="s2">&quot;with a species name in each line or in the &quot;</span>
                              <span class="s2">&quot;command line separated by whitespace&quot;</span><span class="p">)</span>
    <span class="c1"># manipulation.add_argument(&quot;-outgroup&quot;, dest=&quot;outgroup_taxa&quot;, nargs=&quot;*&quot;,</span>
    <span class="c1">#                           help=&quot;Provide taxon names/number for the &quot;</span>
    <span class="c1">#                           &quot;outgroup  (This option is only supported for &quot;</span>
    <span class="c1">#                           &quot;NEXUS output format files)&quot;)</span>

    <span class="n">utilities</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Utilities&quot;</span><span class="p">)</span>
    <span class="n">utilities</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--get-taxa&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;get_taxa&quot;</span><span class="p">,</span>
                           <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span> <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                           <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Writes all taxa names into a&quot;</span>
                           <span class="s2">&quot; .csv file&quot;</span><span class="p">)</span>

    <span class="n">miscellaneous</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;Miscellaneous&quot;</span><span class="p">)</span>
    <span class="n">miscellaneous</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-quiet&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;quiet&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_const&quot;</span><span class="p">,</span>
                               <span class="n">const</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Removes all &quot;</span>
                               <span class="s2">&quot;terminal output&quot;</span><span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">arg_list</span><span class="p">)</span>

    <span class="c1"># Print help when no arguments are provided</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">unittest</span><span class="p">:</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">print_help</span><span class="p">()</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../trifusion.TriSeq.html#trifusion.TriSeq.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">arguments</span> <span class="o">=</span> <span class="n">get_args</span><span class="p">()</span>
    <span class="n">triseq_arg_check</span><span class="p">(</span><span class="n">arguments</span><span class="p">)</span>
    <span class="n">main_parser</span><span class="p">(</span><span class="n">arguments</span><span class="p">,</span> <span class="n">arguments</span><span class="o">.</span><span class="n">infile</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>

    <span class="n">main</span><span class="p">()</span>


<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Diogo N. Silva&quot;</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Diogo N. Silva.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.5.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>