

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>trifusion.process.sequence module &mdash; TriFusion 0.5.5 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="TriFusion 0.5.5 documentation" href="index.html"/>
        <link rel="up" title="trifusion.process package" href="trifusion.process.html"/>
        <link rel="prev" title="trifusion.process.error_handling module" href="trifusion.process.error_handling.html"/> 

  
  <script src="static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> TriFusion
          

          
          </a>

          
            
            
              <div class="version">
                0.5.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="trifusion.html">TriFusion modules and API reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="trifusion.html#what-is-trifusion">What is TriFusion</a></li>
<li class="toctree-l2"><a class="reference internal" href="trifusion.html#how-can-trifusion-be-used">How can TriFusion be used</a></li>
<li class="toctree-l2"><a class="reference internal" href="trifusion.html#components-of-trifusion">Components of TriFusion</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="trifusion.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="trifusion.base.html">trifusion.base package</a></li>
<li class="toctree-l3"><a class="reference internal" href="trifusion.data.html">trifusion.data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="trifusion.ortho.html">trifusion.ortho package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="trifusion.process.html">trifusion.process package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="trifusion.process.html#introduction-to-trifusion-s-process-module">Introduction to TriFusion&#8217;s process module</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="trifusion.process.html#submodules">Submodules</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="trifusion.html#submodules">Submodules</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TriFusion</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="trifusion.html">TriFusion modules and API reference</a> &raquo;</li>
        
          <li><a href="trifusion.process.html">trifusion.process package</a> &raquo;</li>
        
      <li>trifusion.process.sequence module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="sources/trifusion.process.sequence.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-trifusion.process.sequence">
<span id="trifusion-process-sequence-module"></span><h1>trifusion.process.sequence module<a class="headerlink" href="#module-trifusion.process.sequence" title="Permalink to this headline">¶</a></h1>
<p>The <cite>sequence</cite> module of TriFusion contains the main classes handling
alignment sequence data. These are <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>. Here follows a brief explanation of how these
classes work and how to deal with the sqlite database.</p>
<div class="section" id="alignment-class">
<h2><cite>Alignment</cite> class<a class="headerlink" href="#alignment-class" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> class is the main interface with single alignment
files. It can be viewed as the building block of an <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>
object, which can have one or more <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> objects. It contains
all methods and attributes that pertain to a given alignment and are used to
retrieve information or modify it. <strong>However, it is NOT meant to be used
independently</strong>, but rather within the context of an <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>
object. The data from each alignment is stored in a single sqlite database
during the execution of TriFusion or the TriSeq/TriStats CLI programs. The
connection to this database is automatically handled by
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> for all <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> objects included in it.
In this way, we can use the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> class to handle the
setup of the sqlite3 database, and focus on single alignment data handling
in general in this class.</p>
<p>The main types of methods defined in this class are:</p>
<div class="section" id="parsers">
<h3>Parsers<a class="headerlink" href="#parsers" title="Permalink to this headline">¶</a></h3>
<p>Parsing methods are defined for each format with <cite>_read_&lt;format&gt;</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_phylip" title="trifusion.process.sequence.Alignment._read_phylip"><code class="xref py py-meth docutils literal"><span class="pre">_read_phylip()</span></code></a>: Parses phylip format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_fasta" title="trifusion.process.sequence.Alignment._read_fasta"><code class="xref py py-meth docutils literal"><span class="pre">_read_fasta()</span></code></a>: Parses fasta format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_nexus" title="trifusion.process.sequence.Alignment._read_nexus"><code class="xref py py-meth docutils literal"><span class="pre">_read_nexus()</span></code></a>: Parses nexus format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_loci" title="trifusion.process.sequence.Alignment._read_loci"><code class="xref py py-meth docutils literal"><span class="pre">_read_loci()</span></code></a>: Parses pyRAD/ipyrad loci format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_stockholm" title="trifusion.process.sequence.Alignment._read_stockholm"><code class="xref py py-meth docutils literal"><span class="pre">_read_stockholm()</span></code></a>: Parses stockholm format.</li>
</ul>
</div></blockquote>
<p>They are always called from the <a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-meth docutils literal"><span class="pre">read_alignment()</span></code></a> method,
and not directly. When an <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object is instantiated with a
path to an alignment file, it automatically detects the format of the
alignment and keeps that information on the <a class="reference internal" href="#trifusion.process.sequence.Alignment.input_format" title="trifusion.process.sequence.Alignment.input_format"><code class="xref py py-attr docutils literal"><span class="pre">input_format</span></code></a>
attribute. The <a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-meth docutils literal"><span class="pre">read_alignment()</span></code></a> method then calls the
parsing method corresponding to that format. That information is stored in a
dictionary:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">parsing_methods</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_phylip</span><span class="p">,</span>
    <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_fasta</span><span class="p">,</span>
    <span class="s2">&quot;loci&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_loci</span><span class="p">,</span>
    <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_nexus</span><span class="p">,</span>
    <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_stockholm</span>
<span class="p">}</span>

<span class="c1"># Call the appropriate method</span>
<span class="n">parsing_methods</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">input_format</span><span class="p">]()</span>
</pre></div>
</div>
<p>Each format has its own parsing method (which can be modified directly). To
add a new format, it is necessary to add it to the automatic format
recognition in <a class="reference internal" href="trifusion.process.base.html#trifusion.process.base.Base.autofinder" title="trifusion.process.base.Base.autofinder"><code class="xref py py-meth docutils literal"><span class="pre">autofinder()</span></code></a>. Then,
create the new parsing method, using the same <cite>_read_&lt;format&gt;</cite> notation and
add it to the <cite>parsing_methods</cite> dictionary in
<a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-meth docutils literal"><span class="pre">read_alignment()</span></code></a>.</p>
<p>New parsers must insert alignment data into a table in the sqlite database.
This table is automatically created when the <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object
is instantiated, and its name is stored in the <a class="reference internal" href="#trifusion.process.sequence.Alignment.table_name" title="trifusion.process.sequence.Alignment.table_name"><code class="xref py py-attr docutils literal"><span class="pre">table_name</span></code></a>
attribute (see <a class="reference internal" href="#trifusion.process.sequence.Alignment._create_table" title="trifusion.process.sequence.Alignment._create_table"><code class="xref py py-meth docutils literal"><span class="pre">_create_table()</span></code></a>):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;CREATE TABLE [</span><span class="si">{}</span><span class="s2">](&quot;</span>
            <span class="s2">&quot;txId INT,&quot;</span>
            <span class="s2">&quot;taxon TEXT,&quot;</span>
            <span class="s2">&quot;seq TEXT)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">table_name</span><span class="p">))</span>
</pre></div>
</div>
<p>To insert data into the database, a taxon id (<cite>txId</cite>), taxon name (<cite>taxon</cite>)
and sequence (<cite>seq</cite>) must be provided. For example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="s2">&quot;INSERT INTO [</span><span class="si">{}</span><span class="s2">] VALUES (?, ?, ?)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">table_name</span><span class="p">),</span>
            <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;spa&quot;</span><span class="p">,</span> <span class="s2">&quot;AAA&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="section" id="data-fetching">
<h3>Data fetching<a class="headerlink" href="#data-fetching" title="Permalink to this headline">¶</a></h3>
<p>To facilitate fetching alignment data from the database, several generators
and data retrieval methods are defined:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_columns" title="trifusion.process.sequence.Alignment.iter_columns"><code class="xref py py-meth docutils literal"><span class="pre">iter_columns()</span></code></a>: Iterates over the columns of the
alignment.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_columns_uniq" title="trifusion.process.sequence.Alignment.iter_columns_uniq"><code class="xref py py-meth docutils literal"><span class="pre">iter_columns_uniq()</span></code></a>: Iterates over the columns of
the alignment but yields only unique characters.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_sequences" title="trifusion.process.sequence.Alignment.iter_sequences"><code class="xref py py-meth docutils literal"><span class="pre">iter_sequences()</span></code></a>: Iterates over each sequence in the
alignment.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_alignment" title="trifusion.process.sequence.Alignment.iter_alignment"><code class="xref py py-meth docutils literal"><span class="pre">iter_alignment()</span></code></a>: Iterates over both taxon name
and sequence in the alignment.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.get_sequence" title="trifusion.process.sequence.Alignment.get_sequence"><code class="xref py py-meth docutils literal"><span class="pre">get_sequence()</span></code></a>: Return the sequence from a particular
taxon.</li>
</ul>
</div></blockquote>
<p>These should always use the
<a class="reference internal" href="#trifusion.process.sequence.setup_intable" title="trifusion.process.sequence.setup_intable"><code class="xref py py-func docutils literal"><span class="pre">setup_intable()</span></code></a> decorator and defined with the <cite>table_suffix</cite>, <cite>table_name</cite>
and <cite>table</cite> arguments. For instance, the <a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_sequences" title="trifusion.process.sequence.Alignment.iter_sequences"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.iter_sequences()</span></code></a> method is a
generator that allows the iteration over the sequences in the alignment
and is defined as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="nd">@setup_intable</span>
<span class="k">def</span> <span class="nf">iter_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">table_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</pre></div>
</div>
<p>When calling these methods, only the <cite>table_suffix</cite> and <cite>table_name</cite> have to
be provided. In fact, the value provided to the <cite>table</cite> argument at calling
time is ignored. The decorator will check the values of both <cite>table_suffix</cite>
and <cite>table_name</cite> and evaluate the database table that will be used. This
final table name will then be provided as the <cite>table</cite> argument value within
the decorator. In this way, these methods can be called like:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">(</span><span class="n">table_suffix</span><span class="o">=</span><span class="s2">&quot;_collapse&quot;</span><span class="p">):</span>
    <span class="c1"># Do stuff</span>
</pre></div>
</div>
<p>In this case, the <a class="reference internal" href="#trifusion.process.sequence.setup_intable" title="trifusion.process.sequence.setup_intable"><code class="xref py py-func docutils literal"><span class="pre">setup_intable()</span></code></a> decorator will append the
<cite>table_suffix</cite> to the name of the original alignment table. If
<code class="xref py py-attr docutils literal"><span class="pre">Alignment.table_name`=&quot;main&quot;</span></code>, then the final table name in this
case will be &#8220;main_collapse&#8221;.</p>
<p>Alternatively, we can use <cite>table_name</cite>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">iter_sequences</span><span class="p">(</span><span class="n">table_name</span><span class="o">=</span><span class="s2">&quot;main_2&quot;</span><span class="p">):</span>
    <span class="c1"># Do stuff</span>
</pre></div>
</div>
<p>In this case, the final table name will be &#8220;main_2&#8221;.</p>
<p>If the final table name does not exist, the method falls back to the original
table name defined by the <cite>table_name</cite> attribute.</p>
</div>
<div class="section" id="alignment-modifiers">
<h3>Alignment modifiers<a class="headerlink" href="#alignment-modifiers" title="Permalink to this headline">¶</a></h3>
<p>Methods that perform modifications to the alignment are also defined here.
These include:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-meth docutils literal"><span class="pre">collapse()</span></code></a>: Collapses alignment into unique sequences.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.consensus" title="trifusion.process.sequence.Alignment.consensus"><code class="xref py py-meth docutils literal"><span class="pre">consensus()</span></code></a>: Merges alignment into a single sequence.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_codon_positions" title="trifusion.process.sequence.Alignment.filter_codon_positions"><code class="xref py py-meth docutils literal"><span class="pre">filter_codon_positions()</span></code></a>: Filters alignment columns
according to codon position.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="trifusion.process.sequence.Alignment.filter_missing_data"><code class="xref py py-meth docutils literal"><span class="pre">filter_missing_data()</span></code></a>: Filters columns according to
missing data.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment.code_gaps" title="trifusion.process.sequence.Alignment.code_gaps"><code class="xref py py-meth docutils literal"><span class="pre">code_gaps()</span></code></a>: Code alignment&#8217;s indel patterns as a
binary matrix.</li>
</ul>
</div></blockquote>
<p>For example, the <a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.collapse()</span></code></a> method transforms the original
alignment into a new one that contains only unique sequences. An important
factor to take into account with alignment modifying methods, is that it may
be important to preserve the original alignment data for future operations.
In TriFusion, the original alignment must be available at all times since
users may perform any number of process executions in a single session.
Therefore, all methods that can potentially modify the original alignment
need to be decorated with the <a class="reference internal" href="#trifusion.process.sequence.setup_database" title="trifusion.process.sequence.setup_database"><code class="xref py py-func docutils literal"><span class="pre">setup_database()</span></code></a> function, and must be
defined with at least the <cite>table_in</cite> and <cite>table_out</cite> arguments. The
decorator and these arguments will work together to determine the database&#8217;s
table from where the data will be fetched, and to where the modified
alignment will be written. For instance, the <a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.collapse()</span></code></a>
method is defined as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="nd">@setup_database</span>
<span class="k">def</span> <span class="nf">collapse</span><span class="p">(</span><span class="o">...</span><span class="p">,</span> <span class="n">table_in</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;collapsed&quot;</span><span class="p">,</span>
             <span class="n">use_main_table</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
</pre></div>
</div>
<p>If we want to perform a collapse of the original alignment and store the
modified alignment in a new table, we could call <a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.collapse()</span></code></a>
like:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">collapse</span><span class="p">(</span><span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;new_table&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="#trifusion.process.sequence.setup_database" title="trifusion.process.sequence.setup_database"><code class="xref py py-func docutils literal"><span class="pre">setup_database()</span></code></a> decorator interprets <cite>table_in=None</cite> as an
instruction to use the table with the original alignment, and stores the
modified alignment in <cite>&#8220;new_table&#8221;</cite>.</p>
<p>However, we may want to perform a collapse operation after a previous
modification from other method. In that case, we can specify a
<cite>table_in</cite>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">collapse</span><span class="p">(</span><span class="n">table_in</span><span class="o">=</span><span class="s2">&quot;other_table&quot;</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="s2">&quot;collapse&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>One issue with this approach is that we do not know <em>a priori</em> which
operations will be requested by the user nor the order. If one execution
performs, say, a <cite>consensus</cite> and a <cite>collapse</cite>, the new table should be
created in <cite>consensus</cite> and then used as input in <cite>collapse</cite>. However,
if only <cite>collapse</cite> is called, then the new table should only be created
there. To solve this issue, the <cite>setup_database</cite> decorator is smart about
its arguments. We can create a sequence of operations with the same
<cite>table_in</cite> and <cite>table_out</cite> arguments:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">new_table</span> <span class="o">=</span> <span class="s2">&quot;master_table&quot;</span>
<span class="k">if</span> <span class="s2">&quot;consensus&quot;</span> <span class="ow">in</span> <span class="n">operations</span><span class="p">:</span>
    <span class="n">consensus</span><span class="p">(</span><span class="n">table_in</span><span class="o">=</span><span class="n">new_table</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="n">new_table</span><span class="p">)</span>
<span class="k">if</span> <span class="s2">&quot;collapse&quot;</span> <span class="ow">in</span> <span class="n">operations</span><span class="p">:</span>
    <span class="n">collapse</span><span class="p">(</span><span class="n">table_in</span><span class="o">=</span><span class="n">new_table</span><span class="p">,</span> <span class="n">table_out</span><span class="o">=</span><span class="n">new_table</span><span class="p">)</span>
</pre></div>
</div>
<p>In this simple pipeline, the user may perform either a <cite>consensus</cite>,
a <cite>collapse</cite>, or both. When the first method is called,
the <a class="reference internal" href="#trifusion.process.sequence.setup_database" title="trifusion.process.sequence.setup_database"><code class="xref py py-func docutils literal"><span class="pre">setup_database()</span></code></a> decorator will check if the table provided in
<cite>table_in</cite> exists. In the first called method it will not exist, so instead
of returning an error, it falls back to the original alignment table and
then writes the modified alignment to &#8220;master_table&#8221;. In the second method,
<cite>table_in</cite> already exists, so it fetches alignment data from the
&#8220;master_table&#8221;. This will work whether these methods are called individually
or in combination.</p>
<p>When there is no need to keep the original alignment data (in single
execution of TriSeq, for instance), the special <cite>use_main_table</cite> argument
can be provided to tell the method to use the original table as the input
and output table. If this argument is True, it supersedes any information
provided by <cite>table_in</cite> or <cite>table_out</cite>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">collapse</span><span class="p">(</span><span class="n">use_main_table</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="writers">
<h3>Writers<a class="headerlink" href="#writers" title="Permalink to this headline">¶</a></h3>
<p>Like parsers, writer methods are defined with <cite>_write_&lt;format&gt;</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_fasta" title="trifusion.process.sequence.Alignment._write_fasta"><code class="xref py py-meth docutils literal"><span class="pre">_write_fasta()</span></code></a>: Writes to fasta format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_phylip" title="trifusion.process.sequence.Alignment._write_phylip"><code class="xref py py-meth docutils literal"><span class="pre">_write_phylip()</span></code></a>: Writes to phylip format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_nexus" title="trifusion.process.sequence.Alignment._write_nexus"><code class="xref py py-meth docutils literal"><span class="pre">_write_nexus()</span></code></a>: Writes to nexus format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_stockholm" title="trifusion.process.sequence.Alignment._write_stockholm"><code class="xref py py-meth docutils literal"><span class="pre">_write_stockholm()</span></code></a>: Writes to stockholm format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_gphocs" title="trifusion.process.sequence.Alignment._write_gphocs"><code class="xref py py-meth docutils literal"><span class="pre">_write_gphocs()</span></code></a>: Writes to gphocs format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_ima2" title="trifusion.process.sequence.Alignment._write_ima2"><code class="xref py py-meth docutils literal"><span class="pre">_write_ima2()</span></code></a>: Writes to IMa2 format.</li>
<li><a class="reference internal" href="#trifusion.process.sequence.Alignment._write_mcmctree" title="trifusion.process.sequence.Alignment._write_mcmctree"><code class="xref py py-meth docutils literal"><span class="pre">_write_mcmctree()</span></code></a>: Writes to MCMCTree format.</li>
</ul>
</div></blockquote>
<p>They are always called from the <a class="reference internal" href="#trifusion.process.sequence.Alignment.write_to_file" title="trifusion.process.sequence.Alignment.write_to_file"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.write_to_file()</span></code></a> method,
not directly. When the <a class="reference internal" href="#trifusion.process.sequence.Alignment.write_to_file" title="trifusion.process.sequence.Alignment.write_to_file"><code class="xref py py-meth docutils literal"><span class="pre">Alignment.write_to_file()</span></code></a> method is called,
a list with the requested output formats is also provided as an argument.
For each format specified in the argument, the corresponding writer method
is called. That method is responsible for fetching the data from the
database and write it to an output file in the appropriate format. The map
between the formats and the methods is stored in a dictionary:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">write_methods</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_fasta</span><span class="p">,</span>
    <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_phylip</span><span class="p">,</span>
    <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_nexus</span><span class="p">,</span>
    <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_stockholm</span><span class="p">,</span>
    <span class="s2">&quot;gphocs&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_gphocs</span><span class="p">,</span>
    <span class="s2">&quot;ima2&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_ima2</span><span class="p">,</span>
    <span class="s2">&quot;mcmctree&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_write_mcmctree</span>
<span class="p">}</span>

<span class="c1"># Call apropriate method for each output format</span>
<span class="k">for</span> <span class="n">fmt</span> <span class="ow">in</span> <span class="n">output_format</span><span class="p">:</span>
    <span class="n">write_methods</span><span class="p">[</span><span class="n">fmt</span><span class="p">](</span><span class="n">output_file</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>
</div>
<p>The <cite>output_file</cite> and a <cite>kwargs</cite> dictionary are provided as arguments to
each of these methods. The <cite>kwargs</cite> dictionary contains all keyword
arguments used when calling <cite>write_to_file</cite> and each writer method fetches
the ones relevant to the format. For instance, in the beginning of the
<a class="reference internal" href="#trifusion.process.sequence.Alignment._write_fasta" title="trifusion.process.sequence.Alignment._write_fasta"><code class="xref py py-meth docutils literal"><span class="pre">_write_fasta()</span></code></a> method, the relevant keyword arguments are
retrieved:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Get relevant keyword arguments</span>
<span class="n">ld_hat</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ld_hat&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">interleave</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;interleave&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
<span class="n">table_suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_suffix&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">table_name</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;table_name&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">ns_pipe</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ns_pipe&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">pbar</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;pbar&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<p>In addition to the <cite>write_methods</cite> dictionary, a dictionary mapping the
formats and their corresponding file extensions is also defined:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">format_ext</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ima2&quot;</span><span class="p">:</span> <span class="s2">&quot;.txt&quot;</span><span class="p">,</span>
              <span class="s2">&quot;mcmctree&quot;</span><span class="p">:</span> <span class="s2">&quot;_mcmctree.phy&quot;</span><span class="p">,</span>
              <span class="s2">&quot;phylip&quot;</span><span class="p">:</span> <span class="s2">&quot;.phy&quot;</span><span class="p">,</span>
              <span class="s2">&quot;nexus&quot;</span><span class="p">:</span> <span class="s2">&quot;.nex&quot;</span><span class="p">,</span>
              <span class="s2">&quot;fasta&quot;</span><span class="p">:</span> <span class="s2">&quot;.fas&quot;</span><span class="p">,</span>
              <span class="s2">&quot;stockholm&quot;</span><span class="p">:</span> <span class="s2">&quot;.stockholm&quot;</span><span class="p">,</span>
              <span class="s2">&quot;gphocs&quot;</span><span class="p">:</span> <span class="s2">&quot;.txt&quot;</span><span class="p">}</span>
</pre></div>
</div>
<p>To add a new output format writer, simply create the method using the
<cite>_write_&lt;format&gt;</cite> notation and include it in the <cite>write_methods</cite> dictionary,
along with the extension in the <cite>format_ext</cite> variable.</p>
</div>
</div>
<div class="section" id="alignmentlist-class">
<h2><cite>AlignmentList</cite> class<a class="headerlink" href="#alignmentlist-class" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> is the main interface between the user and the
alignment data. It may contain one or more <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> objects,
which are considered the building blocks of the data set. These
<a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> objects are bound by the same sqlite database connection.</p>
<div class="section" id="how-to-create-an-instance">
<h3>How to create an instance<a class="headerlink" href="#how-to-create-an-instance" title="Permalink to this headline">¶</a></h3>
<p>An <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> instance can be created with a single argument,
which is a list of paths to alignment files:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">aln_obj</span> <span class="o">=</span> <span class="n">AlignmentList</span><span class="p">([</span><span class="s2">&quot;file1.fas&quot;</span><span class="p">,</span> <span class="s2">&quot;file2.fas&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Even when no information is provided, an sqlite database connection is
established automatically (generating a &#8220;trifusion.sqlite3&#8221; file in the
current working directory by default). However, it is possible and advisable
to specify the path to the sqlite database:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">aln_obj</span> <span class="o">=</span> <span class="n">AlignmentList</span><span class="p">([</span><span class="s2">&quot;file1.fas&quot;</span><span class="p">,</span> <span class="s2">&quot;file2.fas&quot;</span><span class="p">],</span> <span class="n">sql_path</span><span class="o">=</span><span class="s2">&quot;.sql.db&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>A connection to the database can also be provided when instantiating the
class, so it&#8217;s perhaps more useful to see how the
<code class="xref py py-meth docutils literal"><span class="pre">__init__()</span></code> works:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alignment_list</span><span class="p">,</span> <span class="n">sql_db</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">db_cur</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">db_con</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
             <span class="n">pbar</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

    <span class="k">if</span> <span class="n">db_cur</span> <span class="ow">and</span> <span class="n">db_con</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">con</span> <span class="o">=</span> <span class="n">db_con</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cur</span> <span class="o">=</span> <span class="n">db_cur</span>
    <span class="k">elif</span> <span class="n">sql_db</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span> <span class="o">=</span> <span class="n">sql_db</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sql_path</span> <span class="o">=</span> <span class="s2">&quot;trifusion.sqlite3&quot;</span>
</pre></div>
</div>
<p>As we can see, if a database connection is provided via the <cite>db_cur</cite> and
<cite>db_con</cite> arguments, the <cite>sql_path</cite> is ignored. If no sqlite information
is provided, the <cite>sql_path</cite> attribute defaults to &#8220;trifusion.sqlite3&#8221;.</p>
</div>
<div class="section" id="adding-alignment-data">
<h3>Adding alignment data<a class="headerlink" href="#adding-alignment-data" title="Permalink to this headline">¶</a></h3>
<p>Alignment data can be loaded when initializing the class as above and/or
added later via the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignments" title="trifusion.process.sequence.AlignmentList.add_alignments"><code class="xref py py-meth docutils literal"><span class="pre">add_alignments()</span></code></a> and
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignment_files" title="trifusion.process.sequence.AlignmentList.add_alignment_files"><code class="xref py py-meth docutils literal"><span class="pre">add_alignment_files()</span></code></a> methods (in fact,
the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignment_files" title="trifusion.process.sequence.AlignmentList.add_alignment_files"><code class="xref py py-meth docutils literal"><span class="pre">add_alignment_files()</span></code></a> method is the one called at
<code class="xref py py-meth docutils literal"><span class="pre">__init__()</span></code>):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">aln_obj</span><span class="o">.</span><span class="n">add_alignment_files</span><span class="p">([</span><span class="s2">&quot;file3.fas&quot;</span><span class="p">,</span> <span class="s2">&quot;file4.fas&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>The most important aspects when adding alignment data is how each alignment
is processed and how errors and exceptions are handled. Briefly, the flow is:</p>
<blockquote>
<div><ol class="arabic">
<li><p class="first">Check for file duplications within the loaded file list. Duplications
are stored in the <cite>duplicate_alignments</cite> attribute.</p>
</li>
<li><p class="first">Check for file duplications between the loaded files and any
alignment already present. The duplications go to the same attribute
as above.</p>
</li>
<li><p class="first">For each provided file, create an <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object. Errors
that occur when creating an <cite>Alignment</cite> object are stored in its
<a class="reference internal" href="#trifusion.process.sequence.Alignment.e" title="trifusion.process.sequence.Alignment.e"><code class="xref py py-attr docutils literal"><span class="pre">e</span></code></a>
attribute. This attribute is then checked before adding the alignment to
the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> object.</p>
<blockquote>
<div><ol class="arabic simple">
<li>Check for <cite>InputError</cite> exception. These are malformated files
and are stored in the <cite>bad_alignments</cite> attribute.</li>
<li>Check for <cite>AlignmentUnequalLength</cite> exception. These are sequence
sets of
unequal length (not alignments) and are stored in the
<cite>non_alignments</cite> attribute.</li>
<li>Check for <cite>EmptyAlignment</cite> exception. These are empty alignments
and are
stored in the <cite>bad_alignments</cite> attribute.</li>
<li>Check the sequence code of the <cite>Alignment</cite> object. For now,
the <cite>AlignmentList</cite> object only accepts alignments of the same
sequence type (e.g. DNA, protein).</li>
<li>If all of the previous checks pass, add the <cite>Alignment</cite> object
to the <cite>alignments</cite> attribute.</li>
</ol>
</div></blockquote>
</li>
<li><p class="first">Update the overall taxa names of the full data set after the
inclusion of the new alignments.</p>
</li>
</ol>
</div></blockquote>
</div>
<div class="section" id="wrapping-alignment-methods">
<h3>Wrapping <cite>Alignment</cite> methods<a class="headerlink" href="#wrapping-alignment-methods" title="Permalink to this headline">¶</a></h3>
<p>The majority of the methods defined in the <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object can
also be accessed in the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> object. These are defined
roughly with the same arguments in both classes so that their behavior is
the same. These can be simple wrappers that call the respective
<a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> method for each alignment in the
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList.alignments" title="trifusion.process.sequence.AlignmentList.alignments"><code class="xref py py-attr docutils literal"><span class="pre">alignments</span></code></a> attribute. For instance,
the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.change_taxon_name" title="trifusion.process.sequence.AlignmentList.change_taxon_name"><code class="xref py py-meth docutils literal"><span class="pre">change_taxon_name()</span></code></a> method is simply:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">change_taxon_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">):</span>

    <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="n">alignment_obj</span><span class="o">.</span><span class="n">change_taxon_name</span><span class="p">(</span><span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">new_name</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">old_name</span> <span class="k">else</span> <span class="n">x</span>
                       <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">taxa_names</span><span class="p">]</span>
</pre></div>
</div>
<p>To avoid duplicating the argument list, the wrapping method can use <cite>args</cite>
and <cite>kwargs</cite> to transfer arguments. This ensures that if the argument list
is modified in the <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> method, it doesn&#8217;t need any
modification in the wrapper method. For instance,
the <a class="reference internal" href="#trifusion.process.sequence.Alignment.write_to_file" title="trifusion.process.sequence.Alignment.write_to_file"><code class="xref py py-meth docutils literal"><span class="pre">write_to_file()</span></code></a> method of <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> accepts
a large number of positional and keyword arguments, which would be an hassle
to define an maintain in the wrapper method of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>. So,
the <code class="xref py py-meth docutils literal"><span class="pre">write_to_file()</span></code> method of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>
is simply defined as:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">write_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_format</span><span class="p">,</span> <span class="n">conversion_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                  <span class="n">output_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
</pre></div>
</div>
<p>The <cite>output_format</cite>, <cite>conversion_suffix</cite> and <cite>output_suffix</cite> are the only
positional arguments required when calling this wrapper. All the remaining
arguments are packed in the <cite>args</cite> and <cite>kwargs</cite> objects and used
normally by the wrapped method.</p>
</div>
<div class="section" id="exclusive-alignmentlist-methods">
<h3>Exclusive <cite>AlignmentList</cite> methods<a class="headerlink" href="#exclusive-alignmentlist-methods" title="Permalink to this headline">¶</a></h3>
<p>Some methods are exclusive of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> because they only make
sense to be applied to lists of alignments (e.g.
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList.concatenate" title="trifusion.process.sequence.AlignmentList.concatenate"><code class="xref py py-meth docutils literal"><span class="pre">concatenate()</span></code></a>). These have more freedom in how they
are defined and called.</p>
</div>
<div class="section" id="active-and-inactive-datasets">
<h3>Active and inactive datasets<a class="headerlink" href="#active-and-inactive-datasets" title="Permalink to this headline">¶</a></h3>
<p>Taxa and/or alignments can become &#8216;inactive&#8217;, that is, they are temporarily
removed from their respective attributes, <cite>taxa_names</cite> and <cite>alignments</cite>.
This means that these &#8216;inactive&#8217; elements are ignored when performing most
operations. To change the &#8216;active&#8217; status of alignments,
the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_active_alignments" title="trifusion.process.sequence.AlignmentList.update_active_alignments"><code class="xref py py-meth docutils literal"><span class="pre">update_active_alignments()</span></code></a> and
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_active_alignment" title="trifusion.process.sequence.AlignmentList.update_active_alignment"><code class="xref py py-meth docutils literal"><span class="pre">update_active_alignment()</span></code></a> methods are available. For
taxa, the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_taxa_names" title="trifusion.process.sequence.AlignmentList.update_taxa_names"><code class="xref py py-meth docutils literal"><span class="pre">update_taxa_names()</span></code></a> method can be used:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Set only two active alignments</span>
<span class="bp">self</span><span class="o">.</span><span class="n">update_active_alignments</span><span class="p">([</span><span class="s2">&quot;file1.fas&quot;</span><span class="p">,</span> <span class="s2">&quot;file2.fas&quot;</span><span class="p">])</span>

<span class="c1"># Set only two active taxa</span>
<span class="bp">self</span><span class="o">.</span><span class="n">update_taxa_names</span><span class="p">([</span><span class="s2">&quot;taxonA&quot;</span><span class="p">,</span> <span class="s2">&quot;taxonB&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Note that all these modifications are reversible. &#8216;Inactive&#8217; elements are
stored in the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.shelve_alignments" title="trifusion.process.sequence.AlignmentList.shelve_alignments"><code class="xref py py-attr docutils literal"><span class="pre">shelve_alignments</span></code></a> attribute for
alignments, and <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.shelved_taxa" title="trifusion.process.sequence.AlignmentList.shelved_taxa"><code class="xref py py-attr docutils literal"><span class="pre">shelved_taxa</span></code></a> for taxa.</p>
</div>
<div class="section" id="updating-alignment-objects">
<h3>Updating <cite>Alignment</cite> objects<a class="headerlink" href="#updating-alignment-objects" title="Permalink to this headline">¶</a></h3>
<p>Some tasks perform changes to core attributes of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a>,
but they may also be necessary on each <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object. For
instance, the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.remove_taxa" title="trifusion.process.sequence.AlignmentList.remove_taxa"><code class="xref py py-meth docutils literal"><span class="pre">remove_taxa()</span></code></a> method is used to remove a
list of taxa from the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> object. It is easy to change
only the relevant <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> attributes, but this change also
requires those particular taxa to be removed in all <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a>
objects. For this reason, such methods should be defined in the same way in
both classes. Using the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.remove_taxa" title="trifusion.process.sequence.AlignmentList.remove_taxa"><code class="xref py py-meth docutils literal"><span class="pre">remove_taxa()</span></code></a> example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">remove_taxa</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxa_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;remove&quot;</span><span class="p">):</span>

    <span class="c1"># &lt;changes to AlignmentList&gt;</span>

    <span class="k">for</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="n">alignment_obj</span><span class="o">.</span><span class="n">remove_taxa</span><span class="p">(</span><span class="n">taxa_list</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">)</span>
</pre></div>
</div>
<p>As you can see, the usage is the same for both methods.</p>
</div>
<div class="section" id="plot-data-methods">
<h3>Plot data methods<a class="headerlink" href="#plot-data-methods" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> object contains all methods that generate data
for plotting in TriFusion&#8217;s Statistics screen and TriStats CLI program.
However, it&#8217;s important to establish a separation between the generation of
plot data, and the generation of the plot itself. The
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> methods only generate the data and relevant
instructions necessary to to draw the plot. This information is then passed
on to the appropriate plot generation functions, which are defined in
<a class="reference internal" href="trifusion.base.plotter.html#module-trifusion.base.plotter" title="trifusion.base.plotter"><code class="xref py py-mod docutils literal"><span class="pre">trifusion.base.plotter</span></code></a>. The reason for this separation of tasks is
that many different alignment analyses are represented by the same plot.</p>
<p>The complete process of how new plots can be added to TriFusion is
described <a class="reference external" href="https://github.com/ODiogoSilva/TriFusion/wiki/Add-Statistics-plot-analysis">here</a>. In this section, we provide only a few guidelines on what
to expect from these methods.</p>
<p>All plot data methods must be decorated with the <a class="reference internal" href="#trifusion.process.sequence.check_data" title="trifusion.process.sequence.check_data"><code class="xref py py-func docutils literal"><span class="pre">check_data()</span></code></a> decorator
and take at least a <cite>Namespace</cite> argument. In most cases, no more arguments
are required:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="nd">@check_data</span>
<span class="k">def</span> <span class="nf">gene_occupancy</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="#trifusion.process.sequence.check_data" title="trifusion.process.sequence.check_data"><code class="xref py py-func docutils literal"><span class="pre">check_data()</span></code></a> decorator is responsible for performing checks
before and after executing the method. The <cite>Namespace</cite> argument, <cite>ns</cite>,
is used to allow communication between the main and worker threads of
TriFusion.</p>
<p>Additional keyword arguments may be defined, but in that case they must be
provided in TriFusion when the
<a class="reference internal" href="trifusion.app.html#trifusion.app.TriFusionApp.stats_show_plot" title="trifusion.app.TriFusionApp.stats_show_plot"><code class="xref py py-meth docutils literal"><span class="pre">trifusion.app.TriFusionApp.stats_show_plot()</span></code></a> method is called,
using the <cite>additional_args</cite> argument. This object will be passed to the
<a class="reference internal" href="trifusion.data.resources.background_tasks.html#trifusion.data.resources.background_tasks.get_stats_data" title="trifusion.data.resources.background_tasks.get_stats_data"><code class="xref py py-func docutils literal"><span class="pre">get_stats_data()</span></code></a> function
in <a class="reference internal" href="trifusion.data.resources.background_tasks.html#module-trifusion.data.resources.background_tasks" title="trifusion.data.resources.background_tasks"><code class="xref py py-mod docutils literal"><span class="pre">trifusion.data.resources.background_tasks</span></code></a> and used when calling
the plot data methods:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">additional_args</span><span class="p">:</span>
    <span class="n">plot_data</span> <span class="o">=</span> <span class="n">methods</span><span class="p">[</span><span class="n">stats_idx</span><span class="p">](</span><span class="n">ns</span><span class="o">=</span><span class="n">ns</span><span class="p">,</span> <span class="o">**</span><span class="n">additional_args</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">plot_data</span> <span class="o">=</span> <span class="n">methods</span><span class="p">[</span><span class="n">stats_idx</span><span class="p">](</span><span class="n">ns</span><span class="p">)</span>
</pre></div>
</div>
<p>The other requirement of plot data methods, is that they must always return
a single dictionary object. This dictionary must contain at least one
<cite>key:value</cite> with the &#8220;data&#8221; key and a numpy.array with the plot data as the
value. The other entries in the dictionary are optional and refer to
instructions for plot generation. For example, the
<cite>missing_data_distribution</cite> method returns:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">return</span> <span class="p">{</span><span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">,</span>
        <span class="s2">&quot;title&quot;</span><span class="p">:</span> <span class="s2">&quot;Distribution of missing data&quot;</span><span class="p">,</span>
        <span class="s2">&quot;legend&quot;</span><span class="p">:</span> <span class="n">legend</span><span class="p">,</span>
        <span class="s2">&quot;ax_names&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Proportion&quot;</span><span class="p">,</span> <span class="s2">&quot;Number of genes&quot;</span><span class="p">],</span>
        <span class="s2">&quot;table_header&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Bin&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">legend</span><span class="p">}</span>
</pre></div>
</div>
<p>The first entry is the mandatory &#8220;data&#8221; key with the numpy.array <cite>data</cite>.
The other instructions are &#8220;title&#8221;, which sets the title of the plot, &#8220;legend&#8221;,
which provides the labels for the plot legend, &#8220;ax_names&#8221;, which provides
the name of the <cite>x</cite> and <cite>y</cite> axis, and the &#8220;table_header&#8221;, which specified
the header for the table of that plot.</p>
<p>The allowed plot instructions depend on the plot function that will be used
and not all of them need to be specified.</p>
</div>
<div class="section" id="logging-progress">
<h3>Logging progress<a class="headerlink" href="#logging-progress" title="Permalink to this headline">¶</a></h3>
<p>The majority of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> methods support the setup and update of
progress indicators that can be used in TriFusion (GUI) and the CLI programs.
In the case of TriFusion, the progress indicator is provided via a
<cite>multiprocessing.Namespace</cite> object that transfers information between the
main thread (where the GUI changes take place) and the working thread.
In the case of the CLI programs, the indicator is provided via a <cite>ProgressBar</cite>
object. In either case, the setup, update and resetting of the progress
indicators is performed by the same methods.</p>
<p>At the beginning of an operation, the <code class="xref py py-meth docutils literal"><span class="pre">AlignmentList_set_pipes()</span></code>
method is called:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">))</span>
</pre></div>
</div>
<p>The <cite>Namespace</cite> object is defined as <cite>ns</cite> and the <cite>ProgressBar</cite> is defined
as <cite>pbar</cite>. Usually, only one of them is provided, depending on whether it
was called from TriFusion or from a CLI program. We also set the total of
the progress indicator. In this case it&#8217;s the number of alignments. In case
the operation is called from TriFusion using the <cite>Namespace</cite> object,
this method also checks the number of active alignments. If there is only
one active alignment, it sets a Namespace attribute that will silence the
progress logging of the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> object and receive the
information from the <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object instead:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ns</span><span class="o">.</span><span class="n">sa</span> <span class="o">=</span> <span class="kc">False</span>
</pre></div>
</div>
<p>Then, inside the task, we can update the progress within a loop using
the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList._update_pipes" title="trifusion.process.sequence.AlignmentList._update_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_update_pipes()</span></code></a> method:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">p</span><span class="p">,</span> <span class="n">alignment_obj</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alignments</span><span class="o">.</span><span class="n">values</span><span class="p">())):</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">,</span> <span class="n">pbar</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">p</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                       <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Some message&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, we provide the <cite>Namespace</cite> and <cite>ProgressBar</cite> objects as before.
In addition we provide the <cite>value</cite> associated with each iteration.
Optionally, the <cite>msg</cite> argument can be specified, which is used exclusively
by TriFusion to show a custom message.</p>
<p>At the end of a task its good practice to reset the progress indicators
by using the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList._reset_pipes" title="trifusion.process.sequence.AlignmentList._reset_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_reset_pipes()</span></code></a> method:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">_reset_pipes</span><span class="p">(</span><span class="n">ns</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, only the <cite>Namespace</cite> object is necessary, since the <cite>ProgressBar</cite>
indicator is automatically reset on <a class="reference internal" href="#trifusion.process.sequence.AlignmentList._set_pipes" title="trifusion.process.sequence.AlignmentList._set_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_set_pipes()</span></code></a>.</p>
</div>
<div class="section" id="incorporating-a-kill-switch">
<h3>Incorporating a kill switch<a class="headerlink" href="#incorporating-a-kill-switch" title="Permalink to this headline">¶</a></h3>
<p>All time consuming methods of <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> accept a <cite>Namespace</cite>
object when called from TriFusion, allowing communication between the main
thread and the worker thread. Since python <cite>threads</cite> cannot be forcibly
terminated like <cite>processes</cite>, most methods should listen to a kill switch
flag that is actually an attribute of the <cite>Namespace</cite> object. This kill
switch is already incorporated into the <a class="reference internal" href="#trifusion.process.sequence.AlignmentList._set_pipes" title="trifusion.process.sequence.AlignmentList._set_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_set_pipes()</span></code></a>,
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList._update_pipes" title="trifusion.process.sequence.AlignmentList._update_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_update_pipes()</span></code></a> and
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList._reset_pipes" title="trifusion.process.sequence.AlignmentList._reset_pipes"><code class="xref py py-meth docutils literal"><span class="pre">_reset_pipes()</span></code></a> methods.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">ns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ns</span><span class="o">.</span><span class="n">stop</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">KillByUser</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>What it does is to listen to a kill signal from TriFusion&#8217;s window (it can
be clicking on the &#8220;Cancel&#8221; button, for example). When this kill signal is
received in the main thread, it sets the <cite>Namespace.stop</cite> attribute to True
in both main and worker threads. In the worker thread, when the <cite>stop</cite>
attribute evaluates to True, a custom <cite>KillByUser</cite> exception is raised,
immediately stopping the task. The exception is then handled in the
<a class="reference internal" href="trifusion.data.resources.background_tasks.html#trifusion.data.resources.background_tasks.process_execution" title="trifusion.data.resources.background_tasks.process_execution"><code class="xref py py-func docutils literal"><span class="pre">trifusion.data.resources.background_tasks.process_execution()</span></code></a> function
and transmitted to the main thread.</p>
</div>
</div>
<div class="section" id="using-sqlite">
<h2>Using SQLite<a class="headerlink" href="#using-sqlite" title="Permalink to this headline">¶</a></h2>
<p>A great deal of the high performance and memory efficiency of the <cite>sequence</cite>
module comes from the use of sqlite to store and manipulate alignment data
on disk rather than RAM. This means that parsing, modifying and writing
alignment data must be done with great care to ensure that only the
essential data is loaded into memory, while minimizing the number of (
expensive) database queries. This has some implications for the methods of
both <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and <a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> objects with respect to
how parsing, alignment modification and output writing is performed.</p>
<div class="section" id="implications-for-parsing">
<h3>Implications for parsing<a class="headerlink" href="#implications-for-parsing" title="Permalink to this headline">¶</a></h3>
<p>When writing or modifying parsing methods it is important to take into
account that alignment files can be very large (&gt; 1Gb) and loading the
entire data into memory should be avoided. Whenever possible, only the data
of a single taxon should be kept in memory before inserting it into the
database and then releasing that memory. For most formats, particularly
leave formats, it&#8217;s fairly straightforward to do this. However, interleave
formats can fragment the data from each taxon across the entire file. Since
database insertions and updates are expensive, loading the data in each line
can greatly decrease the performance in these formats. Therefore,
it&#8217;s preferable to read the alignment file once per taxon, join the entire
sequence of that taxon, and then insert it into the database. This ensures
that only sequence data from one taxon is kept in memory at any given time
and only a minimal number of database insertions are performed. It will also
result in the same file being parsed N times, where N is the number of taxa.
However, the decrease in speed is marginal, since most lines are actually
skipped, whereas the efficiency increase is substantial.</p>
</div>
<div class="section" id="implications-for-fetching-data">
<h3>Implications for fetching data<a class="headerlink" href="#implications-for-fetching-data" title="Permalink to this headline">¶</a></h3>
<p>Retrieving data from an sqlite database is not as simple as accessing python
native data structure. Therefore, a set of methods and generators have
been defined in the <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> object to facilitate the interface with
the data in the database (See <a class="reference internal" href="#data-fetching">Data fetching</a>). When some kind of data
is required from the database, it is preferable to modify or create a
dedicated method, instead of interacting with the database directly. This
creates a layer of abstraction between the database and <a class="reference internal" href="#trifusion.process.sequence.Alignment" title="trifusion.process.sequence.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a>/
<a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-class docutils literal"><span class="pre">AlignmentList</span></code></a> methods that greatly facilitates future updates.</p>
</div>
<div class="section" id="implications-for-alignment-modification">
<h3>Implications for alignment modification<a class="headerlink" href="#implications-for-alignment-modification" title="Permalink to this headline">¶</a></h3>
<p>When performing modification to the alignment data it is important to take
into account that the original alignment data may need to be preserved for
future operations. These methods must be defined and called using
appropriate decorators and arguments that establish the name of the
database table from where information will be retrieved, and the name of the
table where the information will be written (See <a class="reference internal" href="#alignment-modifiers">Alignment modifiers</a>).</p>
</div>
<div class="section" id="implications-for-writing-files">
<h3>Implications for writing files<a class="headerlink" href="#implications-for-writing-files" title="Permalink to this headline">¶</a></h3>
<p>When writing alignment data into new output files, the same caution of the
alignment parsing is advised. It&#8217;s quite easy to let the entire alignment
be loaded into RAM, particularly when writing in interleave format.</p>
<dl class="class">
<dt id="trifusion.process.sequence.Alignment">
<em class="property">class </em><code class="descclassname">trifusion.process.sequence.</code><code class="descname">Alignment</code><span class="sig-paren">(</span><em>input_alignment</em>, <em>input_format=None</em>, <em>partitions=None</em>, <em>locus_length=None</em>, <em>sequence_code=None</em>, <em>taxa_list=None</em>, <em>taxa_idx=None</em>, <em>sql_cursor=None</em>, <em>sql_con=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="trifusion.process.base.html#trifusion.process.base.Base" title="trifusion.process.base.Base"><code class="xref py py-class docutils literal"><span class="pre">trifusion.process.base.Base</span></code></a></p>
<p>Main interface for single alignment files.</p>
<p>The <cite>Alignment</cite> class is the main interface for single alignment files,
providing methods that parse, query, retrieve and modify alignment data.
It is usually created within the context of <cite>AlignmentList</cite>,
which handles the connection with the sqlite database. Nevertheless,
<cite>Alignment</cite> instances can be created by providing the <cite>input_alignment</cite>
argument, which can be one of the following:</p>
<ul class="simple">
<li>path to alignment file. In this case, the file is parsed and
information stored in a new sqlite table.</li>
<li>sqlite table name. In this case, the table name must be present
in the database.</li>
</ul>
<p>In either case, the sqlite <cite>Connection</cite> and <cite>Cursor</cite> objects should
be provided.</p>
<p>When an <cite>Alignment</cite> object is instantiated, it first generates the 
<cite>table_name</cite> based on the <cite>input_alignment</cite> string, filtering all
characters that are not alpha numeric. Then, it queries the database
to check is a table already exists with that name. If yes, it is assumed
that the alignment data is stored in the provided table name. If there
is no table with that name, it is assumed that <cite>input_alignment</cite> is a 
path to the alignment file and the regular parsing ensues. An empty
table is created, the sequence type, format and missing data symbol are
automatically detected and the alignment is parsed according to the
detected format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>input_alignment</strong> : str</p>
<blockquote>
<div><p>Can be either the path to an alignment file or the name of a table
in the sqlite database.</p>
</div></blockquote>
<p><strong>sql_cursor</strong> : sqlite3.Cursor</p>
<blockquote>
<div><p>Cursor object of the sqlite database.</p>
</div></blockquote>
<p><strong>sql_con</strong> : sqlite3.Connection</p>
<blockquote>
<div><p>Connection object of the sqlite database.</p>
</div></blockquote>
<p><strong>input_format</strong> : str, optional</p>
<blockquote>
<div><p>File format of <cite>input_alignment</cite>. If <cite>input_alignment</cite> is a
file path, the format will be automatically detect from the file.
The value provided with this argument overrides the automatic 
detection&#8217;s result.</p>
</div></blockquote>
<p><strong>partitions</strong> : <cite>trifusion.process.data.Partitions</cite>, optional</p>
<blockquote>
<div><p>If provided, it will set the <cite>partitions</cite> attribute. This should
be used only when <cite>input_alignment</cite> is a database table name.</p>
</div></blockquote>
<p><strong>locus_length</strong> : int, optional</p>
<blockquote>
<div><p>Sets the length of the current alignment, stored in the <cite>locus_length</cite>
attribute. This option should only be used when <cite>input_alignment</cite>
is a database table name. Otherwise, it is automatically set during
alignment parsing.</p>
</div></blockquote>
<p><strong>sequence_code</strong> : tuple, optional</p>
<blockquote>
<div><p>Sets the <cite>sequence_code</cite> attribute with the information on
(&lt;sequence_type&gt;, &lt;missing data symbol&gt;). This option should only be</p>
<blockquote>
<div><p>used when <cite>input_alignment</cite> is a database table name. Otherwise, 
it is automatically set during alignment parsing.</p>
</div></blockquote>
</div></blockquote>
<p><strong>taxa_list</strong> : list, optional</p>
<blockquote>
<div><p>Sets the list attribute <cite>taxa_list</cite> with the names of the taxa
present in the alignment. This option should only be
used when <cite>input_alignment</cite> is a database table name. Otherwise,
it is automatically set during alignment parsing.</p>
</div></blockquote>
<p><strong>taxa_idx</strong> : dict, optional</p>
<blockquote class="last">
<div><p>Sets the dictionary attribute <cite>taxa_idx</cite> that maps the taxon names
to their index in the sqlite database table. This option should only
be used when <cite>input_alignment</cite> is a database table name. Otherwise,
it is automatically set during alignment parsing.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.AlignmentList" title="trifusion.process.sequence.AlignmentList"><code class="xref py py-obj docutils literal"><span class="pre">AlignmentList</span></code></a>, <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignment_files" title="trifusion.process.sequence.AlignmentList.add_alignment_files"><code class="xref py py-obj docutils literal"><span class="pre">AlignmentList.add_alignment_files</span></code></a>, <a class="reference internal" href="#trifusion.process.sequence.AlignmentList.retrieve_alignment" title="trifusion.process.sequence.AlignmentList.retrieve_alignment"><code class="xref py py-obj docutils literal"><span class="pre">AlignmentList.retrieve_alignment</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The <cite>Alignment</cite> class is not meant to be used directly (although it is
quite possible to do so if you handle the database connections before).
Instead, use the <cite>AlignmentList</cite> class, even if there is only one
alignment. All <cite>Alignment</cite> methods are available from the <cite>AlignmentList</cite>
and it is also possible to retrieve specific <cite>Alignment</cite> objects.</p>
<p>The <cite>Alignment</cite> class was designed to be a lightweight, fast and
powerful interface between alignment data and a set of manipulation
and transformation methods.
For performance and efficiency purposes, all alignment data is stored 
in a sqlite database that prevents the entire alignment from being
loaded into memory. To facilitate the retrieval and iteration over the
alignment data, several methods (<cite>iter_columns</cite>, &#8216;iter_sequences`, etc)
are available to handle the interface with the database and retrieve
only the necessary information.</p>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="93%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>cur</td>
<td>(sqlite3.Cursor) Cursor object of the sqlite database.</td>
</tr>
<tr class="row-even"><td>con</td>
<td>(sqlite3.Connection) Connection object of the sqlite database.</td>
</tr>
<tr class="row-odd"><td>table_name</td>
<td>(str) Name of the sqlite database&#8217;s table storing the sequence data.</td>
</tr>
<tr class="row-even"><td>tables</td>
<td>(list) Lists the sqlite database&#8217;s tables created for the <cite>Alignment</cite> instance. The &#8220;master&#8221; table is always present and represents the original alignment.</td>
</tr>
<tr class="row-odd"><td>partitions</td>
<td>(<cite>trifusion.process.data.Partitions</cite>) Stores the partitions and substitution model definition for the alignment.</td>
</tr>
<tr class="row-even"><td>locus_length</td>
<td>(int) Length of the alignment in base pairs or residues.</td>
</tr>
<tr class="row-odd"><td>restriction_range</td>
<td>(str) Only used when gaps are coded in a binary matrix. Stores a string with the range of the restriction-type data that will encode gaps and will only be used when nexus is in the output format.</td>
</tr>
<tr class="row-even"><td>e</td>
<td>(None or Exception) Stores any exceptions that occur during the parsing of the alignment file. It remains None unless something wrong happens.</td>
</tr>
<tr class="row-odd"><td>taxa_list</td>
<td>(list) List with the active taxon names.</td>
</tr>
<tr class="row-even"><td>taxa_idx</td>
<td>(dict) Maps the taxon names to their corresponding index in the sqlite database. The index is not retrieved from the position of the taxon in <cite>taxa_list</cite> to prevent messing up when taxa are removed from the <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-odd"><td>shelved_taxa</td>
<td>(list) List of ignored taxon names.</td>
</tr>
<tr class="row-even"><td>path</td>
<td>(str) Full path to alignment file.</td>
</tr>
<tr class="row-odd"><td>sname</td>
<td>(str) Basename of the alignment file without the extension.</td>
</tr>
<tr class="row-even"><td>name</td>
<td>(str) Basename of the alignment file with the extension.</td>
</tr>
<tr class="row-odd"><td>sequence_code</td>
<td>(tuple) Contains information on (&lt;sequence type&gt;, &lt;missing data symbol&gt;), e.g. (&#8220;Protein&#8221;, &#8220;x&#8221;).</td>
</tr>
<tr class="row-even"><td>interleave_data</td>
<td>(bool) Attribute that is set to True when interleave data has been created for the alignment data.</td>
</tr>
<tr class="row-odd"><td>input_format</td>
<td>(str) Format of the input alignment file.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">autofinder</span></code>(reference_file)</td>
<td>Autodetects format, missing data symbol and sequence type.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.change_taxon_name" title="trifusion.process.sequence.Alignment.change_taxon_name"><code class="xref py py-obj docutils literal"><span class="pre">change_taxon_name</span></code></a>(old_name,&nbsp;new_name)</td>
<td>Changes the name of a particular taxon.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.code_gaps" title="trifusion.process.sequence.Alignment.code_gaps"><code class="xref py py-obj docutils literal"><span class="pre">code_gaps</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Code gap patterns as binary matrix at the end of alignment.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-obj docutils literal"><span class="pre">collapse</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Collapses equal sequences into unique haplotypes.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.consensus" title="trifusion.process.sequence.Alignment.consensus"><code class="xref py py-obj docutils literal"><span class="pre">consensus</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates a consensus sequence from the alignment.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">duplicate_taxa</span></code>(taxa_list)</td>
<td>Identified duplicate items in a list.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_codon_positions" title="trifusion.process.sequence.Alignment.filter_codon_positions"><code class="xref py py-obj docutils literal"><span class="pre">filter_codon_positions</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filter alignment according to codon positions.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_informative_sites" title="trifusion.process.sequence.Alignment.filter_informative_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_informative_sites</span></code></a>(min_val,&nbsp;max_val[,&nbsp;...])</td>
<td>Checks if the variable sites from alignment are within range.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="trifusion.process.sequence.Alignment.filter_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">filter_missing_data</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filters alignment according to missing data.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_segregating_sites" title="trifusion.process.sequence.Alignment.filter_segregating_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_segregating_sites</span></code></a>(min_val,&nbsp;max_val[,&nbsp;...])</td>
<td>Checks if the segregating sites from alignment are within range.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">get_loci_taxa</span></code>(loci_file)</td>
<td>Get the list of taxa from a .loci file.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.get_sequence" title="trifusion.process.sequence.Alignment.get_sequence"><code class="xref py py-obj docutils literal"><span class="pre">get_sequence</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Returns the sequence string for a given taxon.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">guess_code</span></code>(sequence)</td>
<td>Guess the sequence type, i.e.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_alignment" title="trifusion.process.sequence.Alignment.iter_alignment"><code class="xref py py-obj docutils literal"><span class="pre">iter_alignment</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Generator for (taxon, sequence) tuples.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_columns" title="trifusion.process.sequence.Alignment.iter_columns"><code class="xref py py-obj docutils literal"><span class="pre">iter_columns</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Generator over alignment columns.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_columns_uniq" title="trifusion.process.sequence.Alignment.iter_columns_uniq"><code class="xref py py-obj docutils literal"><span class="pre">iter_columns_uniq</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Generator over unique characters of an alignment column.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.iter_sequences" title="trifusion.process.sequence.Alignment.iter_sequences"><code class="xref py py-obj docutils literal"><span class="pre">iter_sequences</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Generator over sequence strings in the alignment.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a>([size_check])</td>
<td>Main alignment parser method.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">read_basic_csv</span></code>(file_handle)</td>
<td>Reads a basic CSV into a list.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.remove_alignment" title="trifusion.process.sequence.Alignment.remove_alignment"><code class="xref py py-obj docutils literal"><span class="pre">remove_alignment</span></code></a>()</td>
<td>Removes alignment data from the database.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.remove_taxa" title="trifusion.process.sequence.Alignment.remove_taxa"><code class="xref py py-obj docutils literal"><span class="pre">remove_taxa</span></code></a>(taxa_list_file[,&nbsp;mode])</td>
<td>Removes taxa from the <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.reverse_concatenate" title="trifusion.process.sequence.Alignment.reverse_concatenate"><code class="xref py py-obj docutils literal"><span class="pre">reverse_concatenate</span></code></a>([table_in,&nbsp;db_con,&nbsp;ns,&nbsp;pbar])</td>
<td>Reverses the alignment according to the <cite>partitions</cite> attribute.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">rm_illegal</span></code>(taxon_string)</td>
<td>Removes illegal characters from taxon name.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.set_partitions" title="trifusion.process.sequence.Alignment.set_partitions"><code class="xref py py-obj docutils literal"><span class="pre">set_partitions</span></code></a>(partitions)</td>
<td>Updates the <cite>partition</cite> attribute.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.shelve_taxa" title="trifusion.process.sequence.Alignment.shelve_taxa"><code class="xref py py-obj docutils literal"><span class="pre">shelve_taxa</span></code></a>(lst)</td>
<td>Shelves taxa from <cite>Alignment</cite> methods.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.write_loci_correspondence" title="trifusion.process.sequence.Alignment.write_loci_correspondence"><code class="xref py py-obj docutils literal"><span class="pre">write_loci_correspondence</span></code></a>(hap_dict,&nbsp;output_file)</td>
<td>Writes the file mapping taxa to unique haplotypes for <cite>collapse</cite>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.Alignment.write_to_file" title="trifusion.process.sequence.Alignment.write_to_file"><code class="xref py py-obj docutils literal"><span class="pre">write_to_file</span></code></a>(output_format,&nbsp;output_file,&nbsp;...)</td>
<td>Writes alignment data into an output file.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="trifusion.process.sequence.Alignment._check_partitions">
<code class="descname">_check_partitions</code><span class="sig-paren">(</span><em>partition_obj</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._check_partitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._check_partitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Consistency check of the <cite>partition_obj</cite> for <cite>Alignment</cite> object.</p>
<p>Performs checks to ensure that a <cite>partition_obj</cite> is consistent
with the alignment data of the <cite>Alignment</cite> object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>partition_obj</strong> : trifusion.process.data.Partitions</p>
<blockquote class="last">
<div><p>Partitions object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="trifusion.process.data.html#trifusion.process.data.Partitions" title="trifusion.process.data.Partitions"><code class="xref py py-obj docutils literal"><span class="pre">trifusion.process.data.Partitions</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._create_table">
<code class="descname">_create_table</code><span class="sig-paren">(</span><em>table_name</em>, <em>cur=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._create_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._create_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a new table in the database.</p>
<p>Convenience method that creates a new table in the sqlite database.
It accepts a custom Cursor object, which overrides the <cite>cur</cite>
attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_name</strong> : str</p>
<blockquote>
<div><p>Name of the table to be created.</p>
</div></blockquote>
<p><strong>cur</strong> : sqlite3.Cursor, optional</p>
<blockquote class="last">
<div><p>Custom Cursor object used to query the database.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._eval_missing_symbol">
<code class="descname">_eval_missing_symbol</code><span class="sig-paren">(</span><em>sequence</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._eval_missing_symbol"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._eval_missing_symbol" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluates missing data symbol from sequence string.</p>
<p>This method is performed when sequence data is being parsed from the
alignment and only executes when the missing data symbol stored in
<cite>Alignment.sequence_code</cite> is not defined. It attempts to count
the regular characters used to denote missing data. First, tries to
find &#8221;?&#8221; characters, which are set as the symbol if they exist.
Then, it finds &#8220;n&#8221; characters and if sequence type is set to DNA,
this character is set as the symbol.
Finally, it finds &#8220;x&#8221; characters and if sequence type is set to
Protein, this character is set as the symbol.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sequence</strong> : str</p>
<blockquote class="last">
<div><p>Sequence string</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._filter_columns">
<code class="descname">_filter_columns</code><span class="sig-paren">(</span><em>gap_threshold</em>, <em>missing_threshold</em>, <em>table_in</em>, <em>table_out</em>, <em>ns=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._filter_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._filter_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters alignment columns based on missing data content.</p>
<p>Filters alignment columns based on the amount of missing data and
gap content. If the calculated content is higher than a defined
threshold, the column is removed from the final alignment. Setting
<cite>gap_threshold</cite> and <cite>missing_threshold</cite> to 0 results in an alignment
with no missing data, while setting them to 100 will not remove any
column.</p>
<p>This method should not be called directly. Instead it is used by the
<cite>filter_missing_data</cite> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gap_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap percentage is removed.</p>
</div></blockquote>
<p><strong>missing_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap+missing percentage is removed.</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;collapsed&#8221;).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocessing.Manager.Namespace, optional</p>
<blockquote class="last">
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="trifusion.process.sequence.Alignment.filter_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">filter_missing_data</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._filter_terminals">
<code class="descname">_filter_terminals</code><span class="sig-paren">(</span><em>table_in</em>, <em>table_out</em>, <em>ns=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._filter_terminals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._filter_terminals" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace gaps at alignment&#8217;s extremities with missing data.</p>
<p>This method should not be called directly. Instead it is used by the
<cite>filter_missing_data</cite> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;collapsed&#8221;).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote class="last">
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="trifusion.process.sequence.Alignment.filter_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">filter_missing_data</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._get_interleave_data">
<code class="descname">_get_interleave_data</code><span class="sig-paren">(</span><em>table_name=None</em>, <em>table_suffix=None</em>, <em>ns_pipe=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._get_interleave_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._get_interleave_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds an interleave matrix on the database.</p>
<p>When the user requests an interleave output format, this method
rearranges the data matrix (which is leave in the database) in
interleave format in a new table. This should be performed only once,
even when multiple formats are specified in interleave. The first one
calls this class, and the subsequent ones use the this table.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_fasta">
<code class="descname">_read_fasta</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for fasta format.</p>
<p>Parses a fasta alignment file and stores taxa and sequence data in
the database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_interleave_nexus">
<code class="descname">_read_interleave_nexus</code><span class="sig-paren">(</span><em>ntaxa</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_interleave_nexus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_interleave_nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for interleave nexus format.</p>
<p>Parses an interleave nexus alignment file and stores taxa and
sequence data in the database.This method is only called from the
<cite>_read_nexus</cite> method, when the regular nexus parser detects that the
file is in interleave format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ntaxa</strong> : int</p>
<blockquote>
<div><p>Number of taxa contained in the alignment file.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_nexus" title="trifusion.process.sequence.Alignment._read_nexus"><code class="xref py py-obj docutils literal"><span class="pre">_read_nexus</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The nexus interleave format splits the alignment into blocks of
a certain lenght (usually 90 characters) separated by blank lines.
This means that to gather the complete sequence of a given taxon,
the parser has to read the entire file. To prevent the entire
alignment to be loaded into memory, we actually iterate over a range
determined by the number of taxa in the alignment. In each iteration,
we open a file handle and we retrieve only the sequence of a
particular taxon. This ensures that only sequence data for
a single taxon is store in memory at any given time. This also means
that the alignment file has to be read N times, where N = number of
taxa. However, since the vast majority of lines are actually skipped
in each iteration, the decrease in speed is marginal, while the
gains in memory efficient are much larger.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_interleave_phylip">
<code class="descname">_read_interleave_phylip</code><span class="sig-paren">(</span><em>ntaxa</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_interleave_phylip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_interleave_phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for interleave phylip format.</p>
<p>Parses an interleave phylip alignment file and stores taxa and
sequence data in the database. This method is only called from the
<cite>_read_phylip</cite> method, when the regular phylip parser detects that the
file is in interleave format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ntaxa</strong> : int</p>
<blockquote>
<div><p>Number of taxa contained in the alignment file.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment._read_phylip" title="trifusion.process.sequence.Alignment._read_phylip"><code class="xref py py-obj docutils literal"><span class="pre">_read_phylip</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The phylip interleave format splits the alignment into blocks of
a certain lenght (usually 90 characters) separated by blank lines.
This means that to gather the complete sequence of a given taxon,
the parser has to read the entire file. To prevent the entire
alignment to be loaded into memory, we actually iterate over a range
determined by the number of taxa in the alignment. In each iteration,
we open a file handle and we retrieve only the sequence of
a particular taxon. This ensures that only sequence data for
a single taxon is store in memory at any given time. This also means
that the alignment file has to be read N times, where N = number of
taxa. However, since the vast majority of lines are actually skipped
in each iteration, the decrease in speed is marginal, while the
gains in memory efficient are much larger.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_loci">
<code class="descname">_read_loci</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_loci"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_loci" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for pyRAD and ipyrad loci format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_nexus">
<code class="descname">_read_nexus</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_nexus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for nexus format.</p>
<p>Parses a nexus alignment file and stores taxa and sequence data in
the database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_phylip">
<code class="descname">_read_phylip</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_phylip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for phylip format.</p>
<p>Parses a phylip alignment file and stored taxa and sequence data in
the database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._read_stockholm">
<code class="descname">_read_stockholm</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._read_stockholm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._read_stockholm" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment parser for stockholm format.</p>
<p>Parses a stockholm alignment file and stores taxa and sequence data in
the database.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>size_list</strong> : list</p>
<blockquote class="last">
<div><p>List of sequence size (int) for each taxon in the alignment file.
Used to check size consistency at the end of the alignment parsing
in <cite>read_alignment</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.read_alignment" title="trifusion.process.sequence.Alignment.read_alignment"><code class="xref py py-obj docutils literal"><span class="pre">read_alignment</span></code></a></p>
</div>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.Alignment._reset_pipes">
<em class="property">static </em><code class="descname">_reset_pipes</code><span class="sig-paren">(</span><em>ns</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._reset_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._reset_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset progress indicators for both GUI and CLI task executions.</p>
<p>This should be done at the end of any task that initialized the
progress objects, but it only affects the Namespace object. It
resets all Namespace attributes to None.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote class="last">
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment._set_pipes" title="trifusion.process.sequence.Alignment._set_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_set_pipes</span></code></a>, <a class="reference internal" href="#trifusion.process.sequence.Alignment._update_pipes" title="trifusion.process.sequence.Alignment._update_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_update_pipes</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._set_format">
<code class="descname">_set_format</code><span class="sig-paren">(</span><em>input_format</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._set_format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._set_format" title="Permalink to this definition">¶</a></dt>
<dd><p>Manually set the input format of the <cite>Alignment</cite> object.</p>
<p>Manually sets the input format associated with the <cite>Alignment</cite> object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>input_format</strong> : str</p>
<blockquote class="last">
<div><p>The input format.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.Alignment._set_pipes">
<em class="property">static </em><code class="descname">_set_pipes</code><span class="sig-paren">(</span><em>ns=None</em>, <em>pbar=None</em>, <em>total=None</em>, <em>msg=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._set_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._set_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Setup of progress indicators for both GUI and CLI task executions.</p>
<blockquote>
<div><p>This handles the setup of the objects responsible for updating
the progress of task&#8217;s execution of both TriFusion (GUI) and
TriSeq (CLI). At the beginning of any given task, these objects
can be initialized by providing either the Namespace object (<cite>ns</cite>)
in the case of TriFusion, or the ProgressBar object (<cite>pbar</cite>), in
the case of TriSeq. Along with one of these objects, the expected
<cite>total</cite> of the progress should also be provided. The <cite>ns</cite> and
<cite>pbar</cite> objects are updated at each iteration of a given task,
and the <cite>total</cite> is used to get a measure of the progress.</p>
<p>Optionally, a message can be also provided for the Namespace object
that will be used by TriFusion.</p>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>total</strong> : int</p>
<blockquote>
<div><p>Expected total of the task&#8217;s progress.</p>
</div></blockquote>
<p><strong>msg</strong> : str, optional</p>
<blockquote class="last">
<div><p>A secondary message that appears in TriFusion&#8217;s progress dialogs.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment._reset_pipes" title="trifusion.process.sequence.Alignment._reset_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_reset_pipes</span></code></a>, <a class="reference internal" href="#trifusion.process.sequence.Alignment._update_pipes" title="trifusion.process.sequence.Alignment._update_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_update_pipes</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The progress of any given task can be provided by either an
<cite>Alignment</cite> or <cite>AlignmentList</cite> instance. Generally, the tasks follow
the progress of the <cite>AlignmentList</cite> instance, unless that instance
contains only one <cite>Alignment</cite> object. In that case, the progress
information is piped from the <cite>Alignment</cite> instance. For that reason,
and for the Namespace (<cite>ns</cite>) object only, this method first checks
if the <cite>ns.sa</cite> attribute exists. If it does, it means that the parent
<cite>AlignmentList</cite> instance contains only one <cite>Alignment</cite> object and,
therefore, the progress information shoud come from here.</p>
<p class="rubric">Examples</p>
<p>Start a progress counter  for a task that will make 100 iterations:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="n">ns_obj</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar_obj</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
<span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Some message&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._table_exists">
<code class="descname">_table_exists</code><span class="sig-paren">(</span><em>table_name</em>, <em>cur=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._table_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._table_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if a table exists in the database.</p>
<p>Convenience method that checks if a table exists in the database.
It accepts a custom Cursor object, which overrides the <cite>cur</cite>
attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_name</strong> : str</p>
<blockquote>
<div><p>Name of the table.</p>
</div></blockquote>
<p><strong>cur: sqlite3.Cursor, optional</strong></p>
<blockquote>
<div><p>Custom Cursor object used to query the database.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>res</strong> : list</p>
<blockquote class="last">
<div><p>List with the results of a query for &#8216;table&#8217; type with 
<cite>table_name</cite> name. Is empty when the table does not exist.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>This returns a list that will contain one item if the table exists
in the database. If it doesn&#8217;t exist, returns an empty list.
Therefore, this can be used simply like:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="s2">&quot;my_table&quot;</span><span class="p">):</span>
     <span class="c1"># Do stuff`</span>
</pre></div>
</div>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.Alignment._test_range">
<em class="property">static </em><code class="descname">_test_range</code><span class="sig-paren">(</span><em>s</em>, <em>min_val</em>, <em>max_val</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._test_range"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._test_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if a given <cite>s</cite> integer in inside a specified range.</p>
<p>Wrapper for the tests that determine whether a certain value
(<cite>s</cite>) is within the range provided by <cite>min_val</cite> and <cite>max_val</cite>.
Both of these arguments can be None, in which case there are not
lower and/or upper boundaries, respectively.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>s</strong> : int</p>
<blockquote>
<div><p>Test value</p>
</div></blockquote>
<p><strong>min_val</strong> : int</p>
<blockquote>
<div><p>Minimum range allowed for the test value.</p>
</div></blockquote>
<p><strong>max_val</strong> : int</p>
<blockquote>
<div><p>Maximum range allowed for the test value.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : bool</p>
<blockquote class="last">
<div><p>True if <cite>s</cite> within boundaries. False if not.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.Alignment._update_pipes">
<em class="property">static </em><code class="descname">_update_pipes</code><span class="sig-paren">(</span><em>ns=None</em>, <em>pbar=None</em>, <em>value=None</em>, <em>msg=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._update_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._update_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Update progress indicators for both GUI and CLI task executions.</p>
<p>This method provides a single interface for updating the progress
objects <cite>ns</cite> or <cite>pbar</cite>, which should have been initialized at the
beginning of the task with the <cite>_set_pipes</cite> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>value</strong> : int</p>
<blockquote>
<div><p>Value of the current progress index</p>
</div></blockquote>
<p><strong>msg</strong> : str, optional</p>
<blockquote class="last">
<div><p>A secondary message that appears in TriFusion&#8217;s progress dialogs.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment._set_pipes" title="trifusion.process.sequence.Alignment._set_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_set_pipes</span></code></a>, <a class="reference internal" href="#trifusion.process.sequence.Alignment._reset_pipes" title="trifusion.process.sequence.Alignment._reset_pipes"><code class="xref py py-obj docutils literal"><span class="pre">_reset_pipes</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>The progress of any given task can be provided by either an
<cite>Alignment</cite> or <cite>AlignmentList</cite> instance. Generally, the tasks follow
the progress of the <cite>AlignmentList</cite> instance, unless that instance
contains only one <cite>Alignment</cite> object. In that case, the progress
information is piped from the <cite>Alignment</cite> instance. For that reason,
and for the Namespace (<cite>ns</cite>) object only, this method first checks
if the <cite>ns.sa</cite> attribute exists. If it does, it means that the parent
<cite>AlignmentList</cite> instance contains only one <cite>Alignment</cite> object and,
therefore, the progress information shoud come from here.</p>
<p class="rubric">Examples</p>
<p>Update the counter in an iteration of 100:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">):</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="n">ns_obj</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar_obj</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                       <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Some string&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_fasta">
<code class="descname">_write_fasta</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into fasta output format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of the output file. If using <cite>ns_pipe</cite> in TriFusion,
it will prompt the user if the output file already exists.</p>
</div></blockquote>
<p><strong>interleave</strong> : bool, optional</p>
<blockquote>
<div><p>Determines whether the output alignment
will be in leave (False) or interleave (True) format. Not all
output formats support this option.</p>
</div></blockquote>
<p><strong>ld_hat</strong> : bool, optional</p>
<blockquote>
<div><p>If True, the fasta output format will include a first line
compliant with the format of LD Hat and will truncate sequence
names and sequence length per line accordingly.</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_gphocs">
<code class="descname">_write_gphocs</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_gphocs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_gphocs" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into gphocs file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_ima2">
<code class="descname">_write_ima2</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_ima2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_ima2" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into ima2 file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>tx_space_ima2</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in ima2
format (default is 10).</p>
</div></blockquote>
<p><strong>cut_space_ima2</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
ima2 format. Longer  names are truncated (default is 8).</p>
</div></blockquote>
<p><strong>ima2_params</strong> : list</p>
<blockquote>
<div><p>A list with the additional information required for the ima2
output format. The list should contains the following information:</p>
<blockquote>
<div><ol class="arabic simple">
<li>(str) path to file containing the species and populations.</li>
<li>(str) Population tree in newick format, e.g. (0,1):2.</li>
<li>(str) Mutational model for all alignments.</li>
<li>(str) inheritance scalar.</li>
</ol>
</div></blockquote>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_mcmctree">
<code class="descname">_write_mcmctree</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_mcmctree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_mcmctree" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into mcmctree file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>tx_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in phylip
format (default is 40).</p>
</div></blockquote>
<p><strong>cut_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
phylip format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_nexus">
<code class="descname">_write_nexus</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_nexus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_nexus" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into nexus file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_format</strong> : list</p>
<blockquote>
<div><p>List with the output formats to generate. Options are:
{&#8220;fasta&#8221;, &#8220;phylip&#8221;, &#8220;nexus&#8221;, &#8220;stockholm&#8221;, &#8220;gphocs&#8221;, &#8220;ima2&#8221;}.</p>
</div></blockquote>
<p><strong>tx_space_nex</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in nexus
format (default is 40).</p>
</div></blockquote>
<p><strong>cut_space_nex</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
nexus format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>interleave</strong> : bool, optional</p>
<blockquote>
<div><p>Determines whether the output alignment
will be in leave (False) or interleave (True) format. Not all
output formats support this option.</p>
</div></blockquote>
<p><strong>gap</strong> : str, optional</p>
<blockquote>
<div><p>Symbol for alignment gaps (default is &#8216;-&#8216;).</p>
</div></blockquote>
<p><strong>use_charset</strong> : bool, optional</p>
<blockquote>
<div><p>If True, partitions from the <cite>Partitions</cite> object will be written
in the nexus output format (default is True).</p>
</div></blockquote>
<p><strong>use_nexus_models</strong> : bool, optional</p>
<blockquote>
<div><p>If True, writes the partitions charset block in nexus format.</p>
</div></blockquote>
<p><strong>outgroup_list</strong> : list, optional</p>
<blockquote>
<div><p>Specify list of taxon names that will be defined as the outgroup
in the nexus output format. This may be useful for analyses with
MrBayes or other software that may require outgroups.</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_phylip">
<code class="descname">_write_phylip</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_phylip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into phylip format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of the output file. If using <cite>ns_pipe</cite> in TriFusion,
it will prompt the user if the output file already exists.</p>
</div></blockquote>
<p><strong>tx_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in phylip
format (default is 40).</p>
</div></blockquote>
<p><strong>cut_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
phylip format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>interleave</strong> : bool, optional</p>
<blockquote>
<div><p>Determines whether the output alignment
will be in leave (False) or interleave (True) format. Not all
output formats support this option.</p>
</div></blockquote>
<p><strong>phy_truncate_names</strong> : bool, optional</p>
<blockquote>
<div><p>If True, taxon names in phylip format are truncated to a maximum
of 10 characters (default is False).</p>
</div></blockquote>
<p><strong>partition_file</strong> : bool, optional</p>
<blockquote>
<div><p>If True, the auxiliary partitions file will be written (default
is True).</p>
</div></blockquote>
<p><strong>model_phylip</strong> : str, optional</p>
<blockquote>
<div><p>Substitution model for the auxiliary partition file of phylip
format, compliant with RAxML.</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment._write_stockholm">
<code class="descname">_write_stockholm</code><span class="sig-paren">(</span><em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment._write_stockholm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment._write_stockholm" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> object into stockholm file format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.change_taxon_name">
<code class="descname">change_taxon_name</code><span class="sig-paren">(</span><em>old_name</em>, <em>new_name</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.change_taxon_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.change_taxon_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the name of a particular taxon.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>old_name</strong> : str</p>
<blockquote>
<div><p>Original taxon name.</p>
</div></blockquote>
<p><strong>new_name</strong> : str</p>
<blockquote class="last">
<div><p>New taxon name.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.code_gaps">
<code class="descname">code_gaps</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.code_gaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.code_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Code gap patterns as binary matrix at the end of alignment.</p>
<p>This method codes gaps present in the alignment in binary format,
according to the method of Simmons and Ochoterena (2000), to be read
by phylogenetic programs such as MrBayes. The resultant alignment,
however, can only be output in the Nexus format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;gaps&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool, optional</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupDatabase</span></code></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.collapse">
<code class="descname">collapse</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.collapse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.collapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Collapses equal sequences into unique haplotypes.</p>
<p>Collapses equal sequences into unique haplotypes. This method fetches
the sequences for the current alignment and creates a new database
table with the collapsed haplotypes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>write_haplotypes</strong> : bool, optional</p>
<blockquote>
<div><p>If True, a file mapping the taxa names name to their respective
haplotype name will be generated (default is True).</p>
</div></blockquote>
<p><strong>haplotypes_file</strong> : string, optional</p>
<blockquote>
<div><p>Name of the file mapping taxa names to their respective haplotype.
Only used when <cite>write_haplotypes</cite> is True. If it not provided and
<cite>write_haplotypes</cite> is True, the file name will be determined
from <cite>Alignment.name</cite>.</p>
</div></blockquote>
<p><strong>dest</strong> : string, optional</p>
<blockquote>
<div><p>Path of directory where <cite>haplotypes_file</cite> will be generated
(default is &#8221;.&#8221;).</p>
</div></blockquote>
<p><strong>conversion_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix appended to the haplotypes file. Only used when
<cite>write_haplotypes</cite> is True and <cite>haplotypes_file</cite> is None.</p>
</div></blockquote>
<p><strong>haplotype_name</strong> : string, optional</p>
<blockquote>
<div><p>Prefix of the haplotype string. The final haplotype name will be
<cite>haplotype_name</cite> + &lt;int&gt; (default is &#8220;Hap&#8221;).</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;collapsed&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool, optional</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupDatabase</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>In order to collapse an alignment, all sequence would need to be
compared and somehow stored until the last sequence is processed.
This creates the immediate issue that, if all sequences are different,
the entire alignment could have to be stored in memory. To increase
memory efficiency and increase performance, sequences are converted
into hashes, and these hashes are compared among each other instead
of the actual sequences. This means that instead of comparing
potentially very long string and maintaining them in memory, only a
small hash string is maintained and compared. In this way, only a
fraction of the information is stored in memory and the comparisons
are much faster.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.consensus">
<code class="descname">consensus</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a consensus sequence from the alignment.</p>
<p>Converts the current <cite>Alignment</cite> alignment data into a single
consensus sequence. The <cite>consensus_type</cite> argument determines how
variation in the original alignment is handled for the generation
of the consensus sequence.</p>
<p>The options for handling sequence variation are:</p>
<blockquote>
<div><ul class="simple">
<li>IUPAC: Converts variable sites according to the corresponding
IUPAC symbols (DNA sequence type only)</li>
<li>Soft mask: Converts variable sites into missing data</li>
<li>Remove: Removes variable sites</li>
<li>First sequence: Uses the first sequence in the dictionary</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>consensus_type</strong> : {&#8220;IUPAC&#8221;, &#8220;Soft mask&#8221;, &#8220;Remove&#8221;, &#8220;First sequence&#8221;}</p>
<blockquote>
<div><p>Type of variation handling. See summary above.</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;consensus&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool, optional</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupDatabase</span></code></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.e">
<code class="descname">e</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.e" title="Permalink to this definition">¶</a></dt>
<dd><p>The <cite>e</cite> attribute will store any exceptions that occur during the
parsing of the alignment object. It remains None unless something
wrong happens.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.filter_codon_positions">
<code class="descname">filter_codon_positions</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.filter_codon_positions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.filter_codon_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter alignment according to codon positions.</p>
<p>Filters codon positions from <cite>Alignment</cite> object with DNA sequence
type. The <cite>position_list</cite> argument determines which codon positions
will be stored. It should be a three element list, corresponding to
the three codon position, and the positions that are saved are the
ones with a True value.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>position_list</strong> : list</p>
<blockquote>
<div><p>List of three bool elements that correspond to each codon position.
Ex. [True, True, True] will save all positions while
[True, True, False] will exclude the third codon position</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;filter&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool, optional</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupDatabase</span></code></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.filter_informative_sites">
<code class="descname">filter_informative_sites</code><span class="sig-paren">(</span><em>min_val</em>, <em>max_val</em>, <em>table_in=None</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.filter_informative_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.filter_informative_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if the variable sites from alignment are within range.</p>
<p>Similar to <cite>filter_segregating_sites</cite> method, but only considers
informative sites (variable sites present in more than 2 taxa).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>min_val</strong> : int</p>
<blockquote>
<div><p>Minimum number of informative sites for the alignment to pass.
Can be None, in which case there is no lower bound.</p>
</div></blockquote>
<p><strong>max_val</strong> : int</p>
<blockquote>
<div><p>Maximum number of informative sites for the alignment to pass.
Can be None, in which case there is no upper bound.</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : bool</p>
<blockquote class="last">
<div><p>True if the number of informative sites are within the specified
range. False if not.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_segregating_sites" title="trifusion.process.sequence.Alignment.filter_segregating_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_segregating_sites</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.filter_missing_data">
<code class="descname">filter_missing_data</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.filter_missing_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters alignment according to missing data.</p>
<p>Filters gaps and true missing data from the alignment using tolerance
thresholds for each type of missing data. Both thresholds are maximum
percentages of sites in an alignment column containing the type of
missing data. If gap_threshold=50, for example, alignment columns with
more than 50% of sites with gaps are removed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gap_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap percentage is removed.</p>
</div></blockquote>
<p><strong>missing_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap+missing percentage is removed.</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;filter&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool, optional</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote class="last">
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.filter_segregating_sites">
<code class="descname">filter_segregating_sites</code><span class="sig-paren">(</span><em>min_val</em>, <em>max_val</em>, <em>table_in=None</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.filter_segregating_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.filter_segregating_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if the segregating sites from alignment are within range.</p>
<p>Evaluates the number of segregating sites of the current alignment
and returns True if it falls between the range specified by the
<cite>min_val</cite> and <cite>max_val</cite> arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>min_val</strong> : int</p>
<blockquote>
<div><p>Minimum number of segregating sites for the alignment to pass.
Can be None, in which case there is no lower bound.</p>
</div></blockquote>
<p><strong>max_val</strong> : int</p>
<blockquote>
<div><p>Maximum number of segregating sites for the alignment to pass.
Can be None, in which case there is no upper bound.</p>
</div></blockquote>
<p><strong>table_in</strong> : string, optional</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : bool</p>
<blockquote class="last">
<div><p>True if the number of segregating sites are within the specified
range. False if not.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_informative_sites" title="trifusion.process.sequence.Alignment.filter_informative_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_informative_sites</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.get_sequence">
<code class="descname">get_sequence</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.get_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.get_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sequence string for a given taxon.</p>
<p>Returns the sequence string of the corresponding <cite>taxon</cite>. If the
taxon does not exist in the table, raises a KeyError.
The sequence is retrieved from a database table specified either by
the <cite>table_name</cite> or <cite>table_suffix</cite> arguments. <cite>table_name</cite> will always
take precedence over <cite>table_suffix</cite> if both are provided. If none
are provided, the default <cite>Alignment.table_name</cite> is used. If the
table name provided by either <cite>table_name</cite> or <cite>table_suffix</cite> is
invalid, the default <cite>Alignment.table_name</cite> is also used.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxon</strong> : str</p>
<blockquote>
<div><p>Name of the taxon.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote>
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
<p><strong>table</strong> : string</p>
<blockquote>
<div><p>This argument will be automatically setup by the <cite>SetupInTable</cite>
decorator. Do not use directly.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>seq</strong> : str</p>
<blockquote>
<div><p>Sequence string.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><p class="first"><strong>KeyError</strong></p>
<blockquote class="last">
<div><p>If the taxon does not exist in the specified table.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>Ignores data associated with taxa present in the <cite>shelved_taxa</cite>
attribute.</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.input_format">
<code class="descname">input_format</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.input_format" title="Permalink to this definition">¶</a></dt>
<dd><p>Format of the alignment file.</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.interleave_data">
<code class="descname">interleave_data</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.interleave_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute that is set to True when interleave data has been
created for the alignment data. Buiding the interleave matrix is a bit
costly, so when it is requested by the user it is built once and
stored in the database, and then further usages will use that table.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.iter_alignment">
<code class="descname">iter_alignment</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.iter_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.iter_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator for (taxon, sequence) tuples.</p>
<p>Generator that yields (taxon, sequence) tuples from the database.
Sequence data is retrieved from a database table specified either by
the <cite>table_name</cite> or <cite>table_suffix</cite> arguments. <cite>table_name</cite> will always
take precedence over <cite>table_suffix</cite> if both are provided. If none
are provided, the default <cite>Alignment.table_name</cite> is used. If the
table name provided by either <cite>table_name</cite> or <cite>table_suffix</cite> is
invalid, the default <cite>Alignment.table_name</cite> is also used.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote>
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
<p><strong>table</strong> : string</p>
<blockquote>
<div><p>This argument will be automatically setup by the <cite>SetupInTable</cite>
decorator. Do not use directly.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first"><strong>tx</strong> : string</p>
<blockquote>
<div><p>Taxon name.</p>
</div></blockquote>
<p><strong>seq</strong> : string</p>
<blockquote class="last">
<div><p>Sequence string.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupInTable</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>All generators ignore data associated with taxa present in the
<cite>shelved_taxa</cite> attribute.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.iter_columns">
<code class="descname">iter_columns</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.iter_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.iter_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator over alignment columns.</p>
<p>Generator that yields the alignment columns in a list of characters.
Sequence data is retrieved from a database table specified either by
the <cite>table_name</cite> or <cite>table_suffix</cite> arguments. <cite>table_name</cite> will always
take precedence over <cite>table_suffix</cite> if both are provided. If none
are provided, the default <cite>Alignment.table_name</cite> is used. If the
table name provided by either <cite>table_name</cite> or <cite>table_suffix</cite> is
invalid, the default <cite>Alignment.table_name</cite> is also used.</p>
<p>The <cite>table</cite> variable will be automatically setup in the <cite>SetupInTable</cite>
decorator according to <cite>table_suffix</cite> and <cite>table_name</cite>, and should
not be used directly (values provided when calling it are actually
ignored). The decorator will set this variable</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is appended to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote>
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
<p><strong>table</strong> : string</p>
<blockquote>
<div><p>This argument will be automatically setup by the <cite>SetupInTable</cite>
decorator. Do not use directly.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first"><strong>i</strong> : list</p>
<blockquote class="last">
<div><p>List with alignment column characters as elements
([&#8220;A&#8221;, &#8220;A&#8221;, &#8220;C&#8221;, &#8220;A&#8221;, &#8220;A&#8221;]).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupInTable</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>To prevent the query from loading the entire alignment data into
memory, two techniques are employed. First, the iteration over the
query results is lazy, using <cite>itertools.izip</cite>. Second, a <cite>range</cite>
iteration is performed to ensure that only blocks of 100k alignment
sites are fetch at any given time. Several tests were performed to
ensure that this block size provided a good trade-off between
RAM usage and speed.</p>
<p>All generators ignore data associated with taxa present in the
<cite>shelved_taxa</cite> attribute.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.iter_columns_uniq">
<code class="descname">iter_columns_uniq</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.iter_columns_uniq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.iter_columns_uniq" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator over unique characters of an alignment column.</p>
<p>Generator that yields unique elements of alignment columns in a list.
Sequence data is retrieved from a database table specified either by
the <cite>table_name</cite> or <cite>table_suffix</cite> arguments. <cite>table_name</cite> will always
take precedence over <cite>table_suffix</cite> if both are provided. If none
are provided, the default <cite>Alignment.table_name</cite> is used. If the
table name provided by either <cite>table_name</cite> or <cite>table_suffix</cite> is
invalid, the default <cite>Alignment.table_name</cite> is also used.</p>
<p>The <cite>table</cite> variable will be automatically setup in the <cite>SetupInTable</cite>
decorator according to <cite>table_suffix</cite> and <cite>table_name</cite>, and should
not be used directly (values provided when calling it are actually
ignored). The decorator will set this variable</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is appended to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote>
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
<p><strong>table</strong> : string</p>
<blockquote>
<div><p>This argument will be automatically setup by the <cite>SetupInTable</cite>
decorator. Do not use directly.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first"><strong>i</strong> : list</p>
<blockquote class="last">
<div><p>List with alignment column characters as elements ([&#8220;A&#8221;, &#8220;C&#8221;]).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupInTable</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>To prevent the query from loading the entire alignment data into
memory, two techniques are employed. First, the iteration over the
query results is lazy, using <cite>itertools.izip</cite>. Second, a <cite>range</cite>
iteration is performed to ensure that only blocks of 100k alignment
sites are fetch at any given time. Several tests were performed to
ensure that this block size provided a good trade-off between
RAM usage and speed.</p>
<p>All generators ignore data associated with taxa present in the
<cite>shelved_taxa</cite> attribute.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.iter_sequences">
<code class="descname">iter_sequences</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.iter_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.iter_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator over sequence strings in the alignment.</p>
<p>Generator for sequence data of the <cite>Alignment</cite> object.
Sequence data is retrieved from a database table specified either by
the <cite>table_name</cite> or <cite>table_suffix</cite> arguments. <cite>table_name</cite> will always
take precedence over <cite>table_suffix</cite> if both are provided. If none
are provided, the default <cite>Alignment.table_name</cite> is used. If the
table name provided by either <cite>table_name</cite> or <cite>table_suffix</cite> is
invalid, the default <cite>Alignment.table_name</cite> is also used.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is appended to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote>
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
<p><strong>table</strong> : string</p>
<blockquote>
<div><p>This argument will be automatically setup by the <cite>SetupInTable</cite>
decorator. Do not use directly.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Yields:</th><td class="field-body"><p class="first"><strong>seq</strong> : string</p>
<blockquote class="last">
<div><p>Sequence string for a given taxon</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><code class="xref py py-obj docutils literal"><span class="pre">SetupInTable</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>All generators ignore data associated with taxa present in the
<cite>shelved_taxa</cite> attribute.</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.locus_length">
<code class="descname">locus_length</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.locus_length" title="Permalink to this definition">¶</a></dt>
<dd><p>The length of the alignment object. Even if the current alignment
object is partitioned, this will return the length of the entire
alignment</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.name">
<code class="descname">name</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute with basename of alignment file</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.partitions">
<code class="descname">partitions</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.partitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializing a Partitions instance for the current alignment. By
default, the Partitions instance will assume that the whole Alignment
object is one single partition. However, if the current alignment is the
result of a concatenation, or if a partitions file is provided, the
partitions argument can be used. Partitions can be later changed using
the set_partitions method. Substitution models objects are associated
with each partition and they default to None</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.path">
<code class="descname">path</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.path" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute with full path to alignment file</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.read_alignment">
<code class="descname">read_alignment</code><span class="sig-paren">(</span><em>size_check=True</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.read_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.read_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Main alignment parser method.</p>
<p>This is the main alignment parsing method that is called when the
<cite>Alignment</cite> object is instantiated with a file path as the argument.
Given the alignment format automatically detected in <cite>__init__</cite>,
it calls the specific method that parses that alignment format.
After the execution of this method, all attributes of the class will
be set and the full range of methods can be applied.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>size_check</strong> : bool, optional</p>
<blockquote class="last">
<div><p>If True, perform a check for sequence size consistency across
taxa (default is True).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.remove_alignment">
<code class="descname">remove_alignment</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.remove_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.remove_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes alignment data from the database.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.remove_taxa">
<code class="descname">remove_taxa</code><span class="sig-paren">(</span><em>taxa_list_file</em>, <em>mode='remove'</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.remove_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.remove_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes taxa from the <cite>Alignment</cite> object.</p>
<p>Removes taxa based on the  <cite>taxa_list_file</cite> argument from the
<cite>Alignment</cite> object.</p>
<p>This method supports a list or path to CSV file as the
<cite>taxa_list_file</cite> argument. In case of CSV path, it parses the CSV
file into a list. The CSV file should be a simple text file containing
taxon names in each line.</p>
<dl class="docutils">
<dt>This method also supports two removal modes.</dt>
<dd><ul class="first last simple">
<li>remove: removes the specified taxa</li>
<li>inverse: removes all but the specified taxa</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxa_list_file</strong> : list or string</p>
<blockquote>
<div><p>List containing taxa names or string with path to a CSV file
containig the taxa names.</p>
</div></blockquote>
<p><strong>mode</strong> : {&#8220;remove&#8221;, &#8220;inverse&#8221;}, optional</p>
<blockquote class="last">
<div><p>The removal mode (default is remove).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.restriction_range">
<code class="descname">restriction_range</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.restriction_range" title="Permalink to this definition">¶</a></dt>
<dd><p>This option is only relevant when gaps are coded. This will store a
string with the range of the restriction-type data that will encode
gaps and will only be used when nexus is in the output format</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.reverse_concatenate">
<code class="descname">reverse_concatenate</code><span class="sig-paren">(</span><em>table_in=''</em>, <em>db_con=None</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.reverse_concatenate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.reverse_concatenate" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverses the alignment according to the <cite>partitions</cite> attribute.</p>
<p>This method splits the current alignment according to the
partitions set in the <cite>partitions</cite> attribute. Returns an
<cite>AlignmentList</cite> object where each partition is an <cite>Alignment</cite> object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>db_con</strong> : sqlite3.Connection</p>
<blockquote>
<div><p>Database connection object that will be provided to each new
<cite>Alignment</cite> object.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.set_partitions">
<code class="descname">set_partitions</code><span class="sig-paren">(</span><em>partitions</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.set_partitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.set_partitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the <cite>partition</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>partitions</strong> : trifusion.process.data.Partitions</p>
<blockquote class="last">
<div><p>Partitions object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.shelve_taxa">
<code class="descname">shelve_taxa</code><span class="sig-paren">(</span><em>lst</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.shelve_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.shelve_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Shelves taxa from <cite>Alignment</cite> methods.</p>
<p>The taxa provided in the <cite>lst</cite> argument will be ignored in
subsequent operations. In practice, taxa from the <cite>taxa_list</cite>
attribute that are present in <cite>lst</cite> will be moved to the
<cite>shelved_taxa</cite> attribute and ignored in all methods of the class.
To revert all taxa to the <cite>taxa_list</cite> attribute, simply call this
method with an empty list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>lst</strong> : list</p>
<blockquote class="last">
<div><p>List with taxa names that should be ignored.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.shelved_taxa">
<code class="descname">shelved_taxa</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.shelved_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute that will store shelved taxa. When retrieving taxa from
the database, this list will be checked and present taxa will
be ignored</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.sname">
<code class="descname">sname</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.sname" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute with basename of alignment file without extension</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.table_name">
<code class="descname">table_name</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.table_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a database table for the current alignment. This will be the
link attribute between the sqlite database and the remaining methods
that require information on the sequence data. This also removes
any non alpha numeric characters the table name might have and ensures
that it starts with a aphabetic character to avoid a database error</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.tables">
<code class="descname">tables</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.tables" title="Permalink to this definition">¶</a></dt>
<dd><p>Lists the currently active tables for the Alignment object. The
&#8216;master&#8217; table is always present and represents the original
alignment. Additional tables may be added as needed, and then dropped
when no longer necessary. When all tables, except the master, need to
be removed, this attribute can be used to quickly drop all derived
tables</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.taxa_idx">
<code class="descname">taxa_idx</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.taxa_idx" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute that will store the index of the taxa in the sql database.
This index is stored in a dictionary instead of being retrieved
from the index position of the taxa_list list because taxa may
be removed (which would mess up the link between the two indexes)</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.Alignment.taxa_list">
<code class="descname">taxa_list</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.Alignment.taxa_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute that will store the taxa names.</p>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.Alignment.write_loci_correspondence">
<em class="property">static </em><code class="descname">write_loci_correspondence</code><span class="sig-paren">(</span><em>hap_dict</em>, <em>output_file</em>, <em>dest='./'</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.write_loci_correspondence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.write_loci_correspondence" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the file mapping taxa to unique haplotypes for <cite>collapse</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>hap_dict</strong> : dict</p>
<blockquote>
<div><p>Dictionary mapping the haplotype (key) to a list of taxa (value)
sharing the same haplotype.</p>
</div></blockquote>
<p><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of the haplotype correspondence file</p>
</div></blockquote>
<p><strong>dest</strong> : str, optional</p>
<blockquote class="last">
<div><p>Path to directory where the <cite>output_file</cite> will be written.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.Alignment.write_to_file">
<code class="descname">write_to_file</code><span class="sig-paren">(</span><em>output_format</em>, <em>output_file</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#Alignment.write_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.Alignment.write_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes alignment data into an output file.</p>
<p>Writes the alignment object into a specified output file,
automatically adding the extension, according to the specified
output formats.</p>
<p>This function supports the writing of both converted (no partitions)
and concatenated (partitioned files). The choice between these modes
is determined by the Partitions object associated with the Alignment
object. If it contains multiple partitions, it will produce a
concatenated alignment and the auxiliary partition files where
necessary. Otherwise it will treat the alignment as a single partition.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_format</strong> : list</p>
<blockquote>
<div><p>List with the output formats to generate. Options are:
{&#8220;fasta&#8221;, &#8220;phylip&#8221;, &#8220;nexus&#8221;, &#8220;stockholm&#8221;, &#8220;gphocs&#8221;, &#8220;ima2&#8221;}.</p>
</div></blockquote>
<p><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of the output file. If using <cite>ns_pipe</cite> in TriFusion,
it will prompt the user if the output file already exists.</p>
</div></blockquote>
<p><strong>tx_space_nex</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in nexus
format (default is 40).</p>
</div></blockquote>
<p><strong>tx_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in phylip
format (default is 40).</p>
</div></blockquote>
<p><strong>tx_space_ima2</strong> : int, optional</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in ima2
format (default is 10).</p>
</div></blockquote>
<p><strong>cut_space_nex</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
nexus format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>cut_space_phy</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
phylip format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>cut_space_ima2</strong> : int, optional</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
ima2 format. Longer  names are truncated (default is 8).</p>
</div></blockquote>
<p><strong>interleave</strong> : bool, optional</p>
<blockquote>
<div><p>Determines whether the output alignment
will be in leave (False) or interleave (True) format. Not all
output formats support this option.</p>
</div></blockquote>
<p><strong>gap</strong> : str, optional</p>
<blockquote>
<div><p>Symbol for alignment gaps (default is &#8216;-&#8216;).</p>
</div></blockquote>
<p><strong>model_phylip</strong> : str, optional</p>
<blockquote>
<div><p>Substitution model for the auxiliary partition file of phylip
format, compliant with RAxML.</p>
</div></blockquote>
<p><strong>outgroup_list</strong> : list, optional</p>
<blockquote>
<div><p>Specify list of taxon names that will be defined as the outgroup
in the nexus output format. This may be useful for analyses with
MrBayes or other software that may require outgroups.</p>
</div></blockquote>
<p><strong>ima2_params</strong> : list, optional</p>
<blockquote>
<div><p>A list with the additional information required for the ima2
output format. The list should contains the following information:</p>
<blockquote>
<div><ol class="arabic simple">
<li>(str) path to file containing the species and populations.</li>
<li>(str) Population tree in newick format, e.g. (0,1):2.</li>
<li>(str) Mutational model for all alignments.</li>
<li>(str) inheritance scalar.</li>
</ol>
</div></blockquote>
</div></blockquote>
<p><strong>use_charset</strong> : bool, optional</p>
<blockquote>
<div><p>If True, partitions from the <cite>Partitions</cite> object will be written
in the nexus output format (default is True).</p>
</div></blockquote>
<p><strong>partition_file</strong> : bool, optional</p>
<blockquote>
<div><p>If True, the auxiliary partitions file will be written (default
is True).</p>
</div></blockquote>
<p><strong>output_dir</strong> : str, optional</p>
<blockquote>
<div><p>If provided, the output file will be written on the specified
directory.</p>
</div></blockquote>
<p><strong>phy_truncate_names</strong> : bool, optional</p>
<blockquote>
<div><p>If True, taxon names in phylip format are truncated to a maximum
of 10 characters (default is False).</p>
</div></blockquote>
<p><strong>ld_hat</strong> : bool, optional</p>
<blockquote>
<div><p>If True, the fasta output format will include a first line
compliant with the format of LD Hat and will truncate sequence
names and sequence length per line accordingly.</p>
</div></blockquote>
<p><strong>use_nexus_models</strong> : bool, optional</p>
<blockquote>
<div><p>If True, writes the partitions charset block in nexus format.</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace, optional</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string, optional</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string, optional</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="trifusion.process.sequence.AlignmentException">
<em class="property">exception </em><code class="descclassname">trifusion.process.sequence.</code><code class="descname">AlignmentException</code><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentException"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentException" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></p>
<p>Generic Alignment object exception.</p>
</dd></dl>

<dl class="class">
<dt id="trifusion.process.sequence.AlignmentList">
<em class="property">class </em><code class="descclassname">trifusion.process.sequence.</code><code class="descname">AlignmentList</code><span class="sig-paren">(</span><em>alignment_list</em>, <em>sql_db=None</em>, <em>db_cur=None</em>, <em>db_con=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="trifusion.process.base.html#trifusion.process.base.Base" title="trifusion.process.base.Base"><code class="xref py py-class docutils literal"><span class="pre">trifusion.process.base.Base</span></code></a></p>
<p>Main interface for groups of <cite>Alignment</cite> objects.</p>
<p>The <cite>AlignmentList</cite> class can be seen as a group of <cite>Alignment</cite> objects
that is treated as a single cohesive data set. It is the main class
used to perform alignment operations on TriFusion and TriSeq programs,
even when there is only a single alignment.</p>
<p>An instance can be created by simply providing a list of paths to
alignment files, which can be empty (and alignments added later). It
is recommended that the path to the sqlite database be specified
via the <cite>sql_db</cite> argument, but this is optional.</p>
<p>Individual alignment files can be provided when instantiating the class
or added later, and each is processed and stored as an <cite>Alignment</cite> object.
The <cite>AlignmentList</cite> object provides attributes that are relative to the
global data set. For instance, each <cite>Alignment</cite> object may have their
individual and unique list of taxa, but the corresponding attribute in the
<cite>AlignmentList</cite> object contains the list of all taxa that are found
in the <cite>Alignment</cite> object list.</p>
<p>Several methods of the <cite>Alignment</cite> class are also present in this class
with the same usage (e.g. <cite>collapse</cite>). These are basically wrappers that
apply the method to all <cite>Alignment</cite> objects and may have some
modifications adapted to multiple alignments.</p>
<p>Other methods are exclusive and desgined to deal with multiple alignments,
such as concatenation (<cite>concatenate</cite>), filter by taxa (<cite>filter_by_taxa</cite>),
etc.</p>
<p>Plotting methods are exclusive to <cite>AlignmentList</cite>, even the
ones that are focused on single alignments. In this case, the
<cite>Alignment</cite> object can be specified and processed individually directly
from this class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>alignment_list</strong> : list</p>
<blockquote>
<div><p>List with paths to alignment files. Can be empty.</p>
</div></blockquote>
<p><strong>sql_db</strong> : str, optional</p>
<blockquote>
<div><p>Path where the sqlite3 database will be created.</p>
</div></blockquote>
<p><strong>db_cur</strong> : sqlite3.Cursor, optional</p>
<blockquote>
<div><p>Provide a database Cursor object in conjunction with the Connection
object (<cite>db_con</cite>) to connect to an existing database.</p>
</div></blockquote>
<p><strong>db_con</strong> : sqlite3.Connection, optional</p>
<blockquote>
<div><p>Provide a database Connection object in conjunction with the Cursor
object (<cite>db_cur</cite>) to connect to an existing database.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar, optional</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="93%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>con</td>
<td>(sqlite3.Connection) Connection object of the sqlite database.</td>
</tr>
<tr class="row-even"><td>cur</td>
<td>(sqlite3.Cursor) Cursor object of the sqlite database.</td>
</tr>
<tr class="row-odd"><td>sql_path</td>
<td>(str) Path to sqlite3 database file.</td>
</tr>
<tr class="row-even"><td>alignments</td>
<td>(collections.OrderedDict) Stores the &#8216;active&#8217; <cite>Alignment objects</cite> as values, with the corresponding key being <cite>Alignment.path</cite>.</td>
</tr>
<tr class="row-odd"><td>shelve_alignments</td>
<td>(collections.OrderedDict) Stores the &#8216;inactive&#8217; or &#8216;shelved&#8217; <cite>Alignment</cite> objects. <cite>AlignmentList</cite> methods will operate only on the <cite>alignments</cite> attribute, unless explicitly stated otherwise. The key:value is the same as in the <cite>alignments</cite> attribute.</td>
</tr>
<tr class="row-even"><td>bad_alignments</td>
<td>(list) Stores the <cite>Alignment.name</cite> attribute of badly formatted alignments.</td>
</tr>
<tr class="row-odd"><td>duplicate_alignments</td>
<td>(list) Stores the <cite>Alignment.name</cite> attribute of duplicated alignment paths.</td>
</tr>
<tr class="row-even"><td>non_alignments</td>
<td>(list) Stores the <cite>Alignment.name</cite> attribute of sequence sets of unequal length.</td>
</tr>
<tr class="row-odd"><td>taxa_names</td>
<td>(list) Stores the name of all taxa across the <cite>Alignment</cite> object list.</td>
</tr>
<tr class="row-even"><td>shelved_taxa</td>
<td>(list) Stores the &#8216;inactive&#8217; or &#8216;shelved&#8217; taxa. <cite>AlignmentList</cite> methods will operate only on the <cite>taxa_names</cite> attribute, unless explicitly stated otherwise.</td>
</tr>
<tr class="row-odd"><td>path_list</td>
<td>(list) List of <cite>Alignment.paths</cite> for all <cite>Alignment</cite> objects.</td>
</tr>
<tr class="row-even"><td>filtered_alignments</td>
<td>(collections.OrderedDict) Ordered dictionary of four key:values used to count the number of <cite>Alignment</cite> objects that were filtered by four operations. The keys are {&#8220;By minimum taxa&#8221;, &#8220;By taxa&#8221;, &#8220;By variable sites&#8221;, &#8220;By informative sites&#8221;}.</td>
</tr>
<tr class="row-odd"><td>sequence_code</td>
<td>(tuple) Contains information on (&lt;sequence type&gt;, &lt;missing data symbol&gt;), e.g. (&#8220;Protein&#8221;, &#8220;x&#8221;).</td>
</tr>
<tr class="row-even"><td>gap_symbol</td>
<td>(str) Symbol used to denote gaps in the alignment (default is &#8216;-&#8216;).</td>
</tr>
<tr class="row-odd"><td>summary_stats</td>
<td>(dict) Dictionary that stores several summary statistics that are calculated once the <cite>get_summary_stats</cite> method is executed.</td>
</tr>
<tr class="row-even"><td>summary_gene_table</td>
<td>(pandas.DataFrame) DataFrame containing the summary statistics for each <cite>Alignment</cite> object. Also populated when the <cite>get_summary_stats</cite> method is executed.</td>
</tr>
<tr class="row-odd"><td>active_tables</td>
<td>(list) List of active database tables associated with the <cite>AlignmentList</cite> instance.</td>
</tr>
<tr class="row-even"><td>partitions</td>
<td>(trifusion.process.data.Partitions) Partitions object that refers to the total <cite>AlignmentList</cite>.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignment_files" title="trifusion.process.sequence.AlignmentList.add_alignment_files"><code class="xref py py-obj docutils literal"><span class="pre">add_alignment_files</span></code></a>(file_name_list[,&nbsp;pbar,&nbsp;ns])</td>
<td>Adds a list of alignment files to the current <cite>AlignmentList</cite>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.add_alignments" title="trifusion.process.sequence.AlignmentList.add_alignments"><code class="xref py py-obj docutils literal"><span class="pre">add_alignments</span></code></a>(alignment_obj_list[,&nbsp;...])</td>
<td>Add a list of <cite>Alignment</cite> objects to the current <cite>AlignmentList</cite>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum" title="trifusion.process.sequence.AlignmentList.allele_frequency_spectrum"><code class="xref py py-obj docutils literal"><span class="pre">allele_frequency_spectrum</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for the allele frequency spectrum.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum_gene" title="trifusion.process.sequence.AlignmentList.allele_frequency_spectrum_gene"><code class="xref py py-obj docutils literal"><span class="pre">allele_frequency_spectrum_gene</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for the allele frequency spectrum of an <cite>Alignment</cite>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.aln_names" title="trifusion.process.sequence.AlignmentList.aln_names"><code class="xref py py-obj docutils literal"><span class="pre">aln_names</span></code></a>()</td>
<td>Returns list of basenames of <cite>alignments</cite> file paths.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">autofinder</span></code>(reference_file)</td>
<td>Autodetects format, missing data symbol and sequence type.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.average_seqsize" title="trifusion.process.sequence.AlignmentList.average_seqsize"><code class="xref py py-obj docutils literal"><span class="pre">average_seqsize</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for the average sequence size for the entire data set.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.average_seqsize_per_species" title="trifusion.process.sequence.AlignmentList.average_seqsize_per_species"><code class="xref py py-obj docutils literal"><span class="pre">average_seqsize_per_species</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for average sequence size per taxon.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.change_taxon_name" title="trifusion.process.sequence.AlignmentList.change_taxon_name"><code class="xref py py-obj docutils literal"><span class="pre">change_taxon_name</span></code></a>(old_name,&nbsp;new_name)</td>
<td>Changes the name of a taxon.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.characters_proportion" title="trifusion.process.sequence.AlignmentList.characters_proportion"><code class="xref py py-obj docutils literal"><span class="pre">characters_proportion</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for the proportion of nt/aa for the data set.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.characters_proportion_per_species" title="trifusion.process.sequence.AlignmentList.characters_proportion_per_species"><code class="xref py py-obj docutils literal"><span class="pre">characters_proportion_per_species</span></code></a>(*args,&nbsp;...)</td>
<td>Creates data for the proportion of nt/aa per species.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.clear_alignments" title="trifusion.process.sequence.AlignmentList.clear_alignments"><code class="xref py py-obj docutils literal"><span class="pre">clear_alignments</span></code></a>()</td>
<td>Clears all attributes and data from the <cite>AlignmentList</cite> object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.code_gaps" title="trifusion.process.sequence.AlignmentList.code_gaps"><code class="xref py py-obj docutils literal"><span class="pre">code_gaps</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Code gaps in each <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.collapse" title="trifusion.process.sequence.AlignmentList.collapse"><code class="xref py py-obj docutils literal"><span class="pre">collapse</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Collapses equal sequences for each <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.concatenate" title="trifusion.process.sequence.AlignmentList.concatenate"><code class="xref py py-obj docutils literal"><span class="pre">concatenate</span></code></a>([table_in,&nbsp;ns,&nbsp;pbar])</td>
<td>Concatenates alignments into a single <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.consensus" title="trifusion.process.sequence.AlignmentList.consensus"><code class="xref py py-obj docutils literal"><span class="pre">consensus</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates consensus sequences for each alignment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.cumulative_missing_genes" title="trifusion.process.sequence.AlignmentList.cumulative_missing_genes"><code class="xref py py-obj docutils literal"><span class="pre">cumulative_missing_genes</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for a distribution of the maximum number of genes available for consecutive thresholds of missing data.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">duplicate_taxa</span></code>(taxa_list)</td>
<td>Identified duplicate items in a list.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_by_taxa" title="trifusion.process.sequence.AlignmentList.filter_by_taxa"><code class="xref py py-obj docutils literal"><span class="pre">filter_by_taxa</span></code></a>(taxa_list,&nbsp;filter_mode[,&nbsp;ns,&nbsp;...])</td>
<td>Filters <cite>alignments</cite> if they contain or exclude certain taxa.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_codon_positions" title="trifusion.process.sequence.AlignmentList.filter_codon_positions"><code class="xref py py-obj docutils literal"><span class="pre">filter_codon_positions</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filters codon positions in each <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_informative_sites" title="trifusion.process.sequence.AlignmentList.filter_informative_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_informative_sites</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filters <cite>Alignment</cite> objects according to informative sites number.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_min_taxa" title="trifusion.process.sequence.AlignmentList.filter_min_taxa"><code class="xref py py-obj docutils literal"><span class="pre">filter_min_taxa</span></code></a>(min_taxa[,&nbsp;ns,&nbsp;pbar])</td>
<td>Filters <cite>alignments</cite> by minimum taxa proportion.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_missing_data" title="trifusion.process.sequence.AlignmentList.filter_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">filter_missing_data</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filters missing data in each <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.filter_segregating_sites" title="trifusion.process.sequence.AlignmentList.filter_segregating_sites"><code class="xref py py-obj docutils literal"><span class="pre">filter_segregating_sites</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Filters <cite>Alignment</cite> objects according to segregating sites number.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.format_list" title="trifusion.process.sequence.AlignmentList.format_list"><code class="xref py py-obj docutils literal"><span class="pre">format_list</span></code></a>()</td>
<td>Returns list of unique sequence types from <cite>Alignment</cite> objects.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.gene_occupancy" title="trifusion.process.sequence.AlignmentList.gene_occupancy"><code class="xref py py-obj docutils literal"><span class="pre">gene_occupancy</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for gene occupancy plot.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.get_gene_table_stats" title="trifusion.process.sequence.AlignmentList.get_gene_table_stats"><code class="xref py py-obj docutils literal"><span class="pre">get_gene_table_stats</span></code></a>([active_alignments,&nbsp;...])</td>
<td>Returns summary statistics for each <cite>Alignment</cite>.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">get_loci_taxa</span></code>(loci_file)</td>
<td>Get the list of taxa from a .loci file.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.get_summary_stats" title="trifusion.process.sequence.AlignmentList.get_summary_stats"><code class="xref py py-obj docutils literal"><span class="pre">get_summary_stats</span></code></a>([active_alignments,&nbsp;ns])</td>
<td>Calculates summary statistics for the &#8216;active&#8217; alignments.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.get_tables" title="trifusion.process.sequence.AlignmentList.get_tables"><code class="xref py py-obj docutils literal"><span class="pre">get_tables</span></code></a>()</td>
<td>Return list with <cite>table_name</cite> of <em>all</em> <cite>Alignment</cite> objects.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">guess_code</span></code>(sequence)</td>
<td>Guess the sequence type, i.e.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.iter_alignment_files" title="trifusion.process.sequence.AlignmentList.iter_alignment_files"><code class="xref py py-obj docutils literal"><span class="pre">iter_alignment_files</span></code></a>()</td>
<td>Returns an iterable on <cite>Alignment.path</cite> of each alignment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.length_polymorphism_correlation" title="trifusion.process.sequence.AlignmentList.length_polymorphism_correlation"><code class="xref py py-obj docutils literal"><span class="pre">length_polymorphism_correlation</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for correlation between alignment length and informative sites.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.missing_data_distribution" title="trifusion.process.sequence.AlignmentList.missing_data_distribution"><code class="xref py py-obj docutils literal"><span class="pre">missing_data_distribution</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for overall distribution of missing data plot.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.missing_data_per_species" title="trifusion.process.sequence.AlignmentList.missing_data_per_species"><code class="xref py py-obj docutils literal"><span class="pre">missing_data_per_species</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for missing data proportion per species plot.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.missing_genes_average" title="trifusion.process.sequence.AlignmentList.missing_genes_average"><code class="xref py py-obj docutils literal"><span class="pre">missing_genes_average</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates histogram data with average missing taxa.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.missing_genes_per_species" title="trifusion.process.sequence.AlignmentList.missing_genes_per_species"><code class="xref py py-obj docutils literal"><span class="pre">missing_genes_per_species</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data with distribution of missing genes per species plot.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_missing_data" title="trifusion.process.sequence.AlignmentList.outlier_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">outlier_missing_data</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for missing data alignment outliers.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_missing_data_sp" title="trifusion.process.sequence.AlignmentList.outlier_missing_data_sp"><code class="xref py py-obj docutils literal"><span class="pre">outlier_missing_data_sp</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for missing data taxa outliers.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_segregating" title="trifusion.process.sequence.AlignmentList.outlier_segregating"><code class="xref py py-obj docutils literal"><span class="pre">outlier_segregating</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for segregating site alignment outliers.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_segregating_sp" title="trifusion.process.sequence.AlignmentList.outlier_segregating_sp"><code class="xref py py-obj docutils literal"><span class="pre">outlier_segregating_sp</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for segregating sites taxa outliers.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_sequence_size" title="trifusion.process.sequence.AlignmentList.outlier_sequence_size"><code class="xref py py-obj docutils literal"><span class="pre">outlier_sequence_size</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for sequence size alignment outliers.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.outlier_sequence_size_sp" title="trifusion.process.sequence.AlignmentList.outlier_sequence_size_sp"><code class="xref py py-obj docutils literal"><span class="pre">outlier_sequence_size_sp</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for sequence size taxa outliers.</td>
</tr>
<tr class="row-even"><td><code class="xref py py-obj docutils literal"><span class="pre">read_basic_csv</span></code>(file_handle)</td>
<td>Reads a basic CSV into a list.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.remove_file" title="trifusion.process.sequence.AlignmentList.remove_file"><code class="xref py py-obj docutils literal"><span class="pre">remove_file</span></code></a>(filename_list)</td>
<td>Removes alignments.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.remove_tables" title="trifusion.process.sequence.AlignmentList.remove_tables"><code class="xref py py-obj docutils literal"><span class="pre">remove_tables</span></code></a>([preserve_tables,&nbsp;trash_tables])</td>
<td>Drops tables from the database.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.remove_taxa" title="trifusion.process.sequence.AlignmentList.remove_taxa"><code class="xref py py-obj docutils literal"><span class="pre">remove_taxa</span></code></a>(taxa_list[,&nbsp;mode])</td>
<td>Removes the specified taxa.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.resume_database" title="trifusion.process.sequence.AlignmentList.resume_database"><code class="xref py py-obj docutils literal"><span class="pre">resume_database</span></code></a>()</td>
<td>Reconnects to the sqlite database.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.retrieve_alignment" title="trifusion.process.sequence.AlignmentList.retrieve_alignment"><code class="xref py py-obj docutils literal"><span class="pre">retrieve_alignment</span></code></a>(name)</td>
<td>Return <cite>Alignment</cite> object with a given <cite>name</cite>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.reverse_concatenate" title="trifusion.process.sequence.AlignmentList.reverse_concatenate"><code class="xref py py-obj docutils literal"><span class="pre">reverse_concatenate</span></code></a>([ns])</td>
<td>Reverse a concatenated file according to the partitions.</td>
</tr>
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">rm_illegal</span></code>(taxon_string)</td>
<td>Removes illegal characters from taxon name.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.select_by_taxa" title="trifusion.process.sequence.AlignmentList.select_by_taxa"><code class="xref py py-obj docutils literal"><span class="pre">select_by_taxa</span></code></a>(taxa_list[,&nbsp;mode])</td>
<td>Selects a list of <cite>Alignment</cite> objects by taxa.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_segregation" title="trifusion.process.sequence.AlignmentList.sequence_segregation"><code class="xref py py-obj docutils literal"><span class="pre">sequence_segregation</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for overall distribution of segregating sites.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_segregation_gene" title="trifusion.process.sequence.AlignmentList.sequence_segregation_gene"><code class="xref py py-obj docutils literal"><span class="pre">sequence_segregation_gene</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Create data for a sliding window analysis of segregating sites.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_segregation_per_species" title="trifusion.process.sequence.AlignmentList.sequence_segregation_per_species"><code class="xref py py-obj docutils literal"><span class="pre">sequence_segregation_per_species</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for a triangular matrix of sequence segregation for pairs of taxa.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_similarity" title="trifusion.process.sequence.AlignmentList.sequence_similarity"><code class="xref py py-obj docutils literal"><span class="pre">sequence_similarity</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for average sequence similarity plot.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_similarity_gene" title="trifusion.process.sequence.AlignmentList.sequence_similarity_gene"><code class="xref py py-obj docutils literal"><span class="pre">sequence_similarity_gene</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for sliding window sequence similarity for alignment.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.sequence_similarity_per_species" title="trifusion.process.sequence.AlignmentList.sequence_similarity_per_species"><code class="xref py py-obj docutils literal"><span class="pre">sequence_similarity_per_species</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for sequence similarity per species pair plot.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.set_database_connections" title="trifusion.process.sequence.AlignmentList.set_database_connections"><code class="xref py py-obj docutils literal"><span class="pre">set_database_connections</span></code></a>(cur,&nbsp;con)</td>
<td>Provides Connection and Cursor to <cite>Alignment</cite> objects.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.set_partition_from_alignment" title="trifusion.process.sequence.AlignmentList.set_partition_from_alignment"><code class="xref py py-obj docutils literal"><span class="pre">set_partition_from_alignment</span></code></a>(alignment_obj)</td>
<td>Updates <cite>partitions</cite> object with the provided <cite>Alignment</cite> object.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.taxa_distribution" title="trifusion.process.sequence.AlignmentList.taxa_distribution"><code class="xref py py-obj docutils literal"><span class="pre">taxa_distribution</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Creates data for a distribution of taxa frequency across alignments.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_active_alignment" title="trifusion.process.sequence.AlignmentList.update_active_alignment"><code class="xref py py-obj docutils literal"><span class="pre">update_active_alignment</span></code></a>(aln_name,&nbsp;direction)</td>
<td>Updates the &#8216;active&#8217; status of a single alignment.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_active_alignments" title="trifusion.process.sequence.AlignmentList.update_active_alignments"><code class="xref py py-obj docutils literal"><span class="pre">update_active_alignments</span></code></a>([aln_list,&nbsp;all_files])</td>
<td>Sets the active alignments.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.update_taxa_names" title="trifusion.process.sequence.AlignmentList.update_taxa_names"><code class="xref py py-obj docutils literal"><span class="pre">update_taxa_names</span></code></a>([taxa_list,&nbsp;all_taxa])</td>
<td>Sets the active taxa.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.write_taxa_to_file" title="trifusion.process.sequence.AlignmentList.write_taxa_to_file"><code class="xref py py-obj docutils literal"><span class="pre">write_taxa_to_file</span></code></a>([file_name])</td>
<td>Writes the taxa in <cite>taxa_list</cite> to a file.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#trifusion.process.sequence.AlignmentList.write_to_file" title="trifusion.process.sequence.AlignmentList.write_to_file"><code class="xref py py-obj docutils literal"><span class="pre">write_to_file</span></code></a>(output_format[,&nbsp;...])</td>
<td>Writes <cite>Alignment</cite> objects into files.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._create_table">
<code class="descname">_create_table</code><span class="sig-paren">(</span><em>table_name</em>, <em>index=None</em>, <em>cur=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._create_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._create_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a new table in the database.</p>
<p>Convenience method that creates a new table in the sqlite database.
It accepts a custom Cursor object, which overrides the <cite>cur</cite>
attribute. Optionally, an index can also be specified.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_name</strong> : str</p>
<blockquote>
<div><p>Name of the table to be created.</p>
</div></blockquote>
<p><strong>index</strong> : list, optional</p>
<blockquote>
<div><p>Provide a list with [&lt;name of index&gt;, &lt;columns to be indexed&gt;].
(e.g., [&#8220;concindex&#8221;, &#8220;txId&#8221;]).</p>
</div></blockquote>
<p><strong>cur</strong> : sqlite3.Cursor, optional</p>
<blockquote class="last">
<div><p>Custom Cursor object used to query the database.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._get_filename_list">
<code class="descname">_get_filename_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._get_filename_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._get_filename_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list with the <cite>Alignment.name</cite> of alignments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : list</p>
<blockquote class="last">
<div><p>List of <cite>Alignment.name</cite> of alignments.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._get_informative_sites">
<code class="descname">_get_informative_sites</code><span class="sig-paren">(</span><em>aln</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._get_informative_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._get_informative_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates number of informative sites in an <cite>Alignment</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>aln</strong> : trifusion.process.sequence.Alignment</p>
<blockquote>
<div><p><cite>Alignment</cite> object.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>informative_sites</strong> : int</p>
<blockquote class="last">
<div><p>Number of informative sites for <cite>Alignment</cite> object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList._get_similarity">
<code class="descname">_get_similarity</code><em class="property"> = &lt;functools.partial object&gt;</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._get_similarity" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._get_taxa_list">
<code class="descname">_get_taxa_list</code><span class="sig-paren">(</span><em>only_active=False</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._get_taxa_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._get_taxa_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the global taxa list from all alignments.</p>
<p>If called with no arguments, it returns the taxa list from <em>all</em>
alignments, including the &#8216;inactive&#8217; ones. The <cite>only_active</cite>
argument can be set to True to return only the taxa list from the
&#8216;active&#8217; files.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>only_active</strong> : bool</p>
<blockquote>
<div><p>If True, only returns the taxa list from the &#8216;active&#8217; alignments.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>full_taxa</strong> : list</p>
<blockquote class="last">
<div><p>List with taxon names as strings.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.AlignmentList._mad_based_outlier">
<em class="property">static </em><code class="descname">_mad_based_outlier</code><span class="sig-paren">(</span><em>p</em>, <em>threshold=3.5</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._mad_based_outlier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._mad_based_outlier" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate MAD based outliers.</p>
<p>An outlier detection method based on median absolute deviation. This
code was adapted from <a class="reference external" href="http://stackoverflow.com/a/22357811/1990165">http://stackoverflow.com/a/22357811/1990165</a>.
The usage of the median is much less biased that the mean and is
robust to smaller data sets.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>p</strong> : numpy.array</p>
<blockquote>
<div><p>Array with data observations.</p>
</div></blockquote>
<p><strong>threshold</strong> : float</p>
<blockquote class="last">
<div><p>Modified Z-score to use as a threshold.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.AlignmentList._reset_pipes">
<em class="property">static </em><code class="descname">_reset_pipes</code><span class="sig-paren">(</span><em>ns</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._reset_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._reset_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset progress indicators for both GUI and CLI task executions.</p>
<p>This should be done at the end of any task that initialized the
progress objects, but it only affects the Namespace object. It
resets all Namespace attributes to None.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote class="last">
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._reset_summary_stats">
<code class="descname">_reset_summary_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._reset_summary_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._reset_summary_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets the <cite>summary_stats</cite> attribute.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._set_pipes">
<code class="descname">_set_pipes</code><span class="sig-paren">(</span><em>ns=None</em>, <em>pbar=None</em>, <em>total=None</em>, <em>msg=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._set_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._set_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Setup of progress indicators for both GUI and CLI task executions.</p>
<p>This handles the setup of the objects responsible for updating
the progress of task&#8217;s execution of both TriFusion (GUI) and
TriSeq (CLI). At the beginning of any given task, these objects
can be initialized by providing either the Namespace object (<cite>ns</cite>)
in the case of TriFusion, or the ProgressBar object (<cite>pbar</cite>), in
the case of TriSeq. Along with one of these objects, the expected
<cite>total</cite> of the progress should also be provided. The <cite>ns</cite> and
<cite>pbar</cite> objects are updated at each iteration of a given task,
and the <cite>total</cite> is used to get a measure of the progress.</p>
<p>Optionally, a message can be also provided for the Namespace object
that will be used by TriFusion.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>total</strong> : int</p>
<blockquote>
<div><p>Expected total of the task&#8217;s progress.</p>
</div></blockquote>
<p><strong>msg</strong> : str, optional</p>
<blockquote class="last">
<div><p>A secondary message that appears in TriFusion&#8217;s progress dialogs.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>The progress of any given task can be provided by either an
<cite>Alignment</cite> or <cite>AlignmentList</cite> instance. Generally, the tasks follow
the progress of the <cite>AlignmentList</cite> instance, unless that instance
contains only one <cite>Alignment</cite> object. In that case, the progress
information is piped from the <cite>Alignment</cite> instance. For that reason,
and for the Namespace (<cite>ns</cite>) object only, this method checks if
there is only one &#8216;active&#8217; <cite>Alignment</cite> object. If yes, set the
<cite>Namepsace.sa</cite> flag to True, so that the progress indicators
are piped from the <cite>Alignment</cite> object instead.</p>
<p class="rubric">Examples</p>
<p>Start a progress counter  for a task that will make 100 iterations:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">_set_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="n">ns_obj</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar_obj</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Some message&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList._table_exists">
<code class="descname">_table_exists</code><span class="sig-paren">(</span><em>table_name</em>, <em>cur=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._table_exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._table_exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if a table exists in the database.</p>
<p>Convenience method that checks if a table exists in the database.
It accepts a custom Cursor object, which overrides the <cite>cur</cite>
attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_name</strong> : str</p>
<blockquote>
<div><p>Name of the table.</p>
</div></blockquote>
<p><strong>cur: sqlite3.Cursor, optional</strong></p>
<blockquote>
<div><p>Custom Cursor object used to query the database.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>res</strong> : list</p>
<blockquote class="last">
<div><p>List with the results of a query for &#8216;table&#8217; type with
<cite>table_name</cite> name. Is empty when the table does not exist.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>This returns a list that will contain one item if the table exists
in the database. If it doesn&#8217;t exist, returns an empty list.
Therefore, this can be used simply like:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_table_exists</span><span class="p">(</span><span class="s2">&quot;my_table&quot;</span><span class="p">):</span>
     <span class="c1"># Do stuff</span>
</pre></div>
</div>
</dd></dl>

<dl class="staticmethod">
<dt id="trifusion.process.sequence.AlignmentList._update_pipes">
<em class="property">static </em><code class="descname">_update_pipes</code><span class="sig-paren">(</span><em>ns=None</em>, <em>pbar=None</em>, <em>value=None</em>, <em>msg=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList._update_pipes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList._update_pipes" title="Permalink to this definition">¶</a></dt>
<dd><p>Update progress indicators for both GUI and CLI task executions.</p>
<p>This method provides a single interface for updating the progress
objects <cite>ns</cite> or <cite>pbar</cite>, which should have been initialized at the
beginning of the task with the <cite>_set_pipes</cite> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>value</strong> : int</p>
<blockquote>
<div><p>Value of the current progress index</p>
</div></blockquote>
<p><strong>msg</strong> : str, optional</p>
<blockquote class="last">
<div><p>A secondary message that appears in TriFusion&#8217;s progress dialogs.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>The progress of any given task can be provided by either an
<cite>Alignment</cite> or <cite>AlignmentList</cite> instance. Generally, the tasks follow
the progress of the <cite>AlignmentList</cite> instance, unless that instance
contains only one <cite>Alignment</cite> object. In that case, the progress
information is piped from the <cite>Alignment</cite> instance. For that reason,
and for the Namespace (<cite>ns</cite>) object only, if the <cite>Namespace.sa</cite>
attribute is set to True, then to not update the progress. Progress
is being piped directly from the <cite>Alignment</cite> object.</p>
<p class="rubric">Examples</p>
<p>Update the counter in an iteration of 100:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">):</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_update_pipes</span><span class="p">(</span><span class="n">ns</span><span class="o">=</span><span class="n">ns_obj</span><span class="p">,</span> <span class="n">pbar</span><span class="o">=</span><span class="n">pbar_obj</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                       <span class="n">msg</span><span class="o">=</span><span class="s2">&quot;Some string&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.add_alignment_files">
<code class="descname">add_alignment_files</code><span class="sig-paren">(</span><em>file_name_list</em>, <em>pbar=None</em>, <em>ns=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.add_alignment_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.add_alignment_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a list of alignment files to the current <cite>AlignmentList</cite>.</p>
<p>Adds a list of alignment paths to the current <cite>AlignmentList</cite>. Each
path in the list is used to instantiate an <cite>Alignment</cite> object and
several checks are performed to ensure that the alignments are
correct and compliant with the other members of the <cite>AlignmentList</cite>
object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>file_name_list</strong> : list</p>
<blockquote>
<div><p>List with paths to sequence alignment files.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.add_alignments">
<code class="descname">add_alignments</code><span class="sig-paren">(</span><em>alignment_obj_list</em>, <em>ignore_paths=False</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.add_alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.add_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a list of <cite>Alignment</cite> objects to the current <cite>AlignmentList</cite>.</p>
<p>Adds a list of alignments, already as <cite>Alignment</cite> objects, to the
current <cite>AlignmentList</cite>. This method performs several checks
to assure that the <cite>Alignment</cite> objects being added are correct. If
all checks out, updates all attributes of the class with the new
alignments.</p>
<p>The <cite>ignore_paths</cite> argument can be used to prevent <cite>Alignment</cite>
objects with short paths (only basename, for instance) from being
added if they have the same basename and another in the <cite>alignments</cite>
dict.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>alignment_obj_list</strong> : list</p>
<blockquote>
<div><p>List with <cite>Alignment</cite> objects.</p>
</div></blockquote>
<p><strong>ignore_paths</strong> : bool</p>
<blockquote class="last">
<div><p>If True, the <cite>Alignment.path</cite> of the new alignments are not
checked with the loaded alignments.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.alignments">
<code class="descname">alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores the &#8220;active&#8221; <cite>Alignment</cite> objects for the current
<cite>AlignmentList</cite>. Keys will be the <cite>Alignment.path</cite> for quick lookup of
<cite>Alignment</cite> object values</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.allele_frequency_spectrum">
<code class="descname">allele_frequency_spectrum</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.allele_frequency_spectrum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for the allele frequency spectrum.</p>
<p>Calculates the allele frequency spectrum of the entire
alignment data set. Here, multiple alignments are effectively treated
as a single one. This method is exclusive of DNA sequence type and
supports IUPAC ambiguity codes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>proportions</strong> : bool</p>
<blockquote>
<div><p>If True, use proportions instead of absolute values</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.allele_frequency_spectrum_gene">
<code class="descname">allele_frequency_spectrum_gene</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.allele_frequency_spectrum_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.allele_frequency_spectrum_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for the allele frequency spectrum of an <cite>Alignment</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gene_name</strong> : str</p>
<blockquote>
<div><p><cite>Alignment.name</cite> of an alignment.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table,</p>
<p>&#8220;real_bin_num&#8221;:</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.aln_names">
<code class="descname">aln_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.aln_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.aln_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of basenames of <cite>alignments</cite> file paths.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.average_seqsize">
<code class="descname">average_seqsize</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.average_seqsize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.average_seqsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for the average sequence size for the entire data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.average_seqsize_per_species">
<code class="descname">average_seqsize_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.average_seqsize_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.average_seqsize_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for average sequence size per taxon.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.bad_alignments">
<code class="descname">bad_alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.bad_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute list that stores the Alignment.name attribute of badly
formatted alignments</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.change_taxon_name">
<code class="descname">change_taxon_name</code><span class="sig-paren">(</span><em>old_name</em>, <em>new_name</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.change_taxon_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.change_taxon_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the name of a taxon.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.characters_proportion">
<code class="descname">characters_proportion</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.characters_proportion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.characters_proportion" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for the proportion of nt/aa for the data set.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table,</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.characters_proportion_per_species">
<code class="descname">characters_proportion_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.characters_proportion_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.characters_proportion_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for the proportion of nt/aa per species.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;legend&#8221;: mpl Legend object,</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table}</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.clear_alignments">
<code class="descname">clear_alignments</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.clear_alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.clear_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Clears all attributes and data from the <cite>AlignmentList</cite> object.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.code_gaps">
<code class="descname">code_gaps</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.code_gaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.code_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Code gaps in each <cite>Alignment</cite> object.</p>
<p>This wraps the execution of <cite>code_gaps</cite> method for
each <cite>Alignment</cite> object in the <cite>alignments</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;gaps&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.code_gaps" title="trifusion.process.sequence.Alignment.code_gaps"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.code_gaps</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.collapse">
<code class="descname">collapse</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.collapse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.collapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Collapses equal sequences for each <cite>Alignment</cite> object.</p>
<p>This wraps the execution of <cite>collapse</cite> method for
each <cite>Alignment</cite> object in the <cite>alignments</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>write_haplotypes</strong> : bool</p>
<blockquote>
<div><p>If True, a file mapping the taxa names name to their respective
haplotype name will be generated (default is True).</p>
</div></blockquote>
<p><strong>haplotypes_file</strong> : string</p>
<blockquote>
<div><p>Name of the file mapping taxa names to their respective haplotype.
Only used when <cite>write_haplotypes</cite> is True. If it not provided and
<cite>write_haplotypes</cite> is True, the file name will be determined
from <cite>Alignment.name</cite>.</p>
</div></blockquote>
<p><strong>dest</strong> : string</p>
<blockquote>
<div><p>Path of directory where <cite>haplotypes_file</cite> will be generated
(default is &#8221;.&#8221;).</p>
</div></blockquote>
<p><strong>conversion_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix appended to the haplotypes file. Only used when
<cite>write_haplotypes</cite> is True and <cite>haplotypes_file</cite> is None.</p>
</div></blockquote>
<p><strong>haplotype_name</strong> : string</p>
<blockquote>
<div><p>Prefix of the haplotype string. The final haplotype name will be
<cite>haplotype_name</cite> + &lt;int&gt; (default is &#8220;Hap&#8221;).</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;collapsed&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.collapse" title="trifusion.process.sequence.Alignment.collapse"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.collapse</span></code></a></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.con">
<code class="descname">con</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.con" title="Permalink to this definition">¶</a></dt>
<dd><p>Database Connection object</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.concatenate">
<code class="descname">concatenate</code><span class="sig-paren">(</span><em>table_in=''</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.concatenate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.concatenate" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates alignments into a single <cite>Alignment</cite> object.</p>
<p>Concatenates multiple sequence alignments creating a single <cite>Alignment</cite>
object and the auxiliary Partitions object defining the partitions
of the concatenated alignment</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>concatenated_alignment</strong> : trifusion.process.sequence.Alignment</p>
<blockquote class="last">
<div><p>The <cite>Alignment</cite> object with the concatenated data and partitions.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>Since sqlite queries are quite expensive, the previous approach
of querying the sequence for each taxa AND alignment table was
dropped. In this approach, sqlite will do all the heavy lifting.
First, a temporary table with the same definition as all alignment
tables is populated with data from all taxa and alignments. It is
important that and index is created on the new txId column, which
will redifine the txId of individual alignments according to the
global <cite>taxa_names</cite> attribute. In the first procedure, there
will be only one query per alignment. When the temporary table is
complete, some sql operations are used to group sequences from each
taxon for all alignments and then concatenated them, all in a single
query. This query returns the concatenated sequences, corrected txId
and taxon information, ready to populate the final concatenation
table. This approach is an order o magnitude faster than the
previous one where thousands of queries could be performed.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.consensus">
<code class="descname">consensus</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates consensus sequences for each alignment.</p>
<p>This wraps the execution of <cite>consensus</cite> method for
each <cite>Alignment</cite> object in the <cite>alignments</cite> attribute. It has an
additional argument, <cite>single_file</cite>, which can be used to merge
the consensus sequence from each alignment into a new <cite>Alignment</cite>
object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>consensus_type</strong> : {&#8220;IUPAC&#8221;, &#8220;Soft mask&#8221;, &#8220;Remove&#8221;, &#8220;First sequence&#8221;}</p>
<blockquote>
<div><p>Type of variation handling. See summary above.</p>
</div></blockquote>
<p><strong>single_file</strong> : bool</p>
<blockquote>
<div><p>If True, the consensus sequence of each <cite>Alignment</cite> object will
be merged into a single <cite>Alignment</cite> object.</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;consensus&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>Returns</strong></p>
<p><strong>consensus_aln</strong> : trifusion.process.sequence.Alignment</p>
<blockquote class="last">
<div><p>Alignment object with all the consensus sequences merged. It&#8217;s
only returned when the <cite>single_file</cite> parameter is set to True.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.consensus" title="trifusion.process.sequence.Alignment.consensus"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.consensus</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.cumulative_missing_genes">
<code class="descname">cumulative_missing_genes</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.cumulative_missing_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.cumulative_missing_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for a distribution of the maximum number of genes
available for consecutive thresholds of missing data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table}</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.cur">
<code class="descname">cur</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.cur" title="Permalink to this definition">¶</a></dt>
<dd><p>Database Cursor object</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.duplicate_alignments">
<code class="descname">duplicate_alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.duplicate_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute list that stores duplicate Alignment.name.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_by_taxa">
<code class="descname">filter_by_taxa</code><span class="sig-paren">(</span><em>taxa_list</em>, <em>filter_mode</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_by_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_by_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters <cite>alignments</cite> if they contain or exclude certain taxa.</p>
<p>Filters the <cite>alignments</cite> attribute by a given taxa list.
The filtering may be performed to filter alignments
depending on whether they include or exclude certain taxa.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxa_list</strong> : list</p>
<blockquote>
<div><p>List of taxa names.</p>
</div></blockquote>
<p><strong>filter_mode</strong> : str</p>
<blockquote>
<div><p>Filter mode. Can be &#8220;contain&#8221; or &#8220;exclude&#8221;.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_codon_positions">
<code class="descname">filter_codon_positions</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_codon_positions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_codon_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters codon positions in each <cite>Alignment</cite> object.</p>
<p>This wraps the execution of <cite>filter_codon_positions</cite> method for
each <cite>Alignment</cite> object in the <cite>alignments</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>position_list</strong> : list</p>
<blockquote>
<div><p>List of three bool elements that correspond to each codon position.
Ex. [True, True, True] will save all positions while
[True, True, False] will exclude the third codon position</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;filter&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_codon_positions" title="trifusion.process.sequence.Alignment.filter_codon_positions"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.filter_codon_positions</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_informative_sites">
<code class="descname">filter_informative_sites</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_informative_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_informative_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters <cite>Alignment</cite> objects according to informative sites number.</p>
<p>Filters <cite>alignments</cite> according to whether or not their number
of informative sites is within a specified range. <cite>Alignment</cite> objects
with a number of informative sites outside the range are move to
the <cite>shelve_alignments</cite> and the counter of <cite>filtered_alignments</cite> is
updated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>min_val</strong> : int</p>
<blockquote>
<div><p>Minimum number of informative sites for the alignment to pass.
Can be None, in which case there is no lower bound.</p>
</div></blockquote>
<p><strong>max_val</strong> : int</p>
<blockquote>
<div><p>Maximum number of informative sites for the alignment to pass.
Can be None, in which case there is no upper bound.</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_informative_sites" title="trifusion.process.sequence.Alignment.filter_informative_sites"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.filter_informative_sites</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_min_taxa">
<code class="descname">filter_min_taxa</code><span class="sig-paren">(</span><em>min_taxa</em>, <em>ns=None</em>, <em>pbar=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_min_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_min_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters <cite>alignments</cite> by minimum taxa proportion.</p>
<p>Filters <cite>Alignment</cite> objects based on a minimum taxa representation
threshold. Alignments with less that the specified minimum taxa
percentage will be moved to the <cite>filtered_alignments</cite> attribute.</p>
<p>NOTE: Since this filtering is meant to be performed when executing
the process operations, it will permanently change the AlignmentList
object, which means both self.alignments and self.partitions. Not doing
so and removing/adding the partitions would create a great deal of
conflicts that can be easily avoided by simply copying the
AlignmentList object and modifying this object for the process
execution</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>min_taxa</strong> : int</p>
<blockquote>
<div><p>Percentage of minimum allowed taxa representation (e.g. 50).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_missing_data">
<code class="descname">filter_missing_data</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_missing_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_missing_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters missing data in each <cite>Alignment</cite> object.</p>
<p>This wraps the execution of <cite>filter_missing_data</cite> method for
each <cite>Alignment</cite> object in the <cite>alignments</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gap_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap percentage is removed.</p>
</div></blockquote>
<p><strong>missing_threshold</strong> : int</p>
<blockquote>
<div><p>Integer between 0 and 100 defining the percentage above which
a column with that gap+missing percentage is removed.</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>table_out</strong> : string</p>
<blockquote>
<div><p>Name of database table where the final alignment will be inserted
(default is &#8220;filter&#8221;).</p>
</div></blockquote>
<p><strong>use_main_table</strong> : bool</p>
<blockquote>
<div><p>If True, both <cite>table_in</cite> and <cite>table_out</cite> are ignore and the main
table <cite>Alignment.table_name</cite> is used as the input and output
table (default is False).</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_missing_data" title="trifusion.process.sequence.Alignment.filter_missing_data"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.filter_missing_data</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.filter_segregating_sites">
<code class="descname">filter_segregating_sites</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.filter_segregating_sites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filter_segregating_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters <cite>Alignment</cite> objects according to segregating sites number.</p>
<p>Filters <cite>alignments</cite> according to whether or not their number
of segregating sites is within a specified range. <cite>Alignment</cite> objects
with a number of segregating sites outside the range are move to
the <cite>shelve_alignments</cite> and the counter of <cite>filtered_alignments</cite> is
updated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>min_val</strong> : int</p>
<blockquote>
<div><p>Minimum number of segregating sites for the alignment to pass.
Can be None, in which case there is no lower bound.</p>
</div></blockquote>
<p><strong>max_val</strong> : int</p>
<blockquote>
<div><p>Maximum number of segregating sites for the alignment to pass.
Can be None, in which case there is no upper bound.</p>
</div></blockquote>
<p><strong>table_in</strong> : string</p>
<blockquote>
<div><p>Name of database table containing the alignment data that is
used for this operation.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote class="last">
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.filter_segregating_sites" title="trifusion.process.sequence.Alignment.filter_segregating_sites"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.filter_segregating_sites</span></code></a></p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.filtered_alignments">
<code class="descname">filtered_alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.filtered_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Records the number of filtered alignments by each individual filter type</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.format_list">
<code class="descname">format_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.format_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.format_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of unique sequence types from <cite>Alignment</cite> objects.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : list</p>
<blockquote class="last">
<div><p>List with unique sequence types for <cite>Alignment</cite> objects as
strings.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.gap_symbol">
<code class="descname">gap_symbol</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.gap_symbol" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary with summary statistics for the active alignments</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.gene_occupancy">
<code class="descname">gene_occupancy</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.gene_occupancy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.gene_occupancy" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for gene occupancy plot.</p>
<p>Creates data for an interpolation plot to visualize the amount of
missing genes in a alignment dataset.</p>
<dl class="docutils">
<dt>ns <span class="classifier-delimiter">:</span> <span class="classifier">multiprocesssing.Manager.Namespace</span></dt>
<dd>A Namespace object used to communicate with the main thread
in TriFusion.</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.get_gene_table_stats">
<code class="descname">get_gene_table_stats</code><span class="sig-paren">(</span><em>active_alignments=None</em>, <em>sortby=None</em>, <em>ascending=True</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.get_gene_table_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.get_gene_table_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns summary statistics for each <cite>Alignment</cite>.</p>
<p>Gets summary statistics for each individual gene for the gene table
view of TriFusion. Summary statistics have already been calculated in
get_summary_stats, so here we only get the active alignments for
showing. Therefore, this function should only be called after
get_summary_stats.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>active_alignments</strong> : list</p>
<blockquote>
<div><p>List of <cite>Alignment.name</cite> objects to show</p>
</div></blockquote>
<p><strong>sortby</strong> : str</p>
<blockquote>
<div><p>Table header used for sorting. Can be {&#8220;nsites&#8221;, &#8220;taxa&#8221;, &#8220;var&#8221;,
&#8220;inf&#8221;, &#8220;gap&#8221;, &#8220;missing&#8221;}.</p>
</div></blockquote>
<p><strong>ascending</strong> : bool</p>
<blockquote>
<div><p>If True, the <cite>sortby</cite> column will be sorted in ascending order
(default is True).</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>summary_gene_table</strong> : pandas.DataFrame</p>
<blockquote>
<div><p>DataFrame with summary statistics for each alignment.</p>
</div></blockquote>
<p><strong>table</strong> : list</p>
<blockquote class="last">
<div><p>List with the data for creating .csv tables.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.get_summary_stats">
<code class="descname">get_summary_stats</code><span class="sig-paren">(</span><em>active_alignments=None</em>, <em>ns=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.get_summary_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.get_summary_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates summary statistics for the &#8216;active&#8217; alignments.</p>
<p>Creates/Updates summary statistics for the active alignments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>active_alignments</strong> : list</p>
<blockquote>
<div><p>List of <cite>Alignment</cite> objects for which the summary statistics
will be calculated</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>summary_stats</strong> : dict</p>
<blockquote>
<div><p>Dictionary with the overall summary statistics</p>
</div></blockquote>
<p><strong>table</strong> : list</p>
<blockquote class="last">
<div><p>List with overall summary statistics for creating .csv tables.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.get_tables">
<code class="descname">get_tables</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.get_tables"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.get_tables" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list with <cite>table_name</cite> of <em>all</em> <cite>Alignment</cite> objects.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : list</p>
<blockquote class="last">
<div><p>List with <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.iter_alignment_files">
<code class="descname">iter_alignment_files</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.iter_alignment_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.iter_alignment_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterable on <cite>Alignment.path</cite> of each alignment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : iter</p>
<blockquote class="last">
<div><p>Iterable over the <cite>Alignment.path</cite> attribute of each alignment.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.length_polymorphism_correlation">
<code class="descname">length_polymorphism_correlation</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.length_polymorphism_correlation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.length_polymorphism_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for correlation between alignment length and
informative sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;correlation&#8221;: bool, whether to perform spearman correlation,</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.missing_data_distribution">
<code class="descname">missing_data_distribution</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.missing_data_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.missing_data_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for overall distribution of missing data plot.</p>
<p>Creates data for overall distribution of missing data. This will
calculate the amount of gaps, missing and actual data.</p>
<dl class="docutils">
<dt>ns <span class="classifier-delimiter">:</span> <span class="classifier">multiprocesssing.Manager.Namespace</span></dt>
<dd>A Namespace object used to communicate with the main thread
in TriFusion.</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;legend&#8221;: mpl Legend object,</p>
<p>&#8220;table_header&#8221;: list with headers of table.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.missing_data_per_species">
<code class="descname">missing_data_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.missing_data_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.missing_data_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for missing data proportion per species plot.</p>
<p>For each taxon in <cite>taxon_list</cite> calculate the proportion of gap,
missing and actual data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;legend&#8221;: mpl Legend object,</p>
<p>&#8220;table_header&#8221;: list with headers of table,</p>
<p>&#8220;normalize&#8221;: bool, whether data values should be normalized,</p>
<p>&#8220;normalize_factor&#8221;: int, factor to use in normalization</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.missing_genes_average">
<code class="descname">missing_genes_average</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.missing_genes_average"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.missing_genes_average" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates histogram data with average missing taxa.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.missing_genes_per_species">
<code class="descname">missing_genes_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.missing_genes_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.missing_genes_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data with distribution of missing genes per species plot.</p>
<p>Calculates, for each taxon, the number of alignments where it is
present.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.non_alignments">
<code class="descname">non_alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.non_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Attribute list that stores sequence sets of unequal lenght</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_missing_data">
<code class="descname">outlier_missing_data</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_missing_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_missing_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for missing data alignment outliers.</p>
<p>Get data for outlier detection of genes based on the distribution of
average missing data per gene. Data points will be based on the
proportion of missing data symbols out of the possible total. For
example, in an alignment with three taxa, each with 100 sites,
the total possible missing data is 300 (100 * 3). Here, missing data
will be gathered from all taxa and a proportion will be calculated
based n the total possible</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_missing_data_sp">
<code class="descname">outlier_missing_data_sp</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_missing_data_sp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_missing_data_sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for missing data taxa outliers.</p>
<p>Gets data for outlier detection of species based on missing data. For
this analysis, genes for which a taxa is completely absent will be
ignored for the calculations of that taxa. The reason for this,
is that including genes where the taxon is absent would bias the
outlier detection towards taxa that have low prevalence in the data
set, even if they have low missing data in the alignments where they
are present.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels}</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_segregating">
<code class="descname">outlier_segregating</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_segregating"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_segregating" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for segregating site alignment outliers.</p>
<p>Generates data for the outlier detection of genes based on
segregating sites. The data will be based on the number of alignments
columns with a variable number of sites, excluding gaps and missing
data</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_segregating_sp">
<code class="descname">outlier_segregating_sp</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_segregating_sp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_segregating_sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for segregating sites taxa outliers.</p>
<p>Generates data for the outlier detection of species based on their
average pair-wise proportion of segregating sites. Comparisons
with gaps or missing data are ignored</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_sequence_size">
<code class="descname">outlier_sequence_size</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_sequence_size"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_sequence_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for sequence size alignment outliers.</p>
<p>Generates data for the outlier detection of genes based on their
sequence length (excluding missing data).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.outlier_sequence_size_sp">
<code class="descname">outlier_sequence_size_sp</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.outlier_sequence_size_sp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.outlier_sequence_size_sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for sequence size taxa outliers.</p>
<p>Generates data for the outlier detection of species based on their
sequence length (excluding missing data)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;outliers&#8221;: list of outlier data points,</p>
<p>&#8220;outliers_labels&#8221;: list of outlier labels</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.path_list">
<code class="descname">path_list</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.path_list" title="Permalink to this definition">¶</a></dt>
<dd><p>List of <cite>Alignment.path</cite></p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.remove_file">
<code class="descname">remove_file</code><span class="sig-paren">(</span><em>filename_list</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.remove_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.remove_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes alignments.</p>
<p>Removes <cite>Alignment</cite> objects from <cite>alignments</cite> and <cite>shelve_alignments</cite>
based on their <cite>name</cite> attribute.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>filename_list</strong> : list</p>
<blockquote class="last">
<div><p>List of <cite>Alignment.name</cite> to be removed.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.remove_tables">
<code class="descname">remove_tables</code><span class="sig-paren">(</span><em>preserve_tables=None</em>, <em>trash_tables=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.remove_tables"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.remove_tables" title="Permalink to this definition">¶</a></dt>
<dd><p>Drops tables from the database.</p>
<p>This method can be used to drop tables from the database.</p>
<p>If no arguments are provided it will remove <em>all</em>  database tables
associated with the <cite>AlignmentList</cite> object.  If the
<cite>preserve_tables</cite> argument is provided, the table names present in
this list will not be dropped. If <cite>trash_tables</cite> is provided,
only the tables names specified in that list are dropped. If
both are provided, <cite>trash_tables</cite> will take precedence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>preserve_tables</strong> : list, optional</p>
<blockquote>
<div><p>If provided, the table names in this list will NOT be dropped.</p>
</div></blockquote>
<p><strong>trash_tables</strong> : list, optional</p>
<blockquote class="last">
<div><p>If provided, only the table names in this list will be dropped.
Takes precedence over <cite>preserve_tables</cite> if both are provided.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.remove_taxa">
<code class="descname">remove_taxa</code><span class="sig-paren">(</span><em>taxa_list</em>, <em>mode='remove'</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.remove_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.remove_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes the specified taxa.</p>
<p>Removes the specified taxa from the global <cite>AlignmentList</cite> attributes
and from each <cite>Alignment</cite> object.</p>
<p>This method supports a list or path to CSV file as the
<cite>taxa_list_file</cite> argument. In case of CSV path, it parses the CSV
file into a list. The CSV file should be a simple text file containing
taxon names in each line.</p>
<dl class="docutils">
<dt>This method also supports two removal modes.</dt>
<dd><ul class="first last simple">
<li>remove: removes the specified taxa</li>
<li>inverse: removes all but the specified taxa</li>
</ul>
</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxa_list_file</strong> : list or string</p>
<blockquote>
<div><p>List containing taxa names or string with path to a CSV file
containig the taxa names.</p>
</div></blockquote>
<p><strong>mode</strong> : {&#8220;remove&#8221;, &#8220;inverse&#8221;}</p>
<blockquote class="last">
<div><p>The removal mode (default is remove).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#trifusion.process.sequence.Alignment.remove_taxa" title="trifusion.process.sequence.Alignment.remove_taxa"><code class="xref py py-obj docutils literal"><span class="pre">Alignment.remove_taxa</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.resume_database">
<code class="descname">resume_database</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.resume_database"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.resume_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Reconnects to the sqlite database.</p>
<p>The connection to the database can be closes or even severed when
passing objects between threads. This methods allows the
reconnection of the database for the <cite>AlignmentList</cite> and
<em>all</em> (even the shelved ones) <cite>Alignment</cite> objects.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.retrieve_alignment">
<code class="descname">retrieve_alignment</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.retrieve_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.retrieve_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Return <cite>Alignment</cite> object with a given <cite>name</cite>.</p>
<p>Retrieves the <cite>Alignment</cite> object with a matching <cite>name</cite> attribute
from <cite>alignments</cite> or <cite>shelve_alignments</cite>. Returns None if no match
is found.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>name</strong> : str</p>
<blockquote>
<div><p><cite>name</cite> attribute of an <cite>Alignment</cite> object.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : trifusion.process.sequence.Alignment</p>
<blockquote class="last">
<div><p><cite>Alignment</cite> object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.reverse_concatenate">
<code class="descname">reverse_concatenate</code><span class="sig-paren">(</span><em>ns=None</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.reverse_concatenate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.reverse_concatenate" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse a concatenated file according to the partitions.</p>
<p>This is basically a wrapper of the <cite>reverse_concatenate</cite> method
of <cite>Alignment</cite> and is rarely useful. It is preferable to retrieve
an <cite>Alignment</cite> object, provide some partitions and then
perform the reverse concatenation. Since the reverse concatenation
operation can only be applied to <cite>Alignment</cite> objects, this method
actually creates a concatenated <cite>Alignment</cite> .</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>reverted_alns</strong> : AlignmentList</p>
<blockquote class="last">
<div><p>AlignmentList with the partitions as <cite>Alignment</cite> objects.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.select_by_taxa">
<code class="descname">select_by_taxa</code><span class="sig-paren">(</span><em>taxa_list</em>, <em>mode='strict'</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.select_by_taxa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.select_by_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Selects a list of <cite>Alignment</cite> objects by taxa.</p>
<p>This method is used to select <cite>Alignments</cite> based on whether they
contain or exclude certain taxa. The selected alignments are
returned in a list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxa_list</strong> : list</p>
<blockquote>
<div><p>List of taxa names.</p>
</div></blockquote>
<p><strong>mode</strong> : str</p>
<blockquote>
<div><p>Selection mode. Can be:</p>
<blockquote>
<div><ul class="simple">
<li>&#8220;strict&#8221;: Selects alignments containing all and only the</li>
</ul>
<p>provide taxa.
- &#8220;inclusive&#8221;: Selects alignments containing the provided
taxa.
- &#8220;relaxed&#8221;: Selects alignments if they contain at least
one of the provided taxa.</p>
</div></blockquote>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>selected_alignments</strong> : list</p>
<blockquote class="last">
<div><p>List of <cite>Alignment</cite> objects.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.sequence_code">
<code class="descname">sequence_code</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_code" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple with the AlignmentList sequence code. Either (&#8220;DNA&#8221;, &#8220;n&#8221;) or
(&#8220;Protein&#8221;, &#8220;x&#8221;)</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_segregation">
<code class="descname">sequence_segregation</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_segregation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_segregation" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for overall distribution of segregating sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>proportions</strong> : bool</p>
<blockquote>
<div><p>If True, use proportions instead of absolute values</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table,</p>
<p>&#8220;real_bin_num&#8221;:</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_segregation_gene">
<code class="descname">sequence_segregation_gene</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_segregation_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_segregation_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Create data for a sliding window analysis of segregating sites.</p>
<p>Retrieves an alignment using <cite>gene_name</cite> and calculates the
number of segregating sites along the alignment length
using a sliding window approach.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gene_name</strong> : str</p>
<blockquote>
<div><p><cite>Alignment.name</cite> from an alignment</p>
</div></blockquote>
<p><strong>window_size</strong> : int</p>
<blockquote>
<div><p>Size of sliding window.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;window_size&#8221;: int, size of sliding window,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_segregation_per_species">
<code class="descname">sequence_segregation_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_segregation_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_segregation_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for a triangular matrix of sequence segregation for
pairs of taxa.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;color_label&#8221;: str, label for colorbar</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_similarity">
<code class="descname">sequence_similarity</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_similarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_similarity" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for average sequence similarity plot.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y]</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_similarity_gene">
<code class="descname">sequence_similarity_gene</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_similarity_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_similarity_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for sliding window sequence similarity for alignment.</p>
<p>Retrieves an alignment using <cite>gene_name</cite> and calculates the
average pair-wise sequence similarity along the alignment length
using a sliding window approach.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>gene_name</strong> : str</p>
<blockquote>
<div><p><cite>Alignment.name</cite> from an alignment</p>
</div></blockquote>
<p><strong>window_size</strong> : int</p>
<blockquote>
<div><p>Size of sliding window.</p>
</div></blockquote>
<p><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;window_size&#8221;: int, size of sliding window,</p>
<p>&#8220;ax_names&#8221;: 2 element list with axis labels [x, y],</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;table_header&#8221;: list with headers of table</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.sequence_similarity_per_species">
<code class="descname">sequence_similarity_per_species</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.sequence_similarity_per_species"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sequence_similarity_per_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for sequence similarity per species pair plot.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;color_label&#8221;: str, label for colorbar</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.set_database_connections">
<code class="descname">set_database_connections</code><span class="sig-paren">(</span><em>cur</em>, <em>con</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.set_database_connections"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.set_database_connections" title="Permalink to this definition">¶</a></dt>
<dd><p>Provides Connection and Cursor to <cite>Alignment</cite> objects.</p>
<p>Sets the database connections manually for <em>all</em> (even the
shelved ones) <cite>Alignment</cite> objects.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>cur</strong> : sqlite3.Cursor</p>
<blockquote>
<div><p>Provide a database Cursor object.</p>
</div></blockquote>
<p><strong>con</strong> : sqlite3.Connection</p>
<blockquote class="last">
<div><p>Provide a database Connection object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.set_partition_from_alignment">
<code class="descname">set_partition_from_alignment</code><span class="sig-paren">(</span><em>alignment_obj</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.set_partition_from_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.set_partition_from_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates <cite>partitions</cite> object with the provided <cite>Alignment</cite> object.</p>
<p>Uses the <cite>partitions</cite> of an <cite>Alignment</cite> object to set the <cite>partitions</cite>
attribute of <cite>AlignmentList</cite></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>alignment_obj</strong> : trifusion.process.sequence.Alignment</p>
<blockquote class="last">
<div><p><cite>Alignment</cite> object.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.shelve_alignments">
<code class="descname">shelve_alignments</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.shelve_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores the &#8220;inactive&#8221; or &#8220;shelved&#8221; <cite>Alignment</cite> objects. All
<cite>AlignmentList</cite> methods will operate only on the alignments
attribute, unless explicitly stated otherwise.
Key-value is the same as the <cite>alignments</cite> attribute.</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.shelved_taxa">
<code class="descname">shelved_taxa</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.shelved_taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>List with non active taxa</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.sql_path">
<code class="descname">sql_path</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.sql_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to sqlite3 database file</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.summary_gene_table">
<code class="descname">summary_gene_table</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.summary_gene_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Lists the currently active tables. This is mainly used for the
conversion of the consensus alignments into a single Alignment object</p>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.summary_stats">
<code class="descname">summary_stats</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.summary_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary with summary statistics for each alignment.</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.taxa_distribution">
<code class="descname">taxa_distribution</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.taxa_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.taxa_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates data for a distribution of taxa frequency across alignments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>ns</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>_</strong> : dict</p>
<blockquote class="last">
<div><p>&#8220;data&#8221;: <cite>numpy.array</cite> with data for plotting,</p>
<p>&#8220;labels&#8221;: list, label for xticks,</p>
<p>&#8220;title&#8221;: str with title,</p>
<p>&#8220;color_label&#8221;: str, label for colorbar,</p>
<p>&#8220;real_bin_num&#8221;:</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="trifusion.process.sequence.AlignmentList.taxa_names">
<code class="descname">taxa_names</code><em class="property"> = None</em><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.taxa_names" title="Permalink to this definition">¶</a></dt>
<dd><p>List with the name of the taxa included in the AlignmentList object</p>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.update_active_alignment">
<code class="descname">update_active_alignment</code><span class="sig-paren">(</span><em>aln_name</em>, <em>direction</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.update_active_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.update_active_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the &#8216;active&#8217; status of a single alignment.</p>
<p>Changes the &#8216;active&#8217; status of a single alignment in a specified
direction.</p>
<p>After updating the <cite>alignments`and `shelve_alignments</cite> attributes,
the <cite>taxa_names</cite> attribute is also updated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>aln_name</strong> : str</p>
<blockquote>
<div><p><cite>Alignment.name</cite> to update.</p>
</div></blockquote>
<p><strong>direction</strong> : str</p>
<blockquote class="last">
<div><p>Direction where <cite>aln_name</cite> will be moved. Can be
{&#8220;shelve&#8221;, &#8220;active&#8221;}.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.update_active_alignments">
<code class="descname">update_active_alignments</code><span class="sig-paren">(</span><em>aln_list=None</em>, <em>all_files=False</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.update_active_alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.update_active_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the active alignments.</p>
<p>Sets the &#8216;active&#8217; alignments stored in the <cite>alignments</cite> attribute
and the &#8216;inactive&#8217; alignments stored in the <cite>shelve_alignments</cite>
attribute, based on the <cite>aln_list</cite> argument. Regardless of the
current &#8216;active&#8217; and &#8216;inactive&#8217; alignments, calling this method
with the <cite>aln_list</cite> argument will set only those as the &#8216;active&#8217;
alignments.</p>
<p>Alternatively, the <cite>all_files</cite> bool argument can be provided to
set all alignments as &#8216;active&#8217;.</p>
<p>After updating the <cite>alignments`and `shelve_alignments</cite> attributes,
the <cite>taxa_names</cite> attribute is also updated.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>aln_list</strong> : list</p>
<blockquote>
<div><p>List of <cite>Alignment.name</cite> that will be set as &#8216;active&#8217;.</p>
</div></blockquote>
<p><strong>all_files</strong> : bool</p>
<blockquote class="last">
<div><p>If True, all alignments become active. Ignores <cite>aln_list</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.update_taxa_names">
<code class="descname">update_taxa_names</code><span class="sig-paren">(</span><em>taxa_list=None</em>, <em>all_taxa=False</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.update_taxa_names"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.update_taxa_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the active taxa.</p>
<p>Sets the &#8216;active&#8217; taxa stored in the <cite>taxa_names</cite> attribute and the
&#8216;inactive&#8217; taxa stored in the <cite>shelved_taxa</cite> attribute, based on
the <cite>taxa_list</cite> argument.</p>
<p>Alternatively, the <cite>all_taxa</cite> bool argument can be provided to
set all taxa as &#8216;active&#8217;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>taxa_list</strong> : list</p>
<blockquote>
<div><p>List of taxon names that will be set as &#8216;active&#8217;.</p>
</div></blockquote>
<p><strong>all_taxa</strong> : bool</p>
<blockquote class="last">
<div><p>If True, all alignments become active. Ignores <cite>taxa_list</cite>.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.write_taxa_to_file">
<code class="descname">write_taxa_to_file</code><span class="sig-paren">(</span><em>file_name='Taxa_list.csv'</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.write_taxa_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.write_taxa_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes the taxa in <cite>taxa_list</cite> to a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>file_name</strong> : string</p>
<blockquote class="last">
<div><p>Path to the generated file (default is &#8216;Taxa_list.csv&#8217;).</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="trifusion.process.sequence.AlignmentList.write_to_file">
<code class="descname">write_to_file</code><span class="sig-paren">(</span><em>output_format</em>, <em>conversion_suffix=''</em>, <em>output_suffix=''</em>, <em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentList.write_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentList.write_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes <cite>Alignment</cite> objects into files.</p>
<p>Wrapper of the <cite>write_to_file</cite> method of the <cite>Alignment</cite> object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>output_format</strong> : list</p>
<blockquote>
<div><p>List with the output formats to generate. Options are:
{&#8220;fasta&#8221;, &#8220;phylip&#8221;, &#8220;nexus&#8221;, &#8220;stockholm&#8221;, &#8220;gphocs&#8221;, &#8220;ima2&#8221;}.</p>
</div></blockquote>
<p><strong>output_file</strong> : str</p>
<blockquote>
<div><p>Name of the output file. If using <cite>ns_pipe</cite> in TriFusion,
it will prompt the user if the output file already exists.</p>
</div></blockquote>
<p><strong>tx_space_nex</strong> : int</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in nexus
format (default is 40).</p>
</div></blockquote>
<p><strong>tx_space_phy</strong> : int</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in phylip
format (default is 40).</p>
</div></blockquote>
<p><strong>tx_space_ima2</strong> : int</p>
<blockquote>
<div><p>Space (in characters) provided for the taxon name in ima2
format (default is 10).</p>
</div></blockquote>
<p><strong>cut_space_nex</strong> : int</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
nexus format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>cut_space_phy</strong> : int</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
phylip format. Longer  names are truncated (default is 39).</p>
</div></blockquote>
<p><strong>cut_space_ima2</strong> : int</p>
<blockquote>
<div><p>Set the maximum allowed character length for taxon names in
ima2 format. Longer  names are truncated (default is 8).</p>
</div></blockquote>
<p><strong>interleave</strong> : bool</p>
<blockquote>
<div><p>Determines whether the output alignment
will be in leave (False) or interleave (True) format. Not all
output formats support this option.</p>
</div></blockquote>
<p><strong>gap</strong> : str</p>
<blockquote>
<div><p>Symbol for alignment gaps (default is &#8216;-&#8216;).</p>
</div></blockquote>
<p><strong>model_phylip</strong> : str</p>
<blockquote>
<div><p>Substitution model for the auxiliary partition file of phylip
format, compliant with RAxML.</p>
</div></blockquote>
<p><strong>outgroup_list</strong> : list</p>
<blockquote>
<div><p>Specify list of taxon names that will be defined as the outgroup
in the nexus output format. This may be useful for analyses with
MrBayes or other software that may require outgroups.</p>
</div></blockquote>
<p><strong>ima2_params</strong> : list</p>
<blockquote>
<div><p>A list with the additional information required for the ima2
output format. The list should contains the following information:</p>
<blockquote>
<div><ol class="arabic simple">
<li>(str) path to file containing the species and populations.</li>
<li>(str) Population tree in newick format, e.g. (0,1):2.</li>
<li>(str) Mutational model for all alignments.</li>
<li>(str) inheritance scalar.</li>
</ol>
</div></blockquote>
</div></blockquote>
<p><strong>use_charset</strong> : bool</p>
<blockquote>
<div><p>If True, partitions from the <cite>Partitions</cite> object will be written
in the nexus output format (default is True).</p>
</div></blockquote>
<p><strong>partition_file</strong> : bool</p>
<blockquote>
<div><p>If True, the auxiliary partitions file will be written (default
is True).</p>
</div></blockquote>
<p><strong>output_dir</strong> : str</p>
<blockquote>
<div><p>If provided, the output file will be written on the specified
directory.</p>
</div></blockquote>
<p><strong>phy_truncate_names</strong> : bool</p>
<blockquote>
<div><p>If True, taxon names in phylip format are truncated to a maximum
of 10 characters (default is False).</p>
</div></blockquote>
<p><strong>ld_hat</strong> : bool</p>
<blockquote>
<div><p>If True, the fasta output format will include a first line
compliant with the format of LD Hat and will truncate sequence
names and sequence length per line accordingly.</p>
</div></blockquote>
<p><strong>use_nexus_models</strong> : bool</p>
<blockquote>
<div><p>If True, writes the partitions charset block in nexus format.</p>
</div></blockquote>
<p><strong>ns_pipe</strong> : multiprocesssing.Manager.Namespace</p>
<blockquote>
<div><p>A Namespace object used to communicate with the main thread
in TriFusion.</p>
</div></blockquote>
<p><strong>pbar</strong> : ProgressBar</p>
<blockquote>
<div><p>A ProgressBar object used to log the progress of TriSeq execution.</p>
</div></blockquote>
<p><strong>table_suffix</strong> : string</p>
<blockquote>
<div><p>Suffix of the table from where the sequence data is fetched.
The suffix is append to <cite>Alignment.table_name</cite>.</p>
</div></blockquote>
<p><strong>table_name</strong> : string</p>
<blockquote class="last">
<div><p>Name of the table from where the sequence data is fetched.</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="trifusion.process.sequence.AlignmentUnequalLength">
<em class="property">exception </em><code class="descclassname">trifusion.process.sequence.</code><code class="descname">AlignmentUnequalLength</code><a class="reference internal" href="modules/trifusion/process/sequence.html#AlignmentUnequalLength"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.AlignmentUnequalLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">exceptions.Exception</span></code></p>
<p>Raised when sequences in alignment have unequal length.</p>
</dd></dl>

<dl class="class">
<dt id="trifusion.process.sequence.LookupDatabase">
<em class="property">class </em><code class="descclassname">trifusion.process.sequence.</code><code class="descname">LookupDatabase</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#LookupDatabase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.LookupDatabase" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Decorator handling hash lookup table with pre-calculated values.</p>
<p>This decorator class is used to decorate class methods that calculate 
pairwise sequence similarity. To ensure proper functionality, the
decorated method should be called first with a single &#8220;connect&#8221; argument
and after finishing all calculations, with a single &#8220;disconnect&#8221; argument.
These special method callings are necessary to setup and close the
database storing the calculated values, respectively.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>func</strong> : function</p>
<blockquote class="last">
<div><p>Decorated function</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>The <cite>con</cite> and <cite>c</cite> attributes are initialized when the decorated function
is called with the single &#8220;connect&#8221; argument 
(e.g. decorated_func(&#8220;connect&#8221;)). When the decorated function is called
with the single &#8220;disconnect&#8221; argument (e.g. decorated_func(&#8220;disconnect&#8221;)),
the database changes are committed, the sqlite Connection is closed and
the Cursor is reset.</p>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="94%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>func</td>
<td>(function) Decorated function</td>
</tr>
<tr class="row-even"><td>con</td>
<td>(None or sqlite.Connection) Connection object of the sqlite database</td>
</tr>
<tr class="row-odd"><td>c</td>
<td>(None or sqlite.Cursor) Cursor object of the sqlite database</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#trifusion.process.sequence.LookupDatabase.__call__" title="trifusion.process.sequence.LookupDatabase.__call__"><code class="xref py py-obj docutils literal"><span class="pre">__call__</span></code></a>(*args)</td>
<td>Wraps the call of <cite>func</cite>.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="trifusion.process.sequence.LookupDatabase.__call__">
<code class="descname">__call__</code><span class="sig-paren">(</span><em>*args</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#LookupDatabase.__call__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.LookupDatabase.__call__" title="Permalink to this definition">¶</a></dt>
<dd><p>Wraps the call of <cite>func</cite>.</p>
<p>When the decorated method is called, this code is wrapped around
its execution. It accepts an arbitrary number of positional arguments
but is is currently designed to decorate the <cite>_get_similarity</cite>
method of the <cite>AlignmentList</cite> class. There are three expected calling
modes:</p>
<blockquote>
<div>1. decorated_func(&#8220;connect&#8221;) : Initializes the Connection and
Cursor objects of the database
2. decorated_func(seq1, seq2, locus_lenght) : The typical main
execution of the method, providing the sequence1 and sequence2
strings along with the integer with their lenght
3. decorated_func(&#8220;disconnect&#8221;) : Commits changes to database
and closes Connection and Cursor objects.</div></blockquote>
<p>The path of this sqlite database is automatically obtained from 
the <cite>AlignmentList</cite> attribute <cite>sql_path</cite>. We use the path to the 
same directory, but with a different name for the database, &#8220;pw.db&#8221;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>args</strong> : list</p>
<blockquote class="last">
<div><p>List of positional arguments for decorated method</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>When the decorated method is called normally (i.e., not with a 
&#8220;connect&#8221; or &#8220;disconnect&#8221; argument&#8221;), it first creates an hash
of the argument list (excluding <cite>self</cite>). This hash is used to query
the database to check if a value for that combination of sequence
strings and sequence length has already been calculated. If yes, 
the result is immediately  returned without executing the decorated
method. If no, the method is executed to perform calculations. The
result of this calculation is then stored in the database and the 
result is returned.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="trifusion.process.sequence.check_data">
<code class="descclassname">trifusion.process.sequence.</code><code class="descname">check_data</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#check_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.check_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator handling the result from AlignmentList plotting methods.</p>
<p>This should decorate all plotting methods from the <cite>AlignmentList</cite> class.
It can be used to control warnings and handle exceptions when calling
the plotting methods. Currently, it ensures that a list of methods are
not executed when the <cite>AlignmentList</cite> instance contains a single 
alignment, and handles cases where the plotting methods return an
empty array of data. Further control can be added here to prevent
methods from being executed in certain conditions and handling 
certain outputs of the plotting methods.</p>
<p>The only requirement is that, even when the plotting methods are not
executed, this should always return a dictionary. In such cases,
this dictionary should contain a single key:value with
{&#8220;exception&#8221;:&lt;exception_type_string&gt;}. The &lt;exception_type_string&gt; should
be a simple string with an informative name that will be handled in
the <cite>stats_write_plot</cite> of the <cite>TriFusionApp</cite> class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>func</strong> : function</p>
<blockquote>
<div><p>Decorated function</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>res</strong> : dict</p>
<blockquote class="last">
<div><p>The value returned by <cite>func</cite> is always a dictionary. If the
decorator prevents the execution of <cite>func</cite> for some reason, ensure
that a dictionary is always return with a single key:value
{&#8220;exception&#8221;: &lt;exception_type_string&gt;}</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="trifusion.process.sequence.setup_database">
<code class="descclassname">trifusion.process.sequence.</code><code class="descname">setup_database</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#setup_database"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.setup_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator handling the active database tables.</p>
<p>Decorates methods from the <cite>Alignment</cite> object that use and
perform modifications to the original alignment data. All methods 
decorated with this must have the keyword arguments <cite>table_in</cite> and
<cite>table_out</cite> (and <cite>use_main_table</cite>, optionally). The values associated with 
these arguments will determine which tables will be used and created 
before the execution of the decorated method (See Notes for the 
rationale). The strings provided as arguments for <cite>table_in</cite> and 
<cite>table_out</cite> will serve as a suffix to the <cite>Alignment</cite> instance attribute
<cite>table_name</cite>.</p>
<p>If <cite>use_main_table</cite> is provided, and is set to True, both <cite>table_in</cite>
and <cite>table_out</cite> will default to <cite>Alignment.table_name</cite>. Values provided
for <cite>table_in</cite> and <cite>table_out</cite> at calling time will be ignored.</p>
<p>The following cases assume <cite>use_main_table</cite> is not provided or set to 
False.</p>
<p>If both <cite>table_in</cite> and <cite>table_out</cite> are provided, <cite>table_out</cite> is modified
so that <cite>table_out = Alignment.table_name + table_out</cite>. If <cite>table_out</cite>
is not provided, it defaults to <cite>Alignment.table_name</cite>.</p>
<p>If the final <cite>table_out</cite> does not exist in the database, create it. 
In this case, if <cite>table_in</cite> will default to <cite>Alignment.table_name</cite>. If 
<cite>table_out</cite> already exists in the database, then <cite>table_in=table_out</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>func</strong> : function</p>
<blockquote class="last">
<div><p>Decorated function</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>In order to fully understand the mechanism behind the setup of the
database tables, one must first know that when an <cite>Alignment</cite> instance
is created, it generates a table in the database containing the
original data from the alignment.
In TriFusion (GUI), this original table MUST NOT be modified, since users
may want to execute several methods on the same <cite>Alignment</cite> object. 
Therefore, when a particular method needs to modify the original alignment,
a new temporary table is created to store the modified version
until the end of the execution. If a second modification is requested,
it will also be necessary to set the input table as the output table
of the previous alignment modification.
In the case of TriSeq (CLI), there is no such requirement, 
so it&#8217;s much simpler to use the same table  for all modifications.</p>
<p>This decorator greatly simplifies this process in the same way for 
all methods of the <cite>Alignment</cite> object that modify the original alignment
data. To accomplish this, all decorated method must have the keyword
arguments: <cite>table_in</cite> and <cite>table_out</cite> (and <cite>use_main_table</cite>, optionally).</p>
<p>For methods called within the execution of TriFusion, the idea is simple.
Since we don&#8217;t know which methods will be used by the user, all chained
methods that will create the same output file can be called with 
<cite>table_in=new_table</cite> and <cite>table_out=new_table</cite>. 
Note that bot arguments have the same value. Whatever is the first method
being called, it will face the fact that &#8220;new_table&#8221; does not yet exist.
In this case, the decorator will create a &#8220;new_table&#8221; in the database and 
reset <cite>table_in</cite> to the original <cite>Alignment.table_name</cite>. This ensures that
the first method will still be able to fetch the alignment data. In the
following methods, <cite>table_in</cite> and <cite>table_out</cite> will be used based on the
original values, ensuring that the alignment data is being fetched
from the last modification made and that the modification chain is 
maintained. In this way, the execution of <cite>Alignment</cite> methods can
be the same, regardless of the order or number of operations requested
by the user.</p>
<p>In the case of TriSeq, the methods should set <cite>use_main_table=True</cite>.
This will always set <cite>table_in=table_out=Alignment.table_name</cite>, which
means that the main database will be used for all methods.</p>
</dd></dl>

<dl class="function">
<dt id="trifusion.process.sequence.setup_intable">
<code class="descclassname">trifusion.process.sequence.</code><code class="descname">setup_intable</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="reference internal" href="modules/trifusion/process/sequence.html#setup_intable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#trifusion.process.sequence.setup_intable" title="Permalink to this definition">¶</a></dt>
<dd><p>Decorator handling active database table when fetching data.</p>
<p>This class is mean to decorate methods of the <cite>Alignment</cite> class that
retrieves alignment data from the database. The requirement is that
these methods have the positional arguments <cite>table_suffix</cite>, 
<cite>table_name</cite> and <cite>table</cite>. These are all optional, and only <cite>table_suffix</cite>
and <cite>table_name</cite> should be used when calling these methods. The values
of these two will be used to define the value of <cite>table</cite>, so any
value provided to this argument when calling the decorated method will
be ignored. In the end, the <cite>table</cite> variable will contain the final
table name and only this variable will be used to retrieve the alignment
data.</p>
<p><cite>table_suffix</cite> is a suffix that is appended to <cite>Alignment.table_name</cite>.
<cite>table_name</cite> defines a complete table name.</p>
<p>If neither <cite>table_suffix</cite> nor <cite>table_name</cite> are provided, then <cite>table</cite>
will default to <cite>Alignment.table_name</cite>.</p>
<p>If <cite>table_name</cite> is provided, <cite>table=table_name</cite>. If <cite>table_suffix</cite>
is provided, <cite>table=Alignment.table_name + table_suffix</cite>.</p>
<p>If both are provided, <cite>table_name</cite> takes precedence over <cite>table_suffix</cite>
and <cite>table=table_name</cite>.</p>
<p>In any case, we test the existence of the final <cite>table</cite> value in the
database. If it does not exist, <cite>table</cite> will revert to 
<cite>Alignment.table_name</cite> to prevent errors.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>func</strong> : function</p>
<blockquote class="last">
<div><p>Decorated function</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="88%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>func</td>
<td>(function) Decorated function</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="trifusion.process.error_handling.html" class="btn btn-neutral" title="trifusion.process.error_handling module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Diogo N. Silva.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.5.5',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="static/jquery.js"></script>
      <script type="text/javascript" src="static/underscore.js"></script>
      <script type="text/javascript" src="static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>